// @spartacus/core/esm5/src/user/connectors/address/user-address.connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { UserAddressAdapter } from './user-address.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./user-address.adapter";
var UserAddressConnector = /** @class */ (function () {
    function UserAddressConnector(adapter) {
        this.adapter = adapter;
    }
    /**
     * @param {?} userId
     * @return {?}
     */
    UserAddressConnector.prototype.getAll = /**
     * @param {?} userId
     * @return {?}
     */
    function (userId) {
        return this.adapter.loadAll(userId);
    };
    /**
     * @param {?} userId
     * @param {?} address
     * @return {?}
     */
    UserAddressConnector.prototype.add = /**
     * @param {?} userId
     * @param {?} address
     * @return {?}
     */
    function (userId, address) {
        return this.adapter.add(userId, address);
    };
    /**
     * @param {?} userId
     * @param {?} addressId
     * @param {?} address
     * @return {?}
     */
    UserAddressConnector.prototype.update = /**
     * @param {?} userId
     * @param {?} addressId
     * @param {?} address
     * @return {?}
     */
    function (userId, addressId, address) {
        return this.adapter.update(userId, addressId, address);
    };
    /**
     * @param {?} userId
     * @param {?} address
     * @return {?}
     */
    UserAddressConnector.prototype.verify = /**
     * @param {?} userId
     * @param {?} address
     * @return {?}
     */
    function (userId, address) {
        return this.adapter.verify(userId, address);
    };
    /**
     * @param {?} userId
     * @param {?} addressId
     * @return {?}
     */
    UserAddressConnector.prototype.delete = /**
     * @param {?} userId
     * @param {?} addressId
     * @return {?}
     */
    function (userId, addressId) {
        return this.adapter.delete(userId, addressId);
    };
    UserAddressConnector.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    UserAddressConnector.ctorParameters = function () { return [
        { type: UserAddressAdapter }
    ]; };
    /** @nocollapse */ UserAddressConnector.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function UserAddressConnector_Factory() { return new UserAddressConnector(i0.ɵɵinject(i1.UserAddressAdapter)); }, token: UserAddressConnector, providedIn: "root" });
    return UserAddressConnector;
}());
export { UserAddressConnector };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    UserAddressConnector.prototype.adapter;
}
