// @spartacus/core/esm5/src/asm/http-interceptors/csagent-token.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
import { switchMap, take } from 'rxjs/operators';
import { InterceptorUtil, USE_CUSTOMER_SUPPORT_AGENT_TOKEN, } from '../../occ/utils/interceptor-util';
import { AsmAuthService } from '../facade/asm-auth.service';
import * as i0 from "@angular/core";
import * as i1 from "../facade/asm-auth.service";
var CustomerSupportAgentTokenInterceptor = /** @class */ (function () {
    function CustomerSupportAgentTokenInterceptor(asmAuthService) {
        this.asmAuthService = asmAuthService;
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    CustomerSupportAgentTokenInterceptor.prototype.intercept = /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    function (request, next) {
        return this.getCustomerSupportAgentToken(request).pipe(take(1), switchMap((/**
         * @param {?} token
         * @return {?}
         */
        function (token) {
            if (token) {
                request = request.clone({
                    setHeaders: {
                        Authorization: token.token_type + " " + token.access_token,
                    },
                });
            }
            return next.handle(request);
        })));
    };
    /**
     * @private
     * @param {?} request
     * @return {?}
     */
    CustomerSupportAgentTokenInterceptor.prototype.getCustomerSupportAgentToken = /**
     * @private
     * @param {?} request
     * @return {?}
     */
    function (request) {
        if (InterceptorUtil.getInterceptorParam(USE_CUSTOMER_SUPPORT_AGENT_TOKEN, request.headers)) {
            return this.asmAuthService.getCustomerSupportAgentToken();
        }
        return of(null);
    };
    CustomerSupportAgentTokenInterceptor.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    CustomerSupportAgentTokenInterceptor.ctorParameters = function () { return [
        { type: AsmAuthService }
    ]; };
    /** @nocollapse */ CustomerSupportAgentTokenInterceptor.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function CustomerSupportAgentTokenInterceptor_Factory() { return new CustomerSupportAgentTokenInterceptor(i0.ɵɵinject(i1.AsmAuthService)); }, token: CustomerSupportAgentTokenInterceptor, providedIn: "root" });
    return CustomerSupportAgentTokenInterceptor;
}());
export { CustomerSupportAgentTokenInterceptor };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CustomerSupportAgentTokenInterceptor.prototype.asmAuthService;
}
