// @spartacus/core/esm2015/src/site-context/store/selectors/currencies.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { getSiteContextState } from './site-context.selector';
/** @type {?} */
const currenciesEntitiesSelector = (/**
 * @param {?} state
 * @return {?}
 */
(state) => state.entities);
const ɵ0 = currenciesEntitiesSelector;
/** @type {?} */
const activeCurrencySelector = (/**
 * @param {?} state
 * @return {?}
 */
(state) => state.activeCurrency);
const ɵ1 = activeCurrencySelector;
const ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
(state) => state.currencies;
/** @type {?} */
export const getCurrenciesState = createSelector(getSiteContextState, (ɵ2));
/** @type {?} */
export const getCurrenciesEntities = createSelector(getCurrenciesState, currenciesEntitiesSelector);
/** @type {?} */
export const getActiveCurrency = createSelector(getCurrenciesState, activeCurrencySelector);
const ɵ3 = /**
 * @param {?} entities
 * @return {?}
 */
entities => {
    return entities
        ? Object.keys(entities).map((/**
         * @param {?} isocode
         * @return {?}
         */
        isocode => entities[isocode]))
        : null;
};
/** @type {?} */
export const getAllCurrencies = createSelector(getCurrenciesEntities, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
