// @spartacus/core/esm2015/src/auth/store/actions/user-token.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const LOAD_USER_TOKEN = '[Auth] Load User Token';
/** @type {?} */
export const LOAD_USER_TOKEN_FAIL = '[Auth] Load User Token Fail';
/** @type {?} */
export const LOAD_USER_TOKEN_SUCCESS = '[Auth] Load User Token Success';
/** @type {?} */
export const REFRESH_USER_TOKEN = '[Auth] Refresh User Token';
/** @type {?} */
export const REFRESH_USER_TOKEN_FAIL = '[Auth] Refresh User Token Fail';
/** @type {?} */
export const REFRESH_USER_TOKEN_SUCCESS = '[Auth] Refresh User Token Success';
/** @type {?} */
export const REVOKE_USER_TOKEN = '[Auth] Revoke User Token';
/** @type {?} */
export const REVOKE_USER_TOKEN_FAIL = '[Auth] Revoke User Token Fail';
/** @type {?} */
export const REVOKE_USER_TOKEN_SUCCESS = '[Auth] Revoke User Token Success';
export class LoadUserToken {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_USER_TOKEN;
    }
}
if (false) {
    /** @type {?} */
    LoadUserToken.prototype.type;
    /** @type {?} */
    LoadUserToken.prototype.payload;
}
export class LoadUserTokenFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_USER_TOKEN_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadUserTokenFail.prototype.type;
    /** @type {?} */
    LoadUserTokenFail.prototype.payload;
}
export class LoadUserTokenSuccess {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_USER_TOKEN_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadUserTokenSuccess.prototype.type;
    /** @type {?} */
    LoadUserTokenSuccess.prototype.payload;
}
export class RefreshUserToken {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = REFRESH_USER_TOKEN;
    }
}
if (false) {
    /** @type {?} */
    RefreshUserToken.prototype.type;
    /** @type {?} */
    RefreshUserToken.prototype.payload;
}
export class RefreshUserTokenSuccess {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = REFRESH_USER_TOKEN_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    RefreshUserTokenSuccess.prototype.type;
    /** @type {?} */
    RefreshUserTokenSuccess.prototype.payload;
}
export class RefreshUserTokenFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = REFRESH_USER_TOKEN_FAIL;
    }
}
if (false) {
    /** @type {?} */
    RefreshUserTokenFail.prototype.type;
    /** @type {?} */
    RefreshUserTokenFail.prototype.payload;
}
export class RevokeUserToken {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = REVOKE_USER_TOKEN;
    }
}
if (false) {
    /** @type {?} */
    RevokeUserToken.prototype.type;
    /** @type {?} */
    RevokeUserToken.prototype.payload;
}
export class RevokeUserTokenSuccess {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = REVOKE_USER_TOKEN_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    RevokeUserTokenSuccess.prototype.type;
    /** @type {?} */
    RevokeUserTokenSuccess.prototype.payload;
}
export class RevokeUserTokenFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = REVOKE_USER_TOKEN_FAIL;
    }
}
if (false) {
    /** @type {?} */
    RevokeUserTokenFail.prototype.type;
    /** @type {?} */
    RevokeUserTokenFail.prototype.payload;
}
