// @spartacus/core/esm2015/src/user/connectors/customer-coupon/customer-coupon.connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { CustomerCouponAdapter } from './customer-coupon.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./customer-coupon.adapter";
export class CustomerCouponConnector {
    /**
     * @param {?} adapter
     */
    constructor(adapter) {
        this.adapter = adapter;
    }
    /**
     * @param {?} userId
     * @param {?} pageSize
     * @param {?} currentPage
     * @param {?} sort
     * @return {?}
     */
    getCustomerCoupons(userId, pageSize, currentPage, sort) {
        return this.adapter.getCustomerCoupons(userId, pageSize, currentPage, sort);
    }
    /**
     * @param {?} userId
     * @param {?} couponCode
     * @return {?}
     */
    turnOnNotification(userId, couponCode) {
        return this.adapter.turnOnNotification(userId, couponCode);
    }
    /**
     * @param {?} userId
     * @param {?} couponCode
     * @return {?}
     */
    turnOffNotification(userId, couponCode) {
        return this.adapter.turnOffNotification(userId, couponCode);
    }
    /**
     * @param {?} userId
     * @param {?} couponCode
     * @return {?}
     */
    claimCustomerCoupon(userId, couponCode) {
        return this.adapter.claimCustomerCoupon(userId, couponCode);
    }
}
CustomerCouponConnector.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
CustomerCouponConnector.ctorParameters = () => [
    { type: CustomerCouponAdapter }
];
/** @nocollapse */ CustomerCouponConnector.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function CustomerCouponConnector_Factory() { return new CustomerCouponConnector(i0.ɵɵinject(i1.CustomerCouponAdapter)); }, token: CustomerCouponConnector, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    CustomerCouponConnector.prototype.adapter;
}
