// @spartacus/core/esm5/src/state/utils/entity-processes-loader/entity-processes-loader.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { entityMeta } from '../entity/entity.action';
import { processesDecrementMeta, processesIncrementMeta, processesLoaderResetMeta, } from '../processes-loader/processes-loader.action';
/** @type {?} */
export var ENTITY_PROCESSES_LOADER_RESET_ACTION = '[ENTITY] PROCESSES LOADER RESET';
/** @type {?} */
export var ENTITY_PROCESSES_INCREMENT_ACTION = '[ENTITY] PROCESSES INCREMENT';
/** @type {?} */
export var ENTITY_PROCESSES_DECREMENT_ACTION = '[ENTITY] PROCESSES DECREMENT';
/**
 * @record
 */
export function EntityProcessesLoaderMeta() { }
/**
 * @record
 */
export function EntityProcessesLoaderAction() { }
if (false) {
    /** @type {?|undefined} */
    EntityProcessesLoaderAction.prototype.payload;
    /** @type {?|undefined} */
    EntityProcessesLoaderAction.prototype.meta;
}
/**
 * @param {?} entityType
 * @param {?} id
 * @return {?}
 */
export function entityProcessesLoaderResetMeta(entityType, id) {
    return tslib_1.__assign({}, processesLoaderResetMeta(entityType), entityMeta(entityType, id));
}
/**
 * @param {?} entityType
 * @param {?} id
 * @return {?}
 */
export function entityProcessesIncrementMeta(entityType, id) {
    return tslib_1.__assign({}, processesIncrementMeta(entityType), entityMeta(entityType, id));
}
/**
 * @param {?} entityType
 * @param {?} id
 * @return {?}
 */
export function entityProcessesDecrementMeta(entityType, id) {
    return tslib_1.__assign({}, processesDecrementMeta(entityType), entityMeta(entityType, id));
}
var EntityProcessesLoaderResetAction = /** @class */ (function () {
    function EntityProcessesLoaderResetAction(entityType, id) {
        this.type = ENTITY_PROCESSES_LOADER_RESET_ACTION;
        this.meta = entityProcessesLoaderResetMeta(entityType, id);
    }
    return EntityProcessesLoaderResetAction;
}());
export { EntityProcessesLoaderResetAction };
if (false) {
    /** @type {?} */
    EntityProcessesLoaderResetAction.prototype.type;
    /** @type {?} */
    EntityProcessesLoaderResetAction.prototype.meta;
}
var EntityProcessesIncrementAction = /** @class */ (function () {
    function EntityProcessesIncrementAction(entityType, id) {
        this.type = ENTITY_PROCESSES_INCREMENT_ACTION;
        this.meta = entityProcessesIncrementMeta(entityType, id);
    }
    return EntityProcessesIncrementAction;
}());
export { EntityProcessesIncrementAction };
if (false) {
    /** @type {?} */
    EntityProcessesIncrementAction.prototype.type;
    /** @type {?} */
    EntityProcessesIncrementAction.prototype.meta;
}
var EntityProcessesDecrementAction = /** @class */ (function () {
    function EntityProcessesDecrementAction(entityType, id) {
        this.type = ENTITY_PROCESSES_DECREMENT_ACTION;
        this.meta = entityProcessesDecrementMeta(entityType, id);
    }
    return EntityProcessesDecrementAction;
}());
export { EntityProcessesDecrementAction };
if (false) {
    /** @type {?} */
    EntityProcessesDecrementAction.prototype.type;
    /** @type {?} */
    EntityProcessesDecrementAction.prototype.meta;
}
