// @spartacus/core/esm5/src/features-config/directives/feature-level.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { FeatureConfigService } from '../services/feature-config.service';
var FeatureLevelDirective = /** @class */ (function () {
    function FeatureLevelDirective(templateRef, viewContainer, featureConfig) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.featureConfig = featureConfig;
        this.hasView = false;
    }
    Object.defineProperty(FeatureLevelDirective.prototype, "cxFeatureLevel", {
        set: /**
         * @param {?} level
         * @return {?}
         */
        function (level) {
            if (this.featureConfig.isLevel(level.toString()) && !this.hasView) {
                this.viewContainer.createEmbeddedView(this.templateRef);
                this.hasView = true;
            }
            else if (!this.featureConfig.isLevel(level.toString()) && this.hasView) {
                this.viewContainer.clear();
                this.hasView = false;
            }
        },
        enumerable: true,
        configurable: true
    });
    FeatureLevelDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[cxFeatureLevel]',
                },] }
    ];
    /** @nocollapse */
    FeatureLevelDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef },
        { type: FeatureConfigService }
    ]; };
    FeatureLevelDirective.propDecorators = {
        cxFeatureLevel: [{ type: Input }]
    };
    return FeatureLevelDirective;
}());
export { FeatureLevelDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FeatureLevelDirective.prototype.hasView;
    /**
     * @type {?}
     * @protected
     */
    FeatureLevelDirective.prototype.templateRef;
    /**
     * @type {?}
     * @protected
     */
    FeatureLevelDirective.prototype.viewContainer;
    /**
     * @type {?}
     * @protected
     */
    FeatureLevelDirective.prototype.featureConfig;
}
