// @spartacus/core/esm5/src/features-config/directives/feature.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { FeatureConfigService } from '../services/feature-config.service';
var FeatureDirective = /** @class */ (function () {
    function FeatureDirective(templateRef, viewContainer, featureConfig) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.featureConfig = featureConfig;
        this.hasView = false;
    }
    Object.defineProperty(FeatureDirective.prototype, "cxFeature", {
        set: /**
         * @param {?} feature
         * @return {?}
         */
        function (feature) {
            if (this.featureConfig.isEnabled(feature) && !this.hasView) {
                this.viewContainer.createEmbeddedView(this.templateRef);
                this.hasView = true;
            }
            else if (!this.featureConfig.isEnabled(feature) && this.hasView) {
                this.viewContainer.clear();
                this.hasView = false;
            }
        },
        enumerable: true,
        configurable: true
    });
    FeatureDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[cxFeature]',
                },] }
    ];
    /** @nocollapse */
    FeatureDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef },
        { type: FeatureConfigService }
    ]; };
    FeatureDirective.propDecorators = {
        cxFeature: [{ type: Input }]
    };
    return FeatureDirective;
}());
export { FeatureDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FeatureDirective.prototype.hasView;
    /**
     * @type {?}
     * @protected
     */
    FeatureDirective.prototype.templateRef;
    /**
     * @type {?}
     * @protected
     */
    FeatureDirective.prototype.viewContainer;
    /**
     * @type {?}
     * @protected
     */
    FeatureDirective.prototype.featureConfig;
}
