// @spartacus/core/esm5/src/features-config/services/feature-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { FeaturesConfig } from '../config/features-config';
import { isFeatureEnabled, isFeatureLevel, } from '../utils/feature-config-utils';
import * as i0 from "@angular/core";
import * as i1 from "../config/features-config";
var FeatureConfigService = /** @class */ (function () {
    function FeatureConfigService(config) {
        this.config = config;
    }
    /**
     * @param {?} version
     * @return {?}
     */
    FeatureConfigService.prototype.isLevel = /**
     * @param {?} version
     * @return {?}
     */
    function (version) {
        return isFeatureLevel(this.config, version);
    };
    /**
     * @param {?} feature
     * @return {?}
     */
    FeatureConfigService.prototype.isEnabled = /**
     * @param {?} feature
     * @return {?}
     */
    function (feature) {
        return isFeatureEnabled(this.config, feature);
    };
    FeatureConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    FeatureConfigService.ctorParameters = function () { return [
        { type: FeaturesConfig }
    ]; };
    /** @nocollapse */ FeatureConfigService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function FeatureConfigService_Factory() { return new FeatureConfigService(i0.ɵɵinject(i1.FeaturesConfig)); }, token: FeatureConfigService, providedIn: "root" });
    return FeatureConfigService;
}());
export { FeatureConfigService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    FeatureConfigService.prototype.config;
}
