// @spartacus/core/esm2015/src/user/connectors/notification-preference/user-notification-preference.connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { UserNotificationPreferenceAdapter } from './user-notification-preference.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./user-notification-preference.adapter";
export class UserNotificationPreferenceConnector {
    /**
     * @param {?} adapter
     */
    constructor(adapter) {
        this.adapter = adapter;
    }
    /**
     * @param {?} userId
     * @return {?}
     */
    loadAll(userId) {
        return this.adapter.loadAll(userId);
    }
    /**
     * @param {?} userId
     * @param {?} preferences
     * @return {?}
     */
    update(userId, preferences) {
        return this.adapter.update(userId, preferences);
    }
}
UserNotificationPreferenceConnector.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
UserNotificationPreferenceConnector.ctorParameters = () => [
    { type: UserNotificationPreferenceAdapter }
];
/** @nocollapse */ UserNotificationPreferenceConnector.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function UserNotificationPreferenceConnector_Factory() { return new UserNotificationPreferenceConnector(i0.ɵɵinject(i1.UserNotificationPreferenceAdapter)); }, token: UserNotificationPreferenceConnector, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    UserNotificationPreferenceConnector.prototype.adapter;
}
