// @spartacus/core/esm5/src/cms/store/actions/component.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { StateEntityLoaderActions } from '../../../state/utils/index';
import { COMPONENT_ENTITY } from '../cms-state';
/** @type {?} */
export var LOAD_CMS_COMPONENT = '[Cms] Load Component';
/** @type {?} */
export var LOAD_CMS_COMPONENT_FAIL = '[Cms] Load Component Fail';
/** @type {?} */
export var LOAD_CMS_COMPONENT_SUCCESS = '[Cms] Load Component Success';
// TODO(issue:6027) - rename the const to `CMS_GET_COMPONENT_FROM_PAGE`
/** @type {?} */
export var CMS_GET_COMPONENET_FROM_PAGE = '[Cms] Get Component from Page';
var LoadCmsComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCmsComponent, _super);
    // TODO(issue:6027) - this action should have only one `payload` property which should encapsulate all of the constructor's arguments
    function LoadCmsComponent(payload, pageContext) {
        var _this = _super.call(this, COMPONENT_ENTITY, payload) || this;
        _this.payload = payload;
        _this.pageContext = pageContext;
        _this.type = LOAD_CMS_COMPONENT;
        return _this;
    }
    return LoadCmsComponent;
}(StateEntityLoaderActions.EntityLoadAction));
export { LoadCmsComponent };
if (false) {
    /** @type {?} */
    LoadCmsComponent.prototype.type;
    /** @type {?} */
    LoadCmsComponent.prototype.payload;
    /** @type {?} */
    LoadCmsComponent.prototype.pageContext;
}
var LoadCmsComponentFail = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCmsComponentFail, _super);
    // TODO(issue:6027) - this action should have only one `payload` property which should encapsulate all of the constructor's arguments
    function LoadCmsComponentFail(uid, payload, pageContext) {
        var _this = _super.call(this, COMPONENT_ENTITY, uid, payload) || this;
        _this.payload = payload;
        _this.pageContext = pageContext;
        _this.type = LOAD_CMS_COMPONENT_FAIL;
        return _this;
    }
    return LoadCmsComponentFail;
}(StateEntityLoaderActions.EntityFailAction));
export { LoadCmsComponentFail };
if (false) {
    /** @type {?} */
    LoadCmsComponentFail.prototype.type;
    /** @type {?} */
    LoadCmsComponentFail.prototype.payload;
    /** @type {?} */
    LoadCmsComponentFail.prototype.pageContext;
}
/**
 * @template T
 */
var /**
 * @template T
 */
LoadCmsComponentSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCmsComponentSuccess, _super);
    // TODO(issue:6027) - this action should have only one `payload` property which should encapsulate all of the constructor's arguments
    function LoadCmsComponentSuccess(payload, uid, pageContext) {
        var _this = _super.call(this, COMPONENT_ENTITY, uid || payload.uid || '') || this;
        _this.payload = payload;
        _this.pageContext = pageContext;
        _this.type = LOAD_CMS_COMPONENT_SUCCESS;
        return _this;
    }
    return LoadCmsComponentSuccess;
}(StateEntityLoaderActions.EntitySuccessAction));
/**
 * @template T
 */
export { LoadCmsComponentSuccess };
if (false) {
    /** @type {?} */
    LoadCmsComponentSuccess.prototype.type;
    /** @type {?} */
    LoadCmsComponentSuccess.prototype.payload;
    /** @type {?} */
    LoadCmsComponentSuccess.prototype.pageContext;
}
/**
 * @template T
 */
var /**
 * @template T
 */
CmsGetComponentFromPage = /** @class */ (function (_super) {
    tslib_1.__extends(CmsGetComponentFromPage, _super);
    // TODO(issue:6027) - this action should have only one `payload` property which should encapsulate all of the constructor's arguments
    function CmsGetComponentFromPage(payload, pageContext) {
        var _this = _super.call(this, COMPONENT_ENTITY, payload.map((/**
         * @param {?} cmp
         * @return {?}
         */
        function (cmp) { return cmp.uid; }))) || this;
        _this.payload = payload;
        _this.pageContext = pageContext;
        _this.type = CMS_GET_COMPONENET_FROM_PAGE;
        return _this;
    }
    return CmsGetComponentFromPage;
}(StateEntityLoaderActions.EntitySuccessAction));
/**
 * @template T
 */
export { CmsGetComponentFromPage };
if (false) {
    /** @type {?} */
    CmsGetComponentFromPage.prototype.type;
    /** @type {?} */
    CmsGetComponentFromPage.prototype.payload;
    /** @type {?} */
    CmsGetComponentFromPage.prototype.pageContext;
}
