// @spartacus/core/esm5/src/cms/store/actions/navigation-entry-item.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { StateEntityLoaderActions } from '../../../state/utils/index';
import { NAVIGATION_DETAIL_ENTITY } from '../cms-state';
/** @type {?} */
export var LOAD_CMS_NAVIGATION_ITEMS = '[Cms] Load NavigationEntry items';
/** @type {?} */
export var LOAD_CMS_NAVIGATION_ITEMS_FAIL = '[Cms] Load NavigationEntry items Fail';
/** @type {?} */
export var LOAD_CMS_NAVIGATION_ITEMS_SUCCESS = '[Cms] Load NavigationEntry items Success';
var LoadCmsNavigationItems = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCmsNavigationItems, _super);
    function LoadCmsNavigationItems(payload) {
        var _this = _super.call(this, NAVIGATION_DETAIL_ENTITY, payload.nodeId) || this;
        _this.payload = payload;
        _this.type = LOAD_CMS_NAVIGATION_ITEMS;
        return _this;
    }
    return LoadCmsNavigationItems;
}(StateEntityLoaderActions.EntityLoadAction));
export { LoadCmsNavigationItems };
if (false) {
    /** @type {?} */
    LoadCmsNavigationItems.prototype.type;
    /** @type {?} */
    LoadCmsNavigationItems.prototype.payload;
}
var LoadCmsNavigationItemsFail = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCmsNavigationItemsFail, _super);
    function LoadCmsNavigationItemsFail(nodeId, payload) {
        var _this = _super.call(this, NAVIGATION_DETAIL_ENTITY, nodeId, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_CMS_NAVIGATION_ITEMS_FAIL;
        return _this;
    }
    return LoadCmsNavigationItemsFail;
}(StateEntityLoaderActions.EntityFailAction));
export { LoadCmsNavigationItemsFail };
if (false) {
    /** @type {?} */
    LoadCmsNavigationItemsFail.prototype.type;
    /** @type {?} */
    LoadCmsNavigationItemsFail.prototype.payload;
}
var LoadCmsNavigationItemsSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCmsNavigationItemsSuccess, _super);
    function LoadCmsNavigationItemsSuccess(payload) {
        var _this = _super.call(this, NAVIGATION_DETAIL_ENTITY, payload.nodeId) || this;
        _this.payload = payload;
        _this.type = LOAD_CMS_NAVIGATION_ITEMS_SUCCESS;
        return _this;
    }
    return LoadCmsNavigationItemsSuccess;
}(StateEntityLoaderActions.EntitySuccessAction));
export { LoadCmsNavigationItemsSuccess };
if (false) {
    /** @type {?} */
    LoadCmsNavigationItemsSuccess.prototype.type;
    /** @type {?} */
    LoadCmsNavigationItemsSuccess.prototype.payload;
}
