// @spartacus/core/esm5/src/cms/store/actions/page.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { StateEntityLoaderActions } from '../../../state/utils/index';
/** @type {?} */
export var LOAD_CMS_PAGE_DATA = '[Cms] Load Page Data';
/** @type {?} */
export var LOAD_CMS_PAGE_DATA_FAIL = '[Cms] Load Page Data Fail';
/** @type {?} */
export var LOAD_CMS_PAGE_DATA_SUCCESS = '[Cms] Load Page Data Success';
/** @type {?} */
export var CMS_SET_PAGE_SUCCESS_INDEX = '[Cms] Set Page Success Index';
/** @type {?} */
export var CMS_SET_PAGE_FAIL_INDEX = '[Cms] Set Page Fail Index';
var LoadCmsPageData = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCmsPageData, _super);
    function LoadCmsPageData(payload) {
        var _this = _super.call(this, payload.type, payload.id) || this;
        _this.payload = payload;
        _this.type = LOAD_CMS_PAGE_DATA;
        return _this;
    }
    return LoadCmsPageData;
}(StateEntityLoaderActions.EntityLoadAction));
export { LoadCmsPageData };
if (false) {
    /** @type {?} */
    LoadCmsPageData.prototype.type;
    /** @type {?} */
    LoadCmsPageData.prototype.payload;
}
var LoadCmsPageDataFail = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCmsPageDataFail, _super);
    function LoadCmsPageDataFail(pageContext, error) {
        var _this = _super.call(this, pageContext.type, pageContext.id, error) || this;
        _this.type = LOAD_CMS_PAGE_DATA_FAIL;
        return _this;
    }
    return LoadCmsPageDataFail;
}(StateEntityLoaderActions.EntityFailAction));
export { LoadCmsPageDataFail };
if (false) {
    /** @type {?} */
    LoadCmsPageDataFail.prototype.type;
}
var LoadCmsPageDataSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCmsPageDataSuccess, _super);
    function LoadCmsPageDataSuccess(pageContext, payload) {
        var _this = _super.call(this, pageContext.type, pageContext.id, payload) || this;
        _this.type = LOAD_CMS_PAGE_DATA_SUCCESS;
        return _this;
    }
    return LoadCmsPageDataSuccess;
}(StateEntityLoaderActions.EntitySuccessAction));
export { LoadCmsPageDataSuccess };
if (false) {
    /** @type {?} */
    LoadCmsPageDataSuccess.prototype.type;
}
var CmsSetPageSuccessIndex = /** @class */ (function (_super) {
    tslib_1.__extends(CmsSetPageSuccessIndex, _super);
    function CmsSetPageSuccessIndex(pageContext, payload) {
        var _this = _super.call(this, pageContext.type, pageContext.id, payload) || this;
        _this.type = CMS_SET_PAGE_SUCCESS_INDEX;
        return _this;
    }
    return CmsSetPageSuccessIndex;
}(StateEntityLoaderActions.EntitySuccessAction));
export { CmsSetPageSuccessIndex };
if (false) {
    /** @type {?} */
    CmsSetPageSuccessIndex.prototype.type;
}
var CmsSetPageFailIndex = /** @class */ (function (_super) {
    tslib_1.__extends(CmsSetPageFailIndex, _super);
    function CmsSetPageFailIndex(pageContext, payload) {
        var _this = _super.call(this, pageContext.type, pageContext.id) || this;
        _this.payload = payload;
        _this.type = CMS_SET_PAGE_FAIL_INDEX;
        return _this;
    }
    return CmsSetPageFailIndex;
}(StateEntityLoaderActions.EntityFailAction));
export { CmsSetPageFailIndex };
if (false) {
    /** @type {?} */
    CmsSetPageFailIndex.prototype.type;
    /** @type {?} */
    CmsSetPageFailIndex.prototype.payload;
}
