// @spartacus/core/esm5/src/product/store/reducers/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
import { SiteContextActions } from '../../../site-context/store/actions/index';
import { PRODUCT_DETAIL_ENTITY } from '../product-state';
import * as fromProductReferences from './product-references.reducer';
import * as fromProductReviews from './product-reviews.reducer';
import * as fromProductsSearch from './product-search.reducer';
import { entityScopedLoaderReducer } from '../../../state/utils/scoped-loader/entity-scoped-loader.reducer';
/**
 * @return {?}
 */
export function getReducers() {
    return {
        search: fromProductsSearch.reducer,
        details: entityScopedLoaderReducer(PRODUCT_DETAIL_ENTITY),
        reviews: fromProductReviews.reducer,
        references: fromProductReferences.reducer,
    };
}
/** @type {?} */
export var reducerToken = new InjectionToken('ProductReducers');
/** @type {?} */
export var reducerProvider = {
    provide: reducerToken,
    useFactory: getReducers,
};
/**
 * @param {?} reducer
 * @return {?}
 */
export function clearProductsState(reducer) {
    return (/**
     * @param {?} state
     * @param {?} action
     * @return {?}
     */
    function (state, action) {
        if (action.type === SiteContextActions.CURRENCY_CHANGE ||
            action.type === SiteContextActions.LANGUAGE_CHANGE) {
            state = undefined;
        }
        return reducer(state, action);
    });
}
/** @type {?} */
export var metaReducers = [clearProductsState];
