// @spartacus/core/esm5/src/product/store/reducers/product-references.reducer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ProductActions } from '../actions/index';
/** @type {?} */
export var initialState = {
    productCode: '',
    list: [],
};
/**
 * @param {?=} state
 * @param {?=} action
 * @return {?}
 */
export function reducer(state, action) {
    if (state === void 0) { state = initialState; }
    switch (action.type) {
        case ProductActions.LOAD_PRODUCT_REFERENCES_SUCCESS: {
            /** @type {?} */
            var productCode = action.payload.productCode;
            /** @type {?} */
            var list = action.payload.list;
            return tslib_1.__assign({}, state, { list: tslib_1.__spread(state.list, (list ? list : [])).reduce((/**
                 * @param {?} productReferences
                 * @param {?} productReference
                 * @return {?}
                 */
                function (productReferences, productReference) {
                    if (!productReferences.some((/**
                     * @param {?} obj
                     * @return {?}
                     */
                    function (obj) {
                        return obj.referenceType === productReference.referenceType &&
                            obj.target.code === productReference.target.code;
                    }))) {
                        productReferences.push(productReference);
                    }
                    return productReferences;
                }), []), productCode: productCode });
        }
        case ProductActions.CLEAN_PRODUCT_REFERENCES: {
            return initialState;
        }
    }
    return state;
}
/** @type {?} */
export var getProductReferenceList = (/**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.list; });
/** @type {?} */
export var getProductReferenceProductCode = (/**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.productCode; });
