// @spartacus/core/esm2015/src/user/services/find-product-page-meta.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { of, combineLatest } from 'rxjs';
import { filter, map, switchMap } from 'rxjs/operators';
import { PageMetaResolver, } from '../../cms/page';
import { TranslationService } from '../../i18n/translation.service';
import { PageType } from '../../model/cms.model';
import { RoutingService } from '../../routing/facade/routing.service';
import { ProductSearchService } from '../../product/facade/product-search.service';
import { AuthService } from '../../auth/facade/auth.service';
import * as i0 from "@angular/core";
import * as i1 from "../../routing/facade/routing.service";
import * as i2 from "../../product/facade/product-search.service";
import * as i3 from "../../i18n/translation.service";
import * as i4 from "../../auth/facade/auth.service";
export class FindProductPageMetaResolver extends PageMetaResolver {
    /**
     * @param {?} routingService
     * @param {?} productSearchService
     * @param {?} translation
     * @param {?} authService
     */
    constructor(routingService, productSearchService, translation, authService) {
        super();
        this.routingService = routingService;
        this.productSearchService = productSearchService;
        this.translation = translation;
        this.authService = authService;
        this.totalAndCode$ = combineLatest([
            this.productSearchService.getResults().pipe(filter((/**
             * @param {?} data
             * @return {?}
             */
            data => !!(data && data.pagination))), map((/**
             * @param {?} results
             * @return {?}
             */
            results => results.pagination.totalResults))),
            this.routingService.getRouterState().pipe(map((/**
             * @param {?} state
             * @return {?}
             */
            state => state.state.queryParams['couponcode'])), filter(Boolean)),
        ]);
        this.pageType = PageType.CONTENT_PAGE;
        this.pageTemplate = 'SearchResultsListPageTemplate';
    }
    /**
     * @deprecated since version 1.3
     *
     * The resolve method is no longer preferred and will be removed with release 2.0.
     * The caller `PageMetaService` service is improved to expect all individual resolvers
     * instead, so that the code is easier extensible.
     * @return {?}
     */
    resolve() {
        return combineLatest([this.resolveTitle(), this.resolveBreadcrumbs()]).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        ([title, breadcrumbs]) => ({
            title,
            breadcrumbs,
        }))));
    }
    /**
     * @return {?}
     */
    resolveBreadcrumbs() {
        /** @type {?} */
        const breadcrumbs = [{ label: 'Home', link: '/' }];
        this.authService.isUserLoggedIn().subscribe((/**
         * @param {?} login
         * @return {?}
         */
        login => {
            if (login)
                breadcrumbs.push({ label: 'My Coupons', link: '/my-account/coupons' });
        }));
        return of(breadcrumbs);
    }
    /**
     * @return {?}
     */
    resolveTitle() {
        return this.totalAndCode$.pipe(switchMap((/**
         * @param {?} __0
         * @return {?}
         */
        ([total, code]) => this.translation.translate('pageMetaResolver.search.findProductTitle', {
            count: total,
            coupon: code,
        }))));
    }
    /**
     * @param {?} page
     * @return {?}
     */
    getScore(page) {
        /** @type {?} */
        let score = 0;
        if (this.pageType) {
            score += page.type === this.pageType ? 1 : -1;
        }
        if (this.pageTemplate) {
            score += page.template === this.pageTemplate ? 1 : -1;
        }
        this.routingService
            .getRouterState()
            .pipe(map((/**
         * @param {?} state
         * @return {?}
         */
        state => {
            return state.state.queryParams;
        })), filter(Boolean))
            .subscribe((/**
         * @param {?} queryParams
         * @return {?}
         */
        (queryParams) => {
            if (queryParams) {
                score += queryParams['couponcode'] ? 1 : -1;
            }
        }))
            .unsubscribe();
        return score;
    }
}
FindProductPageMetaResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
FindProductPageMetaResolver.ctorParameters = () => [
    { type: RoutingService },
    { type: ProductSearchService },
    { type: TranslationService },
    { type: AuthService }
];
/** @nocollapse */ FindProductPageMetaResolver.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function FindProductPageMetaResolver_Factory() { return new FindProductPageMetaResolver(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i2.ProductSearchService), i0.ɵɵinject(i3.TranslationService), i0.ɵɵinject(i4.AuthService)); }, token: FindProductPageMetaResolver, providedIn: "root" });
if (false) {
    /** @type {?} */
    FindProductPageMetaResolver.prototype.totalAndCode$;
    /**
     * @type {?}
     * @protected
     */
    FindProductPageMetaResolver.prototype.routingService;
    /**
     * @type {?}
     * @protected
     */
    FindProductPageMetaResolver.prototype.productSearchService;
    /**
     * @type {?}
     * @protected
     */
    FindProductPageMetaResolver.prototype.translation;
    /**
     * @type {?}
     * @protected
     */
    FindProductPageMetaResolver.prototype.authService;
}
