// @spartacus/core/esm5/src/cart/connectors/cart/cart.adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
CartAdapter = /** @class */ (function () {
    function CartAdapter() {
    }
    return CartAdapter;
}());
/**
 * @abstract
 */
export { CartAdapter };
if (false) {
    /**
     * Abstract method used to load all carts
     *
     * @abstract
     * @param {?} userId
     * @return {?}
     */
    CartAdapter.prototype.loadAll = function (userId) { };
    /**
     * Abstract method used to load cart
     *
     * @abstract
     * @param {?} userId
     * @param {?} cartId
     * @return {?}
     */
    CartAdapter.prototype.load = function (userId, cartId) { };
    /**
     * Abstract method used to create cart. If toMergeCartGuid is specified, cart will be merged with existing one
     *
     * @abstract
     * @param {?} userId
     * @param {?=} oldCartId
     * @param {?=} toMergeCartGuid
     * @return {?}
     */
    CartAdapter.prototype.create = function (userId, oldCartId, toMergeCartGuid) { };
    /**
     * Abstract method used to delete cart
     *
     * @abstract
     * @param {?} userId
     * @param {?} cartId
     * @return {?}
     */
    CartAdapter.prototype.delete = function (userId, cartId) { };
    /**
     * Abstract method to assign an email to the cart. This step is required to make a guest checkout
     * @abstract
     * @param {?} userId
     * @param {?} cartId
     * @param {?} email
     * @return {?}
     */
    CartAdapter.prototype.addEmail = function (userId, cartId, email) { };
}
