// @spartacus/core/esm2015/src/cms/page/page.resolvers.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Resolves the page heading which is used in the UI. The page
 * heading might differ from the page title, which is used to address
 * the page in the browser (history, tabs) and outside the storefront
 * (Goolge, bots, etc).
 * @record
 */
export function PageHeadingResolver() { }
if (false) {
    /**
     * Resolves the page heading.
     *
     * @deprecated since version 1.3
     * Use `resolveHeading()` instead.
     * @param {...?} args
     * @return {?}
     */
    PageHeadingResolver.prototype.resolveHeading = function (args) { };
    /**
     * Resolves the page heading.
     * @return {?}
     */
    PageHeadingResolver.prototype.resolveHeading = function () { };
}
/**
 * Resolves the page title which is first and foremost used
 * for the page title tag, but could also be used for the
 * page heading in the UI.
 * @record
 */
export function PageTitleResolver() { }
if (false) {
    /**
     * Resolves the page title.
     *
     * @deprecated since version 1.3
     * Use `resolveTitle()` instead.
     * @param {...?} args
     * @return {?}
     */
    PageTitleResolver.prototype.resolveTitle = function (args) { };
    /**
     * Resolves the page title.
     * @return {?}
     */
    PageTitleResolver.prototype.resolveTitle = function () { };
}
/**
 * Resolves the page description. The page description is typically used
 * in the Search Engine Result Page (SERP).
 * @record
 */
export function PageDescriptionResolver() { }
if (false) {
    /**
     * Resolves the page description.
     *
     * @deprecated since version 1.3
     * Use `resolveHeading()` instead.
     * @param {...?} args
     * @return {?}
     */
    PageDescriptionResolver.prototype.resolveDescription = function (args) { };
    /**
     * Resolves the page description.
     * @return {?}
     */
    PageDescriptionResolver.prototype.resolveDescription = function () { };
}
/**
 * Resolves breadcrumbs for the page, which is used in the `BreadcrumbComponent`
 * @record
 */
export function PageBreadcrumbResolver() { }
if (false) {
    /**
     * Resolves the breadcrumbs for the page.
     *
     * @deprecated since version 1.3
     * Use `resolveBreadcrumbs()` instead.
     * @param {...?} args
     * @return {?}
     */
    PageBreadcrumbResolver.prototype.resolveBreadcrumbs = function (args) { };
    /**
     * Resolves the breadcrumbs for the page.
     * @return {?}
     */
    PageBreadcrumbResolver.prototype.resolveBreadcrumbs = function () { };
}
/**
 * Provides a method to resolve the the main image for the page.
 * This is typically used for social sharing (for example by using
 * the `og:image` metatag).
 * @record
 */
export function PageImageResolver() { }
if (false) {
    /**
     * Resolves the main image for the page.
     *
     * @deprecated since version 1.3
     * Use `resolveImage()` instead.
     * @param {...?} args
     * @return {?}
     */
    PageImageResolver.prototype.resolveImage = function (args) { };
    /**
     * Resolves the main image for the page.
     * @return {?}
     */
    PageImageResolver.prototype.resolveImage = function () { };
}
/**
 * Resolves the robot information for the page. This is used by
 * search engines to understand whether the page and subsequential links
 * should be indexed.
 *
 * @record
 */
export function PageRobotsResolver() { }
if (false) {
    /**
     * Resolves the robots for the page.
     *
     * @deprecated since version 1.3
     * Use `resolveRobots()` instead.
     * @param {...?} args
     * @return {?}
     */
    PageRobotsResolver.prototype.resolveRobots = function (args) { };
    /**
     * Resolves the robots for the page.
     * @return {?}
     */
    PageRobotsResolver.prototype.resolveRobots = function () { };
}
