// @spartacus/core/esm5/src/occ/interceptor/with-credentials.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { OccConfig } from '../config/occ-config';
import * as i0 from "@angular/core";
import * as i1 from "../config/occ-config";
/**
 * Http interceptor to add cookies to all cross-site requests.
 */
var WithCredentialsInterceptor = /** @class */ (function () {
    function WithCredentialsInterceptor(config) {
        this.config = config;
    }
    /**
     * Intercepts each request and adds the `withCredential` flag to it
     * if it hasn't been added already.
     */
    /**
     * Intercepts each request and adds the `withCredential` flag to it
     * if it hasn't been added already.
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    WithCredentialsInterceptor.prototype.intercept = /**
     * Intercepts each request and adds the `withCredential` flag to it
     * if it hasn't been added already.
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    function (request, next) {
        if (this.requiresWithCredentials(request)) {
            request = request.clone({
                withCredentials: true,
            });
        }
        return next.handle(request);
    };
    /**
     * Indicates whether this is an OCC request.
     */
    /**
     * Indicates whether this is an OCC request.
     * @protected
     * @param {?} request
     * @return {?}
     */
    WithCredentialsInterceptor.prototype.requiresWithCredentials = /**
     * Indicates whether this is an OCC request.
     * @protected
     * @param {?} request
     * @return {?}
     */
    function (request) {
        return (
        /**
         * This `occConfig?.useWithCredentials` flag has only been introduced in
         * version 2.0. Customers must provide this interceptors manually in their
         * app to opt-in to use this feature, as we don't backport the configuration.
         */
        // this.occConfig?.useWithCredentials &&
        request.url.indexOf(this.occConfig.prefix) > -1);
    };
    Object.defineProperty(WithCredentialsInterceptor.prototype, "occConfig", {
        get: /**
         * @private
         * @return {?}
         */
        function () {
            return this.config && this.config.backend && this.config.backend.occ
                ? this.config.backend.occ
                : {};
        },
        enumerable: true,
        configurable: true
    });
    WithCredentialsInterceptor.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    WithCredentialsInterceptor.ctorParameters = function () { return [
        { type: OccConfig }
    ]; };
    /** @nocollapse */ WithCredentialsInterceptor.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function WithCredentialsInterceptor_Factory() { return new WithCredentialsInterceptor(i0.ɵɵinject(i1.OccConfig)); }, token: WithCredentialsInterceptor, providedIn: "root" });
    return WithCredentialsInterceptor;
}());
export { WithCredentialsInterceptor };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    WithCredentialsInterceptor.prototype.config;
}
