// @spartacus/core/esm2015/src/routing/store/selectors/routing.selector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createFeatureSelector, createSelector, } from '@ngrx/store';
import { ROUTING_FEATURE } from '../routing-state';
/** @type {?} */
export const getRouterFeatureState = createFeatureSelector(ROUTING_FEATURE);
const ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
state => state.router;
/** @type {?} */
export const getRouterState = createSelector(getRouterFeatureState, (ɵ0));
const ɵ1 = /**
 * @param {?} routingState
 * @return {?}
 */
(routingState) => (routingState.state && routingState.state.context) || { id: '' };
/** @type {?} */
export const getPageContext = createSelector(getRouterState, (ɵ1));
const ɵ2 = /**
 * @param {?} routingState
 * @return {?}
 */
(routingState) => routingState.nextState && routingState.nextState.context;
/** @type {?} */
export const getNextPageContext = createSelector(getRouterState, (ɵ2));
const ɵ3 = /**
 * @param {?} context
 * @return {?}
 */
context => !!context;
/** @type {?} */
export const isNavigating = createSelector(getNextPageContext, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
