// @spartacus/core/esm5/src/routing/external-routes/external-routes.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { APP_INITIALIZER, NgModule } from '@angular/core';
import { Config } from '../../config/config.module';
import { ExternalRoutesConfig } from './external-routes-config';
import { addExternalRoutesFactory } from './external-routes.providers';
import { ExternalRoutesService } from './external-routes.service';
/**
 * Prepends the external route that redirects to a different storefront system for configured URLs
 */
var ExternalRoutesModule = /** @class */ (function () {
    function ExternalRoutesModule() {
    }
    /**
     * @return {?}
     */
    ExternalRoutesModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: ExternalRoutesModule,
            providers: [
                ExternalRoutesService,
                { provide: ExternalRoutesConfig, useExisting: Config },
                {
                    provide: APP_INITIALIZER,
                    multi: true,
                    useFactory: addExternalRoutesFactory,
                    deps: [ExternalRoutesService],
                },
            ],
        };
    };
    ExternalRoutesModule.decorators = [
        { type: NgModule }
    ];
    return ExternalRoutesModule;
}());
export { ExternalRoutesModule };
