// @spartacus/core/esm5/src/checkout/connectors/payment/checkout-payment.connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { CheckoutPaymentAdapter } from './checkout-payment.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./checkout-payment.adapter";
var CheckoutPaymentConnector = /** @class */ (function () {
    function CheckoutPaymentConnector(adapter) {
        this.adapter = adapter;
    }
    /**
     * @param {?} userId
     * @param {?} cartId
     * @param {?} paymentDetails
     * @return {?}
     */
    CheckoutPaymentConnector.prototype.create = /**
     * @param {?} userId
     * @param {?} cartId
     * @param {?} paymentDetails
     * @return {?}
     */
    function (userId, cartId, paymentDetails) {
        return this.adapter.create(userId, cartId, paymentDetails);
    };
    /**
     * @param {?} userId
     * @param {?} cartId
     * @param {?} paymentDetailsId
     * @return {?}
     */
    CheckoutPaymentConnector.prototype.set = /**
     * @param {?} userId
     * @param {?} cartId
     * @param {?} paymentDetailsId
     * @return {?}
     */
    function (userId, cartId, paymentDetailsId) {
        return this.adapter.set(userId, cartId, paymentDetailsId);
    };
    /**
     * @return {?}
     */
    CheckoutPaymentConnector.prototype.getCardTypes = /**
     * @return {?}
     */
    function () {
        return this.adapter.loadCardTypes();
    };
    CheckoutPaymentConnector.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    CheckoutPaymentConnector.ctorParameters = function () { return [
        { type: CheckoutPaymentAdapter }
    ]; };
    /** @nocollapse */ CheckoutPaymentConnector.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function CheckoutPaymentConnector_Factory() { return new CheckoutPaymentConnector(i0.ɵɵinject(i1.CheckoutPaymentAdapter)); }, token: CheckoutPaymentConnector, providedIn: "root" });
    return CheckoutPaymentConnector;
}());
export { CheckoutPaymentConnector };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    CheckoutPaymentConnector.prototype.adapter;
}
