// @spartacus/core/esm5/src/product/store/actions/product-references.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var LOAD_PRODUCT_REFERENCES = '[Product] Load Product References Data';
/** @type {?} */
export var LOAD_PRODUCT_REFERENCES_FAIL = '[Product] Load Product References Data Fail';
/** @type {?} */
export var LOAD_PRODUCT_REFERENCES_SUCCESS = '[Product] Load Product References Data Success';
/** @type {?} */
export var CLEAN_PRODUCT_REFERENCES = '[Product] Clean Product References';
var LoadProductReferences = /** @class */ (function () {
    function LoadProductReferences(payload) {
        this.payload = payload;
        this.type = LOAD_PRODUCT_REFERENCES;
    }
    return LoadProductReferences;
}());
export { LoadProductReferences };
if (false) {
    /** @type {?} */
    LoadProductReferences.prototype.type;
    /** @type {?} */
    LoadProductReferences.prototype.payload;
}
var LoadProductReferencesFail = /** @class */ (function () {
    function LoadProductReferencesFail(payload) {
        this.payload = payload;
        this.type = LOAD_PRODUCT_REFERENCES_FAIL;
    }
    return LoadProductReferencesFail;
}());
export { LoadProductReferencesFail };
if (false) {
    /** @type {?} */
    LoadProductReferencesFail.prototype.type;
    /** @type {?} */
    LoadProductReferencesFail.prototype.payload;
}
var LoadProductReferencesSuccess = /** @class */ (function () {
    function LoadProductReferencesSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_PRODUCT_REFERENCES_SUCCESS;
    }
    return LoadProductReferencesSuccess;
}());
export { LoadProductReferencesSuccess };
if (false) {
    /** @type {?} */
    LoadProductReferencesSuccess.prototype.type;
    /** @type {?} */
    LoadProductReferencesSuccess.prototype.payload;
}
var CleanProductReferences = /** @class */ (function () {
    function CleanProductReferences() {
        this.type = CLEAN_PRODUCT_REFERENCES;
    }
    return CleanProductReferences;
}());
export { CleanProductReferences };
if (false) {
    /** @type {?} */
    CleanProductReferences.prototype.type;
}
