// @spartacus/core/esm5/src/product/store/actions/product-reviews.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var LOAD_PRODUCT_REVIEWS = '[Product] Load Product Reviews Data';
/** @type {?} */
export var LOAD_PRODUCT_REVIEWS_FAIL = '[Product] Load Product Reviews Data Fail';
/** @type {?} */
export var LOAD_PRODUCT_REVIEWS_SUCCESS = '[Product] Load Product Reviews Data Success';
/** @type {?} */
export var POST_PRODUCT_REVIEW = '[Product] Post Product Review';
/** @type {?} */
export var POST_PRODUCT_REVIEW_FAIL = '[Product] Post Product Review Fail';
/** @type {?} */
export var POST_PRODUCT_REVIEW_SUCCESS = '[Product] Post Product Review Success';
var LoadProductReviews = /** @class */ (function () {
    function LoadProductReviews(payload) {
        this.payload = payload;
        this.type = LOAD_PRODUCT_REVIEWS;
    }
    return LoadProductReviews;
}());
export { LoadProductReviews };
if (false) {
    /** @type {?} */
    LoadProductReviews.prototype.type;
    /** @type {?} */
    LoadProductReviews.prototype.payload;
}
var LoadProductReviewsFail = /** @class */ (function () {
    function LoadProductReviewsFail(payload) {
        this.payload = payload;
        this.type = LOAD_PRODUCT_REVIEWS_FAIL;
    }
    return LoadProductReviewsFail;
}());
export { LoadProductReviewsFail };
if (false) {
    /** @type {?} */
    LoadProductReviewsFail.prototype.type;
    /** @type {?} */
    LoadProductReviewsFail.prototype.payload;
}
var LoadProductReviewsSuccess = /** @class */ (function () {
    function LoadProductReviewsSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_PRODUCT_REVIEWS_SUCCESS;
    }
    return LoadProductReviewsSuccess;
}());
export { LoadProductReviewsSuccess };
if (false) {
    /** @type {?} */
    LoadProductReviewsSuccess.prototype.type;
    /** @type {?} */
    LoadProductReviewsSuccess.prototype.payload;
}
var PostProductReview = /** @class */ (function () {
    function PostProductReview(payload) {
        this.payload = payload;
        this.type = POST_PRODUCT_REVIEW;
    }
    return PostProductReview;
}());
export { PostProductReview };
if (false) {
    /** @type {?} */
    PostProductReview.prototype.type;
    /** @type {?} */
    PostProductReview.prototype.payload;
}
var PostProductReviewFail = /** @class */ (function () {
    function PostProductReviewFail(payload) {
        this.payload = payload;
        this.type = POST_PRODUCT_REVIEW_FAIL;
    }
    return PostProductReviewFail;
}());
export { PostProductReviewFail };
if (false) {
    /** @type {?} */
    PostProductReviewFail.prototype.type;
    /** @type {?} */
    PostProductReviewFail.prototype.payload;
}
var PostProductReviewSuccess = /** @class */ (function () {
    function PostProductReviewSuccess(payload) {
        this.payload = payload;
        this.type = POST_PRODUCT_REVIEW_SUCCESS;
    }
    return PostProductReviewSuccess;
}());
export { PostProductReviewSuccess };
if (false) {
    /** @type {?} */
    PostProductReviewSuccess.prototype.type;
    /** @type {?} */
    PostProductReviewSuccess.prototype.payload;
}
