// @spartacus/core/esm5/src/product/store/actions/product.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { PRODUCT_DETAIL_ENTITY } from '../product-state';
import { EntityScopedLoaderActions } from '../../../state/utils/scoped-loader/entity-scoped-loader.actions';
/** @type {?} */
export var LOAD_PRODUCT = '[Product] Load Product Data';
/** @type {?} */
export var LOAD_PRODUCT_FAIL = '[Product] Load Product Data Fail';
/** @type {?} */
export var LOAD_PRODUCT_SUCCESS = '[Product] Load Product Data Success';
/**
 * @record
 */
export function ProductMeta() { }
if (false) {
    /** @type {?|undefined} */
    ProductMeta.prototype.scope;
}
/**
 * @record
 */
export function EntityScopedLoaderAction() { }
if (false) {
    /** @type {?|undefined} */
    EntityScopedLoaderAction.prototype.payload;
    /** @type {?|undefined} */
    EntityScopedLoaderAction.prototype.meta;
}
var LoadProduct = /** @class */ (function (_super) {
    tslib_1.__extends(LoadProduct, _super);
    function LoadProduct(payload, scope) {
        if (scope === void 0) { scope = ''; }
        var _this = _super.call(this, PRODUCT_DETAIL_ENTITY, payload, scope) || this;
        _this.payload = payload;
        _this.type = LOAD_PRODUCT;
        return _this;
    }
    return LoadProduct;
}(EntityScopedLoaderActions.EntityScopedLoadAction));
export { LoadProduct };
if (false) {
    /** @type {?} */
    LoadProduct.prototype.type;
    /** @type {?} */
    LoadProduct.prototype.payload;
}
var LoadProductFail = /** @class */ (function (_super) {
    tslib_1.__extends(LoadProductFail, _super);
    function LoadProductFail(productCode, payload, scope) {
        if (scope === void 0) { scope = ''; }
        var _this = _super.call(this, PRODUCT_DETAIL_ENTITY, productCode, scope, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_PRODUCT_FAIL;
        return _this;
    }
    return LoadProductFail;
}(EntityScopedLoaderActions.EntityScopedFailAction));
export { LoadProductFail };
if (false) {
    /** @type {?} */
    LoadProductFail.prototype.type;
    /** @type {?} */
    LoadProductFail.prototype.payload;
}
var LoadProductSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(LoadProductSuccess, _super);
    function LoadProductSuccess(payload, scope) {
        if (scope === void 0) { scope = ''; }
        var _this = _super.call(this, PRODUCT_DETAIL_ENTITY, payload.code, scope) || this;
        _this.payload = payload;
        _this.type = LOAD_PRODUCT_SUCCESS;
        return _this;
    }
    return LoadProductSuccess;
}(EntityScopedLoaderActions.EntityScopedSuccessAction));
export { LoadProductSuccess };
if (false) {
    /** @type {?} */
    LoadProductSuccess.prototype.type;
    /** @type {?} */
    LoadProductSuccess.prototype.payload;
}
