// @spartacus/core/esm2015/src/routing/store/actions/router.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const ROUTER_GO = '[Router] Go';
/** @type {?} */
export const ROUTER_GO_BY_URL = '[Router] Go By Url';
/** @type {?} */
export const ROUTER_BACK = '[Router] Back';
/** @type {?} */
export const ROUTER_FORWARD = '[Router] Forward';
export class RouteGoAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = ROUTER_GO;
    }
}
if (false) {
    /** @type {?} */
    RouteGoAction.prototype.type;
    /** @type {?} */
    RouteGoAction.prototype.payload;
}
export class RouteGoByUrlAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = ROUTER_GO_BY_URL;
    }
}
if (false) {
    /** @type {?} */
    RouteGoByUrlAction.prototype.type;
    /** @type {?} */
    RouteGoByUrlAction.prototype.payload;
}
export class RouteBackAction {
    constructor() {
        this.type = ROUTER_BACK;
    }
}
if (false) {
    /** @type {?} */
    RouteBackAction.prototype.type;
}
export class RouteForwardAction {
    constructor() {
        this.type = ROUTER_FORWARD;
    }
}
if (false) {
    /** @type {?} */
    RouteForwardAction.prototype.type;
}
