// @spartacus/core/esm2015/src/user/connectors/address/user-address.adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
export class UserAddressAdapter {
}
if (false) {
    /**
     * @abstract
     * @param {?} userId
     * @return {?}
     */
    UserAddressAdapter.prototype.loadAll = function (userId) { };
    /**
     * @abstract
     * @param {?} userId
     * @param {?} address
     * @return {?}
     */
    UserAddressAdapter.prototype.add = function (userId, address) { };
    /**
     * @abstract
     * @param {?} userId
     * @param {?} addressId
     * @param {?} address
     * @return {?}
     */
    UserAddressAdapter.prototype.update = function (userId, addressId, address) { };
    /**
     * @abstract
     * @param {?} userId
     * @param {?} address
     * @return {?}
     */
    UserAddressAdapter.prototype.verify = function (userId, address) { };
    /**
     * @abstract
     * @param {?} userId
     * @param {?} addressId
     * @return {?}
     */
    UserAddressAdapter.prototype.delete = function (userId, addressId) { };
}
