// @spartacus/core/esm5/src/user/user.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { UserService } from './facade/user.service';
import { UserStoreModule } from './store/user-store.module';
import { PageMetaResolver } from '../cms/page/page-meta.resolver';
import { FindProductPageMetaResolver } from './services/find-product-page-meta.resolver';
var UserModule = /** @class */ (function () {
    function UserModule() {
    }
    /**
     * @return {?}
     */
    UserModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: UserModule,
            providers: [
                UserService,
                {
                    provide: PageMetaResolver,
                    useExisting: FindProductPageMetaResolver,
                    multi: true,
                },
            ],
        };
    };
    UserModule.decorators = [
        { type: NgModule, args: [{
                    imports: [UserStoreModule],
                },] }
    ];
    return UserModule;
}());
export { UserModule };
