// @ng-select/ng-select/esm2015/ng-select.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { NgSelectComponent, SELECTION_MODEL_FACTORY, NgSelectModule, NgSelectConfig } from './index';
export { ConsoleService as ɵr } from './ng-select/console.service';
export { NgDropdownPanelComponent as ɵp } from './ng-select/ng-dropdown-panel.component';
export { NgDropdownPanelService as ɵd } from './ng-select/ng-dropdown-panel.service';
export { NgOptionHighlightDirective as ɵs } from './ng-select/ng-option-highlight.directive';
export { NgOptionComponent as ɵq } from './ng-select/ng-option.component';
export { NgFooterTemplateDirective as ɵj, NgHeaderTemplateDirective as ɵi, NgLabelTemplateDirective as ɵg, NgLoadingSpinnerTemplateDirective as ɵo, NgLoadingTextTemplateDirective as ɵm, NgMultiLabelTemplateDirective as ɵh, NgNotFoundTemplateDirective as ɵk, NgOptgroupTemplateDirective as ɵf, NgOptionTemplateDirective as ɵe, NgTagTemplateDirective as ɵn, NgTypeToSearchTemplateDirective as ɵl } from './ng-select/ng-templates.directive';
export { DefaultSelectionModel as ɵc, DefaultSelectionModelFactory as ɵb } from './ng-select/selection-model';
