// @spartacus/core/esm2015/src/state/utils/entity-processes-loader/entity-processes-loader.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { entityMeta } from '../entity/entity.action';
import { processesDecrementMeta, processesIncrementMeta, processesLoaderResetMeta, } from '../processes-loader/processes-loader.action';
/** @type {?} */
export const ENTITY_PROCESSES_LOADER_RESET_ACTION = '[ENTITY] PROCESSES LOADER RESET';
/** @type {?} */
export const ENTITY_PROCESSES_INCREMENT_ACTION = '[ENTITY] PROCESSES INCREMENT';
/** @type {?} */
export const ENTITY_PROCESSES_DECREMENT_ACTION = '[ENTITY] PROCESSES DECREMENT';
/**
 * @record
 */
export function EntityProcessesLoaderMeta() { }
/**
 * @record
 */
export function EntityProcessesLoaderAction() { }
if (false) {
    /** @type {?|undefined} */
    EntityProcessesLoaderAction.prototype.payload;
    /** @type {?|undefined} */
    EntityProcessesLoaderAction.prototype.meta;
}
/**
 * @param {?} entityType
 * @param {?} id
 * @return {?}
 */
export function entityProcessesLoaderResetMeta(entityType, id) {
    return Object.assign({}, processesLoaderResetMeta(entityType), entityMeta(entityType, id));
}
/**
 * @param {?} entityType
 * @param {?} id
 * @return {?}
 */
export function entityProcessesIncrementMeta(entityType, id) {
    return Object.assign({}, processesIncrementMeta(entityType), entityMeta(entityType, id));
}
/**
 * @param {?} entityType
 * @param {?} id
 * @return {?}
 */
export function entityProcessesDecrementMeta(entityType, id) {
    return Object.assign({}, processesDecrementMeta(entityType), entityMeta(entityType, id));
}
export class EntityProcessesLoaderResetAction {
    /**
     * @param {?} entityType
     * @param {?} id
     */
    constructor(entityType, id) {
        this.type = ENTITY_PROCESSES_LOADER_RESET_ACTION;
        this.meta = entityProcessesLoaderResetMeta(entityType, id);
    }
}
if (false) {
    /** @type {?} */
    EntityProcessesLoaderResetAction.prototype.type;
    /** @type {?} */
    EntityProcessesLoaderResetAction.prototype.meta;
}
export class EntityProcessesIncrementAction {
    /**
     * @param {?} entityType
     * @param {?} id
     */
    constructor(entityType, id) {
        this.type = ENTITY_PROCESSES_INCREMENT_ACTION;
        this.meta = entityProcessesIncrementMeta(entityType, id);
    }
}
if (false) {
    /** @type {?} */
    EntityProcessesIncrementAction.prototype.type;
    /** @type {?} */
    EntityProcessesIncrementAction.prototype.meta;
}
export class EntityProcessesDecrementAction {
    /**
     * @param {?} entityType
     * @param {?} id
     */
    constructor(entityType, id) {
        this.type = ENTITY_PROCESSES_DECREMENT_ACTION;
        this.meta = entityProcessesDecrementMeta(entityType, id);
    }
}
if (false) {
    /** @type {?} */
    EntityProcessesDecrementAction.prototype.type;
    /** @type {?} */
    EntityProcessesDecrementAction.prototype.meta;
}
