// @spartacus/core/esm5/src/user/connectors/user/user.connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { UserAdapter } from './user.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./user.adapter";
var UserConnector = /** @class */ (function () {
    function UserConnector(adapter) {
        this.adapter = adapter;
    }
    /**
     * @param {?} userId
     * @return {?}
     */
    UserConnector.prototype.get = /**
     * @param {?} userId
     * @return {?}
     */
    function (userId) {
        return this.adapter.load(userId);
    };
    /**
     * @param {?} username
     * @param {?} user
     * @return {?}
     */
    UserConnector.prototype.update = /**
     * @param {?} username
     * @param {?} user
     * @return {?}
     */
    function (username, user) {
        return this.adapter.update(username, user);
    };
    /**
     * @param {?} user
     * @return {?}
     */
    UserConnector.prototype.register = /**
     * @param {?} user
     * @return {?}
     */
    function (user) {
        return this.adapter.register(user);
    };
    /**
     * @param {?} guid
     * @param {?} password
     * @return {?}
     */
    UserConnector.prototype.registerGuest = /**
     * @param {?} guid
     * @param {?} password
     * @return {?}
     */
    function (guid, password) {
        return this.adapter.registerGuest(guid, password);
    };
    /**
     * @param {?} userEmailAddress
     * @return {?}
     */
    UserConnector.prototype.requestForgotPasswordEmail = /**
     * @param {?} userEmailAddress
     * @return {?}
     */
    function (userEmailAddress) {
        return this.adapter.requestForgotPasswordEmail(userEmailAddress);
    };
    /**
     * @param {?} token
     * @param {?} newPassword
     * @return {?}
     */
    UserConnector.prototype.resetPassword = /**
     * @param {?} token
     * @param {?} newPassword
     * @return {?}
     */
    function (token, newPassword) {
        return this.adapter.resetPassword(token, newPassword);
    };
    /**
     * @param {?} userId
     * @param {?} currentPassword
     * @param {?} newUserId
     * @return {?}
     */
    UserConnector.prototype.updateEmail = /**
     * @param {?} userId
     * @param {?} currentPassword
     * @param {?} newUserId
     * @return {?}
     */
    function (userId, currentPassword, newUserId) {
        return this.adapter.updateEmail(userId, currentPassword, newUserId);
    };
    /**
     * @param {?} userId
     * @param {?} oldPassword
     * @param {?} newPassword
     * @return {?}
     */
    UserConnector.prototype.updatePassword = /**
     * @param {?} userId
     * @param {?} oldPassword
     * @param {?} newPassword
     * @return {?}
     */
    function (userId, oldPassword, newPassword) {
        return this.adapter.updatePassword(userId, oldPassword, newPassword);
    };
    /**
     * @param {?} userId
     * @return {?}
     */
    UserConnector.prototype.remove = /**
     * @param {?} userId
     * @return {?}
     */
    function (userId) {
        return this.adapter.remove(userId);
    };
    /**
     * @return {?}
     */
    UserConnector.prototype.getTitles = /**
     * @return {?}
     */
    function () {
        return this.adapter.loadTitles();
    };
    UserConnector.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    UserConnector.ctorParameters = function () { return [
        { type: UserAdapter }
    ]; };
    /** @nocollapse */ UserConnector.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function UserConnector_Factory() { return new UserConnector(i0.ɵɵinject(i1.UserAdapter)); }, token: UserConnector, providedIn: "root" });
    return UserConnector;
}());
export { UserConnector };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    UserConnector.prototype.adapter;
}
