// @spartacus/core/esm5/src/user/connectors/user/user.adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
UserAdapter = /** @class */ (function () {
    function UserAdapter() {
    }
    return UserAdapter;
}());
/**
 * @abstract
 */
export { UserAdapter };
if (false) {
    /**
     * @abstract
     * @param {?} userId
     * @return {?}
     */
    UserAdapter.prototype.load = function (userId) { };
    /**
     * @abstract
     * @param {?} username
     * @param {?} user
     * @return {?}
     */
    UserAdapter.prototype.update = function (username, user) { };
    /**
     * @abstract
     * @param {?} user
     * @return {?}
     */
    UserAdapter.prototype.register = function (user) { };
    /**
     * @abstract
     * @param {?} guid
     * @param {?} password
     * @return {?}
     */
    UserAdapter.prototype.registerGuest = function (guid, password) { };
    /**
     * @abstract
     * @param {?} userEmailAddress
     * @return {?}
     */
    UserAdapter.prototype.requestForgotPasswordEmail = function (userEmailAddress) { };
    /**
     * @abstract
     * @param {?} token
     * @param {?} newPassword
     * @return {?}
     */
    UserAdapter.prototype.resetPassword = function (token, newPassword) { };
    /**
     * @abstract
     * @param {?} userId
     * @param {?} currentPassword
     * @param {?} newUserId
     * @return {?}
     */
    UserAdapter.prototype.updateEmail = function (userId, currentPassword, newUserId) { };
    /**
     * @abstract
     * @param {?} userId
     * @param {?} oldPassword
     * @param {?} newPassword
     * @return {?}
     */
    UserAdapter.prototype.updatePassword = function (userId, oldPassword, newPassword) { };
    /**
     * @abstract
     * @param {?} userId
     * @return {?}
     */
    UserAdapter.prototype.remove = function (userId) { };
    /**
     * @abstract
     * @return {?}
     */
    UserAdapter.prototype.loadTitles = function () { };
}
