// @spartacus/core/esm2015/src/cms/store/actions/component.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StateEntityLoaderActions } from '../../../state/utils/index';
import { COMPONENT_ENTITY } from '../cms-state';
/** @type {?} */
export const LOAD_CMS_COMPONENT = '[Cms] Load Component';
/** @type {?} */
export const LOAD_CMS_COMPONENT_FAIL = '[Cms] Load Component Fail';
/** @type {?} */
export const LOAD_CMS_COMPONENT_SUCCESS = '[Cms] Load Component Success';
// TODO(issue:6027) - rename the const to `CMS_GET_COMPONENT_FROM_PAGE`
/** @type {?} */
export const CMS_GET_COMPONENET_FROM_PAGE = '[Cms] Get Component from Page';
export class LoadCmsComponent extends StateEntityLoaderActions.EntityLoadAction {
    // TODO(issue:6027) - this action should have only one `payload` property which should encapsulate all of the constructor's arguments
    /**
     * @param {?} payload
     * @param {?=} pageContext
     */
    constructor(payload, pageContext) {
        super(COMPONENT_ENTITY, payload);
        this.payload = payload;
        this.pageContext = pageContext;
        this.type = LOAD_CMS_COMPONENT;
    }
}
if (false) {
    /** @type {?} */
    LoadCmsComponent.prototype.type;
    /** @type {?} */
    LoadCmsComponent.prototype.payload;
    /** @type {?} */
    LoadCmsComponent.prototype.pageContext;
}
export class LoadCmsComponentFail extends StateEntityLoaderActions.EntityFailAction {
    // TODO(issue:6027) - this action should have only one `payload` property which should encapsulate all of the constructor's arguments
    /**
     * @param {?} uid
     * @param {?} payload
     * @param {?=} pageContext
     */
    constructor(uid, payload, pageContext) {
        super(COMPONENT_ENTITY, uid, payload);
        this.payload = payload;
        this.pageContext = pageContext;
        this.type = LOAD_CMS_COMPONENT_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadCmsComponentFail.prototype.type;
    /** @type {?} */
    LoadCmsComponentFail.prototype.payload;
    /** @type {?} */
    LoadCmsComponentFail.prototype.pageContext;
}
/**
 * @template T
 */
export class LoadCmsComponentSuccess extends StateEntityLoaderActions.EntitySuccessAction {
    // TODO(issue:6027) - this action should have only one `payload` property which should encapsulate all of the constructor's arguments
    /**
     * @param {?} payload
     * @param {?=} uid
     * @param {?=} pageContext
     */
    constructor(payload, uid, pageContext) {
        super(COMPONENT_ENTITY, uid || payload.uid || '');
        this.payload = payload;
        this.pageContext = pageContext;
        this.type = LOAD_CMS_COMPONENT_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadCmsComponentSuccess.prototype.type;
    /** @type {?} */
    LoadCmsComponentSuccess.prototype.payload;
    /** @type {?} */
    LoadCmsComponentSuccess.prototype.pageContext;
}
/**
 * @template T
 */
export class CmsGetComponentFromPage extends StateEntityLoaderActions.EntitySuccessAction {
    // TODO(issue:6027) - this action should have only one `payload` property which should encapsulate all of the constructor's arguments
    /**
     * @param {?} payload
     * @param {?=} pageContext
     */
    constructor(payload, pageContext) {
        super(COMPONENT_ENTITY, payload.map((/**
         * @param {?} cmp
         * @return {?}
         */
        cmp => cmp.uid)));
        this.payload = payload;
        this.pageContext = pageContext;
        this.type = CMS_GET_COMPONENET_FROM_PAGE;
    }
}
if (false) {
    /** @type {?} */
    CmsGetComponentFromPage.prototype.type;
    /** @type {?} */
    CmsGetComponentFromPage.prototype.payload;
    /** @type {?} */
    CmsGetComponentFromPage.prototype.pageContext;
}
