// @spartacus/core/esm5/src/cart/store/selectors/multi-cart.selector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createFeatureSelector, createSelector, } from '@ngrx/store';
import { entityValueSelector } from '../../../state/utils/entity-loader/entity-loader.selectors';
import { entityHasPendingProcessesSelector, entityIsStableSelector, entityProcessesLoaderStateSelector, } from '../../../state/utils/entity-processes-loader/entity-processes-loader.selectors';
import { MULTI_CART_FEATURE, } from '../multi-cart-state';
/** @type {?} */
export var getMultiCartState = createFeatureSelector(MULTI_CART_FEATURE);
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.carts; };
/** @type {?} */
export var getMultiCartEntities = createSelector(getMultiCartState, (ɵ0));
/** @type {?} */
export var getCartEntitySelectorFactory = (/**
 * @param {?} cartId
 * @return {?}
 */
function (cartId) {
    return createSelector(getMultiCartEntities, (/**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        return entityProcessesLoaderStateSelector(state, cartId);
    }));
});
/** @type {?} */
export var getCartSelectorFactory = (/**
 * @param {?} cartId
 * @return {?}
 */
function (cartId) {
    return createSelector(getMultiCartEntities, (/**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        return entityValueSelector(state, cartId);
    }));
});
/** @type {?} */
export var getCartIsStableSelectorFactory = (/**
 * @param {?} cartId
 * @return {?}
 */
function (cartId) {
    return createSelector(getMultiCartEntities, (/**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        return entityIsStableSelector(state, cartId);
    }));
});
/** @type {?} */
export var getCartHasPendingProcessesSelectorFactory = (/**
 * @param {?} cartId
 * @return {?}
 */
function (cartId) {
    return createSelector(getMultiCartEntities, (/**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        return entityHasPendingProcessesSelector(state, cartId);
    }));
});
/** @type {?} */
export var getCartEntriesSelectorFactory = (/**
 * @param {?} cartId
 * @return {?}
 */
function (cartId) {
    return createSelector(getCartSelectorFactory(cartId), (/**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        return state && state.entries ? state.entries : [];
    }));
});
/** @type {?} */
export var getCartEntrySelectorFactory = (/**
 * @param {?} cartId
 * @param {?} productCode
 * @return {?}
 */
function (cartId, productCode) {
    return createSelector(getCartEntriesSelectorFactory(cartId), (/**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        return state
            ? state.find((/**
             * @param {?} entry
             * @return {?}
             */
            function (entry) { return entry.product.code === productCode; }))
            : undefined;
    }));
});
var ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.active; };
/** @type {?} */
export var getActiveCartId = createSelector(getMultiCartState, (ɵ1));
var ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.wishList; };
/** @type {?} */
export var getWishListId = createSelector(getMultiCartState, (ɵ2));
export { ɵ0, ɵ1, ɵ2 };
