// @spartacus/core/esm5/src/auth/guards/auth-redirect.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { RoutingService } from '../../routing/facade/routing.service';
import * as i0 from "@angular/core";
import * as i1 from "../../routing/facade/routing.service";
import * as i2 from "@angular/router";
var AuthRedirectService = /** @class */ (function () {
    /**
     * This service is responsible for redirecting to the last page before authorization. "The last page" can be:
     * 1. Just the previously opened page; or
     * 2. The page that we just tried to open, but AuthGuard cancelled it
     *
     * For example:
     * 1. The user opens the product page, then clicks /login link and signs in
     *    -> Then we should redirect to the product page; or
     * 2. The user opens the product page, then he clicks /my-account link,
     *    but is automatically redirected to the login page by the AuthGuard, and he signs in
     *    -> Then we should redirect to the my-account page, not the product page
     */
    function AuthRedirectService(routing, router) {
        this.routing = routing;
        this.router = router;
        this.ignoredUrls = new Set();
    }
    /**
     * @return {?}
     */
    AuthRedirectService.prototype.redirect = /**
     * @return {?}
     */
    function () {
        if (this.redirectUrl === undefined) {
            this.routing.go('/');
        }
        else {
            this.routing.goByUrl(this.redirectUrl);
        }
        this.redirectUrl = undefined;
        this.lastAuthGuardNavigation = undefined;
    };
    /**
     * @return {?}
     */
    AuthRedirectService.prototype.reportAuthGuard = /**
     * @return {?}
     */
    function () {
        var _a = this.getCurrentNavigation(), url = _a.url, navigationId = _a.navigationId;
        this.lastAuthGuardNavigation = { url: url, navigationId: navigationId };
        this.redirectUrl = url;
    };
    /**
     * @return {?}
     */
    AuthRedirectService.prototype.reportNotAuthGuard = /**
     * @return {?}
     */
    function () {
        var _a = this.getCurrentNavigation(), url = _a.url, initialUrl = _a.initialUrl, navigationId = _a.navigationId;
        this.ignoredUrls.add(url);
        // Don't save redirect url if you've already come from page with NotAuthGuard (i.e. user has come from login to register)
        if (!this.ignoredUrls.has(initialUrl)) {
            // We compare the navigation id to find out if the url cancelled by AuthGuard (i.e. my-account) is more recent
            // than the last opened page
            if (!this.lastAuthGuardNavigation ||
                this.lastAuthGuardNavigation.navigationId < navigationId - 1) {
                this.redirectUrl = initialUrl;
                this.lastAuthGuardNavigation = undefined;
            }
        }
    };
    /**
     * @private
     * @return {?}
     */
    AuthRedirectService.prototype.getCurrentNavigation = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var initialUrl = this.router.url;
        /** @type {?} */
        var navigation = this.router.getCurrentNavigation();
        /** @type {?} */
        var url = this.router.serializeUrl(navigation.finalUrl);
        return {
            navigationId: navigation.id,
            url: url,
            initialUrl: initialUrl,
        };
    };
    AuthRedirectService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    AuthRedirectService.ctorParameters = function () { return [
        { type: RoutingService },
        { type: Router }
    ]; };
    /** @nocollapse */ AuthRedirectService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AuthRedirectService_Factory() { return new AuthRedirectService(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i2.Router)); }, token: AuthRedirectService, providedIn: "root" });
    return AuthRedirectService;
}());
export { AuthRedirectService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthRedirectService.prototype.redirectUrl;
    /**
     * @type {?}
     * @private
     */
    AuthRedirectService.prototype.ignoredUrls;
    /**
     * @type {?}
     * @private
     */
    AuthRedirectService.prototype.lastAuthGuardNavigation;
    /**
     * @type {?}
     * @private
     */
    AuthRedirectService.prototype.routing;
    /**
     * @type {?}
     * @private
     */
    AuthRedirectService.prototype.router;
}
