// @spartacus/core/esm5/src/auth/guards/not-auth.guard.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { RoutingService } from '../../routing/facade/routing.service';
import { AuthService } from '../facade/auth.service';
import { AuthRedirectService } from './auth-redirect.service';
import * as i0 from "@angular/core";
import * as i1 from "../../routing/facade/routing.service";
import * as i2 from "../facade/auth.service";
import * as i3 from "./auth-redirect.service";
var NotAuthGuard = /** @class */ (function () {
    function NotAuthGuard(routingService, authService, authRedirectService) {
        this.routingService = routingService;
        this.authService = authService;
        this.authRedirectService = authRedirectService;
    }
    /**
     * @return {?}
     */
    NotAuthGuard.prototype.canActivate = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.authRedirectService.reportNotAuthGuard();
        // redirect, if user is already logged in:
        return this.authService.getUserToken().pipe(map((/**
         * @param {?} token
         * @return {?}
         */
        function (token) {
            if (token.access_token) {
                _this.routingService.go({ cxRoute: 'home' });
            }
            return !token.access_token;
        })));
    };
    NotAuthGuard.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    NotAuthGuard.ctorParameters = function () { return [
        { type: RoutingService },
        { type: AuthService },
        { type: AuthRedirectService }
    ]; };
    /** @nocollapse */ NotAuthGuard.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function NotAuthGuard_Factory() { return new NotAuthGuard(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i2.AuthService), i0.ɵɵinject(i3.AuthRedirectService)); }, token: NotAuthGuard, providedIn: "root" });
    return NotAuthGuard;
}());
export { NotAuthGuard };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    NotAuthGuard.prototype.routingService;
    /**
     * @type {?}
     * @protected
     */
    NotAuthGuard.prototype.authService;
    /**
     * @type {?}
     * @private
     */
    NotAuthGuard.prototype.authRedirectService;
}
