// @spartacus/core/esm5/src/occ/utils/occ-fields.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { deepMerge, isObject } from '../../config/utils/deep-merge';
/**
 * Merge occ fields parameters
 *
 * @param {?} fields Fields definition as string or object
 * @return {?}
 */
export function mergeFields(fields) {
    /** @type {?} */
    var parsedFields = fields.map((/**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        return typeof f === 'string' ? parseFields(f) : f;
    }));
    /** @type {?} */
    var mergedFields = optimizeFields(deepMerge.apply(void 0, tslib_1.__spread([{}], parsedFields)));
    return stringifyFields(mergedFields);
}
/**
 * Optimize fields definition by removing not needed groups
 *
 * @param {?=} fields
 * @return {?}
 */
export function optimizeFields(fields) {
    if (fields === void 0) { fields = {}; }
    /** @type {?} */
    var keys = Object.keys(fields);
    if (keys.includes('FULL')) {
        delete fields['DEFAULT'];
        delete fields['BASIC'];
    }
    else if (keys.includes('DEFAULT')) {
        delete fields['BASIC'];
    }
    Object.keys(fields).forEach((/**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        fields[key] = optimizeFields(fields[key]);
    }));
    return fields;
}
/**
 * Parse string field definition to an AST object
 *
 * @param {?} fields Fields string definition
 * @param {?=} startIndex Used for recurrence
 * @return {?}
 */
export function parseFields(fields, startIndex) {
    if (startIndex === void 0) { startIndex = 0; }
    /** @type {?} */
    var parsedFields = {};
    /** @type {?} */
    var i = startIndex;
    while (i < fields.length) {
        if (fields[i] === ',') {
            if (i > startIndex) {
                parsedFields[fields.substr(startIndex, i - startIndex)] = {};
            }
            startIndex = i + 1;
        }
        else if (fields[i] === '(') {
            /** @type {?} */
            var subFields = parseFields(fields, i + 1);
            if (Array.isArray(subFields)) {
                parsedFields[fields.substr(startIndex, i - startIndex)] = subFields[0];
                startIndex = subFields[1];
                i = startIndex - 1;
            }
            else {
                return parsedFields;
            }
        }
        else if (fields[i] === ')') {
            if (i > startIndex) {
                parsedFields[fields.substr(startIndex, i - startIndex)] = {};
            }
            return [parsedFields, i + 1];
        }
        i++;
    }
    if (startIndex < fields.length) {
        parsedFields[fields.substr(startIndex, i - startIndex)] = {};
    }
    return parsedFields;
}
/**
 * Convert AST object fields definition to string representation
 *
 * @param {?} fields
 * @return {?}
 */
export function stringifyFields(fields) {
    return Object.keys(fields)
        .map((/**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var subFields = stringifyFields(fields[key]);
        return subFields ? key + "(" + subFields + ")" : key;
    }))
        .join(',');
}
/**
 * Extract part of the object described by fields definition
 *
 * @template T
 * @param {?} data
 * @param {?} fields
 * @return {?}
 */
export function extractFields(data, fields) {
    /** @type {?} */
    var parsedFields = typeof fields === 'string' ? parseFields(fields) : fields;
    return getObjectPart(data, parsedFields);
}
/**
 * @template T
 * @param {?} data
 * @param {?} fields
 * @return {?}
 */
function getObjectPart(data, fields) {
    if (!isObject(data)) {
        return data;
    }
    /** @type {?} */
    var keys = Object.keys(fields);
    if (keys.length === 0 ||
        // we should not extract parts of the object with ambiguous fields definitions
        keys.find((/**
         * @param {?} el
         * @return {?}
         */
        function (el) { return el === 'BASIC' || el === 'DEFAULT' || el === 'FULL'; }))) {
        return data;
    }
    /** @type {?} */
    var result = (/** @type {?} */ ({}));
    keys.forEach((/**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (data.hasOwnProperty(key)) {
            result[key] = getObjectPart(data[key], fields[key]);
        }
    }));
    return result;
}
