// @spartacus/core/esm5/src/occ/config-loader/occ-config-loader.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DOCUMENT, isPlatformBrowser, isPlatformServer } from '@angular/common';
import { Inject, Injectable, isDevMode, Optional, PLATFORM_ID, } from '@angular/core';
import { makeStateKey, TransferState, } from '@angular/platform-browser';
import { of } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { Config } from '../../config/config.module';
import { deepMerge } from '../../config/utils/deep-merge';
import { SERVER_REQUEST_URL } from '../../ssr/ssr.providers';
import { OccLoadedConfigConverter } from './occ-loaded-config-converter';
import { OccSitesConfigLoader } from './occ-sites-config-loader';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../../config/config.module";
import * as i3 from "./occ-sites-config-loader";
import * as i4 from "./occ-loaded-config-converter";
import * as i5 from "@angular/platform-browser";
import * as i6 from "../../ssr/ssr.providers";
/** @type {?} */
export var EXTERNAL_CONFIG_TRANSFER_ID = makeStateKey('cx-external-config');
var OccConfigLoaderService = /** @class */ (function () {
    function OccConfigLoaderService(platform, document, config, sitesConfigLoader, converter, transferState, serverRequestUrl) {
        this.platform = platform;
        this.document = document;
        this.config = config;
        this.sitesConfigLoader = sitesConfigLoader;
        this.converter = converter;
        this.transferState = transferState;
        this.serverRequestUrl = serverRequestUrl;
    }
    Object.defineProperty(OccConfigLoaderService.prototype, "currentUrl", {
        get: /**
         * @private
         * @return {?}
         */
        function () {
            if (isPlatformBrowser(this.platform)) {
                return this.document.location.href;
            }
            if (this.serverRequestUrl) {
                return this.serverRequestUrl;
            }
            if (isDevMode()) {
                console.error("Please provide token 'SERVER_REQUEST_URL' with the requested URL for SSR");
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Initializes the Spartacus config asynchronously basing on the external config
     */
    /**
     * Initializes the Spartacus config asynchronously basing on the external config
     * @return {?}
     */
    OccConfigLoaderService.prototype.loadConfig = /**
     * Initializes the Spartacus config asynchronously basing on the external config
     * @return {?}
     */
    function () {
        var _this = this;
        return this.get()
            .pipe(tap((/**
         * @param {?} externalConfig
         * @return {?}
         */
        function (externalConfig) { return _this.transfer(externalConfig); })), map((/**
         * @param {?} externalConfig
         * @return {?}
         */
        function (externalConfig) {
            return deepMerge.apply(void 0, tslib_1.__spread([{}], _this.getConfigChunks(externalConfig)));
        })))
            .toPromise();
    };
    /**
     * Returns the external config
     */
    /**
     * Returns the external config
     * @protected
     * @return {?}
     */
    OccConfigLoaderService.prototype.get = /**
     * Returns the external config
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var rehydratedExternalConfig = this.rehydrate();
        return rehydratedExternalConfig
            ? of(rehydratedExternalConfig)
            : this.load();
    };
    /**
     * Loads the external config from backend
     */
    /**
     * Loads the external config from backend
     * @protected
     * @return {?}
     */
    OccConfigLoaderService.prototype.load = /**
     * Loads the external config from backend
     * @protected
     * @return {?}
     */
    function () {
        var _this = this;
        return this.sitesConfigLoader
            .load()
            .pipe(map((/**
         * @param {?} baseSites
         * @return {?}
         */
        function (baseSites) {
            return _this.converter.fromOccBaseSites(baseSites, _this.currentUrl);
        })));
    };
    /**
     * Tries to rehydrate external config in the browser from SSR
     */
    /**
     * Tries to rehydrate external config in the browser from SSR
     * @protected
     * @return {?}
     */
    OccConfigLoaderService.prototype.rehydrate = /**
     * Tries to rehydrate external config in the browser from SSR
     * @protected
     * @return {?}
     */
    function () {
        if (this.transferState && isPlatformBrowser(this.platform)) {
            return this.transferState.get(EXTERNAL_CONFIG_TRANSFER_ID, undefined);
        }
    };
    /**
     * Transfers the given external config in SSR to the browser
     *
     * @param externalConfig
     */
    /**
     * Transfers the given external config in SSR to the browser
     *
     * @protected
     * @param {?} externalConfig
     * @return {?}
     */
    OccConfigLoaderService.prototype.transfer = /**
     * Transfers the given external config in SSR to the browser
     *
     * @protected
     * @param {?} externalConfig
     * @return {?}
     */
    function (externalConfig) {
        if (this.transferState &&
            isPlatformServer(this.platform) &&
            externalConfig) {
            this.transferState.set(EXTERNAL_CONFIG_TRANSFER_ID, externalConfig);
        }
    };
    /**
     * @protected
     * @param {?} externalConfig
     * @return {?}
     */
    OccConfigLoaderService.prototype.getConfigChunks = /**
     * @protected
     * @param {?} externalConfig
     * @return {?}
     */
    function (externalConfig) {
        /** @type {?} */
        var chunks = [this.converter.toSiteContextConfig(externalConfig)];
        if (this.shouldReturnI18nChunk()) {
            chunks.push(this.converter.toI18nConfig(externalConfig));
        }
        return chunks;
    };
    /**
     * @private
     * @return {?}
     */
    OccConfigLoaderService.prototype.shouldReturnI18nChunk = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var fallbackLangExists = typeof (this.config &&
            this.config.i18n &&
            this.config.i18n.fallbackLang) !== 'undefined';
        if (fallbackLangExists && isDevMode()) {
            console.warn("There is an already provided static config for 'i18n.fallbackLang', so the value from OCC loaded config is ignored.");
        }
        return !fallbackLangExists;
    };
    OccConfigLoaderService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    OccConfigLoaderService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [Config,] }] },
        { type: OccSitesConfigLoader },
        { type: OccLoadedConfigConverter },
        { type: TransferState, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Optional }, { type: Inject, args: [SERVER_REQUEST_URL,] }] }
    ]; };
    /** @nocollapse */ OccConfigLoaderService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OccConfigLoaderService_Factory() { return new OccConfigLoaderService(i0.ɵɵinject(i0.PLATFORM_ID), i0.ɵɵinject(i1.DOCUMENT), i0.ɵɵinject(i2.Config), i0.ɵɵinject(i3.OccSitesConfigLoader), i0.ɵɵinject(i4.OccLoadedConfigConverter), i0.ɵɵinject(i5.TransferState, 8), i0.ɵɵinject(i6.SERVER_REQUEST_URL, 8)); }, token: OccConfigLoaderService, providedIn: "root" });
    return OccConfigLoaderService;
}());
export { OccConfigLoaderService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    OccConfigLoaderService.prototype.platform;
    /**
     * @type {?}
     * @protected
     */
    OccConfigLoaderService.prototype.document;
    /**
     * @type {?}
     * @protected
     */
    OccConfigLoaderService.prototype.config;
    /**
     * @type {?}
     * @protected
     */
    OccConfigLoaderService.prototype.sitesConfigLoader;
    /**
     * @type {?}
     * @protected
     */
    OccConfigLoaderService.prototype.converter;
    /**
     * @type {?}
     * @protected
     */
    OccConfigLoaderService.prototype.transferState;
    /**
     * @type {?}
     * @protected
     */
    OccConfigLoaderService.prototype.serverRequestUrl;
}
