// @spartacus/core/esm5/src/occ/config-loader/occ-sites-config-loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { map } from 'rxjs/operators';
import { OccConfig } from '../config/occ-config';
import * as i0 from "@angular/core";
import * as i1 from "../config/occ-config";
import * as i2 from "@angular/common/http";
var OccSitesConfigLoader = /** @class */ (function () {
    function OccSitesConfigLoader(config, http) {
        this.config = config;
        this.http = http;
        this.endpoint = 'basesites?fields=baseSites(uid,defaultLanguage(isocode),urlEncodingAttributes,urlPatterns,stores(currencies(isocode),defaultCurrency(isocode),languages(isocode),defaultLanguage(isocode)))';
    }
    Object.defineProperty(OccSitesConfigLoader.prototype, "baseEndpoint", {
        get: /**
         * @private
         * @return {?}
         */
        function () {
            return ((this.config.backend.occ.baseUrl || '') + this.config.backend.occ.prefix);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OccSitesConfigLoader.prototype, "url", {
        get: /**
         * @private
         * @return {?}
         */
        function () {
            return "" + this.baseEndpoint + this.endpoint;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    OccSitesConfigLoader.prototype.load = /**
     * @return {?}
     */
    function () {
        if (!this.config || !this.config.backend || !this.config.backend.occ) {
            return throwError(new Error("Missing config for OCC backend!"));
        }
        return this.http
            .get(this.url)
            .pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var baseSites = _a.baseSites;
            return baseSites;
        })));
    };
    OccSitesConfigLoader.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    OccSitesConfigLoader.ctorParameters = function () { return [
        { type: OccConfig },
        { type: HttpClient }
    ]; };
    /** @nocollapse */ OccSitesConfigLoader.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OccSitesConfigLoader_Factory() { return new OccSitesConfigLoader(i0.ɵɵinject(i1.OccConfig), i0.ɵɵinject(i2.HttpClient)); }, token: OccSitesConfigLoader, providedIn: "root" });
    return OccSitesConfigLoader;
}());
export { OccSitesConfigLoader };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    OccSitesConfigLoader.prototype.endpoint;
    /**
     * @type {?}
     * @protected
     */
    OccSitesConfigLoader.prototype.config;
    /**
     * @type {?}
     * @protected
     */
    OccSitesConfigLoader.prototype.http;
}
