// @spartacus/core/esm5/src/occ/config-loader/occ-config-loader.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CONFIG_INITIALIZER, } from '../../config/config-initializer/config-initializer';
import { SiteContextConfig } from '../../site-context/config/site-context-config';
import { BASE_SITE_CONTEXT_ID } from '../../site-context/providers/context-ids';
import { OccConfigLoaderService } from './occ-config-loader.service';
/**
 * Initializes the Spartacus config asynchronously basing on the external config
 * @param {?} configLoader
 * @param {?} config
 * @return {?}
 */
export function initConfig(configLoader, config) {
    /**
     * Load config for `context` from backend only when there is no static config for `context.baseSite`
     */
    if (!config.context || !config.context[BASE_SITE_CONTEXT_ID]) {
        return {
            scopes: ['context', 'i18n.fallbackLang'],
            configFactory: (/**
             * @return {?}
             */
            function () { return configLoader.loadConfig(); }),
        };
    }
    return null;
}
/**
 * Re-provides the external config chunk given before Angular bootstrap
 */
var OccConfigLoaderModule = /** @class */ (function () {
    function OccConfigLoaderModule() {
    }
    /**
     * @return {?}
     */
    OccConfigLoaderModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: OccConfigLoaderModule,
            providers: [
                {
                    provide: CONFIG_INITIALIZER,
                    useFactory: initConfig,
                    deps: [OccConfigLoaderService, SiteContextConfig],
                    multi: true,
                },
            ],
        };
    };
    OccConfigLoaderModule.decorators = [
        { type: NgModule }
    ];
    return OccConfigLoaderModule;
}());
export { OccConfigLoaderModule };
