// @spartacus/core/esm2015/src/user/store/selectors/customer-coupons.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { loaderValueSelector, loaderSuccessSelector, loaderLoadingSelector, } from '../../../state/utils/loader/loader.selectors';
import { getUserState } from './feature.selector';
const ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
(state) => state.customerCoupons;
/** @type {?} */
export const getCustomerCouponsState = createSelector(getUserState, (ɵ0));
const ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
(state) => loaderSuccessSelector(state);
/** @type {?} */
export const getCustomerCouponsLoaded = createSelector(getCustomerCouponsState, (ɵ1));
const ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
(state) => loaderLoadingSelector(state);
/** @type {?} */
export const getCustomerCouponsLoading = createSelector(getCustomerCouponsState, (ɵ2));
const ɵ3 = /**
 * @param {?} state
 * @return {?}
 */
(state) => loaderValueSelector(state);
/** @type {?} */
export const getCustomerCoupons = createSelector(getCustomerCouponsState, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
