// @spartacus/core/esm2015/src/user/store/selectors/delivery-countries.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { getUserState } from './feature.selector';
const ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
(state) => state.countries;
/** @type {?} */
export const getDeliveryCountriesState = createSelector(getUserState, (ɵ0));
const ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
(state) => state.entities;
/** @type {?} */
export const getDeliveryCountriesEntites = createSelector(getDeliveryCountriesState, (ɵ1));
const ɵ2 = /**
 * @param {?} entites
 * @return {?}
 */
entites => Object.keys(entites).map((/**
 * @param {?} isocode
 * @return {?}
 */
isocode => entites[isocode]));
/** @type {?} */
export const getAllDeliveryCountries = createSelector(getDeliveryCountriesEntites, (ɵ2));
/** @type {?} */
export const countrySelectorFactory = (/**
 * @param {?} isocode
 * @return {?}
 */
(isocode) => createSelector(getDeliveryCountriesEntites, (/**
 * @param {?} entities
 * @return {?}
 */
entities => (Object.keys(entities).length !== 0 ? entities[isocode] : null))));
export { ɵ0, ɵ1, ɵ2 };
