// @spartacus/core/esm2015/src/user/store/selectors/notification-preference.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { getUserState } from './feature.selector';
import { loaderValueSelector, loaderLoadingSelector, } from '../../../state/utils/loader/loader.selectors';
const ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
(state) => state.notificationPreferences;
/** @type {?} */
export const getPreferencesLoaderState = createSelector(getUserState, (ɵ0));
const ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
(state) => loaderValueSelector(state);
/** @type {?} */
export const getPreferences = createSelector(getPreferencesLoaderState, (ɵ1));
const ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
(state) => loaderValueSelector(state).filter((/**
 * @param {?} p
 * @return {?}
 */
p => p.enabled));
/** @type {?} */
export const getEnabledPreferences = createSelector(getPreferencesLoaderState, (ɵ2));
const ɵ3 = /**
 * @param {?} state
 * @return {?}
 */
(state) => loaderLoadingSelector(state);
/** @type {?} */
export const getPreferencesLoading = createSelector(getPreferencesLoaderState, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
