// @spartacus/core/esm2015/src/user/store/selectors/titles.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { getUserState } from './feature.selector';
const ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
(state) => state.titles;
/** @type {?} */
export const getTitlesState = createSelector(getUserState, (ɵ0));
const ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
(state) => state.entities;
/** @type {?} */
export const getTitlesEntites = createSelector(getTitlesState, (ɵ1));
const ɵ2 = /**
 * @param {?} entites
 * @return {?}
 */
entites => Object.keys(entites).map((/**
 * @param {?} code
 * @return {?}
 */
code => entites[code]));
/** @type {?} */
export const getAllTitles = createSelector(getTitlesEntites, (ɵ2));
/** @type {?} */
export const titleSelectorFactory = (/**
 * @param {?} code
 * @return {?}
 */
(code) => createSelector(getTitlesEntites, (/**
 * @param {?} entities
 * @return {?}
 */
entities => (Object.keys(entities).length !== 0 ? entities[code] : null))));
export { ɵ0, ɵ1, ɵ2 };
