// @spartacus/core/esm2015/src/user/store/selectors/user-consents.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/index';
import { getUserState } from './feature.selector';
const ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
(state) => state.consents;
/** @type {?} */
export const getConsentsState = createSelector(getUserState, (ɵ0));
/** @type {?} */
export const getConsentsValue = createSelector(getConsentsState, StateLoaderSelectors.loaderValueSelector);
/** @type {?} */
export const getConsentByTemplateId = (/**
 * @param {?} templateId
 * @return {?}
 */
(templateId) => createSelector(getConsentsValue, (/**
 * @param {?} templates
 * @return {?}
 */
templates => templates.find((/**
 * @param {?} template
 * @return {?}
 */
template => template.id === templateId)))));
/** @type {?} */
export const getConsentsLoading = createSelector(getConsentsState, StateLoaderSelectors.loaderLoadingSelector);
/** @type {?} */
export const getConsentsSuccess = createSelector(getConsentsState, StateLoaderSelectors.loaderSuccessSelector);
/** @type {?} */
export const getConsentsError = createSelector(getConsentsState, StateLoaderSelectors.loaderErrorSelector);
export { ɵ0 };
