// @spartacus/core/esm2015/src/user/store/selectors/user-addresses.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/index';
import { getUserState } from './feature.selector';
const ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
(state) => state.addresses;
/** @type {?} */
export const getAddressesLoaderState = createSelector(getUserState, (ɵ0));
const ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
(state) => StateLoaderSelectors.loaderValueSelector(state);
/** @type {?} */
export const getAddresses = createSelector(getAddressesLoaderState, (ɵ1));
const ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
(state) => StateLoaderSelectors.loaderLoadingSelector(state);
/** @type {?} */
export const getAddressesLoading = createSelector(getAddressesLoaderState, (ɵ2));
const ɵ3 = /**
 * @param {?} state
 * @return {?}
 */
(state) => StateLoaderSelectors.loaderSuccessSelector(state) &&
    !StateLoaderSelectors.loaderLoadingSelector(state);
/** @type {?} */
export const getAddressesLoadedSuccess = createSelector(getAddressesLoaderState, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
