// @spartacus/core/esm5/src/model/address.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Country() { }
if (false) {
    /** @type {?|undefined} */
    Country.prototype.isocode;
    /** @type {?|undefined} */
    Country.prototype.name;
}
/** @enum {string} */
var CountryType = {
    BILLING: 'BILLING',
    SHIPPING: 'SHIPPING',
};
export { CountryType };
/**
 * @record
 */
export function Region() { }
if (false) {
    /** @type {?|undefined} */
    Region.prototype.countryIso;
    /** @type {?|undefined} */
    Region.prototype.isocode;
    /** @type {?|undefined} */
    Region.prototype.isocodeShort;
    /** @type {?|undefined} */
    Region.prototype.name;
}
/**
 * @record
 */
export function Address() { }
if (false) {
    /** @type {?|undefined} */
    Address.prototype.companyName;
    /** @type {?|undefined} */
    Address.prototype.country;
    /** @type {?|undefined} */
    Address.prototype.defaultAddress;
    /** @type {?|undefined} */
    Address.prototype.email;
    /** @type {?|undefined} */
    Address.prototype.firstName;
    /** @type {?|undefined} */
    Address.prototype.formattedAddress;
    /** @type {?|undefined} */
    Address.prototype.id;
    /** @type {?|undefined} */
    Address.prototype.lastName;
    /** @type {?|undefined} */
    Address.prototype.line1;
    /** @type {?|undefined} */
    Address.prototype.line2;
    /** @type {?|undefined} */
    Address.prototype.phone;
    /** @type {?|undefined} */
    Address.prototype.postalCode;
    /** @type {?|undefined} */
    Address.prototype.region;
    /** @type {?|undefined} */
    Address.prototype.shippingAddress;
    /** @type {?|undefined} */
    Address.prototype.title;
    /** @type {?|undefined} */
    Address.prototype.titleCode;
    /** @type {?|undefined} */
    Address.prototype.town;
    /** @type {?|undefined} */
    Address.prototype.visibleInAddressBook;
}
/**
 * @record
 */
export function AddressValidation() { }
if (false) {
    /** @type {?|undefined} */
    AddressValidation.prototype.decision;
    /** @type {?|undefined} */
    AddressValidation.prototype.errors;
    /** @type {?|undefined} */
    AddressValidation.prototype.suggestedAddresses;
}
