// @spartacus/core/esm5/src/model/cart.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PromotionResult() { }
if (false) {
    /** @type {?|undefined} */
    PromotionResult.prototype.consumedEntries;
    /** @type {?|undefined} */
    PromotionResult.prototype.description;
    /** @type {?|undefined} */
    PromotionResult.prototype.promotion;
}
/** @enum {string} */
var PromotionLocation = {
    ActiveCart: 'CART',
    Checkout: 'CHECKOUT',
    Order: 'ORDER',
};
export { PromotionLocation };
/**
 * @record
 */
export function Voucher() { }
if (false) {
    /** @type {?|undefined} */
    Voucher.prototype.appliedValue;
    /** @type {?|undefined} */
    Voucher.prototype.code;
    /** @type {?|undefined} */
    Voucher.prototype.currency;
    /** @type {?|undefined} */
    Voucher.prototype.description;
    /** @type {?|undefined} */
    Voucher.prototype.freeShipping;
    /** @type {?|undefined} */
    Voucher.prototype.name;
    /** @type {?|undefined} */
    Voucher.prototype.value;
    /** @type {?|undefined} */
    Voucher.prototype.valueFormatted;
    /** @type {?|undefined} */
    Voucher.prototype.valueString;
    /** @type {?|undefined} */
    Voucher.prototype.voucherCode;
}
/**
 * @record
 */
export function DeliveryOrderEntryGroup() { }
if (false) {
    /** @type {?|undefined} */
    DeliveryOrderEntryGroup.prototype.deliveryAddress;
    /** @type {?|undefined} */
    DeliveryOrderEntryGroup.prototype.entries;
    /** @type {?|undefined} */
    DeliveryOrderEntryGroup.prototype.quantity;
    /** @type {?|undefined} */
    DeliveryOrderEntryGroup.prototype.totalPriceWithTax;
}
/**
 * @record
 */
export function Principal() { }
if (false) {
    /** @type {?|undefined} */
    Principal.prototype.name;
    /** @type {?|undefined} */
    Principal.prototype.uid;
}
/**
 * @record
 */
export function CardType() { }
if (false) {
    /** @type {?|undefined} */
    CardType.prototype.code;
    /** @type {?|undefined} */
    CardType.prototype.name;
}
/**
 * @record
 */
export function PaymentDetails() { }
if (false) {
    /** @type {?|undefined} */
    PaymentDetails.prototype.accountHolderName;
    /** @type {?|undefined} */
    PaymentDetails.prototype.billingAddress;
    /** @type {?|undefined} */
    PaymentDetails.prototype.cardNumber;
    /** @type {?|undefined} */
    PaymentDetails.prototype.cardType;
    /** @type {?|undefined} */
    PaymentDetails.prototype.cvn;
    /** @type {?|undefined} */
    PaymentDetails.prototype.defaultPayment;
    /** @type {?|undefined} */
    PaymentDetails.prototype.expiryMonth;
    /** @type {?|undefined} */
    PaymentDetails.prototype.expiryYear;
    /** @type {?|undefined} */
    PaymentDetails.prototype.id;
    /** @type {?|undefined} */
    PaymentDetails.prototype.issueNumber;
    /** @type {?|undefined} */
    PaymentDetails.prototype.saved;
    /** @type {?|undefined} */
    PaymentDetails.prototype.startMonth;
    /** @type {?|undefined} */
    PaymentDetails.prototype.startYear;
    /** @type {?|undefined} */
    PaymentDetails.prototype.subscriptionId;
}
/**
 * @record
 */
export function SaveCartResult() { }
if (false) {
    /** @type {?|undefined} */
    SaveCartResult.prototype.savedCartData;
}
/**
 * @record
 */
export function Cart() { }
if (false) {
    /** @type {?|undefined} */
    Cart.prototype.appliedOrderPromotions;
    /** @type {?|undefined} */
    Cart.prototype.appliedProductPromotions;
    /** @type {?|undefined} */
    Cart.prototype.appliedVouchers;
    /** @type {?|undefined} */
    Cart.prototype.calculated;
    /** @type {?|undefined} */
    Cart.prototype.code;
    /** @type {?|undefined} */
    Cart.prototype.deliveryAddress;
    /** @type {?|undefined} */
    Cart.prototype.deliveryCost;
    /** @type {?|undefined} */
    Cart.prototype.deliveryItemsQuantity;
    /** @type {?|undefined} */
    Cart.prototype.deliveryMode;
    /** @type {?|undefined} */
    Cart.prototype.deliveryOrderGroups;
    /** @type {?|undefined} */
    Cart.prototype.description;
    /** @type {?|undefined} */
    Cart.prototype.entries;
    /** @type {?|undefined} */
    Cart.prototype.expirationTime;
    /** @type {?|undefined} */
    Cart.prototype.guid;
    /** @type {?|undefined} */
    Cart.prototype.name;
    /** @type {?|undefined} */
    Cart.prototype.net;
    /** @type {?|undefined} */
    Cart.prototype.orderDiscounts;
    /** @type {?|undefined} */
    Cart.prototype.paymentInfo;
    /** @type {?|undefined} */
    Cart.prototype.pickupItemsQuantity;
    /** @type {?|undefined} */
    Cart.prototype.pickupOrderGroups;
    /** @type {?|undefined} */
    Cart.prototype.potentialOrderPromotions;
    /** @type {?|undefined} */
    Cart.prototype.potentialProductPromotions;
    /** @type {?|undefined} */
    Cart.prototype.productDiscounts;
    /** @type {?|undefined} */
    Cart.prototype.saveTime;
    /** @type {?|undefined} */
    Cart.prototype.savedBy;
    /** @type {?|undefined} */
    Cart.prototype.site;
    /** @type {?|undefined} */
    Cart.prototype.store;
    /** @type {?|undefined} */
    Cart.prototype.subTotal;
    /** @type {?|undefined} */
    Cart.prototype.totalDiscounts;
    /** @type {?|undefined} */
    Cart.prototype.totalItems;
    /** @type {?|undefined} */
    Cart.prototype.totalPrice;
    /** @type {?|undefined} */
    Cart.prototype.totalPriceWithTax;
    /** @type {?|undefined} */
    Cart.prototype.totalTax;
    /** @type {?|undefined} */
    Cart.prototype.totalUnitCount;
    /** @type {?|undefined} */
    Cart.prototype.user;
}
/**
 * @record
 */
export function CartModification() { }
if (false) {
    /** @type {?|undefined} */
    CartModification.prototype.deliveryModeChanged;
    /** @type {?|undefined} */
    CartModification.prototype.entry;
    /** @type {?|undefined} */
    CartModification.prototype.quantity;
    /** @type {?|undefined} */
    CartModification.prototype.quantityAdded;
    /** @type {?|undefined} */
    CartModification.prototype.statusCode;
    /** @type {?|undefined} */
    CartModification.prototype.statusMessage;
}
