// @spartacus/core/esm5/src/model/cms.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function CmsComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsComponent.prototype.modifiedTime;
    /** @type {?|undefined} */
    CmsComponent.prototype.name;
    /** @type {?|undefined} */
    CmsComponent.prototype.otherProperties;
    /** @type {?|undefined} */
    CmsComponent.prototype.typeCode;
    /** @type {?|undefined} */
    CmsComponent.prototype.uid;
}
/** @enum {string} */
var PageType = {
    CONTENT_PAGE: 'ContentPage',
    PRODUCT_PAGE: 'ProductPage',
    CATEGORY_PAGE: 'CategoryPage',
    CATALOG_PAGE: 'CatalogPage',
};
export { PageType };
/**
 * @record
 */
export function CmsLinkComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsLinkComponent.prototype.url;
    /** @type {?|undefined} */
    CmsLinkComponent.prototype.container;
    /** @type {?|undefined} */
    CmsLinkComponent.prototype.external;
    /** @type {?|undefined} */
    CmsLinkComponent.prototype.contentPage;
    /** @type {?|undefined} */
    CmsLinkComponent.prototype.contentPageLabelOrId;
    /** @type {?|undefined} */
    CmsLinkComponent.prototype.linkName;
    /** @type {?|undefined} */
    CmsLinkComponent.prototype.target;
}
/**
 * @record
 */
export function CmsSiteContextSelectorComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsSiteContextSelectorComponent.prototype.context;
}
/**
 * @record
 */
export function CmsSearchBoxComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsSearchBoxComponent.prototype.container;
    /** @type {?|undefined} */
    CmsSearchBoxComponent.prototype.maxSuggestions;
    /** @type {?|undefined} */
    CmsSearchBoxComponent.prototype.maxProducts;
    /** @type {?|undefined} */
    CmsSearchBoxComponent.prototype.displaySuggestions;
    /** @type {?|undefined} */
    CmsSearchBoxComponent.prototype.displayProducts;
    /** @type {?|undefined} */
    CmsSearchBoxComponent.prototype.displayProductImages;
    /** @type {?|undefined} */
    CmsSearchBoxComponent.prototype.waitTimeBeforeRequest;
    /** @type {?|undefined} */
    CmsSearchBoxComponent.prototype.minCharactersBeforeRequest;
}
/**
 * @record
 */
export function CmsParagraphComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsParagraphComponent.prototype.content;
    /** @type {?|undefined} */
    CmsParagraphComponent.prototype.container;
    /** @type {?|undefined} */
    CmsParagraphComponent.prototype.title;
}
/**
 * @record
 */
export function CMSTabParagraphContainer() { }
if (false) {
    /** @type {?|undefined} */
    CMSTabParagraphContainer.prototype.container;
    /** @type {?|undefined} */
    CMSTabParagraphContainer.prototype.components;
}
/**
 * @record
 */
export function CmsBannerComponentMedia() { }
if (false) {
    /** @type {?|undefined} */
    CmsBannerComponentMedia.prototype.altText;
    /** @type {?|undefined} */
    CmsBannerComponentMedia.prototype.code;
    /** @type {?|undefined} */
    CmsBannerComponentMedia.prototype.mime;
    /** @type {?|undefined} */
    CmsBannerComponentMedia.prototype.url;
}
/**
 * @record
 */
export function CmsResponsiveBannerComponentMedia() { }
if (false) {
    /** @type {?|undefined} */
    CmsResponsiveBannerComponentMedia.prototype.desktop;
    /** @type {?|undefined} */
    CmsResponsiveBannerComponentMedia.prototype.mobile;
    /** @type {?|undefined} */
    CmsResponsiveBannerComponentMedia.prototype.tablet;
    /** @type {?|undefined} */
    CmsResponsiveBannerComponentMedia.prototype.widescreen;
}
/**
 * @record
 */
export function CmsBannerComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsBannerComponent.prototype.headline;
    /** @type {?|undefined} */
    CmsBannerComponent.prototype.content;
    /** @type {?|undefined} */
    CmsBannerComponent.prototype.container;
    /** @type {?|undefined} */
    CmsBannerComponent.prototype.media;
    /** @type {?|undefined} */
    CmsBannerComponent.prototype.urlLink;
    /** @type {?|undefined} */
    CmsBannerComponent.prototype.external;
}
/** @enum {string} */
var CmsBannerCarouselEffect = {
    FADE: 'FADE',
    ZOOM: 'ZOOM',
    CURTAIN: 'CURTAINX',
    TURNDOWN: 'TURNDOWN',
};
export { CmsBannerCarouselEffect };
/**
 * @record
 */
export function CmsBannerCarouselComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsBannerCarouselComponent.prototype.banners;
    /** @type {?|undefined} */
    CmsBannerCarouselComponent.prototype.effect;
}
/**
 * @record
 */
export function CmsProductCarouselComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsProductCarouselComponent.prototype.title;
    /** @type {?|undefined} */
    CmsProductCarouselComponent.prototype.productCodes;
    /** @type {?|undefined} */
    CmsProductCarouselComponent.prototype.container;
    /** @type {?|undefined} */
    CmsProductCarouselComponent.prototype.popup;
    /** @type {?|undefined} */
    CmsProductCarouselComponent.prototype.scroll;
}
/**
 * @record
 */
export function CmsProductReferencesComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsProductReferencesComponent.prototype.title;
    /** @type {?|undefined} */
    CmsProductReferencesComponent.prototype.displayProductTitles;
    /** @type {?|undefined} */
    CmsProductReferencesComponent.prototype.displayProductPrices;
    /** @type {?|undefined} */
    CmsProductReferencesComponent.prototype.maximumNumberProducts;
    /** @type {?|undefined} */
    CmsProductReferencesComponent.prototype.productReferenceTypes;
    /** @type {?|undefined} */
    CmsProductReferencesComponent.prototype.container;
}
/**
 * @record
 */
export function CmsMiniCartComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsMiniCartComponent.prototype.container;
    /** @type {?|undefined} */
    CmsMiniCartComponent.prototype.shownProductCount;
    /** @type {?|undefined} */
    CmsMiniCartComponent.prototype.title;
    /** @type {?|undefined} */
    CmsMiniCartComponent.prototype.totalDisplay;
    /** @type {?|undefined} */
    CmsMiniCartComponent.prototype.lightboxBannerComponent;
}
/**
 * @record
 */
export function CmsBreadcrumbsComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsBreadcrumbsComponent.prototype.container;
}
/**
 * @record
 */
export function CmsNavigationNode() { }
if (false) {
    /** @type {?|undefined} */
    CmsNavigationNode.prototype.uid;
    /** @type {?|undefined} */
    CmsNavigationNode.prototype.title;
    /** @type {?|undefined} */
    CmsNavigationNode.prototype.children;
    /** @type {?|undefined} */
    CmsNavigationNode.prototype.entries;
}
/**
 * @record
 */
export function CmsNavigationEntry() { }
if (false) {
    /** @type {?|undefined} */
    CmsNavigationEntry.prototype.itemId;
    /** @type {?|undefined} */
    CmsNavigationEntry.prototype.itemSuperType;
    /** @type {?|undefined} */
    CmsNavigationEntry.prototype.itemType;
}
/**
 * @record
 */
export function CmsNavigationComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsNavigationComponent.prototype.container;
    /** @type {?|undefined} */
    CmsNavigationComponent.prototype.styleClass;
    /** @type {?|undefined} */
    CmsNavigationComponent.prototype.wrapAfter;
    /** @type {?|undefined} */
    CmsNavigationComponent.prototype.notice;
    /** @type {?|undefined} */
    CmsNavigationComponent.prototype.showLanguageCurrency;
    /** @type {?|undefined} */
    CmsNavigationComponent.prototype.navigationNode;
}
/**
 * @record
 */
export function CmsProductFacetNavigationComponent() { }
if (false) {
    /** @type {?|undefined} */
    CmsProductFacetNavigationComponent.prototype.container;
    /** @type {?|undefined} */
    CmsProductFacetNavigationComponent.prototype.activeFacetValueCode;
    /** @type {?|undefined} */
    CmsProductFacetNavigationComponent.prototype.searchResult;
    /** @type {?|undefined} */
    CmsProductFacetNavigationComponent.prototype.minPerFacet;
}
