// @spartacus/core/esm5/src/model/consent.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function ConsentTemplate() { }
if (false) {
    /** @type {?|undefined} */
    ConsentTemplate.prototype.id;
    /** @type {?|undefined} */
    ConsentTemplate.prototype.name;
    /** @type {?|undefined} */
    ConsentTemplate.prototype.description;
    /** @type {?|undefined} */
    ConsentTemplate.prototype.version;
    /** @type {?|undefined} */
    ConsentTemplate.prototype.currentConsent;
}
/**
 * @record
 */
export function Consent() { }
if (false) {
    /** @type {?|undefined} */
    Consent.prototype.code;
    /** @type {?|undefined} */
    Consent.prototype.consentGivenDate;
    /** @type {?|undefined} */
    Consent.prototype.consentWithdrawnDate;
}
/**
 * @record
 */
export function AnonymousConsent() { }
if (false) {
    /** @type {?|undefined} */
    AnonymousConsent.prototype.templateCode;
    /** @type {?|undefined} */
    AnonymousConsent.prototype.version;
    /** @type {?|undefined} */
    AnonymousConsent.prototype.consentState;
}
/** @enum {string} */
var ANONYMOUS_CONSENT_STATUS = {
    GIVEN: 'GIVEN',
    WITHDRAWN: 'WITHDRAWN',
};
export { ANONYMOUS_CONSENT_STATUS };
