// @spartacus/core/esm5/src/model/product-search.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function SearchQuery() { }
if (false) {
    /** @type {?|undefined} */
    SearchQuery.prototype.value;
}
/**
 * @record
 */
export function SearchState() { }
if (false) {
    /** @type {?|undefined} */
    SearchState.prototype.query;
    /** @type {?|undefined} */
    SearchState.prototype.url;
}
/**
 * @record
 */
export function FacetValue() { }
if (false) {
    /** @type {?|undefined} */
    FacetValue.prototype.count;
    /** @type {?|undefined} */
    FacetValue.prototype.name;
    /** @type {?|undefined} */
    FacetValue.prototype.query;
    /** @type {?|undefined} */
    FacetValue.prototype.selected;
}
/**
 * @record
 */
export function Breadcrumb() { }
if (false) {
    /** @type {?|undefined} */
    Breadcrumb.prototype.facetCode;
    /** @type {?|undefined} */
    Breadcrumb.prototype.facetName;
    /** @type {?|undefined} */
    Breadcrumb.prototype.facetValueCode;
    /** @type {?|undefined} */
    Breadcrumb.prototype.facetValueName;
    /** @type {?|undefined} */
    Breadcrumb.prototype.removeQuery;
    /** @type {?|undefined} */
    Breadcrumb.prototype.truncateQuery;
}
/**
 * @record
 */
export function Facet() { }
if (false) {
    /** @type {?|undefined} */
    Facet.prototype.category;
    /** @type {?|undefined} */
    Facet.prototype.multiSelect;
    /** @type {?|undefined} */
    Facet.prototype.name;
    /** @type {?|undefined} */
    Facet.prototype.priority;
    /** @type {?|undefined} */
    Facet.prototype.topValues;
    /** @type {?|undefined} */
    Facet.prototype.values;
    /** @type {?|undefined} */
    Facet.prototype.visible;
}
/**
 * @record
 */
export function SpellingSuggestion() { }
if (false) {
    /** @type {?|undefined} */
    SpellingSuggestion.prototype.query;
    /** @type {?|undefined} */
    SpellingSuggestion.prototype.suggestion;
}
/**
 * @record
 */
export function ProductSearchPage() { }
if (false) {
    /** @type {?|undefined} */
    ProductSearchPage.prototype.breadcrumbs;
    /** @type {?|undefined} */
    ProductSearchPage.prototype.categoryCode;
    /** @type {?|undefined} */
    ProductSearchPage.prototype.currentQuery;
    /** @type {?|undefined} */
    ProductSearchPage.prototype.facets;
    /** @type {?|undefined} */
    ProductSearchPage.prototype.freeTextSearch;
    /** @type {?|undefined} */
    ProductSearchPage.prototype.keywordRedirectUrl;
    /** @type {?|undefined} */
    ProductSearchPage.prototype.pagination;
    /** @type {?|undefined} */
    ProductSearchPage.prototype.products;
    /** @type {?|undefined} */
    ProductSearchPage.prototype.sorts;
    /** @type {?|undefined} */
    ProductSearchPage.prototype.spellingSuggestion;
}
/**
 * @record
 */
export function Suggestion() { }
if (false) {
    /** @type {?|undefined} */
    Suggestion.prototype.value;
}
/**
 * @record
 */
export function ClearSearch() { }
if (false) {
    /** @type {?|undefined} */
    ClearSearch.prototype.clearPageResults;
    /** @type {?|undefined} */
    ClearSearch.prototype.clearSearchboxResults;
}
