// @spartacus/core/esm2015/src/occ/interceptor/with-credentials.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { OccConfig } from '../config/occ-config';
import * as i0 from "@angular/core";
import * as i1 from "../config/occ-config";
/**
 * Http interceptor to add cookies to all cross-site requests.
 */
export class WithCredentialsInterceptor {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
    }
    /**
     * Intercepts each request and adds the `withCredential` flag to it
     * if it hasn't been added already.
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    intercept(request, next) {
        if (this.requiresWithCredentials(request)) {
            request = request.clone({
                withCredentials: true,
            });
        }
        return next.handle(request);
    }
    /**
     * Indicates whether this is an OCC request.
     * @protected
     * @param {?} request
     * @return {?}
     */
    requiresWithCredentials(request) {
        return (
        /**
         * This `occConfig?.useWithCredentials` flag has only been introduced in
         * version 2.0. Customers must provide this interceptors manually in their
         * app to opt-in to use this feature, as we don't backport the configuration.
         */
        // this.occConfig?.useWithCredentials &&
        request.url.indexOf(this.occConfig.prefix) > -1);
    }
    /**
     * @private
     * @return {?}
     */
    get occConfig() {
        return this.config && this.config.backend && this.config.backend.occ
            ? this.config.backend.occ
            : {};
    }
}
WithCredentialsInterceptor.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
WithCredentialsInterceptor.ctorParameters = () => [
    { type: OccConfig }
];
/** @nocollapse */ WithCredentialsInterceptor.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function WithCredentialsInterceptor_Factory() { return new WithCredentialsInterceptor(i0.ɵɵinject(i1.OccConfig)); }, token: WithCredentialsInterceptor, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    WithCredentialsInterceptor.prototype.config;
}
