// @spartacus/core/esm2015/src/checkout/store/checkout-state.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const CHECKOUT_FEATURE = 'checkout';
/** @type {?} */
export const CHECKOUT_DETAILS = '[Checkout] Checkout Details';
/** @type {?} */
export const SET_DELIVERY_ADDRESS_PROCESS_ID = 'setDeliveryAddress';
/** @type {?} */
export const SET_DELIVERY_MODE_PROCESS_ID = 'setDeliveryMode';
/** @type {?} */
export const SET_SUPPORTED_DELIVERY_MODE_PROCESS_ID = 'setSupportedDeliveryMode';
/** @type {?} */
export const SET_PAYMENT_DETAILS_PROCESS_ID = 'setPaymentDetails';
/**
 * @record
 */
export function StateWithCheckout() { }
if (false) {
    /* Skipping unnamed member:
    [CHECKOUT_FEATURE]: CheckoutState;*/
}
/**
 * @record
 */
export function AddressVerificationState() { }
if (false) {
    /** @type {?} */
    AddressVerificationState.prototype.results;
}
/**
 * @record
 */
export function CardTypesState() { }
if (false) {
    /** @type {?} */
    CardTypesState.prototype.entities;
}
/**
 * @record
 */
export function CheckoutStepsState() { }
if (false) {
    /** @type {?} */
    CheckoutStepsState.prototype.address;
    /** @type {?} */
    CheckoutStepsState.prototype.deliveryMode;
    /** @type {?} */
    CheckoutStepsState.prototype.paymentDetails;
    /** @type {?} */
    CheckoutStepsState.prototype.orderDetails;
}
/**
 * @record
 */
export function CheckoutState() { }
if (false) {
    /** @type {?} */
    CheckoutState.prototype.steps;
    /** @type {?} */
    CheckoutState.prototype.cardTypes;
    /** @type {?} */
    CheckoutState.prototype.addressVerification;
}
