// @spartacus/core/esm5/src/cms/store/cms-state.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var CMS_FEATURE = 'cms';
/** @type {?} */
export var NAVIGATION_DETAIL_ENTITY = '[Cms] Navigation Entity';
// TODO(issue:6027) - fix the const value to `[Cms] Component Entity`
/** @type {?} */
export var COMPONENT_ENTITY = '[Cms[ Component Entity';
/**
 * @record
 */
export function StateWithCms() { }
if (false) {
    /* Skipping unnamed member:
    [CMS_FEATURE]: CmsState;*/
}
/**
 * @record
 */
export function ComponentsContext() { }
if (false) {
    /** @type {?} */
    ComponentsContext.prototype.component;
    /**
     * Page context stores an information for which context does the component exist.
     * For example, if `SiteLogoComponent` was successfully loaded for a product page with an ID of 1776948, then this object will contain:
     *
     * ```ts
     * ProductPage-1776948: {
     *  success: true,
     *  loading: false,
     *  error: false,
     *  // The `value` property indicates that the component exists for the given page context.
     *  value: true
     * }
     * ```
     *
     * If the same `SiteLogoComponent` component was tried to be loaded on homepage (page context id is `homepage`),
     * and it doesn't exist for some reason (maybe it has a restriction), then this object will contain:
     *
     * ```ts
     * ProductPage-1776948: {
     *  success: true,
     *  loading: false,
     *  error: false,
     *  // The `value` property indicates that the component exists for the given page context.
     *  value: true
     * },
     * ContentPage-homepage: {
     *  success: true,
     *  loading: false,
     *  error: false,
     *  // The `value` in this case is `false` indicating that the component was tried to be loaded, but it doesn't exist or has a restriction.
     *  value: false
     * }
     * ```
     *
     * @type {?}
     */
    ComponentsContext.prototype.pageContext;
}
/**
 * @record
 */
export function NavigationNodes() { }
/**
 * @record
 */
export function PageState() { }
if (false) {
    /** @type {?} */
    PageState.prototype.pageData;
    /** @type {?} */
    PageState.prototype.index;
}
/**
 * @record
 */
export function CmsState() { }
if (false) {
    /** @type {?} */
    CmsState.prototype.page;
    /** @type {?} */
    CmsState.prototype.component;
    /** @type {?} */
    CmsState.prototype.components;
    /** @type {?} */
    CmsState.prototype.navigation;
}
