// @spartacus/core/esm5/src/occ/occ-models/occ.models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export var Occ;
(function (Occ) {
    /**
     *
     * An interface representing Country.
     * @record
     */
    function Country() { }
    Occ.Country = Country;
    if (false) {
        /**
         * \@member {string} [isocode]
         * @type {?|undefined}
         */
        Country.prototype.isocode;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Country.prototype.name;
    }
    /**
     *
     * An interface representing Region.
     * @record
     */
    function Region() { }
    Occ.Region = Region;
    if (false) {
        /**
         * \@member {string} [countryIso]
         * @type {?|undefined}
         */
        Region.prototype.countryIso;
        /**
         * \@member {string} [isocode]
         * @type {?|undefined}
         */
        Region.prototype.isocode;
        /**
         * \@member {string} [isocodeShort]
         * @type {?|undefined}
         */
        Region.prototype.isocodeShort;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Region.prototype.name;
    }
    /**
     *
     * An interface representing RegionList.
     * @record
     */
    function RegionList() { }
    Occ.RegionList = RegionList;
    if (false) {
        /**
         * \@member {Region[]} [regions]
         * @type {?|undefined}
         */
        RegionList.prototype.regions;
    }
    /**
     *
     * An interface representing Address.
     * @record
     */
    function Address() { }
    Occ.Address = Address;
    if (false) {
        /**
         * \@member {string} [companyName]
         * @type {?|undefined}
         */
        Address.prototype.companyName;
        /**
         * \@member {Country} [country]
         * @type {?|undefined}
         */
        Address.prototype.country;
        /**
         * \@member {boolean} [defaultAddress]
         * @type {?|undefined}
         */
        Address.prototype.defaultAddress;
        /**
         * \@member {string} [email]
         * @type {?|undefined}
         */
        Address.prototype.email;
        /**
         * \@member {string} [firstName]
         * @type {?|undefined}
         */
        Address.prototype.firstName;
        /**
         * \@member {string} [formattedAddress]
         * @type {?|undefined}
         */
        Address.prototype.formattedAddress;
        /**
         * \@member {string} [id]
         * @type {?|undefined}
         */
        Address.prototype.id;
        /**
         * \@member {string} [lastName]
         * @type {?|undefined}
         */
        Address.prototype.lastName;
        /**
         * \@member {string} [line1]
         * @type {?|undefined}
         */
        Address.prototype.line1;
        /**
         * \@member {string} [line2]
         * @type {?|undefined}
         */
        Address.prototype.line2;
        /**
         * \@member {string} [phone]
         * @type {?|undefined}
         */
        Address.prototype.phone;
        /**
         * \@member {string} [postalCode]
         * @type {?|undefined}
         */
        Address.prototype.postalCode;
        /**
         * \@member {Region} [region]
         * @type {?|undefined}
         */
        Address.prototype.region;
        /**
         * \@member {boolean} [shippingAddress]
         * @type {?|undefined}
         */
        Address.prototype.shippingAddress;
        /**
         * \@member {string} [title]
         * @type {?|undefined}
         */
        Address.prototype.title;
        /**
         * \@member {string} [titleCode]
         * @type {?|undefined}
         */
        Address.prototype.titleCode;
        /**
         * \@member {string} [town]
         * @type {?|undefined}
         */
        Address.prototype.town;
        /**
         * \@member {boolean} [visibleInAddressBook]
         * @type {?|undefined}
         */
        Address.prototype.visibleInAddressBook;
    }
    /**
     *
     * An interface representing AddressList.
     * @record
     */
    function AddressList() { }
    Occ.AddressList = AddressList;
    if (false) {
        /**
         * \@member {Address[]} [addresses]
         * @type {?|undefined}
         */
        AddressList.prototype.addresses;
    }
    /**
     *
     * An interface representing ErrorModel.
     * Error message
     *
     * @record
     */
    function ErrorModel() { }
    Occ.ErrorModel = ErrorModel;
    if (false) {
        /**
         * \@member {string} [message] Descriptive, human readable error message.
         * @type {?|undefined}
         */
        ErrorModel.prototype.message;
        /**
         * \@member {string} [reason] Additional classification specific for each
         * error type e.g. 'noStock'.
         * @type {?|undefined}
         */
        ErrorModel.prototype.reason;
        /**
         * \@member {string} [subject] Identifier of the related object e.g. '1'.
         * @type {?|undefined}
         */
        ErrorModel.prototype.subject;
        /**
         * \@member {string} [subjectType] Type of the object related to the error
         * e.g. 'entry'.
         * @type {?|undefined}
         */
        ErrorModel.prototype.subjectType;
        /**
         * \@member {string} [type] Type of the error e.g. 'LowStockError'.
         * @type {?|undefined}
         */
        ErrorModel.prototype.type;
    }
    /**
     *
     * An interface representing ErrorList.
     * List of errors
     *
     * @record
     */
    function ErrorList() { }
    Occ.ErrorList = ErrorList;
    if (false) {
        /**
         * \@member {ErrorModel[]} [errors]
         * @type {?|undefined}
         */
        ErrorList.prototype.errors;
    }
    /**
     *
     * An interface representing AddressValidation.
     * @record
     */
    function AddressValidation() { }
    Occ.AddressValidation = AddressValidation;
    if (false) {
        /**
         * \@member {string} [decision]
         * @type {?|undefined}
         */
        AddressValidation.prototype.decision;
        /**
         * \@member {ErrorList} [errors]
         * @type {?|undefined}
         */
        AddressValidation.prototype.errors;
        /**
         * \@member {Address[]} [suggestedAddresses]
         * @type {?|undefined}
         */
        AddressValidation.prototype.suggestedAddresses;
    }
    /**
     *
     * An interface representing Price.
     * @record
     */
    function Price() { }
    Occ.Price = Price;
    if (false) {
        /**
         * \@member {string} [currencyIso]
         * @type {?|undefined}
         */
        Price.prototype.currencyIso;
        /**
         * \@member {string} [formattedValue]
         * @type {?|undefined}
         */
        Price.prototype.formattedValue;
        /**
         * \@member {number} [maxQuantity]
         * @type {?|undefined}
         */
        Price.prototype.maxQuantity;
        /**
         * \@member {number} [minQuantity]
         * @type {?|undefined}
         */
        Price.prototype.minQuantity;
        /**
         * \@member {PriceType} [priceType] Possible values include: 'BUY', 'FROM'
         * @type {?|undefined}
         */
        Price.prototype.priceType;
        /**
         * \@member {number} [value]
         * @type {?|undefined}
         */
        Price.prototype.value;
    }
    /**
     *
     * An interface representing Stock.
     * @record
     */
    function Stock() { }
    Occ.Stock = Stock;
    if (false) {
        /**
         * \@member {number} [stockLevel]
         * @type {?|undefined}
         */
        Stock.prototype.stockLevel;
        /**
         * \@member {string} [stockLevelStatus]
         * @type {?|undefined}
         */
        Stock.prototype.stockLevelStatus;
    }
    /**
     *
     * An interface representing Image.
     * @record
     */
    function Image() { }
    Occ.Image = Image;
    if (false) {
        /**
         * \@member {string} [altText]
         * @type {?|undefined}
         */
        Image.prototype.altText;
        /**
         * \@member {string} [format]
         * @type {?|undefined}
         */
        Image.prototype.format;
        /**
         * \@member {number} [galleryIndex]
         * @type {?|undefined}
         */
        Image.prototype.galleryIndex;
        /**
         * \@member {ImageType} [imageType] Possible values include: 'PRIMARY',
         * 'GALLERY'
         * @type {?|undefined}
         */
        Image.prototype.imageType;
        /**
         * \@member {string} [url]
         * @type {?|undefined}
         */
        Image.prototype.url;
    }
    /**
     *
     * An interface representing VariantOptionQualifier.
     * @record
     */
    function VariantOptionQualifier() { }
    Occ.VariantOptionQualifier = VariantOptionQualifier;
    if (false) {
        /**
         * \@member {Image} [image]
         * @type {?|undefined}
         */
        VariantOptionQualifier.prototype.image;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        VariantOptionQualifier.prototype.name;
        /**
         * \@member {string} [qualifier]
         * @type {?|undefined}
         */
        VariantOptionQualifier.prototype.qualifier;
        /**
         * \@member {string} [value]
         * @type {?|undefined}
         */
        VariantOptionQualifier.prototype.value;
    }
    /**
     *
     * An interface representing VariantOption.
     * @record
     */
    function VariantOption() { }
    Occ.VariantOption = VariantOption;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        VariantOption.prototype.code;
        /**
         * \@member {Price} [priceData]
         * @type {?|undefined}
         */
        VariantOption.prototype.priceData;
        /**
         * \@member {Stock} [stock]
         * @type {?|undefined}
         */
        VariantOption.prototype.stock;
        /**
         * \@member {string} [url]
         * @type {?|undefined}
         */
        VariantOption.prototype.url;
        /**
         * \@member {VariantOptionQualifier[]} [variantOptionQualifiers]
         * @type {?|undefined}
         */
        VariantOption.prototype.variantOptionQualifiers;
    }
    /**
     *
     * An interface representing BaseOption.
     * @record
     */
    function BaseOption() { }
    Occ.BaseOption = BaseOption;
    if (false) {
        /**
         * \@member {VariantOption[]} [options]
         * @type {?|undefined}
         */
        BaseOption.prototype.options;
        /**
         * \@member {VariantOption} [selected]
         * @type {?|undefined}
         */
        BaseOption.prototype.selected;
        /**
         * \@member {string} [variantType]
         * @type {?|undefined}
         */
        BaseOption.prototype.variantType;
    }
    /**
     *
     * An interface representing SearchQuery.
     * @record
     */
    function SearchQuery() { }
    Occ.SearchQuery = SearchQuery;
    if (false) {
        /**
         * \@member {string} [value]
         * @type {?|undefined}
         */
        SearchQuery.prototype.value;
    }
    /**
     *
     * An interface representing SearchState.
     * @record
     */
    function SearchState() { }
    Occ.SearchState = SearchState;
    if (false) {
        /**
         * \@member {SearchQuery} [query]
         * @type {?|undefined}
         */
        SearchState.prototype.query;
        /**
         * \@member {string} [url]
         * @type {?|undefined}
         */
        SearchState.prototype.url;
    }
    /**
     *
     * An interface representing Breadcrumb.
     * @record
     */
    function Breadcrumb() { }
    Occ.Breadcrumb = Breadcrumb;
    if (false) {
        /**
         * \@member {string} [facetCode]
         * @type {?|undefined}
         */
        Breadcrumb.prototype.facetCode;
        /**
         * \@member {string} [facetName]
         * @type {?|undefined}
         */
        Breadcrumb.prototype.facetName;
        /**
         * \@member {string} [facetValueCode]
         * @type {?|undefined}
         */
        Breadcrumb.prototype.facetValueCode;
        /**
         * \@member {string} [facetValueName]
         * @type {?|undefined}
         */
        Breadcrumb.prototype.facetValueName;
        /**
         * \@member {SearchState} [removeQuery]
         * @type {?|undefined}
         */
        Breadcrumb.prototype.removeQuery;
        /**
         * \@member {SearchState} [truncateQuery]
         * @type {?|undefined}
         */
        Breadcrumb.prototype.truncateQuery;
    }
    /**
     *
     * An interface representing Component.
     * @record
     */
    function Component() { }
    Occ.Component = Component;
    if (false) {
        /**
         * \@member {Date} [modifiedTime]
         * @type {?|undefined}
         */
        Component.prototype.modifiedTime;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Component.prototype.name;
        /**
         * \@member {any} [otherProperties]
         * @type {?|undefined}
         */
        Component.prototype.otherProperties;
        /**
         * \@member {string} [typeCode]
         * @type {?|undefined}
         */
        Component.prototype.typeCode;
        /**
         * \@member {string} [uid]
         * @type {?|undefined}
         */
        Component.prototype.uid;
    }
    /**
     *
     * An interface representing ComponentList.
     * @record
     */
    function ComponentList() { }
    Occ.ComponentList = ComponentList;
    if (false) {
        /**
         * \@member {Component[]} [component]
         * @type {?|undefined}
         */
        ComponentList.prototype.component;
    }
    /**
     *
     * An interface representing ContentSlot.
     * @record
     */
    function ContentSlot() { }
    Occ.ContentSlot = ContentSlot;
    if (false) {
        /**
         * \@member {ComponentList} [components]
         * @type {?|undefined}
         */
        ContentSlot.prototype.components;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        ContentSlot.prototype.name;
        /**
         * \@member {string} [position]
         * @type {?|undefined}
         */
        ContentSlot.prototype.position;
        /**
         * \@member {string} [slotId]
         * @type {?|undefined}
         */
        ContentSlot.prototype.slotId;
        /**
         * \@member {boolean} [slotShared]
         * @type {?|undefined}
         */
        ContentSlot.prototype.slotShared;
        /**
         * \@member {string} [slotStatus]
         * @type {?|undefined}
         */
        ContentSlot.prototype.slotStatus;
    }
    /**
     *
     * An interface representing ContentSlotList.
     * @record
     */
    function ContentSlotList() { }
    Occ.ContentSlotList = ContentSlotList;
    if (false) {
        /**
         * \@member {ContentSlot[]} [contentSlot]
         * @type {?|undefined}
         */
        ContentSlotList.prototype.contentSlot;
    }
    /**
     *
     * An interface representing CMSPage.
     * @record
     */
    function CMSPage() { }
    Occ.CMSPage = CMSPage;
    if (false) {
        /**
         * \@member {ContentSlotList} [contentSlots]
         * @type {?|undefined}
         */
        CMSPage.prototype.contentSlots;
        /**
         * \@member {boolean} [defaultPage]
         * @type {?|undefined}
         */
        CMSPage.prototype.defaultPage;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        CMSPage.prototype.name;
        /**
         * \@member {string} [template]
         * @type {?|undefined}
         */
        CMSPage.prototype.template;
        /**
         * \@member {string} [title]
         * @type {?|undefined}
         */
        CMSPage.prototype.title;
        /**
         * \@member {string} [typeCode]
         * @type {?|undefined}
         */
        CMSPage.prototype.typeCode;
        /**
         * \@member {string} [uid]
         * @type {?|undefined}
         */
        CMSPage.prototype.uid;
    }
    /**
     *
     * An interface representing CardType.
     * @record
     */
    function CardType() { }
    Occ.CardType = CardType;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        CardType.prototype.code;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        CardType.prototype.name;
    }
    /**
     *
     * An interface representing CardTypeList.
     * @record
     */
    function CardTypeList() { }
    Occ.CardTypeList = CardTypeList;
    if (false) {
        /**
         * \@member {CardType[]} [cardTypes]
         * @type {?|undefined}
         */
        CardTypeList.prototype.cardTypes;
    }
    /**
     *
     * An interface representing PromotionOrderEntryConsumed.
     * @record
     */
    function PromotionOrderEntryConsumed() { }
    Occ.PromotionOrderEntryConsumed = PromotionOrderEntryConsumed;
    if (false) {
        /**
         * \@member {number} [adjustedUnitPrice]
         * @type {?|undefined}
         */
        PromotionOrderEntryConsumed.prototype.adjustedUnitPrice;
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        PromotionOrderEntryConsumed.prototype.code;
        /**
         * \@member {number} [orderEntryNumber]
         * @type {?|undefined}
         */
        PromotionOrderEntryConsumed.prototype.orderEntryNumber;
        /**
         * \@member {number} [quantity]
         * @type {?|undefined}
         */
        PromotionOrderEntryConsumed.prototype.quantity;
    }
    /**
     *
     * An interface representing PromotionRestriction.
     * @record
     */
    function PromotionRestriction() { }
    Occ.PromotionRestriction = PromotionRestriction;
    if (false) {
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        PromotionRestriction.prototype.description;
        /**
         * \@member {string} [restrictionType]
         * @type {?|undefined}
         */
        PromotionRestriction.prototype.restrictionType;
    }
    /**
     *
     * An interface representing Promotion.
     * @record
     */
    function Promotion() { }
    Occ.Promotion = Promotion;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Promotion.prototype.code;
        /**
         * \@member {string[]} [couldFireMessages]
         * @type {?|undefined}
         */
        Promotion.prototype.couldFireMessages;
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        Promotion.prototype.description;
        /**
         * \@member {boolean} [enabled]
         * @type {?|undefined}
         */
        Promotion.prototype.enabled;
        /**
         * \@member {Date} [endDate]
         * @type {?|undefined}
         */
        Promotion.prototype.endDate;
        /**
         * \@member {string[]} [firedMessages]
         * @type {?|undefined}
         */
        Promotion.prototype.firedMessages;
        /**
         * \@member {number} [priority]
         * @type {?|undefined}
         */
        Promotion.prototype.priority;
        /**
         * \@member {Image} [productBanner]
         * @type {?|undefined}
         */
        Promotion.prototype.productBanner;
        /**
         * \@member {string} [promotionGroup]
         * @type {?|undefined}
         */
        Promotion.prototype.promotionGroup;
        /**
         * \@member {string} [promotionType]
         * @type {?|undefined}
         */
        Promotion.prototype.promotionType;
        /**
         * \@member {PromotionRestriction[]} [restrictions]
         * @type {?|undefined}
         */
        Promotion.prototype.restrictions;
        /**
         * \@member {Date} [startDate]
         * @type {?|undefined}
         */
        Promotion.prototype.startDate;
        /**
         * \@member {string} [title]
         * @type {?|undefined}
         */
        Promotion.prototype.title;
    }
    /**
     *
     * An interface representing PromotionResult.
     * @record
     */
    function PromotionResult() { }
    Occ.PromotionResult = PromotionResult;
    if (false) {
        /**
         * \@member {PromotionOrderEntryConsumed[]} [consumedEntries]
         * @type {?|undefined}
         */
        PromotionResult.prototype.consumedEntries;
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        PromotionResult.prototype.description;
        /**
         * \@member {Promotion} [promotion]
         * @type {?|undefined}
         */
        PromotionResult.prototype.promotion;
    }
    /**
     *
     * An interface representing Currency.
     * @record
     */
    function Currency() { }
    Occ.Currency = Currency;
    if (false) {
        /**
         * \@member {boolean} [active]
         * @type {?|undefined}
         */
        Currency.prototype.active;
        /**
         * \@member {string} [isocode]
         * @type {?|undefined}
         */
        Currency.prototype.isocode;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Currency.prototype.name;
        /**
         * \@member {string} [symbol]
         * @type {?|undefined}
         */
        Currency.prototype.symbol;
    }
    /**
     *
     * An interface representing Voucher.
     * @record
     */
    function Voucher() { }
    Occ.Voucher = Voucher;
    if (false) {
        /**
         * \@member {Price} [appliedValue]
         * @type {?|undefined}
         */
        Voucher.prototype.appliedValue;
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Voucher.prototype.code;
        /**
         * \@member {Currency} [currency]
         * @type {?|undefined}
         */
        Voucher.prototype.currency;
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        Voucher.prototype.description;
        /**
         * \@member {boolean} [freeShipping]
         * @type {?|undefined}
         */
        Voucher.prototype.freeShipping;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Voucher.prototype.name;
        /**
         * \@member {number} [value]
         * @type {?|undefined}
         */
        Voucher.prototype.value;
        /**
         * \@member {string} [valueFormatted]
         * @type {?|undefined}
         */
        Voucher.prototype.valueFormatted;
        /**
         * \@member {string} [valueString]
         * @type {?|undefined}
         */
        Voucher.prototype.valueString;
        /**
         * \@member {string} [voucherCode]
         * @type {?|undefined}
         */
        Voucher.prototype.voucherCode;
    }
    /**
     *
     * An interface representing DeliveryMode.
     * @record
     */
    function DeliveryMode() { }
    Occ.DeliveryMode = DeliveryMode;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        DeliveryMode.prototype.code;
        /**
         * \@member {Price} [deliveryCost]
         * @type {?|undefined}
         */
        DeliveryMode.prototype.deliveryCost;
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        DeliveryMode.prototype.description;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        DeliveryMode.prototype.name;
    }
    /**
     *
     * An interface representing GeoPoint.
     * @record
     */
    function GeoPoint() { }
    Occ.GeoPoint = GeoPoint;
    if (false) {
        /**
         * \@member {number} [latitude]
         * @type {?|undefined}
         */
        GeoPoint.prototype.latitude;
        /**
         * \@member {number} [longitude]
         * @type {?|undefined}
         */
        GeoPoint.prototype.longitude;
    }
    /**
     *
     * An interface representing Time.
     * @record
     */
    function Time() { }
    Occ.Time = Time;
    if (false) {
        /**
         * \@member {string} [formattedHour]
         * @type {?|undefined}
         */
        Time.prototype.formattedHour;
        /**
         * \@member {number} [hour]
         * @type {?|undefined}
         */
        Time.prototype.hour;
        /**
         * \@member {number} [minute]
         * @type {?|undefined}
         */
        Time.prototype.minute;
    }
    /**
     *
     * An interface representing SpecialOpeningDay.
     * @record
     */
    function SpecialOpeningDay() { }
    Occ.SpecialOpeningDay = SpecialOpeningDay;
    if (false) {
        /**
         * \@member {boolean} [closed]
         * @type {?|undefined}
         */
        SpecialOpeningDay.prototype.closed;
        /**
         * \@member {Time} [closingTime]
         * @type {?|undefined}
         */
        SpecialOpeningDay.prototype.closingTime;
        /**
         * \@member {string} [comment]
         * @type {?|undefined}
         */
        SpecialOpeningDay.prototype.comment;
        /**
         * \@member {Date} [date]
         * @type {?|undefined}
         */
        SpecialOpeningDay.prototype.date;
        /**
         * \@member {string} [formattedDate]
         * @type {?|undefined}
         */
        SpecialOpeningDay.prototype.formattedDate;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        SpecialOpeningDay.prototype.name;
        /**
         * \@member {Time} [openingTime]
         * @type {?|undefined}
         */
        SpecialOpeningDay.prototype.openingTime;
    }
    /**
     *
     * An interface representing WeekdayOpeningDay.
     * @record
     */
    function WeekdayOpeningDay() { }
    Occ.WeekdayOpeningDay = WeekdayOpeningDay;
    if (false) {
        /**
         * \@member {boolean} [closed]
         * @type {?|undefined}
         */
        WeekdayOpeningDay.prototype.closed;
        /**
         * \@member {Time} [closingTime]
         * @type {?|undefined}
         */
        WeekdayOpeningDay.prototype.closingTime;
        /**
         * \@member {Time} [openingTime]
         * @type {?|undefined}
         */
        WeekdayOpeningDay.prototype.openingTime;
        /**
         * \@member {string} [weekDay]
         * @type {?|undefined}
         */
        WeekdayOpeningDay.prototype.weekDay;
    }
    /**
     *
     * An interface representing OpeningSchedule.
     * @record
     */
    function OpeningSchedule() { }
    Occ.OpeningSchedule = OpeningSchedule;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        OpeningSchedule.prototype.code;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        OpeningSchedule.prototype.name;
        /**
         * \@member {SpecialOpeningDay[]} [specialDayOpeningList]
         * @type {?|undefined}
         */
        OpeningSchedule.prototype.specialDayOpeningList;
        /**
         * \@member {WeekdayOpeningDay[]} [weekDayOpeningList]
         * @type {?|undefined}
         */
        OpeningSchedule.prototype.weekDayOpeningList;
    }
    /**
     *
     * An interface representing PointOfService.
     * @record
     */
    function PointOfService() { }
    Occ.PointOfService = PointOfService;
    if (false) {
        /**
         * \@member {Address} [address]
         * @type {?|undefined}
         */
        PointOfService.prototype.address;
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        PointOfService.prototype.description;
        /**
         * \@member {string} [displayName]
         * @type {?|undefined}
         */
        PointOfService.prototype.displayName;
        /**
         * \@member {number} [distanceKm]
         * @type {?|undefined}
         */
        PointOfService.prototype.distanceKm;
        /**
         * \@member {{ [propertyName: string]: string }} [features]
         * @type {?|undefined}
         */
        PointOfService.prototype.features;
        /**
         * \@member {string} [formattedDistance]
         * @type {?|undefined}
         */
        PointOfService.prototype.formattedDistance;
        /**
         * \@member {GeoPoint} [geoPoint]
         * @type {?|undefined}
         */
        PointOfService.prototype.geoPoint;
        /**
         * \@member {Image} [mapIcon]
         * @type {?|undefined}
         */
        PointOfService.prototype.mapIcon;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        PointOfService.prototype.name;
        /**
         * \@member {OpeningSchedule} [openingHours]
         * @type {?|undefined}
         */
        PointOfService.prototype.openingHours;
        /**
         * \@member {string} [storeContent]
         * @type {?|undefined}
         */
        PointOfService.prototype.storeContent;
        /**
         * \@member {Image[]} [storeImages]
         * @type {?|undefined}
         */
        PointOfService.prototype.storeImages;
        /**
         * \@member {string} [url]
         * @type {?|undefined}
         */
        PointOfService.prototype.url;
    }
    /**
     *
     * An interface representing Category.
     * @record
     */
    function Category() { }
    Occ.Category = Category;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Category.prototype.code;
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Category.prototype.name;
        /**
         * \@member {Image} [image]
         * @type {?|undefined}
         */
        Category.prototype.image;
        /**
         * \@member {string} [url]
         * @type {?|undefined}
         */
        Category.prototype.url;
    }
    /**
     *
     * An interface representing FeatureUnit.
     * @record
     */
    function FeatureUnit() { }
    Occ.FeatureUnit = FeatureUnit;
    if (false) {
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        FeatureUnit.prototype.name;
        /**
         * \@member {string} [symbol]
         * @type {?|undefined}
         */
        FeatureUnit.prototype.symbol;
        /**
         * \@member {string} [unitType]
         * @type {?|undefined}
         */
        FeatureUnit.prototype.unitType;
    }
    /**
     *
     * An interface representing FeatureValue.
     * @record
     */
    function FeatureValue() { }
    Occ.FeatureValue = FeatureValue;
    if (false) {
        /**
         * \@member {string} [value]
         * @type {?|undefined}
         */
        FeatureValue.prototype.value;
    }
    /**
     *
     * An interface representing Feature.
     * @record
     */
    function Feature() { }
    Occ.Feature = Feature;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Feature.prototype.code;
        /**
         * \@member {boolean} [comparable]
         * @type {?|undefined}
         */
        Feature.prototype.comparable;
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        Feature.prototype.description;
        /**
         * \@member {FeatureUnit} [featureUnit]
         * @type {?|undefined}
         */
        Feature.prototype.featureUnit;
        /**
         * \@member {FeatureValue[]} [featureValues]
         * @type {?|undefined}
         */
        Feature.prototype.featureValues;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Feature.prototype.name;
        /**
         * \@member {boolean} [range]
         * @type {?|undefined}
         */
        Feature.prototype.range;
        /**
         * \@member {string} [type]
         * @type {?|undefined}
         */
        Feature.prototype.type;
    }
    /**
     *
     * An interface representing Classification.
     * @record
     */
    function Classification() { }
    Occ.Classification = Classification;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Classification.prototype.code;
        /**
         * \@member {Feature[]} [features]
         * @type {?|undefined}
         */
        Classification.prototype.features;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Classification.prototype.name;
    }
    /**
     *
     * An interface representing FutureStock.
     * @record
     */
    function FutureStock() { }
    Occ.FutureStock = FutureStock;
    if (false) {
        /**
         * \@member {Date} [date]
         * @type {?|undefined}
         */
        FutureStock.prototype.date;
        /**
         * \@member {string} [formattedDate]
         * @type {?|undefined}
         */
        FutureStock.prototype.formattedDate;
        /**
         * \@member {Stock} [stock]
         * @type {?|undefined}
         */
        FutureStock.prototype.stock;
    }
    /**
     *
     * An interface representing PriceRange.
     * @record
     */
    function PriceRange() { }
    Occ.PriceRange = PriceRange;
    if (false) {
        /**
         * \@member {Price} [maxPrice]
         * @type {?|undefined}
         */
        PriceRange.prototype.maxPrice;
        /**
         * \@member {Price} [minPrice]
         * @type {?|undefined}
         */
        PriceRange.prototype.minPrice;
    }
    /**
     *
     * An interface representing ProductReference.
     * @record
     */
    function ProductReference() { }
    Occ.ProductReference = ProductReference;
    if (false) {
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        ProductReference.prototype.description;
        /**
         * \@member {boolean} [preselected]
         * @type {?|undefined}
         */
        ProductReference.prototype.preselected;
        /**
         * \@member {number} [quantity]
         * @type {?|undefined}
         */
        ProductReference.prototype.quantity;
        /**
         * \@member {string} [referenceType]
         * @type {?|undefined}
         */
        ProductReference.prototype.referenceType;
        /**
         * \@member {Product} [target]
         * @type {?|undefined}
         */
        ProductReference.prototype.target;
    }
    /**
     *
     * An interface representing Language.
     * @record
     */
    function Language() { }
    Occ.Language = Language;
    if (false) {
        /**
         * \@member {boolean} [active]
         * @type {?|undefined}
         */
        Language.prototype.active;
        /**
         * \@member {string} [isocode]
         * @type {?|undefined}
         */
        Language.prototype.isocode;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Language.prototype.name;
        /**
         * \@member {string} [nativeName]
         * @type {?|undefined}
         */
        Language.prototype.nativeName;
    }
    /**
     *
     * An interface representing User.
     * @record
     */
    function User() { }
    Occ.User = User;
    if (false) {
        /**
         * \@member {Currency} [currency]
         * @type {?|undefined}
         */
        User.prototype.currency;
        /**
         * \@member {string} [customerId]
         * @type {?|undefined}
         */
        User.prototype.customerId;
        /**
         * \@member {Date} [deactivationDate]
         * @type {?|undefined}
         */
        User.prototype.deactivationDate;
        /**
         * \@member {Address} [defaultAddress]
         * @type {?|undefined}
         */
        User.prototype.defaultAddress;
        /**
         * \@member {string} [displayUid]
         * @type {?|undefined}
         */
        User.prototype.displayUid;
        /**
         * \@member {string} [firstName]
         * @type {?|undefined}
         */
        User.prototype.firstName;
        /**
         * \@member {Language} [language]
         * @type {?|undefined}
         */
        User.prototype.language;
        /**
         * \@member {string} [lastName]
         * @type {?|undefined}
         */
        User.prototype.lastName;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        User.prototype.name;
        /**
         * \@member {string} [title]
         * @type {?|undefined}
         */
        User.prototype.title;
        /**
         * \@member {string} [titleCode]
         * @type {?|undefined}
         */
        User.prototype.titleCode;
        /**
         * \@member {string} [uid]
         * @type {?|undefined}
         */
        User.prototype.uid;
    }
    /**
     *
     * An interface representing Review.
     * @record
     */
    function Review() { }
    Occ.Review = Review;
    if (false) {
        /**
         * \@member {string} [alias]
         * @type {?|undefined}
         */
        Review.prototype.alias;
        /**
         * \@member {string} [comment]
         * @type {?|undefined}
         */
        Review.prototype.comment;
        /**
         * \@member {Date} [date]
         * @type {?|undefined}
         */
        Review.prototype.date;
        /**
         * \@member {string} [headline]
         * @type {?|undefined}
         */
        Review.prototype.headline;
        /**
         * \@member {string} [id]
         * @type {?|undefined}
         */
        Review.prototype.id;
        /**
         * \@member {User} [principal]
         * @type {?|undefined}
         */
        Review.prototype.principal;
        /**
         * \@member {number} [rating]
         * @type {?|undefined}
         */
        Review.prototype.rating;
    }
    /**
     *
     * An interface representing VariantCategory.
     * @record
     */
    function VariantCategory() { }
    Occ.VariantCategory = VariantCategory;
    if (false) {
        /**
         * \@member {boolean} [hasImage]
         * @type {?|undefined}
         */
        VariantCategory.prototype.hasImage;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        VariantCategory.prototype.name;
        /**
         * \@member {number} [priority]
         * @type {?|undefined}
         */
        VariantCategory.prototype.priority;
    }
    /**
     *
     * An interface representing VariantValueCategory.
     * @record
     */
    function VariantValueCategory() { }
    Occ.VariantValueCategory = VariantValueCategory;
    if (false) {
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        VariantValueCategory.prototype.name;
        /**
         * \@member {number} [sequence]
         * @type {?|undefined}
         */
        VariantValueCategory.prototype.sequence;
        /**
         * \@member {VariantCategory[]} [superCategories]
         * @type {?|undefined}
         */
        VariantValueCategory.prototype.superCategories;
    }
    /**
     *
     * An interface representing VariantMatrixElement.
     * @record
     */
    function VariantMatrixElement() { }
    Occ.VariantMatrixElement = VariantMatrixElement;
    if (false) {
        /**
         * \@member {VariantMatrixElement[]} [elements]
         * @type {?|undefined}
         */
        VariantMatrixElement.prototype.elements;
        /**
         * \@member {boolean} [isLeaf]
         * @type {?|undefined}
         */
        VariantMatrixElement.prototype.isLeaf;
        /**
         * \@member {VariantCategory} [parentVariantCategory]
         * @type {?|undefined}
         */
        VariantMatrixElement.prototype.parentVariantCategory;
        /**
         * \@member {VariantOption} [variantOption]
         * @type {?|undefined}
         */
        VariantMatrixElement.prototype.variantOption;
        /**
         * \@member {VariantValueCategory} [variantValueCategory]
         * @type {?|undefined}
         */
        VariantMatrixElement.prototype.variantValueCategory;
    }
    /**
     *
     * An interface representing Product.
     * @record
     */
    function Product() { }
    Occ.Product = Product;
    if (false) {
        /**
         * \@member {boolean} [availableForPickup]
         * @type {?|undefined}
         */
        Product.prototype.availableForPickup;
        /**
         * \@member {number} [averageRating]
         * @type {?|undefined}
         */
        Product.prototype.averageRating;
        /**
         * \@member {BaseOption[]} [baseOptions]
         * @type {?|undefined}
         */
        Product.prototype.baseOptions;
        /**
         * \@member {string} [baseProduct]
         * @type {?|undefined}
         */
        Product.prototype.baseProduct;
        /**
         * \@member {Category[]} [categories]
         * @type {?|undefined}
         */
        Product.prototype.categories;
        /**
         * \@member {Classification[]} [classifications]
         * @type {?|undefined}
         */
        Product.prototype.classifications;
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Product.prototype.code;
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        Product.prototype.description;
        /**
         * \@member {FutureStock[]} [futureStocks]
         * @type {?|undefined}
         */
        Product.prototype.futureStocks;
        /**
         * \@member {Image[]} [images]
         * @type {?|undefined}
         */
        Product.prototype.images;
        /**
         * \@member {string} [manufacturer]
         * @type {?|undefined}
         */
        Product.prototype.manufacturer;
        /**
         * \@member {boolean} [multidimensional]
         * @type {?|undefined}
         */
        Product.prototype.multidimensional;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Product.prototype.name;
        /**
         * \@member {number} [numberOfReviews]
         * @type {?|undefined}
         */
        Product.prototype.numberOfReviews;
        /**
         * \@member {Promotion[]} [potentialPromotions]
         * @type {?|undefined}
         */
        Product.prototype.potentialPromotions;
        /**
         * \@member {Price} [price]
         * @type {?|undefined}
         */
        Product.prototype.price;
        /**
         * \@member {PriceRange} [priceRange]
         * @type {?|undefined}
         */
        Product.prototype.priceRange;
        /**
         * \@member {ProductReference[]} [productReferences]
         * @type {?|undefined}
         */
        Product.prototype.productReferences;
        /**
         * \@member {boolean} [purchasable]
         * @type {?|undefined}
         */
        Product.prototype.purchasable;
        /**
         * \@member {Review[]} [reviews]
         * @type {?|undefined}
         */
        Product.prototype.reviews;
        /**
         * \@member {Stock} [stock]
         * @type {?|undefined}
         */
        Product.prototype.stock;
        /**
         * \@member {string} [summary]
         * @type {?|undefined}
         */
        Product.prototype.summary;
        /**
         * \@member {string} [url]
         * @type {?|undefined}
         */
        Product.prototype.url;
        /**
         * \@member {VariantMatrixElement[]} [variantMatrix]
         * @type {?|undefined}
         */
        Product.prototype.variantMatrix;
        /**
         * \@member {VariantOption[]} [variantOptions]
         * @type {?|undefined}
         */
        Product.prototype.variantOptions;
        /**
         * \@member {string} [variantType]
         * @type {?|undefined}
         */
        Product.prototype.variantType;
        /**
         * \@member {Price[]} [volumePrices]
         * @type {?|undefined}
         */
        Product.prototype.volumePrices;
        /**
         * \@member {boolean} [volumePricesFlag]
         * @type {?|undefined}
         */
        Product.prototype.volumePricesFlag;
    }
    /**
     *
     * An interface representing OrderEntry.
     * @record
     */
    function OrderEntry() { }
    Occ.OrderEntry = OrderEntry;
    if (false) {
        /**
         * \@member {Price} [basePrice]
         * @type {?|undefined}
         */
        OrderEntry.prototype.basePrice;
        /**
         * \@member {DeliveryMode} [deliveryMode]
         * @type {?|undefined}
         */
        OrderEntry.prototype.deliveryMode;
        /**
         * \@member {PointOfService} [deliveryPointOfService]
         * @type {?|undefined}
         */
        OrderEntry.prototype.deliveryPointOfService;
        /**
         * \@member {number} [entryNumber]
         * @type {?|undefined}
         */
        OrderEntry.prototype.entryNumber;
        /**
         * \@member {Product} [product]
         * @type {?|undefined}
         */
        OrderEntry.prototype.product;
        /**
         * \@member {number} [quantity]
         * @type {?|undefined}
         */
        OrderEntry.prototype.quantity;
        /**
         * \@member {Price} [totalPrice]
         * @type {?|undefined}
         */
        OrderEntry.prototype.totalPrice;
        /**
         * \@member {boolean} [updateable]
         * @type {?|undefined}
         */
        OrderEntry.prototype.updateable;
    }
    /**
     *
     * An interface representing DeliveryOrderEntryGroup.
     * @record
     */
    function DeliveryOrderEntryGroup() { }
    Occ.DeliveryOrderEntryGroup = DeliveryOrderEntryGroup;
    if (false) {
        /**
         * \@member {Address} [deliveryAddress]
         * @type {?|undefined}
         */
        DeliveryOrderEntryGroup.prototype.deliveryAddress;
        /**
         * \@member {OrderEntry[]} [entries]
         * @type {?|undefined}
         */
        DeliveryOrderEntryGroup.prototype.entries;
        /**
         * \@member {number} [quantity]
         * @type {?|undefined}
         */
        DeliveryOrderEntryGroup.prototype.quantity;
        /**
         * \@member {Price} [totalPriceWithTax]
         * @type {?|undefined}
         */
        DeliveryOrderEntryGroup.prototype.totalPriceWithTax;
    }
    /**
     *
     * An interface representing PaymentDetails.
     * @record
     */
    function PaymentDetails() { }
    Occ.PaymentDetails = PaymentDetails;
    if (false) {
        /**
         * \@member {string} [accountHolderName]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.accountHolderName;
        /**
         * \@member {Address} [billingAddress]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.billingAddress;
        /**
         * \@member {string} [cardNumber]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.cardNumber;
        /**
         * \@member {CardType} [cardType]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.cardType;
        /**
         * \@member {string} [cvn]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.cvn;
        /**
         * \@member {boolean} [defaultPayment]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.defaultPayment;
        /**
         * \@member {string} [expiryMonth]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.expiryMonth;
        /**
         * \@member {string} [expiryYear]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.expiryYear;
        /**
         * \@member {string} [id]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.id;
        /**
         * \@member {string} [issueNumber]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.issueNumber;
        /**
         * \@member {boolean} [saved]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.saved;
        /**
         * \@member {string} [startMonth]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.startMonth;
        /**
         * \@member {string} [startYear]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.startYear;
        /**
         * \@member {string} [subscriptionId]
         * @type {?|undefined}
         */
        PaymentDetails.prototype.subscriptionId;
    }
    /**
     *
     * An interface representing PickupOrderEntryGroup.
     * @record
     */
    function PickupOrderEntryGroup() { }
    Occ.PickupOrderEntryGroup = PickupOrderEntryGroup;
    if (false) {
        /**
         * \@member {PointOfService} [deliveryPointOfService]
         * @type {?|undefined}
         */
        PickupOrderEntryGroup.prototype.deliveryPointOfService;
        /**
         * \@member {number} [distance]
         * @type {?|undefined}
         */
        PickupOrderEntryGroup.prototype.distance;
        /**
         * \@member {OrderEntry[]} [entries]
         * @type {?|undefined}
         */
        PickupOrderEntryGroup.prototype.entries;
        /**
         * \@member {number} [quantity]
         * @type {?|undefined}
         */
        PickupOrderEntryGroup.prototype.quantity;
        /**
         * \@member {Price} [totalPriceWithTax]
         * @type {?|undefined}
         */
        PickupOrderEntryGroup.prototype.totalPriceWithTax;
    }
    /**
     *
     * An interface representing Principal.
     * @record
     */
    function Principal() { }
    Occ.Principal = Principal;
    if (false) {
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Principal.prototype.name;
        /**
         * \@member {string} [uid]
         * @type {?|undefined}
         */
        Principal.prototype.uid;
    }
    /**
     *
     * An interface representing Cart.
     * @record
     */
    function Cart() { }
    Occ.Cart = Cart;
    if (false) {
        /**
         * \@member {PromotionResult[]} [appliedOrderPromotions]
         * @type {?|undefined}
         */
        Cart.prototype.appliedOrderPromotions;
        /**
         * \@member {PromotionResult[]} [appliedProductPromotions]
         * @type {?|undefined}
         */
        Cart.prototype.appliedProductPromotions;
        /**
         * \@member {Voucher[]} [appliedVouchers]
         * @type {?|undefined}
         */
        Cart.prototype.appliedVouchers;
        /**
         * \@member {boolean} [calculated]
         * @type {?|undefined}
         */
        Cart.prototype.calculated;
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Cart.prototype.code;
        /**
         * \@member {Address} [deliveryAddress]
         * @type {?|undefined}
         */
        Cart.prototype.deliveryAddress;
        /**
         * \@member {Price} [deliveryCost]
         * @type {?|undefined}
         */
        Cart.prototype.deliveryCost;
        /**
         * \@member {number} [deliveryItemsQuantity]
         * @type {?|undefined}
         */
        Cart.prototype.deliveryItemsQuantity;
        /**
         * \@member {DeliveryMode} [deliveryMode]
         * @type {?|undefined}
         */
        Cart.prototype.deliveryMode;
        /**
         * \@member {DeliveryOrderEntryGroup[]} [deliveryOrderGroups]
         * @type {?|undefined}
         */
        Cart.prototype.deliveryOrderGroups;
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        Cart.prototype.description;
        /**
         * \@member {OrderEntry[]} [entries]
         * @type {?|undefined}
         */
        Cart.prototype.entries;
        /**
         * \@member {Date} [expirationTime]
         * @type {?|undefined}
         */
        Cart.prototype.expirationTime;
        /**
         * \@member {string} [guid]
         * @type {?|undefined}
         */
        Cart.prototype.guid;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Cart.prototype.name;
        /**
         * \@member {boolean} [net]
         * @type {?|undefined}
         */
        Cart.prototype.net;
        /**
         * \@member {Price} [orderDiscounts]
         * @type {?|undefined}
         */
        Cart.prototype.orderDiscounts;
        /**
         * \@member {PaymentDetails} [paymentInfo]
         * @type {?|undefined}
         */
        Cart.prototype.paymentInfo;
        /**
         * \@member {number} [pickupItemsQuantity]
         * @type {?|undefined}
         */
        Cart.prototype.pickupItemsQuantity;
        /**
         * \@member {PickupOrderEntryGroup[]} [pickupOrderGroups]
         * @type {?|undefined}
         */
        Cart.prototype.pickupOrderGroups;
        /**
         * \@member {PromotionResult[]} [potentialOrderPromotions]
         * @type {?|undefined}
         */
        Cart.prototype.potentialOrderPromotions;
        /**
         * \@member {PromotionResult[]} [potentialProductPromotions]
         * @type {?|undefined}
         */
        Cart.prototype.potentialProductPromotions;
        /**
         * \@member {Price} [productDiscounts]
         * @type {?|undefined}
         */
        Cart.prototype.productDiscounts;
        /**
         * \@member {Date} [saveTime]
         * @type {?|undefined}
         */
        Cart.prototype.saveTime;
        /**
         * \@member {Principal} [savedBy]
         * @type {?|undefined}
         */
        Cart.prototype.savedBy;
        /**
         * \@member {string} [site]
         * @type {?|undefined}
         */
        Cart.prototype.site;
        /**
         * \@member {string} [store]
         * @type {?|undefined}
         */
        Cart.prototype.store;
        /**
         * \@member {Price} [subTotal]
         * @type {?|undefined}
         */
        Cart.prototype.subTotal;
        /**
         * \@member {Price} [totalDiscounts]
         * @type {?|undefined}
         */
        Cart.prototype.totalDiscounts;
        /**
         * \@member {number} [totalItems]
         * @type {?|undefined}
         */
        Cart.prototype.totalItems;
        /**
         * \@member {Price} [totalPrice]
         * @type {?|undefined}
         */
        Cart.prototype.totalPrice;
        /**
         * \@member {Price} [totalPriceWithTax]
         * @type {?|undefined}
         */
        Cart.prototype.totalPriceWithTax;
        /**
         * \@member {Price} [totalTax]
         * @type {?|undefined}
         */
        Cart.prototype.totalTax;
        /**
         * \@member {number} [totalUnitCount]
         * @type {?|undefined}
         */
        Cart.prototype.totalUnitCount;
        /**
         * \@member {Principal} [user]
         * @type {?|undefined}
         */
        Cart.prototype.user;
    }
    /**
     *
     * An interface representing CartList.
     * @record
     */
    function CartList() { }
    Occ.CartList = CartList;
    if (false) {
        /**
         * \@member {Cart[]} [carts]
         * @type {?|undefined}
         */
        CartList.prototype.carts;
    }
    /**
     *
     * An interface representing CartModification.
     * @record
     */
    function CartModification() { }
    Occ.CartModification = CartModification;
    if (false) {
        /**
         * \@member {boolean} [deliveryModeChanged]
         * @type {?|undefined}
         */
        CartModification.prototype.deliveryModeChanged;
        /**
         * \@member {OrderEntry} [entry]
         * @type {?|undefined}
         */
        CartModification.prototype.entry;
        /**
         * \@member {number} [quantity]
         * @type {?|undefined}
         */
        CartModification.prototype.quantity;
        /**
         * \@member {number} [quantityAdded]
         * @type {?|undefined}
         */
        CartModification.prototype.quantityAdded;
        /**
         * \@member {string} [statusCode]
         * @type {?|undefined}
         */
        CartModification.prototype.statusCode;
        /**
         * \@member {string} [statusMessage]
         * @type {?|undefined}
         */
        CartModification.prototype.statusMessage;
    }
    /**
     *
     * An interface representing CategoryHierarchy.
     * @record
     */
    function CategoryHierarchy() { }
    Occ.CategoryHierarchy = CategoryHierarchy;
    if (false) {
        /**
         * \@member {string} [id]
         * @type {?|undefined}
         */
        CategoryHierarchy.prototype.id;
        /**
         * \@member {Date} [lastModified]
         * @type {?|undefined}
         */
        CategoryHierarchy.prototype.lastModified;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        CategoryHierarchy.prototype.name;
        /**
         * \@member {CategoryHierarchy[]} [subcategories]
         * @type {?|undefined}
         */
        CategoryHierarchy.prototype.subcategories;
        /**
         * \@member {string} [url]
         * @type {?|undefined}
         */
        CategoryHierarchy.prototype.url;
    }
    /**
     *
     * An interface representing CatalogVersion.
     * @record
     */
    function CatalogVersion() { }
    Occ.CatalogVersion = CatalogVersion;
    if (false) {
        /**
         * \@member {CategoryHierarchy[]} [categories]
         * @type {?|undefined}
         */
        CatalogVersion.prototype.categories;
        /**
         * \@member {string} [id]
         * @type {?|undefined}
         */
        CatalogVersion.prototype.id;
        /**
         * \@member {Date} [lastModified]
         * @type {?|undefined}
         */
        CatalogVersion.prototype.lastModified;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        CatalogVersion.prototype.name;
        /**
         * \@member {string} [url]
         * @type {?|undefined}
         */
        CatalogVersion.prototype.url;
    }
    /**
     *
     * An interface representing Catalog.
     * @record
     */
    function Catalog() { }
    Occ.Catalog = Catalog;
    if (false) {
        /**
         * \@member {CatalogVersion[]} [catalogVersions]
         * @type {?|undefined}
         */
        Catalog.prototype.catalogVersions;
        /**
         * \@member {string} [id]
         * @type {?|undefined}
         */
        Catalog.prototype.id;
        /**
         * \@member {Date} [lastModified]
         * @type {?|undefined}
         */
        Catalog.prototype.lastModified;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Catalog.prototype.name;
        /**
         * \@member {string} [url]
         * @type {?|undefined}
         */
        Catalog.prototype.url;
    }
    /**
     *
     * An interface representing CatalogList.
     * @record
     */
    function CatalogList() { }
    Occ.CatalogList = CatalogList;
    if (false) {
        /**
         * \@member {Catalog[]} [catalogs]
         * @type {?|undefined}
         */
        CatalogList.prototype.catalogs;
    }
    /**
     *
     * An interface representing ComponentIDList.
     * @record
     */
    function ComponentIDList() { }
    Occ.ComponentIDList = ComponentIDList;
    if (false) {
        /**
         * \@member {string[]} [idList]
         * @type {?|undefined}
         */
        ComponentIDList.prototype.idList;
    }
    /**
     *
     * An interface representing ConsignmentEntry.
     * @record
     */
    function ConsignmentEntry() { }
    Occ.ConsignmentEntry = ConsignmentEntry;
    if (false) {
        /**
         * \@member {OrderEntry} [orderEntry]
         * @type {?|undefined}
         */
        ConsignmentEntry.prototype.orderEntry;
        /**
         * \@member {number} [quantity]
         * @type {?|undefined}
         */
        ConsignmentEntry.prototype.quantity;
        /**
         * \@member {number} [shippedQuantity]
         * @type {?|undefined}
         */
        ConsignmentEntry.prototype.shippedQuantity;
    }
    /**
     *
     * An interface representing Consignment.
     * @record
     */
    function Consignment() { }
    Occ.Consignment = Consignment;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Consignment.prototype.code;
        /**
         * \@member {PointOfService} [deliveryPointOfService]
         * @type {?|undefined}
         */
        Consignment.prototype.deliveryPointOfService;
        /**
         * \@member {ConsignmentEntry[]} [entries]
         * @type {?|undefined}
         */
        Consignment.prototype.entries;
        /**
         * \@member {Address} [shippingAddress]
         * @type {?|undefined}
         */
        Consignment.prototype.shippingAddress;
        /**
         * \@member {string} [status]
         * @type {?|undefined}
         */
        Consignment.prototype.status;
        /**
         * \@member {Date} [statusDate]
         * @type {?|undefined}
         */
        Consignment.prototype.statusDate;
        /**
         * \@member {string} [trackingID]
         * @type {?|undefined}
         */
        Consignment.prototype.trackingID;
    }
    /**
     *
     * An interface representing CountryList.
     * @record
     */
    function CountryList() { }
    Occ.CountryList = CountryList;
    if (false) {
        /**
         * \@member {Country[]} [countries]
         * @type {?|undefined}
         */
        CountryList.prototype.countries;
    }
    /**
     *
     * An interface representing CurrencyList.
     * @record
     */
    function CurrencyList() { }
    Occ.CurrencyList = CurrencyList;
    if (false) {
        /**
         * \@member {Currency[]} [currencies]
         * @type {?|undefined}
         */
        CurrencyList.prototype.currencies;
    }
    /**
     * An interface representing CustomerCoupon
     * @record
     */
    function CustomerCoupon() { }
    Occ.CustomerCoupon = CustomerCoupon;
    if (false) {
        /**
         * \@member {string} [couponId]
         * @type {?|undefined}
         */
        CustomerCoupon.prototype.couponId;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        CustomerCoupon.prototype.name;
        /**
         * \@member {string} [startDate]
         * @type {?|undefined}
         */
        CustomerCoupon.prototype.startDate;
        /**
         * \@member {string} [endDate]
         * @type {?|undefined}
         */
        CustomerCoupon.prototype.endDate;
        /**
         * \@member {string} [endDate]
         * @type {?|undefined}
         */
        CustomerCoupon.prototype.status;
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        CustomerCoupon.prototype.description;
        /**
         * \@member {boolean} [notificationOn]
         * @type {?|undefined}
         */
        CustomerCoupon.prototype.notificationOn;
        /**
         * \@member {boolean} [allProductsApplicable]
         * @type {?|undefined}
         */
        CustomerCoupon.prototype.allProductsApplicable;
    }
    /**
     * An interface representing CustomerCouponSearchResult
     * @record
     */
    function CustomerCouponSearchResult() { }
    Occ.CustomerCouponSearchResult = CustomerCouponSearchResult;
    if (false) {
        /**
         * \@member {CustomerCoupon[]} [coupons]
         * @type {?|undefined}
         */
        CustomerCouponSearchResult.prototype.coupons;
        /**
         * \@member {Sort[]} [sorts]
         * @type {?|undefined}
         */
        CustomerCouponSearchResult.prototype.sorts;
        /**
         * \@member {Pagination} [pagination]
         * @type {?|undefined}
         */
        CustomerCouponSearchResult.prototype.pagination;
    }
    /**
     *
     * An interface representing DeliveryModeList.
     * @record
     */
    function DeliveryModeList() { }
    Occ.DeliveryModeList = DeliveryModeList;
    if (false) {
        /**
         * \@member {DeliveryMode[]} [deliveryModes]
         * @type {?|undefined}
         */
        DeliveryModeList.prototype.deliveryModes;
    }
    /**
     *
     * An interface representing FacetValue.
     * @record
     */
    function FacetValue() { }
    Occ.FacetValue = FacetValue;
    if (false) {
        /**
         * \@member {number} [count]
         * @type {?|undefined}
         */
        FacetValue.prototype.count;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        FacetValue.prototype.name;
        /**
         * \@member {SearchState} [query]
         * @type {?|undefined}
         */
        FacetValue.prototype.query;
        /**
         * \@member {boolean} [selected]
         * @type {?|undefined}
         */
        FacetValue.prototype.selected;
    }
    /**
     *
     * An interface representing Facet.
     * @record
     */
    function Facet() { }
    Occ.Facet = Facet;
    if (false) {
        /**
         * \@member {boolean} [category]
         * @type {?|undefined}
         */
        Facet.prototype.category;
        /**
         * \@member {boolean} [multiSelect]
         * @type {?|undefined}
         */
        Facet.prototype.multiSelect;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Facet.prototype.name;
        /**
         * \@member {number} [priority]
         * @type {?|undefined}
         */
        Facet.prototype.priority;
        /**
         * \@member {FacetValue[]} [topValues]
         * @type {?|undefined}
         */
        Facet.prototype.topValues;
        /**
         * \@member {FacetValue[]} [values]
         * @type {?|undefined}
         */
        Facet.prototype.values;
        /**
         * \@member {boolean} [visible]
         * @type {?|undefined}
         */
        Facet.prototype.visible;
    }
    /**
     *
     * An interface representing LanguageList.
     * @record
     */
    function LanguageList() { }
    Occ.LanguageList = LanguageList;
    if (false) {
        /**
         * \@member {Language[]} [languages]
         * @type {?|undefined}
         */
        LanguageList.prototype.languages;
    }
    /**
     *
     * An interface representing Pagination.
     * Pagination info
     *
     * @record
     */
    function Pagination() { }
    Occ.Pagination = Pagination;
    if (false) {
        /**
         * \@member {number} [count] Number of elements on this page
         * @type {?|undefined}
         */
        Pagination.prototype.count;
        /**
         * \@member {number} [page] Current page number
         * @type {?|undefined}
         */
        Pagination.prototype.page;
        /**
         * \@member {number} [totalCount] Total number of elements
         * @type {?|undefined}
         */
        Pagination.prototype.totalCount;
        /**
         * \@member {number} [totalPages] Total number of pages
         * @type {?|undefined}
         */
        Pagination.prototype.totalPages;
    }
    /**
     *
     * An interface representing Sort.
     * Sort option
     *
     * @record
     */
    function Sort() { }
    Occ.Sort = Sort;
    if (false) {
        /**
         * \@member {boolean} [asc]
         * @type {?|undefined}
         */
        Sort.prototype.asc;
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Sort.prototype.code;
    }
    /**
     *
     * An interface representing ListAdaptedComponents.
     * @record
     */
    function ListAdaptedComponents() { }
    Occ.ListAdaptedComponents = ListAdaptedComponents;
    if (false) {
        /**
         * \@member {any[]} [components]
         * @type {?|undefined}
         */
        ListAdaptedComponents.prototype.components;
        /**
         * \@member {Pagination} [pagination]
         * @type {?|undefined}
         */
        ListAdaptedComponents.prototype.pagination;
        /**
         * \@member {Sort[]} [sorts]
         * @type {?|undefined}
         */
        ListAdaptedComponents.prototype.sorts;
    }
    /**
     *
     * An interface representing MemberList.
     * @record
     */
    function MemberList() { }
    Occ.MemberList = MemberList;
    if (false) {
        /**
         * \@member {Principal[]} [members]
         * @type {?|undefined}
         */
        MemberList.prototype.members;
    }
    /**
     *
     * An interface representing OrderEntryList.
     * @record
     */
    function OrderEntryList() { }
    Occ.OrderEntryList = OrderEntryList;
    if (false) {
        /**
         * \@member {OrderEntry[]} [orderEntries]
         * @type {?|undefined}
         */
        OrderEntryList.prototype.orderEntries;
    }
    /**
     *
     * An interface representing OrderHistory.
     * @record
     */
    function OrderHistory() { }
    Occ.OrderHistory = OrderHistory;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        OrderHistory.prototype.code;
        /**
         * \@member {string} [guid]
         * @type {?|undefined}
         */
        OrderHistory.prototype.guid;
        /**
         * \@member {Date} [placed]
         * @type {?|undefined}
         */
        OrderHistory.prototype.placed;
        /**
         * \@member {string} [status]
         * @type {?|undefined}
         */
        OrderHistory.prototype.status;
        /**
         * \@member {string} [statusDisplay]
         * @type {?|undefined}
         */
        OrderHistory.prototype.statusDisplay;
        /**
         * \@member {Price} [total]
         * @type {?|undefined}
         */
        OrderHistory.prototype.total;
    }
    /**
     *
     * An interface representing PaginationModel.
     * @record
     */
    function PaginationModel() { }
    Occ.PaginationModel = PaginationModel;
    if (false) {
        /**
         * \@member {number} [currentPage]
         * @type {?|undefined}
         */
        PaginationModel.prototype.currentPage;
        /**
         * \@member {number} [pageSize]
         * @type {?|undefined}
         */
        PaginationModel.prototype.pageSize;
        /**
         * \@member {string} [sort]
         * @type {?|undefined}
         */
        PaginationModel.prototype.sort;
        /**
         * \@member {number} [totalPages]
         * @type {?|undefined}
         */
        PaginationModel.prototype.totalPages;
        /**
         * \@member {number} [totalResults]
         * @type {?|undefined}
         */
        PaginationModel.prototype.totalResults;
    }
    /**
     *
     * An interface representing SortModel.
     * @record
     */
    function SortModel() { }
    Occ.SortModel = SortModel;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        SortModel.prototype.code;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        SortModel.prototype.name;
        /**
         * \@member {boolean} [selected]
         * @type {?|undefined}
         */
        SortModel.prototype.selected;
    }
    /**
     *
     * An interface representing OrderHistoryList.
     * @record
     */
    function OrderHistoryList() { }
    Occ.OrderHistoryList = OrderHistoryList;
    if (false) {
        /**
         * \@member {OrderHistory[]} [orders]
         * @type {?|undefined}
         */
        OrderHistoryList.prototype.orders;
        /**
         * \@member {PaginationModel} [pagination]
         * @type {?|undefined}
         */
        OrderHistoryList.prototype.pagination;
        /**
         * \@member {SortModel[]} [sorts]
         * @type {?|undefined}
         */
        OrderHistoryList.prototype.sorts;
    }
    /**
     *
     * An interface representing OrderStatusUpdateElement.
     * @record
     */
    function OrderStatusUpdateElement() { }
    Occ.OrderStatusUpdateElement = OrderStatusUpdateElement;
    if (false) {
        /**
         * \@member {string} [baseSiteId]
         * @type {?|undefined}
         */
        OrderStatusUpdateElement.prototype.baseSiteId;
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        OrderStatusUpdateElement.prototype.code;
        /**
         * \@member {string} [status]
         * @type {?|undefined}
         */
        OrderStatusUpdateElement.prototype.status;
    }
    /**
     *
     * An interface representing OrderStatusUpdateElementList.
     * @record
     */
    function OrderStatusUpdateElementList() { }
    Occ.OrderStatusUpdateElementList = OrderStatusUpdateElementList;
    if (false) {
        /**
         * \@member {OrderStatusUpdateElement[]} [orderStatusUpdateElements]
         * @type {?|undefined}
         */
        OrderStatusUpdateElementList.prototype.orderStatusUpdateElements;
    }
    /**
     *
     * An interface representing Order.
     * @record
     */
    function Order() { }
    Occ.Order = Order;
    if (false) {
        /**
         * \@member {PromotionResult[]} [appliedOrderPromotions]
         * @type {?|undefined}
         */
        Order.prototype.appliedOrderPromotions;
        /**
         * \@member {PromotionResult[]} [appliedProductPromotions]
         * @type {?|undefined}
         */
        Order.prototype.appliedProductPromotions;
        /**
         * \@member {Voucher[]} [appliedVouchers]
         * @type {?|undefined}
         */
        Order.prototype.appliedVouchers;
        /**
         * \@member {boolean} [calculated]
         * @type {?|undefined}
         */
        Order.prototype.calculated;
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Order.prototype.code;
        /**
         * \@member {Consignment[]} [consignments]
         * @type {?|undefined}
         */
        Order.prototype.consignments;
        /**
         * \@member {Date} [created]
         * @type {?|undefined}
         */
        Order.prototype.created;
        /**
         * \@member {Address} [deliveryAddress]
         * @type {?|undefined}
         */
        Order.prototype.deliveryAddress;
        /**
         * \@member {Price} [deliveryCost]
         * @type {?|undefined}
         */
        Order.prototype.deliveryCost;
        /**
         * \@member {number} [deliveryItemsQuantity]
         * @type {?|undefined}
         */
        Order.prototype.deliveryItemsQuantity;
        /**
         * \@member {DeliveryMode} [deliveryMode]
         * @type {?|undefined}
         */
        Order.prototype.deliveryMode;
        /**
         * \@member {DeliveryOrderEntryGroup[]} [deliveryOrderGroups]
         * @type {?|undefined}
         */
        Order.prototype.deliveryOrderGroups;
        /**
         * \@member {string} [deliveryStatus]
         * @type {?|undefined}
         */
        Order.prototype.deliveryStatus;
        /**
         * \@member {string} [deliveryStatusDisplay]
         * @type {?|undefined}
         */
        Order.prototype.deliveryStatusDisplay;
        /**
         * \@member {OrderEntry[]} [entries]
         * @type {?|undefined}
         */
        Order.prototype.entries;
        /**
         * \@member {boolean} [guestCustomer]
         * @type {?|undefined}
         */
        Order.prototype.guestCustomer;
        /**
         * \@member {string} [guid]
         * @type {?|undefined}
         */
        Order.prototype.guid;
        /**
         * \@member {boolean} [net]
         * @type {?|undefined}
         */
        Order.prototype.net;
        /**
         * \@member {Price} [orderDiscounts]
         * @type {?|undefined}
         */
        Order.prototype.orderDiscounts;
        /**
         * \@member {PaymentDetails} [paymentInfo]
         * @type {?|undefined}
         */
        Order.prototype.paymentInfo;
        /**
         * \@member {number} [pickupItemsQuantity]
         * @type {?|undefined}
         */
        Order.prototype.pickupItemsQuantity;
        /**
         * \@member {PickupOrderEntryGroup[]} [pickupOrderGroups]
         * @type {?|undefined}
         */
        Order.prototype.pickupOrderGroups;
        /**
         * \@member {Price} [productDiscounts]
         * @type {?|undefined}
         */
        Order.prototype.productDiscounts;
        /**
         * \@member {string} [site]
         * @type {?|undefined}
         */
        Order.prototype.site;
        /**
         * \@member {string} [status]
         * @type {?|undefined}
         */
        Order.prototype.status;
        /**
         * \@member {string} [statusDisplay]
         * @type {?|undefined}
         */
        Order.prototype.statusDisplay;
        /**
         * \@member {string} [store]
         * @type {?|undefined}
         */
        Order.prototype.store;
        /**
         * \@member {Price} [subTotal]
         * @type {?|undefined}
         */
        Order.prototype.subTotal;
        /**
         * \@member {Price} [totalDiscounts]
         * @type {?|undefined}
         */
        Order.prototype.totalDiscounts;
        /**
         * \@member {number} [totalItems]
         * @type {?|undefined}
         */
        Order.prototype.totalItems;
        /**
         * \@member {Price} [totalPrice]
         * @type {?|undefined}
         */
        Order.prototype.totalPrice;
        /**
         * \@member {Price} [totalPriceWithTax]
         * @type {?|undefined}
         */
        Order.prototype.totalPriceWithTax;
        /**
         * \@member {Price} [totalTax]
         * @type {?|undefined}
         */
        Order.prototype.totalTax;
        /**
         * \@member {OrderEntry[]} [unconsignedEntries]
         * @type {?|undefined}
         */
        Order.prototype.unconsignedEntries;
        /**
         * \@member {Principal} [user]
         * @type {?|undefined}
         */
        Order.prototype.user;
    }
    /**
     *
     * An interface representing ReturnRequest.
     * @record
     */
    function ReturnRequest() { }
    Occ.ReturnRequest = ReturnRequest;
    if (false) {
        /**
         * \@member {boolean} [cancellable]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.cancellable;
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.code;
        /**
         * \@member {Date} [creationTime]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.creationTime;
        /**
         * \@member {Price} [deliveryCost]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.deliveryCost;
        /**
         * \@member {order} [order]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.order;
        /**
         * \@member {boolean} [refundDeliveryCost]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.refundDeliveryCost;
        /**
         * \@member {ReturnRequestEntry[]} [returnEntries]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.returnEntries;
        /**
         * \@member {string} [returnLabelDownloadUrl]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.returnLabelDownloadUrl;
        /**
         * \@member {string} [rma]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.rma;
        /**
         * \@member {string} [status]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.status;
        /**
         * \@member {Price} [subTotal]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.subTotal;
        /**
         * \@member {Price} [totalPrice]
         * @type {?|undefined}
         */
        ReturnRequest.prototype.totalPrice;
    }
    /**
     *
     * An interface representing ReturnRequestEntry.
     * @record
     */
    function ReturnRequestEntry() { }
    Occ.ReturnRequestEntry = ReturnRequestEntry;
    if (false) {
        /**
         * \@member {OrderEntry} [orderEntry]
         * @type {?|undefined}
         */
        ReturnRequestEntry.prototype.orderEntry;
        /**
         * \@member {number} [expectedQuantity]
         * @type {?|undefined}
         */
        ReturnRequestEntry.prototype.expectedQuantity;
        /**
         * \@member {Price} [refundAmount]
         * @type {?|undefined}
         */
        ReturnRequestEntry.prototype.refundAmount;
    }
    /**
     *
     * An interface representing PaymentDetailsList.
     * @record
     */
    function PaymentDetailsList() { }
    Occ.PaymentDetailsList = PaymentDetailsList;
    if (false) {
        /**
         * \@member {PaymentDetails[]} [payments]
         * @type {?|undefined}
         */
        PaymentDetailsList.prototype.payments;
    }
    /**
     *
     * An interface representing PointOfServiceStock.
     * @record
     */
    function PointOfServiceStock() { }
    Occ.PointOfServiceStock = PointOfServiceStock;
    if (false) {
        /**
         * \@member {Address} [address]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.address;
        /**
         * \@member {string} [description]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.description;
        /**
         * \@member {string} [displayName]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.displayName;
        /**
         * \@member {number} [distanceKm]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.distanceKm;
        /**
         * \@member {{ [propertyName: string]: string }} [features]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.features;
        /**
         * \@member {string} [formattedDistance]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.formattedDistance;
        /**
         * \@member {GeoPoint} [geoPoint]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.geoPoint;
        /**
         * \@member {Image} [mapIcon]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.mapIcon;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.name;
        /**
         * \@member {OpeningSchedule} [openingHours]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.openingHours;
        /**
         * \@member {Stock} [stockInfo]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.stockInfo;
        /**
         * \@member {string} [storeContent]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.storeContent;
        /**
         * \@member {Image[]} [storeImages]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.storeImages;
        /**
         * \@member {string} [url]
         * @type {?|undefined}
         */
        PointOfServiceStock.prototype.url;
    }
    /**
     *
     * An interface representing ProductExpressUpdateElement.
     * @record
     */
    function ProductExpressUpdateElement() { }
    Occ.ProductExpressUpdateElement = ProductExpressUpdateElement;
    if (false) {
        /**
         * \@member {string} [catalogId]
         * @type {?|undefined}
         */
        ProductExpressUpdateElement.prototype.catalogId;
        /**
         * \@member {string} [catalogVersion]
         * @type {?|undefined}
         */
        ProductExpressUpdateElement.prototype.catalogVersion;
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        ProductExpressUpdateElement.prototype.code;
    }
    /**
     *
     * An interface representing ProductExpressUpdateElementList.
     * @record
     */
    function ProductExpressUpdateElementList() { }
    Occ.ProductExpressUpdateElementList = ProductExpressUpdateElementList;
    if (false) {
        /**
         * \@member {ProductExpressUpdateElement[]} [productExpressUpdateElements]
         * @type {?|undefined}
         */
        ProductExpressUpdateElementList.prototype.productExpressUpdateElements;
    }
    /**
     *
     * An interface representing ProductList.
     * @record
     */
    function ProductList() { }
    Occ.ProductList = ProductList;
    if (false) {
        /**
         * \@member {string} [catalog]
         * @type {?|undefined}
         */
        ProductList.prototype.catalog;
        /**
         * \@member {number} [currentPage]
         * @type {?|undefined}
         */
        ProductList.prototype.currentPage;
        /**
         * \@member {Product[]} [products]
         * @type {?|undefined}
         */
        ProductList.prototype.products;
        /**
         * \@member {number} [totalPageCount]
         * @type {?|undefined}
         */
        ProductList.prototype.totalPageCount;
        /**
         * \@member {number} [totalProductCount]
         * @type {?|undefined}
         */
        ProductList.prototype.totalProductCount;
        /**
         * \@member {string} [version]
         * @type {?|undefined}
         */
        ProductList.prototype.version;
    }
    /**
     *
     * An interface representing ProductReferenceList.
     * @record
     */
    function ProductReferenceList() { }
    Occ.ProductReferenceList = ProductReferenceList;
    if (false) {
        /**
         * \@member {ProductReference[]} [references]
         * @type {?|undefined}
         */
        ProductReferenceList.prototype.references;
    }
    /**
     *
     * An interface representing SpellingSuggestion.
     * @record
     */
    function SpellingSuggestion() { }
    Occ.SpellingSuggestion = SpellingSuggestion;
    if (false) {
        /**
         * \@member {string} [query]
         * @type {?|undefined}
         */
        SpellingSuggestion.prototype.query;
        /**
         * \@member {string} [suggestion]
         * @type {?|undefined}
         */
        SpellingSuggestion.prototype.suggestion;
    }
    /**
     *
     * An interface representing ProductSearchPage.
     * @record
     */
    function ProductSearchPage() { }
    Occ.ProductSearchPage = ProductSearchPage;
    if (false) {
        /**
         * \@member {Breadcrumb[]} [breadcrumbs]
         * @type {?|undefined}
         */
        ProductSearchPage.prototype.breadcrumbs;
        /**
         * \@member {string} [categoryCode]
         * @type {?|undefined}
         */
        ProductSearchPage.prototype.categoryCode;
        /**
         * \@member {SearchState} [currentQuery]
         * @type {?|undefined}
         */
        ProductSearchPage.prototype.currentQuery;
        /**
         * \@member {Facet[]} [facets]
         * @type {?|undefined}
         */
        ProductSearchPage.prototype.facets;
        /**
         * \@member {string} [freeTextSearch]
         * @type {?|undefined}
         */
        ProductSearchPage.prototype.freeTextSearch;
        /**
         * \@member {string} [keywordRedirectUrl]
         * @type {?|undefined}
         */
        ProductSearchPage.prototype.keywordRedirectUrl;
        /**
         * \@member {PaginationModel} [pagination]
         * @type {?|undefined}
         */
        ProductSearchPage.prototype.pagination;
        /**
         * \@member {Product[]} [products]
         * @type {?|undefined}
         */
        ProductSearchPage.prototype.products;
        /**
         * \@member {SortModel[]} [sorts]
         * @type {?|undefined}
         */
        ProductSearchPage.prototype.sorts;
        /**
         * \@member {SpellingSuggestion} [spellingSuggestion]
         * @type {?|undefined}
         */
        ProductSearchPage.prototype.spellingSuggestion;
    }
    /**
     *
     * An interface representing PromotionList.
     * @record
     */
    function PromotionList() { }
    Occ.PromotionList = PromotionList;
    if (false) {
        /**
         * \@member {Promotion[]} [promotions]
         * @type {?|undefined}
         */
        PromotionList.prototype.promotions;
    }
    /**
     *
     * An interface representing PromotionResultList.
     * @record
     */
    function PromotionResultList() { }
    Occ.PromotionResultList = PromotionResultList;
    if (false) {
        /**
         * \@member {PromotionResult[]} [promotions]
         * @type {?|undefined}
         */
        PromotionResultList.prototype.promotions;
    }
    /**
     *
     * An interface representing ReviewList.
     * @record
     */
    function ReviewList() { }
    Occ.ReviewList = ReviewList;
    if (false) {
        /**
         * \@member {Review[]} [reviews]
         * @type {?|undefined}
         */
        ReviewList.prototype.reviews;
    }
    /**
     *
     * An interface representing SaveCartResult.
     * @record
     */
    function SaveCartResult() { }
    Occ.SaveCartResult = SaveCartResult;
    if (false) {
        /**
         * \@member {Cart} [savedCartData]
         * @type {?|undefined}
         */
        SaveCartResult.prototype.savedCartData;
    }
    /**
     *
     * An interface representing StoreFinderSearchPage.
     * @record
     */
    function StoreFinderSearchPage() { }
    Occ.StoreFinderSearchPage = StoreFinderSearchPage;
    if (false) {
        /**
         * \@member {number} [boundEastLongitude]
         * @type {?|undefined}
         */
        StoreFinderSearchPage.prototype.boundEastLongitude;
        /**
         * \@member {number} [boundNorthLatitude]
         * @type {?|undefined}
         */
        StoreFinderSearchPage.prototype.boundNorthLatitude;
        /**
         * \@member {number} [boundSouthLatitude]
         * @type {?|undefined}
         */
        StoreFinderSearchPage.prototype.boundSouthLatitude;
        /**
         * \@member {number} [boundWestLongitude]
         * @type {?|undefined}
         */
        StoreFinderSearchPage.prototype.boundWestLongitude;
        /**
         * \@member {string} [locationText]
         * @type {?|undefined}
         */
        StoreFinderSearchPage.prototype.locationText;
        /**
         * \@member {PaginationModel} [pagination]
         * @type {?|undefined}
         */
        StoreFinderSearchPage.prototype.pagination;
        /**
         * \@member {SortModel[]} [sorts]
         * @type {?|undefined}
         */
        StoreFinderSearchPage.prototype.sorts;
        /**
         * \@member {number} [sourceLatitude]
         * @type {?|undefined}
         */
        StoreFinderSearchPage.prototype.sourceLatitude;
        /**
         * \@member {number} [sourceLongitude]
         * @type {?|undefined}
         */
        StoreFinderSearchPage.prototype.sourceLongitude;
        /**
         * \@member {PointOfService[]} [stores]
         * @type {?|undefined}
         */
        StoreFinderSearchPage.prototype.stores;
    }
    /**
     *
     * An interface representing StoreFinderStockSearchPage.
     * @record
     */
    function StoreFinderStockSearchPage() { }
    Occ.StoreFinderStockSearchPage = StoreFinderStockSearchPage;
    if (false) {
        /**
         * \@member {number} [boundEastLongitude]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.boundEastLongitude;
        /**
         * \@member {number} [boundNorthLatitude]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.boundNorthLatitude;
        /**
         * \@member {number} [boundSouthLatitude]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.boundSouthLatitude;
        /**
         * \@member {number} [boundWestLongitude]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.boundWestLongitude;
        /**
         * \@member {string} [locationText]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.locationText;
        /**
         * \@member {PaginationModel} [pagination]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.pagination;
        /**
         * \@member {Product} [product]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.product;
        /**
         * \@member {SortModel[]} [sorts]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.sorts;
        /**
         * \@member {number} [sourceLatitude]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.sourceLatitude;
        /**
         * \@member {number} [sourceLongitude]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.sourceLongitude;
        /**
         * \@member {PointOfServiceStock[]} [stores]
         * @type {?|undefined}
         */
        StoreFinderStockSearchPage.prototype.stores;
    }
    /**
     *
     * An interface representing Suggestion.
     * @record
     */
    function Suggestion() { }
    Occ.Suggestion = Suggestion;
    if (false) {
        /**
         * \@member {string} [value]
         * @type {?|undefined}
         */
        Suggestion.prototype.value;
    }
    /**
     *
     * An interface representing SuggestionList.
     * @record
     */
    function SuggestionList() { }
    Occ.SuggestionList = SuggestionList;
    if (false) {
        /**
         * \@member {Suggestion[]} [suggestions]
         * @type {?|undefined}
         */
        SuggestionList.prototype.suggestions;
    }
    /**
     *
     * An interface representing Title.
     * @record
     */
    function Title() { }
    Occ.Title = Title;
    if (false) {
        /**
         * \@member {string} [code]
         * @type {?|undefined}
         */
        Title.prototype.code;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        Title.prototype.name;
    }
    /**
     *
     * An interface representing TitleList.
     * @record
     */
    function TitleList() { }
    Occ.TitleList = TitleList;
    if (false) {
        /**
         * \@member {Title[]} [titles]
         * @type {?|undefined}
         */
        TitleList.prototype.titles;
    }
    /**
     *
     * An interface representing UserGroup.
     * @record
     */
    function UserGroup() { }
    Occ.UserGroup = UserGroup;
    if (false) {
        /**
         * \@member {Principal[]} [members]
         * @type {?|undefined}
         */
        UserGroup.prototype.members;
        /**
         * \@member {number} [membersCount]
         * @type {?|undefined}
         */
        UserGroup.prototype.membersCount;
        /**
         * \@member {string} [name]
         * @type {?|undefined}
         */
        UserGroup.prototype.name;
        /**
         * \@member {UserGroup[]} [subGroups]
         * @type {?|undefined}
         */
        UserGroup.prototype.subGroups;
        /**
         * \@member {string} [uid]
         * @type {?|undefined}
         */
        UserGroup.prototype.uid;
    }
    /**
     *
     * An interface representing UserGroupList.
     * @record
     */
    function UserGroupList() { }
    Occ.UserGroupList = UserGroupList;
    if (false) {
        /**
         * \@member {number} [currentPage]
         * @type {?|undefined}
         */
        UserGroupList.prototype.currentPage;
        /**
         * \@member {number} [numberOfPages]
         * @type {?|undefined}
         */
        UserGroupList.prototype.numberOfPages;
        /**
         * \@member {number} [pageSize]
         * @type {?|undefined}
         */
        UserGroupList.prototype.pageSize;
        /**
         * \@member {number} [totalNumber]
         * @type {?|undefined}
         */
        UserGroupList.prototype.totalNumber;
        /**
         * \@member {UserGroup[]} [userGroups]
         * @type {?|undefined}
         */
        UserGroupList.prototype.userGroups;
    }
    /**
     *
     * An interface representing UserSignUp.
     * @record
     */
    function UserSignUp() { }
    Occ.UserSignUp = UserSignUp;
    if (false) {
        /**
         * \@member {string} [firstName]
         * @type {?|undefined}
         */
        UserSignUp.prototype.firstName;
        /**
         * \@member {string} [lastName]
         * @type {?|undefined}
         */
        UserSignUp.prototype.lastName;
        /**
         * \@member {string} [password]
         * @type {?|undefined}
         */
        UserSignUp.prototype.password;
        /**
         * \@member {string} [titleCode]
         * @type {?|undefined}
         */
        UserSignUp.prototype.titleCode;
        /**
         * \@member {string} [uid]
         * @type {?|undefined}
         */
        UserSignUp.prototype.uid;
    }
    /**
     * @record
     */
    function StoreCount() { }
    Occ.StoreCount = StoreCount;
    if (false) {
        /** @type {?|undefined} */
        StoreCount.prototype.count;
        /** @type {?|undefined} */
        StoreCount.prototype.isoCode;
        /** @type {?|undefined} */
        StoreCount.prototype.name;
        /** @type {?|undefined} */
        StoreCount.prototype.type;
    }
    /**
     * @record
     */
    function StoreCountList() { }
    Occ.StoreCountList = StoreCountList;
    if (false) {
        /** @type {?|undefined} */
        StoreCountList.prototype.countriesAndRegionsStoreCount;
    }
    /**
     *
     * An interface representing VoucherList.
     * @record
     */
    function VoucherList() { }
    Occ.VoucherList = VoucherList;
    if (false) {
        /**
         * \@member {Voucher[]} [vouchers]
         * @type {?|undefined}
         */
        VoucherList.prototype.vouchers;
    }
    /**
     * Defines values for PriceType.
     * Possible values include: 'BUY', 'FROM'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: PriceType = <PriceType>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var PriceType;
    (function (PriceType) {
        PriceType["BUY"] = "BUY";
        PriceType["FROM"] = "FROM";
    })(PriceType = Occ.PriceType || (Occ.PriceType = {}));
    /**
     * Defines values for ImageType.
     * Possible values include: 'PRIMARY', 'GALLERY'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: ImageType = <ImageType>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var ImageType;
    (function (ImageType) {
        ImageType["PRIMARY"] = "PRIMARY";
        ImageType["GALLERY"] = "GALLERY";
    })(ImageType = Occ.ImageType || (Occ.ImageType = {}));
    /**
     * Defines values for Fields.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields = <Fields>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields;
    (function (Fields) {
        Fields["BASIC"] = "BASIC";
        Fields["DEFAULT"] = "DEFAULT";
        Fields["FULL"] = "FULL";
    })(Fields = Occ.Fields || (Occ.Fields = {}));
    /**
     * Defines values for Fields1.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields1 = <Fields1>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields1;
    (function (Fields1) {
        Fields1["BASIC"] = "BASIC";
        Fields1["DEFAULT"] = "DEFAULT";
        Fields1["FULL"] = "FULL";
    })(Fields1 = Occ.Fields1 || (Occ.Fields1 = {}));
    /**
     * Defines values for Fields2.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields2 = <Fields2>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields2;
    (function (Fields2) {
        Fields2["BASIC"] = "BASIC";
        Fields2["DEFAULT"] = "DEFAULT";
        Fields2["FULL"] = "FULL";
    })(Fields2 = Occ.Fields2 || (Occ.Fields2 = {}));
    /**
     * Defines values for Fields3.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields3 = <Fields3>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields3;
    (function (Fields3) {
        Fields3["BASIC"] = "BASIC";
        Fields3["DEFAULT"] = "DEFAULT";
        Fields3["FULL"] = "FULL";
    })(Fields3 = Occ.Fields3 || (Occ.Fields3 = {}));
    /**
     * Defines values for Fields4.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields4 = <Fields4>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields4;
    (function (Fields4) {
        Fields4["BASIC"] = "BASIC";
        Fields4["DEFAULT"] = "DEFAULT";
        Fields4["FULL"] = "FULL";
    })(Fields4 = Occ.Fields4 || (Occ.Fields4 = {}));
    /**
     * Defines values for Fields5.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields5 = <Fields5>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields5;
    (function (Fields5) {
        Fields5["BASIC"] = "BASIC";
        Fields5["DEFAULT"] = "DEFAULT";
        Fields5["FULL"] = "FULL";
    })(Fields5 = Occ.Fields5 || (Occ.Fields5 = {}));
    /**
     * Defines values for Fields6.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields6 = <Fields6>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields6;
    (function (Fields6) {
        Fields6["BASIC"] = "BASIC";
        Fields6["DEFAULT"] = "DEFAULT";
        Fields6["FULL"] = "FULL";
    })(Fields6 = Occ.Fields6 || (Occ.Fields6 = {}));
    /**
     * Defines values for PageType.
     * Possible values include: 'ContentPage', 'ProductPage', 'CategoryPage',
     * 'CatalogPage'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: PageType = <PageType>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var PageType;
    (function (PageType) {
        PageType["CONTENT_PAGE"] = "ContentPage";
        PageType["PRODUCT_PAGE"] = "ProductPage";
        PageType["CATEGORY_PAGE"] = "CategoryPage";
        PageType["CATALOG_PAGE"] = "CatalogPage";
    })(PageType = Occ.PageType || (Occ.PageType = {}));
    /**
     * Defines values for Fields7.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields7 = <Fields7>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields7;
    (function (Fields7) {
        Fields7["BASIC"] = "BASIC";
        Fields7["DEFAULT"] = "DEFAULT";
        Fields7["FULL"] = "FULL";
    })(Fields7 = Occ.Fields7 || (Occ.Fields7 = {}));
    /**
     * Defines values for Fields8.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields8 = <Fields8>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields8;
    (function (Fields8) {
        Fields8["BASIC"] = "BASIC";
        Fields8["DEFAULT"] = "DEFAULT";
        Fields8["FULL"] = "FULL";
    })(Fields8 = Occ.Fields8 || (Occ.Fields8 = {}));
    /**
     * Defines values for Fields9.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields9 = <Fields9>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields9;
    (function (Fields9) {
        Fields9["BASIC"] = "BASIC";
        Fields9["DEFAULT"] = "DEFAULT";
        Fields9["FULL"] = "FULL";
    })(Fields9 = Occ.Fields9 || (Occ.Fields9 = {}));
    /**
     * Defines values for Fields10.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields10 = <Fields10>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields10;
    (function (Fields10) {
        Fields10["BASIC"] = "BASIC";
        Fields10["DEFAULT"] = "DEFAULT";
        Fields10["FULL"] = "FULL";
    })(Fields10 = Occ.Fields10 || (Occ.Fields10 = {}));
    /**
     * Defines values for Fields11.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields11 = <Fields11>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields11;
    (function (Fields11) {
        Fields11["BASIC"] = "BASIC";
        Fields11["DEFAULT"] = "DEFAULT";
        Fields11["FULL"] = "FULL";
    })(Fields11 = Occ.Fields11 || (Occ.Fields11 = {}));
    /**
     * Defines values for Fields12.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields12 = <Fields12>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields12;
    (function (Fields12) {
        Fields12["BASIC"] = "BASIC";
        Fields12["DEFAULT"] = "DEFAULT";
        Fields12["FULL"] = "FULL";
    })(Fields12 = Occ.Fields12 || (Occ.Fields12 = {}));
    /**
     * Defines values for Fields13.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields13 = <Fields13>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields13;
    (function (Fields13) {
        Fields13["BASIC"] = "BASIC";
        Fields13["DEFAULT"] = "DEFAULT";
        Fields13["FULL"] = "FULL";
    })(Fields13 = Occ.Fields13 || (Occ.Fields13 = {}));
    /**
     * Defines values for Fields14.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields14 = <Fields14>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields14;
    (function (Fields14) {
        Fields14["BASIC"] = "BASIC";
        Fields14["DEFAULT"] = "DEFAULT";
        Fields14["FULL"] = "FULL";
    })(Fields14 = Occ.Fields14 || (Occ.Fields14 = {}));
    /**
     * Defines values for Fields15.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields15 = <Fields15>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields15;
    (function (Fields15) {
        Fields15["BASIC"] = "BASIC";
        Fields15["DEFAULT"] = "DEFAULT";
        Fields15["FULL"] = "FULL";
    })(Fields15 = Occ.Fields15 || (Occ.Fields15 = {}));
    /**
     * Defines values for Fields16.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields16 = <Fields16>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields16;
    (function (Fields16) {
        Fields16["BASIC"] = "BASIC";
        Fields16["DEFAULT"] = "DEFAULT";
        Fields16["FULL"] = "FULL";
    })(Fields16 = Occ.Fields16 || (Occ.Fields16 = {}));
    /**
     * Defines values for SortEnum.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: SortEnum = <SortEnum>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var SortEnum;
    (function (SortEnum) {
        SortEnum["BASIC"] = "BASIC";
        SortEnum["DEFAULT"] = "DEFAULT";
        SortEnum["FULL"] = "FULL";
    })(SortEnum = Occ.SortEnum || (Occ.SortEnum = {}));
    /**
     * Defines values for Fields17.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields17 = <Fields17>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields17;
    (function (Fields17) {
        Fields17["BASIC"] = "BASIC";
        Fields17["DEFAULT"] = "DEFAULT";
        Fields17["FULL"] = "FULL";
    })(Fields17 = Occ.Fields17 || (Occ.Fields17 = {}));
    /**
     * Defines values for Fields18.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields18 = <Fields18>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields18;
    (function (Fields18) {
        Fields18["BASIC"] = "BASIC";
        Fields18["DEFAULT"] = "DEFAULT";
        Fields18["FULL"] = "FULL";
    })(Fields18 = Occ.Fields18 || (Occ.Fields18 = {}));
    /**
     * Defines values for Fields19.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields19 = <Fields19>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields19;
    (function (Fields19) {
        Fields19["BASIC"] = "BASIC";
        Fields19["DEFAULT"] = "DEFAULT";
        Fields19["FULL"] = "FULL";
    })(Fields19 = Occ.Fields19 || (Occ.Fields19 = {}));
    /**
     * Defines values for Fields20.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields20 = <Fields20>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields20;
    (function (Fields20) {
        Fields20["BASIC"] = "BASIC";
        Fields20["DEFAULT"] = "DEFAULT";
        Fields20["FULL"] = "FULL";
    })(Fields20 = Occ.Fields20 || (Occ.Fields20 = {}));
    /**
     * Defines values for Fields21.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields21 = <Fields21>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields21;
    (function (Fields21) {
        Fields21["BASIC"] = "BASIC";
        Fields21["DEFAULT"] = "DEFAULT";
        Fields21["FULL"] = "FULL";
    })(Fields21 = Occ.Fields21 || (Occ.Fields21 = {}));
    /**
     * Defines values for Fields22.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields22 = <Fields22>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields22;
    (function (Fields22) {
        Fields22["BASIC"] = "BASIC";
        Fields22["DEFAULT"] = "DEFAULT";
        Fields22["FULL"] = "FULL";
    })(Fields22 = Occ.Fields22 || (Occ.Fields22 = {}));
    /**
     * Defines values for Fields23.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields23 = <Fields23>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields23;
    (function (Fields23) {
        Fields23["BASIC"] = "BASIC";
        Fields23["DEFAULT"] = "DEFAULT";
        Fields23["FULL"] = "FULL";
    })(Fields23 = Occ.Fields23 || (Occ.Fields23 = {}));
    /**
     * Defines values for Fields24.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields24 = <Fields24>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields24;
    (function (Fields24) {
        Fields24["BASIC"] = "BASIC";
        Fields24["DEFAULT"] = "DEFAULT";
        Fields24["FULL"] = "FULL";
    })(Fields24 = Occ.Fields24 || (Occ.Fields24 = {}));
    /**
     * Defines values for Fields25.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields25 = <Fields25>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields25;
    (function (Fields25) {
        Fields25["BASIC"] = "BASIC";
        Fields25["DEFAULT"] = "DEFAULT";
        Fields25["FULL"] = "FULL";
    })(Fields25 = Occ.Fields25 || (Occ.Fields25 = {}));
    /**
     * Defines values for Fields26.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields26 = <Fields26>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields26;
    (function (Fields26) {
        Fields26["BASIC"] = "BASIC";
        Fields26["DEFAULT"] = "DEFAULT";
        Fields26["FULL"] = "FULL";
    })(Fields26 = Occ.Fields26 || (Occ.Fields26 = {}));
    /**
     * Defines values for Fields27.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields27 = <Fields27>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields27;
    (function (Fields27) {
        Fields27["BASIC"] = "BASIC";
        Fields27["DEFAULT"] = "DEFAULT";
        Fields27["FULL"] = "FULL";
    })(Fields27 = Occ.Fields27 || (Occ.Fields27 = {}));
    /**
     * Defines values for Fields28.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields28 = <Fields28>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields28;
    (function (Fields28) {
        Fields28["BASIC"] = "BASIC";
        Fields28["DEFAULT"] = "DEFAULT";
        Fields28["FULL"] = "FULL";
    })(Fields28 = Occ.Fields28 || (Occ.Fields28 = {}));
    /**
     * Defines values for Fields29.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields29 = <Fields29>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields29;
    (function (Fields29) {
        Fields29["BASIC"] = "BASIC";
        Fields29["DEFAULT"] = "DEFAULT";
        Fields29["FULL"] = "FULL";
    })(Fields29 = Occ.Fields29 || (Occ.Fields29 = {}));
    /**
     * Defines values for Fields30.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields30 = <Fields30>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields30;
    (function (Fields30) {
        Fields30["BASIC"] = "BASIC";
        Fields30["DEFAULT"] = "DEFAULT";
        Fields30["FULL"] = "FULL";
    })(Fields30 = Occ.Fields30 || (Occ.Fields30 = {}));
    /**
     * Defines values for Fields31.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields31 = <Fields31>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields31;
    (function (Fields31) {
        Fields31["BASIC"] = "BASIC";
        Fields31["DEFAULT"] = "DEFAULT";
        Fields31["FULL"] = "FULL";
    })(Fields31 = Occ.Fields31 || (Occ.Fields31 = {}));
    /**
     * Defines values for Fields32.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields32 = <Fields32>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields32;
    (function (Fields32) {
        Fields32["BASIC"] = "BASIC";
        Fields32["DEFAULT"] = "DEFAULT";
        Fields32["FULL"] = "FULL";
    })(Fields32 = Occ.Fields32 || (Occ.Fields32 = {}));
    /**
     * Defines values for Fields33.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields33 = <Fields33>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields33;
    (function (Fields33) {
        Fields33["BASIC"] = "BASIC";
        Fields33["DEFAULT"] = "DEFAULT";
        Fields33["FULL"] = "FULL";
    })(Fields33 = Occ.Fields33 || (Occ.Fields33 = {}));
    /**
     * Defines values for Fields34.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields34 = <Fields34>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields34;
    (function (Fields34) {
        Fields34["BASIC"] = "BASIC";
        Fields34["DEFAULT"] = "DEFAULT";
        Fields34["FULL"] = "FULL";
    })(Fields34 = Occ.Fields34 || (Occ.Fields34 = {}));
    /**
     * Defines values for Fields35.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields35 = <Fields35>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields35;
    (function (Fields35) {
        Fields35["BASIC"] = "BASIC";
        Fields35["DEFAULT"] = "DEFAULT";
        Fields35["FULL"] = "FULL";
    })(Fields35 = Occ.Fields35 || (Occ.Fields35 = {}));
    /**
     * Defines values for Fields36.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields36 = <Fields36>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields36;
    (function (Fields36) {
        Fields36["BASIC"] = "BASIC";
        Fields36["DEFAULT"] = "DEFAULT";
        Fields36["FULL"] = "FULL";
    })(Fields36 = Occ.Fields36 || (Occ.Fields36 = {}));
    /**
     * Defines values for Fields37.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields37 = <Fields37>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields37;
    (function (Fields37) {
        Fields37["BASIC"] = "BASIC";
        Fields37["DEFAULT"] = "DEFAULT";
        Fields37["FULL"] = "FULL";
    })(Fields37 = Occ.Fields37 || (Occ.Fields37 = {}));
    /**
     * Defines values for Fields38.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields38 = <Fields38>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields38;
    (function (Fields38) {
        Fields38["BASIC"] = "BASIC";
        Fields38["DEFAULT"] = "DEFAULT";
        Fields38["FULL"] = "FULL";
    })(Fields38 = Occ.Fields38 || (Occ.Fields38 = {}));
    /**
     * Defines values for Fields39.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields39 = <Fields39>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields39;
    (function (Fields39) {
        Fields39["BASIC"] = "BASIC";
        Fields39["DEFAULT"] = "DEFAULT";
        Fields39["FULL"] = "FULL";
    })(Fields39 = Occ.Fields39 || (Occ.Fields39 = {}));
    /**
     * Defines values for Fields40.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields40 = <Fields40>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields40;
    (function (Fields40) {
        Fields40["BASIC"] = "BASIC";
        Fields40["DEFAULT"] = "DEFAULT";
        Fields40["FULL"] = "FULL";
    })(Fields40 = Occ.Fields40 || (Occ.Fields40 = {}));
    /**
     * Defines values for Fields41.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields41 = <Fields41>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields41;
    (function (Fields41) {
        Fields41["BASIC"] = "BASIC";
        Fields41["DEFAULT"] = "DEFAULT";
        Fields41["FULL"] = "FULL";
    })(Fields41 = Occ.Fields41 || (Occ.Fields41 = {}));
    /**
     * Defines values for Fields42.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields42 = <Fields42>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields42;
    (function (Fields42) {
        Fields42["BASIC"] = "BASIC";
        Fields42["DEFAULT"] = "DEFAULT";
        Fields42["FULL"] = "FULL";
    })(Fields42 = Occ.Fields42 || (Occ.Fields42 = {}));
    /**
     * Defines values for Fields43.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields43 = <Fields43>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields43;
    (function (Fields43) {
        Fields43["BASIC"] = "BASIC";
        Fields43["DEFAULT"] = "DEFAULT";
        Fields43["FULL"] = "FULL";
    })(Fields43 = Occ.Fields43 || (Occ.Fields43 = {}));
    /**
     * Defines values for Fields44.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields44 = <Fields44>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields44;
    (function (Fields44) {
        Fields44["BASIC"] = "BASIC";
        Fields44["DEFAULT"] = "DEFAULT";
        Fields44["FULL"] = "FULL";
    })(Fields44 = Occ.Fields44 || (Occ.Fields44 = {}));
    /**
     * Defines values for Fields45.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields45 = <Fields45>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields45;
    (function (Fields45) {
        Fields45["BASIC"] = "BASIC";
        Fields45["DEFAULT"] = "DEFAULT";
        Fields45["FULL"] = "FULL";
    })(Fields45 = Occ.Fields45 || (Occ.Fields45 = {}));
    /**
     * Defines values for Fields46.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields46 = <Fields46>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields46;
    (function (Fields46) {
        Fields46["BASIC"] = "BASIC";
        Fields46["DEFAULT"] = "DEFAULT";
        Fields46["FULL"] = "FULL";
    })(Fields46 = Occ.Fields46 || (Occ.Fields46 = {}));
    /**
     * Defines values for Fields47.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields47 = <Fields47>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields47;
    (function (Fields47) {
        Fields47["BASIC"] = "BASIC";
        Fields47["DEFAULT"] = "DEFAULT";
        Fields47["FULL"] = "FULL";
    })(Fields47 = Occ.Fields47 || (Occ.Fields47 = {}));
    /**
     * Defines values for Fields48.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields48 = <Fields48>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields48;
    (function (Fields48) {
        Fields48["BASIC"] = "BASIC";
        Fields48["DEFAULT"] = "DEFAULT";
        Fields48["FULL"] = "FULL";
    })(Fields48 = Occ.Fields48 || (Occ.Fields48 = {}));
    /**
     * Defines values for Fields49.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields49 = <Fields49>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields49;
    (function (Fields49) {
        Fields49["BASIC"] = "BASIC";
        Fields49["DEFAULT"] = "DEFAULT";
        Fields49["FULL"] = "FULL";
    })(Fields49 = Occ.Fields49 || (Occ.Fields49 = {}));
    /**
     * Defines values for Fields50.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields50 = <Fields50>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields50;
    (function (Fields50) {
        Fields50["BASIC"] = "BASIC";
        Fields50["DEFAULT"] = "DEFAULT";
        Fields50["FULL"] = "FULL";
    })(Fields50 = Occ.Fields50 || (Occ.Fields50 = {}));
    /**
     * Defines values for Fields51.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields51 = <Fields51>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields51;
    (function (Fields51) {
        Fields51["BASIC"] = "BASIC";
        Fields51["DEFAULT"] = "DEFAULT";
        Fields51["FULL"] = "FULL";
    })(Fields51 = Occ.Fields51 || (Occ.Fields51 = {}));
    /**
     * Defines values for Fields52.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields52 = <Fields52>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields52;
    (function (Fields52) {
        Fields52["BASIC"] = "BASIC";
        Fields52["DEFAULT"] = "DEFAULT";
        Fields52["FULL"] = "FULL";
    })(Fields52 = Occ.Fields52 || (Occ.Fields52 = {}));
    /**
     * Defines values for Fields53.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields53 = <Fields53>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields53;
    (function (Fields53) {
        Fields53["BASIC"] = "BASIC";
        Fields53["DEFAULT"] = "DEFAULT";
        Fields53["FULL"] = "FULL";
    })(Fields53 = Occ.Fields53 || (Occ.Fields53 = {}));
    /**
     * Defines values for Fields54.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields54 = <Fields54>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields54;
    (function (Fields54) {
        Fields54["BASIC"] = "BASIC";
        Fields54["DEFAULT"] = "DEFAULT";
        Fields54["FULL"] = "FULL";
    })(Fields54 = Occ.Fields54 || (Occ.Fields54 = {}));
    /**
     * Defines values for Fields55.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields55 = <Fields55>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields55;
    (function (Fields55) {
        Fields55["BASIC"] = "BASIC";
        Fields55["DEFAULT"] = "DEFAULT";
        Fields55["FULL"] = "FULL";
    })(Fields55 = Occ.Fields55 || (Occ.Fields55 = {}));
    /**
     * Defines values for Fields56.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields56 = <Fields56>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields56;
    (function (Fields56) {
        Fields56["BASIC"] = "BASIC";
        Fields56["DEFAULT"] = "DEFAULT";
        Fields56["FULL"] = "FULL";
    })(Fields56 = Occ.Fields56 || (Occ.Fields56 = {}));
    /**
     * Defines values for Fields57.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields57 = <Fields57>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields57;
    (function (Fields57) {
        Fields57["BASIC"] = "BASIC";
        Fields57["DEFAULT"] = "DEFAULT";
        Fields57["FULL"] = "FULL";
    })(Fields57 = Occ.Fields57 || (Occ.Fields57 = {}));
    /**
     * Defines values for Fields58.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields58 = <Fields58>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields58;
    (function (Fields58) {
        Fields58["BASIC"] = "BASIC";
        Fields58["DEFAULT"] = "DEFAULT";
        Fields58["FULL"] = "FULL";
    })(Fields58 = Occ.Fields58 || (Occ.Fields58 = {}));
    /**
     * Defines values for Fields59.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields59 = <Fields59>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields59;
    (function (Fields59) {
        Fields59["BASIC"] = "BASIC";
        Fields59["DEFAULT"] = "DEFAULT";
        Fields59["FULL"] = "FULL";
    })(Fields59 = Occ.Fields59 || (Occ.Fields59 = {}));
    /**
     * Defines values for Fields60.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields60 = <Fields60>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields60;
    (function (Fields60) {
        Fields60["BASIC"] = "BASIC";
        Fields60["DEFAULT"] = "DEFAULT";
        Fields60["FULL"] = "FULL";
    })(Fields60 = Occ.Fields60 || (Occ.Fields60 = {}));
    /**
     * Defines values for Fields61.
     * Possible values include: 'BASIC', 'DEFAULT', 'FULL'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Fields61 = <Fields61>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Fields61;
    (function (Fields61) {
        Fields61["BASIC"] = "BASIC";
        Fields61["DEFAULT"] = "DEFAULT";
        Fields61["FULL"] = "FULL";
    })(Fields61 = Occ.Fields61 || (Occ.Fields61 = {}));
    /**
     * Defines values for Type.
     * Possible values include: 'all', 'product', 'order'
     * There could be more values for this enum apart from the ones defined here.If
     * you want to set a value that is not from the known values then you can do
     * the following:
     * let param: Type = <Type>"someUnknownValueThatWillStillBeValid";
     * @readonly
     * @enum {string}
     */
    var Type;
    (function (Type) {
        Type["All"] = "all";
        Type["Product"] = "product";
        Type["Order"] = "order";
    })(Type = Occ.Type || (Occ.Type = {}));
    /**
     * @record
     */
    function AnonymousConsent() { }
    Occ.AnonymousConsent = AnonymousConsent;
    if (false) {
        /** @type {?|undefined} */
        AnonymousConsent.prototype.templateCode;
        /** @type {?|undefined} */
        AnonymousConsent.prototype.version;
        /** @type {?|undefined} */
        AnonymousConsent.prototype.consentState;
    }
    var CONSENT_STATUS;
    (function (CONSENT_STATUS) {
        CONSENT_STATUS["ANONYMOUS_CONSENT_GIVEN"] = "GIVEN";
        CONSENT_STATUS["ANONYMOUS_CONSENT_WITHDRAWN"] = "WITHDRAWN";
    })(CONSENT_STATUS = Occ.CONSENT_STATUS || (Occ.CONSENT_STATUS = {}));
    /**
     * @record
     */
    function ConsentTemplate() { }
    Occ.ConsentTemplate = ConsentTemplate;
    if (false) {
        /** @type {?|undefined} */
        ConsentTemplate.prototype.id;
        /** @type {?|undefined} */
        ConsentTemplate.prototype.name;
        /** @type {?|undefined} */
        ConsentTemplate.prototype.description;
        /** @type {?|undefined} */
        ConsentTemplate.prototype.version;
        /** @type {?|undefined} */
        ConsentTemplate.prototype.currentConsent;
    }
    /**
     * @record
     */
    function Consent() { }
    Occ.Consent = Consent;
    if (false) {
        /** @type {?|undefined} */
        Consent.prototype.code;
        /** @type {?|undefined} */
        Consent.prototype.consentGivenDate;
        /** @type {?|undefined} */
        Consent.prototype.consentWithdrawnDate;
    }
    /**
     * @record
     */
    function ConsentTemplateList() { }
    Occ.ConsentTemplateList = ConsentTemplateList;
    if (false) {
        /** @type {?|undefined} */
        ConsentTemplateList.prototype.consentTemplates;
    }
    /**
     * @record
     */
    function BaseSites() { }
    Occ.BaseSites = BaseSites;
    if (false) {
        /** @type {?|undefined} */
        BaseSites.prototype.baseSites;
    }
    /**
     * @record
     */
    function BaseSite() { }
    Occ.BaseSite = BaseSite;
    if (false) {
        /** @type {?|undefined} */
        BaseSite.prototype.channel;
        /** @type {?|undefined} */
        BaseSite.prototype.defaultLanguage;
        /** @type {?|undefined} */
        BaseSite.prototype.defaultPreviewCatalogId;
        /** @type {?|undefined} */
        BaseSite.prototype.defaultPreviewCategoryCode;
        /** @type {?|undefined} */
        BaseSite.prototype.defaultPreviewProductCode;
        /** @type {?|undefined} */
        BaseSite.prototype.locale;
        /** @type {?|undefined} */
        BaseSite.prototype.name;
        /** @type {?|undefined} */
        BaseSite.prototype.theme;
        /** @type {?|undefined} */
        BaseSite.prototype.uid;
        /** @type {?|undefined} */
        BaseSite.prototype.stores;
        /** @type {?|undefined} */
        BaseSite.prototype.urlPatterns;
        /** @type {?|undefined} */
        BaseSite.prototype.urlEncodingAttributes;
    }
    /**
     * @record
     */
    function BaseStore() { }
    Occ.BaseStore = BaseStore;
    if (false) {
        /** @type {?|undefined} */
        BaseStore.prototype.currencies;
        /** @type {?|undefined} */
        BaseStore.prototype.defaultCurrency;
        /** @type {?|undefined} */
        BaseStore.prototype.languages;
        /** @type {?|undefined} */
        BaseStore.prototype.defaultLanguage;
    }
    /**
     * @record
     */
    function ProductInterestEntry() { }
    Occ.ProductInterestEntry = ProductInterestEntry;
    if (false) {
        /** @type {?|undefined} */
        ProductInterestEntry.prototype.interestType;
        /** @type {?|undefined} */
        ProductInterestEntry.prototype.dateAdded;
        /** @type {?|undefined} */
        ProductInterestEntry.prototype.expirationDate;
    }
    /**
     * @record
     */
    function ProductInterestEntryRelation() { }
    Occ.ProductInterestEntryRelation = ProductInterestEntryRelation;
    if (false) {
        /** @type {?|undefined} */
        ProductInterestEntryRelation.prototype.product;
        /** @type {?|undefined} */
        ProductInterestEntryRelation.prototype.productInterestEntry;
    }
    /**
     * @record
     */
    function ProductInterestSearchResult() { }
    Occ.ProductInterestSearchResult = ProductInterestSearchResult;
    if (false) {
        /** @type {?|undefined} */
        ProductInterestSearchResult.prototype.results;
        /** @type {?|undefined} */
        ProductInterestSearchResult.prototype.sorts;
        /** @type {?|undefined} */
        ProductInterestSearchResult.prototype.pagination;
    }
    var NotificationType;
    (function (NotificationType) {
        NotificationType["BACK_IN_STOCK"] = "BACK_IN_STOCK";
    })(NotificationType = Occ.NotificationType || (Occ.NotificationType = {}));
})(Occ || (Occ = {}));
