// @spartacus/core/esm2015/src/user/store/effects/product-interests.effect.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Observable, of } from 'rxjs';
import { map, switchMap, catchError } from 'rxjs/operators';
import { UserActions } from '../actions/index';
import { UserInterestsConnector } from '../../connectors/interests/user-interests.connector';
import { makeErrorSerializable } from '../../../util/serialization-utils';
export class ProductInterestsEffect {
    /**
     * @param {?} actions$
     * @param {?} userInterestsConnector
     */
    constructor(actions$, userInterestsConnector) {
        this.actions$ = actions$;
        this.userInterestsConnector = userInterestsConnector;
        this.loadProductInteres$ = this.actions$.pipe(ofType(UserActions.LOAD_PRODUCT_INTERESTS), map((/**
         * @param {?} action
         * @return {?}
         */
        (action) => action.payload)), switchMap((/**
         * @param {?} payload
         * @return {?}
         */
        payload => {
            return this.userInterestsConnector
                .getInterests(payload.userId, payload.pageSize, payload.currentPage, payload.sort, payload.productCode, payload.notificationType)
                .pipe(map((/**
             * @param {?} interests
             * @return {?}
             */
            (interests) => {
                return new UserActions.LoadProductInterestsSuccess(interests);
            })), catchError((/**
             * @param {?} error
             * @return {?}
             */
            error => of(new UserActions.LoadProductInterestsFail(makeErrorSerializable(error))))));
        })));
        this.removeProductInterest$ = this.actions$.pipe(ofType(UserActions.REMOVE_PRODUCT_INTEREST), map((/**
         * @param {?} action
         * @return {?}
         */
        (action) => action.payload)), switchMap((/**
         * @param {?} payload
         * @return {?}
         */
        payload => this.userInterestsConnector
            .removeInterest(payload.userId, payload.item)
            .pipe(switchMap((/**
         * @param {?} data
         * @return {?}
         */
        data => [
            new UserActions.LoadProductInterests(payload.singleDelete
                ? {
                    userId: payload.userId,
                    productCode: payload.item.product.code,
                    notificationType: payload.item.productInterestEntry[0].interestType,
                }
                : { userId: payload.userId }),
            new UserActions.RemoveProductInterestSuccess(data),
        ])), catchError((/**
         * @param {?} error
         * @return {?}
         */
        error => of(new UserActions.RemoveProductInterestFail(makeErrorSerializable(error)))))))));
        this.addProductInterest$ = this.actions$.pipe(ofType(UserActions.ADD_PRODUCT_INTEREST), map((/**
         * @param {?} action
         * @return {?}
         */
        (action) => action.payload)), switchMap((/**
         * @param {?} payload
         * @return {?}
         */
        payload => this.userInterestsConnector
            .addInterest(payload.userId, payload.productCode, payload.notificationType)
            .pipe(switchMap((/**
         * @param {?} res
         * @return {?}
         */
        (res) => [
            new UserActions.LoadProductInterests({
                userId: payload.userId,
                productCode: payload.productCode,
                notificationType: payload.notificationType,
            }),
            new UserActions.AddProductInterestSuccess(res),
        ])), catchError((/**
         * @param {?} error
         * @return {?}
         */
        error => of(new UserActions.AddProductInterestFail(makeErrorSerializable(error)))))))));
    }
}
ProductInterestsEffect.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ProductInterestsEffect.ctorParameters = () => [
    { type: Actions },
    { type: UserInterestsConnector }
];
tslib_1.__decorate([
    Effect(),
    tslib_1.__metadata("design:type", Observable)
], ProductInterestsEffect.prototype, "loadProductInteres$", void 0);
tslib_1.__decorate([
    Effect(),
    tslib_1.__metadata("design:type", Observable)
], ProductInterestsEffect.prototype, "removeProductInterest$", void 0);
tslib_1.__decorate([
    Effect(),
    tslib_1.__metadata("design:type", Observable)
], ProductInterestsEffect.prototype, "addProductInterest$", void 0);
if (false) {
    /** @type {?} */
    ProductInterestsEffect.prototype.loadProductInteres$;
    /** @type {?} */
    ProductInterestsEffect.prototype.removeProductInterest$;
    /** @type {?} */
    ProductInterestsEffect.prototype.addProductInterest$;
    /**
     * @type {?}
     * @private
     */
    ProductInterestsEffect.prototype.actions$;
    /**
     * @type {?}
     * @private
     */
    ProductInterestsEffect.prototype.userInterestsConnector;
}
