// @spartacus/core/esm2015/src/cart/store/multi-cart-store.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { StoreModule } from '@ngrx/store';
import { StateModule } from '../../state/state.module';
import { MULTI_CART_FEATURE } from './multi-cart-state';
import { multiCartReducerProvider, multiCartReducerToken, multiCartMetaReducers, } from './reducers/index';
import { ConfigModule } from '../../config/config.module';
import { StorageSyncType } from '../../state/config/state-config';
import { MultiCartEffects } from './effects/multi-cart.effect';
import { EffectsModule } from '@ngrx/effects';
/**
 * @return {?}
 */
export function multiCartStoreConfigFactory() {
    /** @type {?} */
    const config = {
        state: {
            storageSync: {
                keys: {
                    [`${MULTI_CART_FEATURE}.active`]: StorageSyncType.LOCAL_STORAGE,
                },
            },
        },
    };
    return config;
}
export class MultiCartStoreModule {
}
MultiCartStoreModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    StateModule,
                    StoreModule.forFeature(MULTI_CART_FEATURE, multiCartReducerToken, {
                        metaReducers: multiCartMetaReducers,
                    }),
                    EffectsModule.forFeature([MultiCartEffects]),
                    ConfigModule.withConfigFactory(multiCartStoreConfigFactory),
                ],
                providers: [multiCartReducerProvider],
            },] }
];
