// @spartacus/core/esm2015/src/ssr/ng-express-engine-decorator.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SERVER_REQUEST_ORIGIN, SERVER_REQUEST_URL } from './ssr.providers';
/**
 * These are the allowed options for the engine
 * @record
 */
export function NgSetupOptions() { }
if (false) {
    /** @type {?} */
    NgSetupOptions.prototype.bootstrap;
    /** @type {?|undefined} */
    NgSetupOptions.prototype.providers;
}
/**
 * These are the allowed options for the render
 * @record
 */
export function RenderOptions() { }
if (false) {
    /** @type {?} */
    RenderOptions.prototype.req;
    /** @type {?|undefined} */
    RenderOptions.prototype.res;
    /** @type {?|undefined} */
    RenderOptions.prototype.url;
    /** @type {?|undefined} */
    RenderOptions.prototype.document;
}
/**
 * The wrapper over the standard ngExpressEngine, that provides tokens for Spartacus
 * @param ngExpressEngine
 */
export class NgExpressEngineDecorator {
    /**
     * Returns the higher order ngExpressEngine with provided tokens for Spartacus
     *
     * @param {?} ngExpressEngine
     * @return {?}
     */
    static get(ngExpressEngine) {
        /** @type {?} */
        const result = (/**
         * @param {?} setupOptions
         * @return {?}
         */
        function cxNgExpressEngine(setupOptions) {
            return (/**
             * @param {?} filePath
             * @param {?} options
             * @param {?} callback
             * @return {?}
             */
            (filePath, options, callback) => {
                /** @type {?} */
                const engineInstance = ngExpressEngine(Object.assign({}, setupOptions, { providers: [
                        ...getServerRequestProviders(options),
                        ...(setupOptions.providers || []),
                    ] }));
                return engineInstance(filePath, options, callback);
            });
        });
        return result;
    }
}
/**
 * Returns Spartacus' providers to be passed to the Angular express engine (in SSR)
 *
 * @param {?} options
 * @return {?}
 */
export function getServerRequestProviders(options) {
    return [
        {
            provide: SERVER_REQUEST_URL,
            useValue: getRequestUrl(options.req),
        },
        {
            provide: SERVER_REQUEST_ORIGIN,
            useValue: getRequestOrigin(options.req),
        },
    ];
}
/**
 * @param {?} req
 * @return {?}
 */
function getRequestUrl(req) {
    return getRequestOrigin(req) + req.originalUrl;
}
/**
 * @param {?} req
 * @return {?}
 */
function getRequestOrigin(req) {
    return req.protocol + '://' + req.get('host');
}
