// @spartacus/core/esm5/src/checkout/facade/checkout-delivery.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { filter, shareReplay, tap, pluck, withLatestFrom, } from 'rxjs/operators';
import { CartDataService } from '../../cart/facade/cart-data.service';
import { OCC_USER_ID_ANONYMOUS } from '../../occ/utils/occ-constants';
import { getProcessStateFactory } from '../../process/store/selectors/process-group.selectors';
import { CheckoutActions } from '../store/actions/index';
import { SET_DELIVERY_ADDRESS_PROCESS_ID, SET_DELIVERY_MODE_PROCESS_ID, SET_SUPPORTED_DELIVERY_MODE_PROCESS_ID, } from '../store/checkout-state';
import { CheckoutSelectors } from '../store/selectors/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "../../cart/facade/cart-data.service";
var CheckoutDeliveryService = /** @class */ (function () {
    function CheckoutDeliveryService(checkoutStore, cartData) {
        this.checkoutStore = checkoutStore;
        this.cartData = cartData;
    }
    /**
     * Get supported delivery modes
     */
    /**
     * Get supported delivery modes
     * @return {?}
     */
    CheckoutDeliveryService.prototype.getSupportedDeliveryModes = /**
     * Get supported delivery modes
     * @return {?}
     */
    function () {
        var _this = this;
        return this.checkoutStore.pipe(select(CheckoutSelectors.getSupportedDeliveryModes), withLatestFrom(this.checkoutStore.pipe(select(getProcessStateFactory(SET_SUPPORTED_DELIVERY_MODE_PROCESS_ID)))), tap((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), loadingState = _b[1];
            if (!(loadingState.loading || loadingState.success || loadingState.error)) {
                _this.loadSupportedDeliveryModes();
            }
        })), pluck(0), shareReplay({ bufferSize: 1, refCount: true }));
    };
    /**
     * Get selected delivery mode
     */
    /**
     * Get selected delivery mode
     * @return {?}
     */
    CheckoutDeliveryService.prototype.getSelectedDeliveryMode = /**
     * Get selected delivery mode
     * @return {?}
     */
    function () {
        return this.checkoutStore.pipe(select(CheckoutSelectors.getSelectedDeliveryMode));
    };
    /**
     * Get selected delivery mode code
     */
    /**
     * Get selected delivery mode code
     * @return {?}
     */
    CheckoutDeliveryService.prototype.getSelectedDeliveryModeCode = /**
     * Get selected delivery mode code
     * @return {?}
     */
    function () {
        return this.checkoutStore.pipe(select(CheckoutSelectors.getSelectedDeliveryModeCode));
    };
    /**
     * Get delivery address
     */
    /**
     * Get delivery address
     * @return {?}
     */
    CheckoutDeliveryService.prototype.getDeliveryAddress = /**
     * Get delivery address
     * @return {?}
     */
    function () {
        return this.checkoutStore.pipe(select(CheckoutSelectors.getDeliveryAddress));
    };
    /**
     * Get status about successfully set Delivery Address
     */
    /**
     * Get status about successfully set Delivery Address
     * @return {?}
     */
    CheckoutDeliveryService.prototype.getSetDeliveryAddressProcess = /**
     * Get status about successfully set Delivery Address
     * @return {?}
     */
    function () {
        return this.checkoutStore.pipe(select(getProcessStateFactory(SET_DELIVERY_ADDRESS_PROCESS_ID)));
    };
    /**
     * Clear info about process of setting Delivery Address
     */
    /**
     * Clear info about process of setting Delivery Address
     * @return {?}
     */
    CheckoutDeliveryService.prototype.resetSetDeliveryAddressProcess = /**
     * Clear info about process of setting Delivery Address
     * @return {?}
     */
    function () {
        this.checkoutStore.dispatch(new CheckoutActions.ResetSetDeliveryAddressProcess());
    };
    /**
     * Get status about of set Delivery Mode process
     */
    /**
     * Get status about of set Delivery Mode process
     * @return {?}
     */
    CheckoutDeliveryService.prototype.getSetDeliveryModeProcess = /**
     * Get status about of set Delivery Mode process
     * @return {?}
     */
    function () {
        return this.checkoutStore.pipe(select(getProcessStateFactory(SET_DELIVERY_MODE_PROCESS_ID)));
    };
    /**
     * Clear info about process of setting Delivery Mode
     */
    /**
     * Clear info about process of setting Delivery Mode
     * @return {?}
     */
    CheckoutDeliveryService.prototype.resetSetDeliveryModeProcess = /**
     * Clear info about process of setting Delivery Mode
     * @return {?}
     */
    function () {
        this.checkoutStore.dispatch(new CheckoutActions.ResetSetDeliveryModeProcess());
    };
    /**
     * Clear info about process of setting Supported Delivery Modes
     */
    /**
     * Clear info about process of setting Supported Delivery Modes
     * @return {?}
     */
    CheckoutDeliveryService.prototype.resetLoadSupportedDeliveryModesProcess = /**
     * Clear info about process of setting Supported Delivery Modes
     * @return {?}
     */
    function () {
        this.checkoutStore.dispatch(new CheckoutActions.ResetLoadSupportedDeliveryModesProcess());
    };
    /**
     * Get status about of set supported Delivery Modes process
     */
    /**
     * Get status about of set supported Delivery Modes process
     * @return {?}
     */
    CheckoutDeliveryService.prototype.getLoadSupportedDeliveryModeProcess = /**
     * Get status about of set supported Delivery Modes process
     * @return {?}
     */
    function () {
        return this.checkoutStore.pipe(select(getProcessStateFactory(SET_SUPPORTED_DELIVERY_MODE_PROCESS_ID)));
    };
    /**
     * Clear supported delivery modes loaded in last checkout process
     */
    /**
     * Clear supported delivery modes loaded in last checkout process
     * @return {?}
     */
    CheckoutDeliveryService.prototype.clearCheckoutDeliveryModes = /**
     * Clear supported delivery modes loaded in last checkout process
     * @return {?}
     */
    function () {
        this.checkoutStore.dispatch(new CheckoutActions.ClearSupportedDeliveryModes());
    };
    /**
     * Get address verification results
     */
    /**
     * Get address verification results
     * @return {?}
     */
    CheckoutDeliveryService.prototype.getAddressVerificationResults = /**
     * Get address verification results
     * @return {?}
     */
    function () {
        return this.checkoutStore.pipe(select(CheckoutSelectors.getAddressVerificationResults), filter((/**
         * @param {?} results
         * @return {?}
         */
        function (results) { return Object.keys(results).length !== 0; })));
    };
    /**
     * Create and set a delivery address using the address param
     * @param address : the Address to be created and set
     */
    /**
     * Create and set a delivery address using the address param
     * @param {?} address : the Address to be created and set
     * @return {?}
     */
    CheckoutDeliveryService.prototype.createAndSetAddress = /**
     * Create and set a delivery address using the address param
     * @param {?} address : the Address to be created and set
     * @return {?}
     */
    function (address) {
        if (this.actionAllowed()) {
            this.checkoutStore.dispatch(new CheckoutActions.AddDeliveryAddress({
                userId: this.cartData.userId,
                cartId: this.cartData.cartId,
                address: address,
            }));
        }
    };
    /**
     * Load supported delivery modes
     */
    /**
     * Load supported delivery modes
     * @return {?}
     */
    CheckoutDeliveryService.prototype.loadSupportedDeliveryModes = /**
     * Load supported delivery modes
     * @return {?}
     */
    function () {
        if (this.actionAllowed()) {
            this.checkoutStore.dispatch(new CheckoutActions.LoadSupportedDeliveryModes({
                userId: this.cartData.userId,
                cartId: this.cartData.cartId,
            }));
        }
    };
    /**
     * Set delivery mode
     * @param mode : The delivery mode to be set
     */
    /**
     * Set delivery mode
     * @param {?} mode : The delivery mode to be set
     * @return {?}
     */
    CheckoutDeliveryService.prototype.setDeliveryMode = /**
     * Set delivery mode
     * @param {?} mode : The delivery mode to be set
     * @return {?}
     */
    function (mode) {
        if (this.actionAllowed()) {
            this.checkoutStore.dispatch(new CheckoutActions.SetDeliveryMode({
                userId: this.cartData.userId,
                cartId: this.cartData.cartId,
                selectedModeId: mode,
            }));
        }
    };
    /**
     * Verifies the address
     * @param address : the address to be verified
     */
    /**
     * Verifies the address
     * @param {?} address : the address to be verified
     * @return {?}
     */
    CheckoutDeliveryService.prototype.verifyAddress = /**
     * Verifies the address
     * @param {?} address : the address to be verified
     * @return {?}
     */
    function (address) {
        if (this.actionAllowed()) {
            this.checkoutStore.dispatch(new CheckoutActions.VerifyAddress({
                userId: this.cartData.userId,
                address: address,
            }));
        }
    };
    /**
     * Set delivery address
     * @param address : The address to be set
     */
    /**
     * Set delivery address
     * @param {?} address : The address to be set
     * @return {?}
     */
    CheckoutDeliveryService.prototype.setDeliveryAddress = /**
     * Set delivery address
     * @param {?} address : The address to be set
     * @return {?}
     */
    function (address) {
        if (this.actionAllowed()) {
            this.checkoutStore.dispatch(new CheckoutActions.SetDeliveryAddress({
                userId: this.cartData.userId,
                cartId: this.cartData.cart.code,
                address: address,
            }));
        }
    };
    /**
     * Clear address verification results
     */
    /**
     * Clear address verification results
     * @return {?}
     */
    CheckoutDeliveryService.prototype.clearAddressVerificationResults = /**
     * Clear address verification results
     * @return {?}
     */
    function () {
        this.checkoutStore.dispatch(new CheckoutActions.ClearAddressVerificationResults());
    };
    /**
     * Clear address already setup in last checkout process
     */
    /**
     * Clear address already setup in last checkout process
     * @return {?}
     */
    CheckoutDeliveryService.prototype.clearCheckoutDeliveryAddress = /**
     * Clear address already setup in last checkout process
     * @return {?}
     */
    function () {
        this.checkoutStore.dispatch(new CheckoutActions.ClearCheckoutDeliveryAddress({
            userId: this.cartData.userId,
            cartId: this.cartData.cartId,
        }));
    };
    /**
     * Clear selected delivery mode setup in last checkout process
     */
    /**
     * Clear selected delivery mode setup in last checkout process
     * @return {?}
     */
    CheckoutDeliveryService.prototype.clearCheckoutDeliveryMode = /**
     * Clear selected delivery mode setup in last checkout process
     * @return {?}
     */
    function () {
        this.checkoutStore.dispatch(new CheckoutActions.ClearCheckoutDeliveryMode({
            userId: this.cartData.userId,
            cartId: this.cartData.cartId,
        }));
    };
    /**
     * Clear address and delivery mode already setup in last checkout process
     */
    /**
     * Clear address and delivery mode already setup in last checkout process
     * @return {?}
     */
    CheckoutDeliveryService.prototype.clearCheckoutDeliveryDetails = /**
     * Clear address and delivery mode already setup in last checkout process
     * @return {?}
     */
    function () {
        this.clearCheckoutDeliveryAddress();
        this.clearCheckoutDeliveryMode();
        this.clearCheckoutDeliveryModes();
    };
    /**
     * @protected
     * @return {?}
     */
    CheckoutDeliveryService.prototype.actionAllowed = /**
     * @protected
     * @return {?}
     */
    function () {
        return (this.cartData.userId !== OCC_USER_ID_ANONYMOUS ||
            this.cartData.isGuestCart);
    };
    CheckoutDeliveryService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    CheckoutDeliveryService.ctorParameters = function () { return [
        { type: Store },
        { type: CartDataService }
    ]; };
    /** @nocollapse */ CheckoutDeliveryService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function CheckoutDeliveryService_Factory() { return new CheckoutDeliveryService(i0.ɵɵinject(i1.Store), i0.ɵɵinject(i2.CartDataService)); }, token: CheckoutDeliveryService, providedIn: "root" });
    return CheckoutDeliveryService;
}());
export { CheckoutDeliveryService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    CheckoutDeliveryService.prototype.checkoutStore;
    /**
     * @type {?}
     * @protected
     */
    CheckoutDeliveryService.prototype.cartData;
}
