// @spartacus/core/esm5/src/asm/store/actions/csagent-token.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { StateLoaderActions } from '../../../state/utils/index';
import { CSAGENT_TOKEN_DATA } from '../asm-state';
import { LOGOUT_CUSTOMER_SUPPORT_AGENT } from '../../../auth/store/actions/login-logout.action';
/** @type {?} */
export var LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN = '[Auth] Load Customer Service Agent Token';
/** @type {?} */
export var LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN_FAIL = '[Auth] Load Customer Service Agent Token Fail';
/** @type {?} */
export var LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN_SUCCESS = '[Auth] Load Customer Service Agent Token Success';
var LoadCustomerSupportAgentToken = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCustomerSupportAgentToken, _super);
    function LoadCustomerSupportAgentToken(payload) {
        var _this = _super.call(this, CSAGENT_TOKEN_DATA) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN;
        return _this;
    }
    return LoadCustomerSupportAgentToken;
}(StateLoaderActions.LoaderLoadAction));
export { LoadCustomerSupportAgentToken };
if (false) {
    /** @type {?} */
    LoadCustomerSupportAgentToken.prototype.type;
    /** @type {?} */
    LoadCustomerSupportAgentToken.prototype.payload;
}
var LoadCustomerSupportAgentTokenFail = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCustomerSupportAgentTokenFail, _super);
    function LoadCustomerSupportAgentTokenFail(payload) {
        var _this = _super.call(this, CSAGENT_TOKEN_DATA) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN_FAIL;
        return _this;
    }
    return LoadCustomerSupportAgentTokenFail;
}(StateLoaderActions.LoaderFailAction));
export { LoadCustomerSupportAgentTokenFail };
if (false) {
    /** @type {?} */
    LoadCustomerSupportAgentTokenFail.prototype.type;
    /** @type {?} */
    LoadCustomerSupportAgentTokenFail.prototype.payload;
}
var LoadCustomerSupportAgentTokenSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCustomerSupportAgentTokenSuccess, _super);
    function LoadCustomerSupportAgentTokenSuccess(payload) {
        var _this = _super.call(this, CSAGENT_TOKEN_DATA) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_SUPPORT_AGENT_TOKEN_SUCCESS;
        return _this;
    }
    return LoadCustomerSupportAgentTokenSuccess;
}(StateLoaderActions.LoaderSuccessAction));
export { LoadCustomerSupportAgentTokenSuccess };
if (false) {
    /** @type {?} */
    LoadCustomerSupportAgentTokenSuccess.prototype.type;
    /** @type {?} */
    LoadCustomerSupportAgentTokenSuccess.prototype.payload;
}
var LogoutCustomerSupportAgent = /** @class */ (function () {
    function LogoutCustomerSupportAgent() {
        this.type = LOGOUT_CUSTOMER_SUPPORT_AGENT;
    }
    return LogoutCustomerSupportAgent;
}());
export { LogoutCustomerSupportAgent };
if (false) {
    /** @type {?} */
    LogoutCustomerSupportAgent.prototype.type;
}
