// @spartacus/core/esm5/src/state/utils/processes-loader/processes-loader.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { resetMeta as loaderResetMeta, } from '../loader/loader.action';
/** @type {?} */
export var PROCESSES_INCREMENT_ACTION = '[PROCESSES LOADER] INCREMENT';
/** @type {?} */
export var PROCESSES_DECREMENT_ACTION = '[PROCESSES LOADER] DECREMENT';
/** @type {?} */
export var PROCESSES_LOADER_RESET_ACTION = '[PROCESSES LOADER] RESET';
/**
 * @record
 */
export function ProcessesLoaderMeta() { }
if (false) {
    /** @type {?} */
    ProcessesLoaderMeta.prototype.entityType;
    /** @type {?|undefined} */
    ProcessesLoaderMeta.prototype.processesCountDiff;
}
/**
 * @record
 */
export function ProcessesLoaderAction() { }
if (false) {
    /** @type {?|undefined} */
    ProcessesLoaderAction.prototype.payload;
    /** @type {?|undefined} */
    ProcessesLoaderAction.prototype.meta;
}
/**
 * @param {?} entityType
 * @return {?}
 */
export function processesIncrementMeta(entityType) {
    return {
        entityType: entityType,
        loader: undefined,
        processesCountDiff: 1,
    };
}
/**
 * @param {?} entityType
 * @return {?}
 */
export function processesDecrementMeta(entityType) {
    return {
        entityType: entityType,
        loader: undefined,
        processesCountDiff: -1,
    };
}
/**
 * @param {?} entityType
 * @return {?}
 */
export function processesLoaderResetMeta(entityType) {
    // processes reset action is a reset action for loader reducer, but not the other way around
    return tslib_1.__assign({}, loaderResetMeta(entityType), { processesCountDiff: null });
}
var ProcessesLoaderResetAction = /** @class */ (function () {
    function ProcessesLoaderResetAction(entityType) {
        this.type = PROCESSES_LOADER_RESET_ACTION;
        this.meta = processesLoaderResetMeta(entityType);
    }
    return ProcessesLoaderResetAction;
}());
export { ProcessesLoaderResetAction };
if (false) {
    /** @type {?} */
    ProcessesLoaderResetAction.prototype.type;
    /** @type {?} */
    ProcessesLoaderResetAction.prototype.meta;
}
var ProcessesIncrementAction = /** @class */ (function () {
    function ProcessesIncrementAction(entityType) {
        this.type = PROCESSES_INCREMENT_ACTION;
        this.meta = processesIncrementMeta(entityType);
    }
    return ProcessesIncrementAction;
}());
export { ProcessesIncrementAction };
if (false) {
    /** @type {?} */
    ProcessesIncrementAction.prototype.type;
    /** @type {?} */
    ProcessesIncrementAction.prototype.meta;
}
var ProcessesDecrementAction = /** @class */ (function () {
    function ProcessesDecrementAction(entityType) {
        this.type = PROCESSES_DECREMENT_ACTION;
        this.meta = processesDecrementMeta(entityType);
    }
    return ProcessesDecrementAction;
}());
export { ProcessesDecrementAction };
if (false) {
    /** @type {?} */
    ProcessesDecrementAction.prototype.type;
    /** @type {?} */
    ProcessesDecrementAction.prototype.meta;
}
