// @spartacus/core/esm5/src/cart/store/actions/cart-voucher.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { PROCESS_FEATURE } from '../../../process/store/process-state';
import { EntityFailAction, EntityLoadAction, EntityResetAction, EntitySuccessAction, } from '../../../state/utils/entity-loader/entity-loader.action';
import { LoaderFailAction, LoaderLoadAction, LoaderSuccessAction, } from '../../../state/utils/loader/loader.action';
import { ADD_VOUCHER_PROCESS_ID, CART_DATA } from '../cart-state';
/** @type {?} */
export var CART_ADD_VOUCHER = '[Cart-voucher] Add Cart Vouchers';
/** @type {?} */
export var CART_ADD_VOUCHER_FAIL = '[Cart-voucher] Add Cart Voucher Fail';
/** @type {?} */
export var CART_ADD_VOUCHER_SUCCESS = '[Cart-voucher] Add Cart Voucher Success';
/** @type {?} */
export var CART_RESET_ADD_VOUCHER = '[Cart-voucher] Reset Add Cart Voucher';
/** @type {?} */
export var CART_REMOVE_VOUCHER = '[Cart-voucher] Remove Cart Voucher';
/** @type {?} */
export var CART_REMOVE_VOUCHER_FAIL = '[Cart-voucher] Remove Cart Voucher Fail';
/** @type {?} */
export var CART_REMOVE_VOUCHER_SUCCESS = '[Cart-voucher] Remove Cart Voucher Success';
// Adding cart voucher actions
var 
// Adding cart voucher actions
CartAddVoucher = /** @class */ (function (_super) {
    tslib_1.__extends(CartAddVoucher, _super);
    function CartAddVoucher(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, ADD_VOUCHER_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = CART_ADD_VOUCHER;
        return _this;
    }
    return CartAddVoucher;
}(EntityLoadAction));
// Adding cart voucher actions
export { CartAddVoucher };
if (false) {
    /** @type {?} */
    CartAddVoucher.prototype.type;
    /** @type {?} */
    CartAddVoucher.prototype.payload;
}
var CartAddVoucherFail = /** @class */ (function (_super) {
    tslib_1.__extends(CartAddVoucherFail, _super);
    function CartAddVoucherFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, ADD_VOUCHER_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = CART_ADD_VOUCHER_FAIL;
        return _this;
    }
    return CartAddVoucherFail;
}(EntityFailAction));
export { CartAddVoucherFail };
if (false) {
    /** @type {?} */
    CartAddVoucherFail.prototype.type;
    /** @type {?} */
    CartAddVoucherFail.prototype.payload;
}
var CartAddVoucherSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(CartAddVoucherSuccess, _super);
    function CartAddVoucherSuccess(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, ADD_VOUCHER_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = CART_ADD_VOUCHER_SUCCESS;
        return _this;
    }
    return CartAddVoucherSuccess;
}(EntitySuccessAction));
export { CartAddVoucherSuccess };
if (false) {
    /** @type {?} */
    CartAddVoucherSuccess.prototype.type;
    /** @type {?} */
    CartAddVoucherSuccess.prototype.payload;
}
var CartResetAddVoucher = /** @class */ (function (_super) {
    tslib_1.__extends(CartResetAddVoucher, _super);
    function CartResetAddVoucher() {
        var _this = _super.call(this, PROCESS_FEATURE, ADD_VOUCHER_PROCESS_ID) || this;
        _this.type = CART_RESET_ADD_VOUCHER;
        return _this;
    }
    return CartResetAddVoucher;
}(EntityResetAction));
export { CartResetAddVoucher };
if (false) {
    /** @type {?} */
    CartResetAddVoucher.prototype.type;
}
// Deleting cart voucher
var 
// Deleting cart voucher
CartRemoveVoucher = /** @class */ (function (_super) {
    tslib_1.__extends(CartRemoveVoucher, _super);
    function CartRemoveVoucher(payload) {
        var _this = _super.call(this, CART_DATA) || this;
        _this.payload = payload;
        _this.type = CART_REMOVE_VOUCHER;
        return _this;
    }
    return CartRemoveVoucher;
}(LoaderLoadAction));
// Deleting cart voucher
export { CartRemoveVoucher };
if (false) {
    /** @type {?} */
    CartRemoveVoucher.prototype.type;
    /** @type {?} */
    CartRemoveVoucher.prototype.payload;
}
var CartRemoveVoucherFail = /** @class */ (function (_super) {
    tslib_1.__extends(CartRemoveVoucherFail, _super);
    function CartRemoveVoucherFail(payload) {
        var _this = _super.call(this, CART_DATA, payload) || this;
        _this.payload = payload;
        _this.type = CART_REMOVE_VOUCHER_FAIL;
        return _this;
    }
    return CartRemoveVoucherFail;
}(LoaderFailAction));
export { CartRemoveVoucherFail };
if (false) {
    /** @type {?} */
    CartRemoveVoucherFail.prototype.type;
    /** @type {?} */
    CartRemoveVoucherFail.prototype.payload;
}
var CartRemoveVoucherSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(CartRemoveVoucherSuccess, _super);
    function CartRemoveVoucherSuccess(payload) {
        var _this = _super.call(this, CART_DATA) || this;
        _this.payload = payload;
        _this.type = CART_REMOVE_VOUCHER_SUCCESS;
        return _this;
    }
    return CartRemoveVoucherSuccess;
}(LoaderSuccessAction));
export { CartRemoveVoucherSuccess };
if (false) {
    /** @type {?} */
    CartRemoveVoucherSuccess.prototype.type;
    /** @type {?} */
    CartRemoveVoucherSuccess.prototype.payload;
}
