// @spartacus/core/esm5/src/site-context/store/selectors/currencies.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { getSiteContextState } from './site-context.selector';
/** @type {?} */
var currenciesEntitiesSelector = (/**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.entities; });
var ɵ0 = currenciesEntitiesSelector;
/** @type {?} */
var activeCurrencySelector = (/**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.activeCurrency; });
var ɵ1 = activeCurrencySelector;
var ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.currencies; };
/** @type {?} */
export var getCurrenciesState = createSelector(getSiteContextState, (ɵ2));
/** @type {?} */
export var getCurrenciesEntities = createSelector(getCurrenciesState, currenciesEntitiesSelector);
/** @type {?} */
export var getActiveCurrency = createSelector(getCurrenciesState, activeCurrencySelector);
var ɵ3 = /**
 * @param {?} entities
 * @return {?}
 */
function (entities) {
    return entities
        ? Object.keys(entities).map((/**
         * @param {?} isocode
         * @return {?}
         */
        function (isocode) { return entities[isocode]; }))
        : null;
};
/** @type {?} */
export var getAllCurrencies = createSelector(getCurrenciesEntities, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
