// @spartacus/core/esm5/src/site-context/store/selectors/languages.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { getSiteContextState } from './site-context.selector';
/** @type {?} */
var activeLanguageSelector = (/**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.activeLanguage; });
var ɵ0 = activeLanguageSelector;
/** @type {?} */
var languagesEntitiesSelector = (/**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.entities; });
var ɵ1 = languagesEntitiesSelector;
var ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.languages; };
/** @type {?} */
export var getLanguagesState = createSelector(getSiteContextState, (ɵ2));
/** @type {?} */
export var getLanguagesEntities = createSelector(getLanguagesState, languagesEntitiesSelector);
/** @type {?} */
export var getActiveLanguage = createSelector(getLanguagesState, activeLanguageSelector);
var ɵ3 = /**
 * @param {?} entities
 * @return {?}
 */
function (entities) {
    return entities
        ? Object.keys(entities).map((/**
         * @param {?} isocode
         * @return {?}
         */
        function (isocode) { return entities[isocode]; }))
        : null;
};
/** @type {?} */
export var getAllLanguages = createSelector(getLanguagesEntities, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
