// @spartacus/core/esm2015/src/model/misc.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Currency() { }
if (false) {
    /** @type {?|undefined} */
    Currency.prototype.active;
    /** @type {?|undefined} */
    Currency.prototype.isocode;
    /** @type {?|undefined} */
    Currency.prototype.name;
    /** @type {?|undefined} */
    Currency.prototype.symbol;
}
/**
 * @record
 */
export function Time() { }
if (false) {
    /** @type {?|undefined} */
    Time.prototype.formattedHour;
    /** @type {?|undefined} */
    Time.prototype.hour;
    /** @type {?|undefined} */
    Time.prototype.minute;
}
/**
 * @record
 */
export function GeoPoint() { }
if (false) {
    /** @type {?|undefined} */
    GeoPoint.prototype.latitude;
    /** @type {?|undefined} */
    GeoPoint.prototype.longitude;
}
/**
 * @record
 */
export function Language() { }
if (false) {
    /** @type {?|undefined} */
    Language.prototype.active;
    /** @type {?|undefined} */
    Language.prototype.isocode;
    /** @type {?|undefined} */
    Language.prototype.name;
    /** @type {?|undefined} */
    Language.prototype.nativeName;
}
/**
 * @record
 */
export function User() { }
if (false) {
    /** @type {?|undefined} */
    User.prototype.currency;
    /** @type {?|undefined} */
    User.prototype.customerId;
    /** @type {?|undefined} */
    User.prototype.deactivationDate;
    /** @type {?|undefined} */
    User.prototype.defaultAddress;
    /** @type {?|undefined} */
    User.prototype.displayUid;
    /** @type {?|undefined} */
    User.prototype.firstName;
    /** @type {?|undefined} */
    User.prototype.language;
    /** @type {?|undefined} */
    User.prototype.lastName;
    /** @type {?|undefined} */
    User.prototype.name;
    /** @type {?|undefined} */
    User.prototype.title;
    /** @type {?|undefined} */
    User.prototype.titleCode;
    /** @type {?|undefined} */
    User.prototype.uid;
}
/**
 * @record
 */
export function PaginationModel() { }
if (false) {
    /** @type {?|undefined} */
    PaginationModel.prototype.currentPage;
    /** @type {?|undefined} */
    PaginationModel.prototype.pageSize;
    /** @type {?|undefined} */
    PaginationModel.prototype.sort;
    /** @type {?|undefined} */
    PaginationModel.prototype.totalPages;
    /** @type {?|undefined} */
    PaginationModel.prototype.totalResults;
}
/**
 * @record
 */
export function SortModel() { }
if (false) {
    /** @type {?|undefined} */
    SortModel.prototype.code;
    /** @type {?|undefined} */
    SortModel.prototype.name;
    /** @type {?|undefined} */
    SortModel.prototype.selected;
}
/**
 * @record
 */
export function Title() { }
if (false) {
    /** @type {?|undefined} */
    Title.prototype.code;
    /** @type {?|undefined} */
    Title.prototype.name;
}
/**
 * @record
 */
export function ErrorModel() { }
if (false) {
    /** @type {?|undefined} */
    ErrorModel.prototype.message;
    /** @type {?|undefined} */
    ErrorModel.prototype.reason;
    /** @type {?|undefined} */
    ErrorModel.prototype.subject;
    /** @type {?|undefined} */
    ErrorModel.prototype.subjectType;
    /** @type {?|undefined} */
    ErrorModel.prototype.type;
}
/**
 * @record
 */
export function HttpErrorModel() { }
if (false) {
    /** @type {?|undefined} */
    HttpErrorModel.prototype.message;
    /** @type {?|undefined} */
    HttpErrorModel.prototype.error;
    /** @type {?|undefined} */
    HttpErrorModel.prototype.status;
    /** @type {?|undefined} */
    HttpErrorModel.prototype.statusText;
    /** @type {?|undefined} */
    HttpErrorModel.prototype.url;
}
/**
 * @record
 */
export function BaseStore() { }
if (false) {
    /** @type {?|undefined} */
    BaseStore.prototype.currencies;
    /** @type {?|undefined} */
    BaseStore.prototype.defaultCurrency;
    /** @type {?|undefined} */
    BaseStore.prototype.languages;
    /** @type {?|undefined} */
    BaseStore.prototype.defaultLanguage;
}
/**
 * @record
 */
export function BaseSite() { }
if (false) {
    /** @type {?|undefined} */
    BaseSite.prototype.channel;
    /** @type {?|undefined} */
    BaseSite.prototype.defaultLanguage;
    /** @type {?|undefined} */
    BaseSite.prototype.defaultPreviewCatalogId;
    /** @type {?|undefined} */
    BaseSite.prototype.defaultPreviewCategoryCode;
    /** @type {?|undefined} */
    BaseSite.prototype.defaultPreviewProductCode;
    /** @type {?|undefined} */
    BaseSite.prototype.locale;
    /** @type {?|undefined} */
    BaseSite.prototype.name;
    /** @type {?|undefined} */
    BaseSite.prototype.theme;
    /** @type {?|undefined} */
    BaseSite.prototype.uid;
    /** @type {?|undefined} */
    BaseSite.prototype.stores;
    /** @type {?|undefined} */
    BaseSite.prototype.urlPatterns;
    /** @type {?|undefined} */
    BaseSite.prototype.urlEncodingAttributes;
}
/**
 * @record
 */
export function UserSignUp() { }
if (false) {
    /** @type {?|undefined} */
    UserSignUp.prototype.firstName;
    /** @type {?|undefined} */
    UserSignUp.prototype.lastName;
    /** @type {?|undefined} */
    UserSignUp.prototype.password;
    /** @type {?|undefined} */
    UserSignUp.prototype.titleCode;
    /** @type {?|undefined} */
    UserSignUp.prototype.uid;
}
