// @spartacus/core/esm2015/src/model/order.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function DeliveryMode() { }
if (false) {
    /** @type {?|undefined} */
    DeliveryMode.prototype.code;
    /** @type {?|undefined} */
    DeliveryMode.prototype.deliveryCost;
    /** @type {?|undefined} */
    DeliveryMode.prototype.description;
    /** @type {?|undefined} */
    DeliveryMode.prototype.name;
}
/**
 * @record
 */
export function OrderEntry() { }
if (false) {
    /** @type {?|undefined} */
    OrderEntry.prototype.basePrice;
    /** @type {?|undefined} */
    OrderEntry.prototype.deliveryMode;
    /** @type {?|undefined} */
    OrderEntry.prototype.deliveryPointOfService;
    /** @type {?|undefined} */
    OrderEntry.prototype.entryNumber;
    /** @type {?|undefined} */
    OrderEntry.prototype.product;
    /** @type {?|undefined} */
    OrderEntry.prototype.quantity;
    /** @type {?|undefined} */
    OrderEntry.prototype.totalPrice;
    /** @type {?|undefined} */
    OrderEntry.prototype.updateable;
    /** @type {?|undefined} */
    OrderEntry.prototype.returnedItemsPrice;
    /** @type {?|undefined} */
    OrderEntry.prototype.returnableQuantity;
    /** @type {?|undefined} */
    OrderEntry.prototype.cancelledItemsPrice;
    /** @type {?|undefined} */
    OrderEntry.prototype.cancellableQuantity;
}
/**
 * @record
 */
export function CancelOrReturnRequestEntryInput() { }
if (false) {
    /** @type {?|undefined} */
    CancelOrReturnRequestEntryInput.prototype.orderEntryNumber;
    /** @type {?|undefined} */
    CancelOrReturnRequestEntryInput.prototype.quantity;
}
/**
 * @record
 */
export function ReturnRequestEntryInputList() { }
if (false) {
    /** @type {?|undefined} */
    ReturnRequestEntryInputList.prototype.orderCode;
    /** @type {?|undefined} */
    ReturnRequestEntryInputList.prototype.returnRequestEntryInputs;
}
/**
 * @record
 */
export function CancellationRequestEntryInputList() { }
if (false) {
    /** @type {?|undefined} */
    CancellationRequestEntryInputList.prototype.cancellationRequestEntryInputs;
}
/**
 * @record
 */
export function ReturnRequestEntry() { }
if (false) {
    /** @type {?|undefined} */
    ReturnRequestEntry.prototype.orderEntry;
    /** @type {?|undefined} */
    ReturnRequestEntry.prototype.expectedQuantity;
    /** @type {?|undefined} */
    ReturnRequestEntry.prototype.refundAmount;
}
/**
 * @record
 */
export function ReturnRequest() { }
if (false) {
    /** @type {?|undefined} */
    ReturnRequest.prototype.cancellable;
    /** @type {?|undefined} */
    ReturnRequest.prototype.code;
    /** @type {?|undefined} */
    ReturnRequest.prototype.creationTime;
    /** @type {?|undefined} */
    ReturnRequest.prototype.deliveryCost;
    /** @type {?|undefined} */
    ReturnRequest.prototype.order;
    /** @type {?|undefined} */
    ReturnRequest.prototype.refundDeliveryCost;
    /** @type {?|undefined} */
    ReturnRequest.prototype.returnEntries;
    /** @type {?|undefined} */
    ReturnRequest.prototype.returnLabelDownloadUrl;
    /** @type {?|undefined} */
    ReturnRequest.prototype.rma;
    /** @type {?|undefined} */
    ReturnRequest.prototype.status;
    /** @type {?|undefined} */
    ReturnRequest.prototype.subTotal;
    /** @type {?|undefined} */
    ReturnRequest.prototype.totalPrice;
}
/**
 * @record
 */
export function ReturnRequestList() { }
if (false) {
    /** @type {?|undefined} */
    ReturnRequestList.prototype.returnRequests;
    /** @type {?|undefined} */
    ReturnRequestList.prototype.pagination;
    /** @type {?|undefined} */
    ReturnRequestList.prototype.sorts;
}
/**
 * @record
 */
export function ReturnRequestModification() { }
if (false) {
    /** @type {?|undefined} */
    ReturnRequestModification.prototype.status;
}
/**
 * @record
 */
export function PickupOrderEntryGroup() { }
if (false) {
    /** @type {?|undefined} */
    PickupOrderEntryGroup.prototype.deliveryPointOfService;
    /** @type {?|undefined} */
    PickupOrderEntryGroup.prototype.distance;
    /** @type {?|undefined} */
    PickupOrderEntryGroup.prototype.entries;
    /** @type {?|undefined} */
    PickupOrderEntryGroup.prototype.quantity;
    /** @type {?|undefined} */
    PickupOrderEntryGroup.prototype.totalPriceWithTax;
}
/**
 * @record
 */
export function PromotionOrderEntryConsumed() { }
if (false) {
    /** @type {?|undefined} */
    PromotionOrderEntryConsumed.prototype.adjustedUnitPrice;
    /** @type {?|undefined} */
    PromotionOrderEntryConsumed.prototype.code;
    /** @type {?|undefined} */
    PromotionOrderEntryConsumed.prototype.orderEntryNumber;
    /** @type {?|undefined} */
    PromotionOrderEntryConsumed.prototype.quantity;
}
/**
 * @record
 */
export function ConsignmentEntry() { }
if (false) {
    /** @type {?|undefined} */
    ConsignmentEntry.prototype.orderEntry;
    /** @type {?|undefined} */
    ConsignmentEntry.prototype.quantity;
    /** @type {?|undefined} */
    ConsignmentEntry.prototype.shippedQuantity;
}
/**
 * @record
 */
export function Consignment() { }
if (false) {
    /** @type {?|undefined} */
    Consignment.prototype.code;
    /** @type {?|undefined} */
    Consignment.prototype.deliveryPointOfService;
    /** @type {?|undefined} */
    Consignment.prototype.entries;
    /** @type {?|undefined} */
    Consignment.prototype.shippingAddress;
    /** @type {?|undefined} */
    Consignment.prototype.status;
    /** @type {?|undefined} */
    Consignment.prototype.statusDate;
    /** @type {?|undefined} */
    Consignment.prototype.trackingID;
}
/**
 * @record
 */
export function OrderHistory() { }
if (false) {
    /** @type {?|undefined} */
    OrderHistory.prototype.code;
    /** @type {?|undefined} */
    OrderHistory.prototype.guid;
    /** @type {?|undefined} */
    OrderHistory.prototype.placed;
    /** @type {?|undefined} */
    OrderHistory.prototype.status;
    /** @type {?|undefined} */
    OrderHistory.prototype.statusDisplay;
    /** @type {?|undefined} */
    OrderHistory.prototype.total;
}
/**
 * @record
 */
export function OrderHistoryList() { }
if (false) {
    /** @type {?|undefined} */
    OrderHistoryList.prototype.orders;
    /** @type {?|undefined} */
    OrderHistoryList.prototype.pagination;
    /** @type {?|undefined} */
    OrderHistoryList.prototype.sorts;
}
/**
 * @record
 */
export function Order() { }
if (false) {
    /** @type {?|undefined} */
    Order.prototype.appliedOrderPromotions;
    /** @type {?|undefined} */
    Order.prototype.appliedProductPromotions;
    /** @type {?|undefined} */
    Order.prototype.appliedVouchers;
    /** @type {?|undefined} */
    Order.prototype.calculated;
    /** @type {?|undefined} */
    Order.prototype.code;
    /** @type {?|undefined} */
    Order.prototype.consignments;
    /** @type {?|undefined} */
    Order.prototype.created;
    /** @type {?|undefined} */
    Order.prototype.deliveryAddress;
    /** @type {?|undefined} */
    Order.prototype.deliveryCost;
    /** @type {?|undefined} */
    Order.prototype.deliveryItemsQuantity;
    /** @type {?|undefined} */
    Order.prototype.deliveryMode;
    /** @type {?|undefined} */
    Order.prototype.deliveryOrderGroups;
    /** @type {?|undefined} */
    Order.prototype.deliveryStatus;
    /** @type {?|undefined} */
    Order.prototype.deliveryStatusDisplay;
    /** @type {?|undefined} */
    Order.prototype.entries;
    /** @type {?|undefined} */
    Order.prototype.guestCustomer;
    /** @type {?|undefined} */
    Order.prototype.guid;
    /** @type {?|undefined} */
    Order.prototype.net;
    /** @type {?|undefined} */
    Order.prototype.orderDiscounts;
    /** @type {?|undefined} */
    Order.prototype.paymentInfo;
    /** @type {?|undefined} */
    Order.prototype.pickupItemsQuantity;
    /** @type {?|undefined} */
    Order.prototype.pickupOrderGroups;
    /** @type {?|undefined} */
    Order.prototype.productDiscounts;
    /** @type {?|undefined} */
    Order.prototype.site;
    /** @type {?|undefined} */
    Order.prototype.status;
    /** @type {?|undefined} */
    Order.prototype.statusDisplay;
    /** @type {?|undefined} */
    Order.prototype.store;
    /** @type {?|undefined} */
    Order.prototype.subTotal;
    /** @type {?|undefined} */
    Order.prototype.totalDiscounts;
    /** @type {?|undefined} */
    Order.prototype.totalItems;
    /** @type {?|undefined} */
    Order.prototype.totalPrice;
    /** @type {?|undefined} */
    Order.prototype.totalPriceWithTax;
    /** @type {?|undefined} */
    Order.prototype.totalTax;
    /** @type {?|undefined} */
    Order.prototype.unconsignedEntries;
    /** @type {?|undefined} */
    Order.prototype.user;
    /** @type {?|undefined} */
    Order.prototype.returnable;
    /** @type {?|undefined} */
    Order.prototype.cancellable;
}
