// @spartacus/core/esm2015/src/model/product.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function VariantOptionQualifier() { }
if (false) {
    /** @type {?|undefined} */
    VariantOptionQualifier.prototype.image;
    /** @type {?|undefined} */
    VariantOptionQualifier.prototype.name;
    /** @type {?|undefined} */
    VariantOptionQualifier.prototype.qualifier;
    /** @type {?|undefined} */
    VariantOptionQualifier.prototype.value;
}
/**
 * @record
 */
export function PromotionRestriction() { }
if (false) {
    /** @type {?|undefined} */
    PromotionRestriction.prototype.description;
    /** @type {?|undefined} */
    PromotionRestriction.prototype.restrictionType;
}
/**
 * @record
 */
export function FeatureUnit() { }
if (false) {
    /** @type {?|undefined} */
    FeatureUnit.prototype.name;
    /** @type {?|undefined} */
    FeatureUnit.prototype.symbol;
    /** @type {?|undefined} */
    FeatureUnit.prototype.unitType;
}
/**
 * @record
 */
export function FeatureValue() { }
if (false) {
    /** @type {?|undefined} */
    FeatureValue.prototype.value;
}
/**
 * @record
 */
export function Feature() { }
if (false) {
    /** @type {?|undefined} */
    Feature.prototype.code;
    /** @type {?|undefined} */
    Feature.prototype.comparable;
    /** @type {?|undefined} */
    Feature.prototype.description;
    /** @type {?|undefined} */
    Feature.prototype.featureUnit;
    /** @type {?|undefined} */
    Feature.prototype.featureValues;
    /** @type {?|undefined} */
    Feature.prototype.name;
    /** @type {?|undefined} */
    Feature.prototype.range;
    /** @type {?|undefined} */
    Feature.prototype.type;
}
/**
 * @record
 */
export function VariantCategory() { }
if (false) {
    /** @type {?|undefined} */
    VariantCategory.prototype.hasImage;
    /** @type {?|undefined} */
    VariantCategory.prototype.name;
    /** @type {?|undefined} */
    VariantCategory.prototype.priority;
}
/**
 * @record
 */
export function VariantValueCategory() { }
if (false) {
    /** @type {?|undefined} */
    VariantValueCategory.prototype.name;
    /** @type {?|undefined} */
    VariantValueCategory.prototype.sequence;
    /** @type {?|undefined} */
    VariantValueCategory.prototype.superCategories;
}
/** @enum {string} */
const VariantType = {
    SIZE: 'ApparelSizeVariantProduct',
    STYLE: 'ApparelStyleVariantProduct',
    COLOR: 'ElectronicsColorVariantProduct',
};
export { VariantType };
/** @enum {string} */
const PriceType = {
    BUY: 'BUY',
    FROM: 'FROM',
};
export { PriceType };
/**
 * @record
 */
export function Price() { }
if (false) {
    /** @type {?|undefined} */
    Price.prototype.currencyIso;
    /** @type {?|undefined} */
    Price.prototype.formattedValue;
    /** @type {?|undefined} */
    Price.prototype.maxQuantity;
    /** @type {?|undefined} */
    Price.prototype.minQuantity;
    /** @type {?|undefined} */
    Price.prototype.priceType;
    /** @type {?|undefined} */
    Price.prototype.value;
}
/**
 * @record
 */
export function Stock() { }
if (false) {
    /** @type {?|undefined} */
    Stock.prototype.stockLevel;
    /** @type {?|undefined} */
    Stock.prototype.stockLevelStatus;
}
/**
 * @record
 */
export function VariantOption() { }
if (false) {
    /** @type {?|undefined} */
    VariantOption.prototype.code;
    /** @type {?|undefined} */
    VariantOption.prototype.priceData;
    /** @type {?|undefined} */
    VariantOption.prototype.stock;
    /** @type {?|undefined} */
    VariantOption.prototype.url;
    /** @type {?|undefined} */
    VariantOption.prototype.variantOptionQualifiers;
}
/**
 * @record
 */
export function Promotion() { }
if (false) {
    /** @type {?|undefined} */
    Promotion.prototype.code;
    /** @type {?|undefined} */
    Promotion.prototype.couldFireMessages;
    /** @type {?|undefined} */
    Promotion.prototype.description;
    /** @type {?|undefined} */
    Promotion.prototype.enabled;
    /** @type {?|undefined} */
    Promotion.prototype.endDate;
    /** @type {?|undefined} */
    Promotion.prototype.firedMessages;
    /** @type {?|undefined} */
    Promotion.prototype.priority;
    /** @type {?|undefined} */
    Promotion.prototype.productBanner;
    /** @type {?|undefined} */
    Promotion.prototype.promotionGroup;
    /** @type {?|undefined} */
    Promotion.prototype.promotionType;
    /** @type {?|undefined} */
    Promotion.prototype.restrictions;
    /** @type {?|undefined} */
    Promotion.prototype.startDate;
    /** @type {?|undefined} */
    Promotion.prototype.title;
}
/**
 * @record
 */
export function Category() { }
if (false) {
    /** @type {?|undefined} */
    Category.prototype.code;
    /** @type {?|undefined} */
    Category.prototype.name;
    /** @type {?|undefined} */
    Category.prototype.image;
    /** @type {?|undefined} */
    Category.prototype.url;
}
/**
 * @record
 */
export function Classification() { }
if (false) {
    /** @type {?|undefined} */
    Classification.prototype.code;
    /** @type {?|undefined} */
    Classification.prototype.features;
    /** @type {?|undefined} */
    Classification.prototype.name;
}
/**
 * @record
 */
export function FutureStock() { }
if (false) {
    /** @type {?|undefined} */
    FutureStock.prototype.date;
    /** @type {?|undefined} */
    FutureStock.prototype.formattedDate;
    /** @type {?|undefined} */
    FutureStock.prototype.stock;
}
/**
 * @record
 */
export function PriceRange() { }
if (false) {
    /** @type {?|undefined} */
    PriceRange.prototype.maxPrice;
    /** @type {?|undefined} */
    PriceRange.prototype.minPrice;
}
/**
 * @record
 */
export function ProductReference() { }
if (false) {
    /** @type {?|undefined} */
    ProductReference.prototype.description;
    /** @type {?|undefined} */
    ProductReference.prototype.preselected;
    /** @type {?|undefined} */
    ProductReference.prototype.quantity;
    /** @type {?|undefined} */
    ProductReference.prototype.referenceType;
    /** @type {?|undefined} */
    ProductReference.prototype.target;
}
/**
 * @record
 */
export function Review() { }
if (false) {
    /** @type {?|undefined} */
    Review.prototype.alias;
    /** @type {?|undefined} */
    Review.prototype.comment;
    /** @type {?|undefined} */
    Review.prototype.date;
    /** @type {?|undefined} */
    Review.prototype.headline;
    /** @type {?|undefined} */
    Review.prototype.id;
    /** @type {?|undefined} */
    Review.prototype.principal;
    /** @type {?|undefined} */
    Review.prototype.rating;
}
/**
 * @record
 */
export function VariantMatrixElement() { }
if (false) {
    /** @type {?|undefined} */
    VariantMatrixElement.prototype.elements;
    /** @type {?|undefined} */
    VariantMatrixElement.prototype.isLeaf;
    /** @type {?|undefined} */
    VariantMatrixElement.prototype.parentVariantCategory;
    /** @type {?|undefined} */
    VariantMatrixElement.prototype.variantOption;
    /** @type {?|undefined} */
    VariantMatrixElement.prototype.variantValueCategory;
}
/**
 * @record
 */
export function ProductReferences() { }
/**
 * @record
 */
export function BaseOption() { }
if (false) {
    /** @type {?|undefined} */
    BaseOption.prototype.options;
    /** @type {?|undefined} */
    BaseOption.prototype.selected;
    /** @type {?|undefined} */
    BaseOption.prototype.variantType;
}
/**
 * @record
 */
export function Product() { }
if (false) {
    /** @type {?|undefined} */
    Product.prototype.availableForPickup;
    /** @type {?|undefined} */
    Product.prototype.averageRating;
    /** @type {?|undefined} */
    Product.prototype.baseOptions;
    /** @type {?|undefined} */
    Product.prototype.baseProduct;
    /** @type {?|undefined} */
    Product.prototype.categories;
    /** @type {?|undefined} */
    Product.prototype.classifications;
    /** @type {?|undefined} */
    Product.prototype.code;
    /** @type {?|undefined} */
    Product.prototype.description;
    /** @type {?|undefined} */
    Product.prototype.futureStocks;
    /** @type {?|undefined} */
    Product.prototype.images;
    /** @type {?|undefined} */
    Product.prototype.manufacturer;
    /** @type {?|undefined} */
    Product.prototype.multidimensional;
    /** @type {?|undefined} */
    Product.prototype.name;
    /** @type {?|undefined} */
    Product.prototype.nameHtml;
    /** @type {?|undefined} */
    Product.prototype.numberOfReviews;
    /** @type {?|undefined} */
    Product.prototype.potentialPromotions;
    /** @type {?|undefined} */
    Product.prototype.price;
    /** @type {?|undefined} */
    Product.prototype.priceRange;
    /** @type {?|undefined} */
    Product.prototype.productReferences;
    /** @type {?|undefined} */
    Product.prototype.purchasable;
    /** @type {?|undefined} */
    Product.prototype.reviews;
    /** @type {?|undefined} */
    Product.prototype.stock;
    /** @type {?|undefined} */
    Product.prototype.summary;
    /** @type {?|undefined} */
    Product.prototype.url;
    /** @type {?|undefined} */
    Product.prototype.variantMatrix;
    /** @type {?|undefined} */
    Product.prototype.variantOptions;
    /** @type {?|undefined} */
    Product.prototype.variantType;
    /** @type {?|undefined} */
    Product.prototype.volumePrices;
    /** @type {?|undefined} */
    Product.prototype.volumePricesFlag;
}
/** @enum {string} */
const VariantQualifier = {
    SIZE: 'size',
    STYLE: 'style',
    COLOR: 'color',
    THUMBNAIL: 'thumbnail',
    PRODUCT: 'product',
    ROLLUP_PROPERTY: 'rollupProperty',
};
export { VariantQualifier };
