// @spartacus/core/esm2015/src/model/unused.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function CategoryHierarchy() { }
if (false) {
    /** @type {?|undefined} */
    CategoryHierarchy.prototype.id;
    /** @type {?|undefined} */
    CategoryHierarchy.prototype.lastModified;
    /** @type {?|undefined} */
    CategoryHierarchy.prototype.name;
    /** @type {?|undefined} */
    CategoryHierarchy.prototype.subcategories;
    /** @type {?|undefined} */
    CategoryHierarchy.prototype.url;
}
/**
 * @record
 */
export function CatalogVersion() { }
if (false) {
    /** @type {?|undefined} */
    CatalogVersion.prototype.categories;
    /** @type {?|undefined} */
    CatalogVersion.prototype.id;
    /** @type {?|undefined} */
    CatalogVersion.prototype.lastModified;
    /** @type {?|undefined} */
    CatalogVersion.prototype.name;
    /** @type {?|undefined} */
    CatalogVersion.prototype.url;
}
/**
 * @record
 */
export function Catalog() { }
if (false) {
    /** @type {?|undefined} */
    Catalog.prototype.catalogVersions;
    /** @type {?|undefined} */
    Catalog.prototype.id;
    /** @type {?|undefined} */
    Catalog.prototype.lastModified;
    /** @type {?|undefined} */
    Catalog.prototype.name;
    /** @type {?|undefined} */
    Catalog.prototype.url;
}
/**
 * @record
 */
export function Pagination() { }
if (false) {
    /** @type {?|undefined} */
    Pagination.prototype.count;
    /** @type {?|undefined} */
    Pagination.prototype.page;
    /** @type {?|undefined} */
    Pagination.prototype.totalCount;
    /** @type {?|undefined} */
    Pagination.prototype.totalPages;
}
/**
 * @record
 */
export function Sort() { }
if (false) {
    /** @type {?|undefined} */
    Sort.prototype.asc;
    /** @type {?|undefined} */
    Sort.prototype.code;
}
/**
 * @record
 */
export function ListAdaptedComponents() { }
if (false) {
    /** @type {?|undefined} */
    ListAdaptedComponents.prototype.components;
    /** @type {?|undefined} */
    ListAdaptedComponents.prototype.pagination;
    /** @type {?|undefined} */
    ListAdaptedComponents.prototype.sorts;
}
/**
 * @record
 */
export function OrderStatusUpdateElement() { }
if (false) {
    /** @type {?|undefined} */
    OrderStatusUpdateElement.prototype.baseSiteId;
    /** @type {?|undefined} */
    OrderStatusUpdateElement.prototype.code;
    /** @type {?|undefined} */
    OrderStatusUpdateElement.prototype.status;
}
/**
 * @record
 */
export function PointOfServiceStock() { }
if (false) {
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.address;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.description;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.displayName;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.distanceKm;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.features;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.formattedDistance;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.geoPoint;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.mapIcon;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.name;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.openingHours;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.stockInfo;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.storeContent;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.storeImages;
    /** @type {?|undefined} */
    PointOfServiceStock.prototype.url;
}
/**
 * @record
 */
export function ProductExpressUpdateElement() { }
if (false) {
    /** @type {?|undefined} */
    ProductExpressUpdateElement.prototype.catalogId;
    /** @type {?|undefined} */
    ProductExpressUpdateElement.prototype.catalogVersion;
    /** @type {?|undefined} */
    ProductExpressUpdateElement.prototype.code;
}
/**
 * @record
 */
export function ProductList() { }
if (false) {
    /** @type {?|undefined} */
    ProductList.prototype.catalog;
    /** @type {?|undefined} */
    ProductList.prototype.currentPage;
    /** @type {?|undefined} */
    ProductList.prototype.products;
    /** @type {?|undefined} */
    ProductList.prototype.totalPageCount;
    /** @type {?|undefined} */
    ProductList.prototype.totalProductCount;
    /** @type {?|undefined} */
    ProductList.prototype.version;
}
/**
 * @record
 */
export function SaveCartResult() { }
if (false) {
    /** @type {?|undefined} */
    SaveCartResult.prototype.savedCartData;
}
/**
 * @record
 */
export function StoreFinderStockSearchPage() { }
if (false) {
    /** @type {?|undefined} */
    StoreFinderStockSearchPage.prototype.boundEastLongitude;
    /** @type {?|undefined} */
    StoreFinderStockSearchPage.prototype.boundSouthLatitude;
    /** @type {?|undefined} */
    StoreFinderStockSearchPage.prototype.boundWestLongitude;
    /** @type {?|undefined} */
    StoreFinderStockSearchPage.prototype.locationText;
    /** @type {?|undefined} */
    StoreFinderStockSearchPage.prototype.pagination;
    /** @type {?|undefined} */
    StoreFinderStockSearchPage.prototype.product;
    /** @type {?|undefined} */
    StoreFinderStockSearchPage.prototype.sorts;
    /** @type {?|undefined} */
    StoreFinderStockSearchPage.prototype.sourceLatitude;
    /** @type {?|undefined} */
    StoreFinderStockSearchPage.prototype.sourceLongitude;
    /** @type {?|undefined} */
    StoreFinderStockSearchPage.prototype.stores;
}
/**
 * @record
 */
export function UserGroup() { }
if (false) {
    /** @type {?|undefined} */
    UserGroup.prototype.members;
    /** @type {?|undefined} */
    UserGroup.prototype.membersCount;
    /** @type {?|undefined} */
    UserGroup.prototype.name;
    /** @type {?|undefined} */
    UserGroup.prototype.subGroups;
    /** @type {?|undefined} */
    UserGroup.prototype.uid;
}
/**
 * @record
 */
export function UserGroupList() { }
if (false) {
    /** @type {?|undefined} */
    UserGroupList.prototype.currentPage;
    /** @type {?|undefined} */
    UserGroupList.prototype.numberOfPages;
    /** @type {?|undefined} */
    UserGroupList.prototype.pageSize;
    /** @type {?|undefined} */
    UserGroupList.prototype.totalNumber;
    /** @type {?|undefined} */
    UserGroupList.prototype.userGroups;
}
