// @spartacus/core/esm5/src/state/utils/entity/entity.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var ENTITY_REMOVE_ACTION = '[ENTITY] REMOVE';
/** @type {?} */
export var ENTITY_REMOVE_ALL_ACTION = '[ENTITY] REMOVE ALL';
/**
 * @record
 */
export function EntityMeta() { }
if (false) {
    /** @type {?} */
    EntityMeta.prototype.entityType;
    /** @type {?} */
    EntityMeta.prototype.entityId;
    /** @type {?|undefined} */
    EntityMeta.prototype.entityRemove;
}
/**
 * @param {?} type
 * @param {?} id
 * @return {?}
 */
export function entityMeta(type, id) {
    return {
        entityType: type,
        entityId: id,
    };
}
/**
 * @param {?} type
 * @param {?} id
 * @return {?}
 */
export function entityRemoveMeta(type, id) {
    return {
        entityId: id,
        entityType: type,
        entityRemove: true,
    };
}
/**
 * @param {?} type
 * @return {?}
 */
export function entityRemoveAllMeta(type) {
    return {
        entityId: null,
        entityType: type,
        entityRemove: true,
    };
}
/**
 * @record
 */
export function EntityAction() { }
if (false) {
    /** @type {?|undefined} */
    EntityAction.prototype.payload;
    /** @type {?|undefined} */
    EntityAction.prototype.meta;
}
var EntityRemoveAction = /** @class */ (function () {
    function EntityRemoveAction(entityType, id) {
        this.type = ENTITY_REMOVE_ACTION;
        this.meta = entityRemoveMeta(entityType, id);
    }
    return EntityRemoveAction;
}());
export { EntityRemoveAction };
if (false) {
    /** @type {?} */
    EntityRemoveAction.prototype.type;
    /** @type {?} */
    EntityRemoveAction.prototype.meta;
}
var EntityRemoveAllAction = /** @class */ (function () {
    function EntityRemoveAllAction(entityType) {
        this.type = ENTITY_REMOVE_ALL_ACTION;
        this.meta = entityRemoveAllMeta(entityType);
    }
    return EntityRemoveAllAction;
}());
export { EntityRemoveAllAction };
if (false) {
    /** @type {?} */
    EntityRemoveAllAction.prototype.type;
    /** @type {?} */
    EntityRemoveAllAction.prototype.meta;
}
