// @spartacus/core/esm5/src/site-context/providers/site-context-params-providers.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { APP_INITIALIZER } from '@angular/core';
import { UrlSerializer } from '@angular/router';
import { ConfigInitializerService } from '../../config/config-initializer/config-initializer.service';
import { SiteContextParamsService } from '../services/site-context-params.service';
import { SiteContextRoutesHandler } from '../services/site-context-routes-handler';
import { SiteContextUrlSerializer } from '../services/site-context-url-serializer';
// functions below should not be exposed in public API:
/**
 * @param {?} siteContextRoutesHandler
 * @param {?} configInit
 * @return {?}
 */
export function initSiteContextRoutesHandler(siteContextRoutesHandler, configInit) {
    return (/**
     * @return {?}
     */
    function () {
        configInit.getStableConfig('context').then((/**
         * @return {?}
         */
        function () {
            siteContextRoutesHandler.init();
        }));
    });
}
/** @type {?} */
export var siteContextParamsProviders = [
    SiteContextParamsService,
    SiteContextUrlSerializer,
    { provide: UrlSerializer, useExisting: SiteContextUrlSerializer },
    {
        provide: APP_INITIALIZER,
        useFactory: initSiteContextRoutesHandler,
        deps: [SiteContextRoutesHandler, ConfigInitializerService],
        multi: true,
    },
];
/**
 * @deprecated since 1.3.1; TODO: remove
 * @param {?} siteContextRoutesHandler
 * @return {?}
 */
export function deprecatedInitSiteContextRoutesHandler(siteContextRoutesHandler) {
    return (/**
     * @return {?}
     */
    function () {
        siteContextRoutesHandler.init();
    });
}
/**
 * @deprecated since 1.3.1; TODO: remove
 * @type {?}
 */
export var deprecatedSiteContextParamsProviders = [
    SiteContextParamsService,
    SiteContextUrlSerializer,
    { provide: UrlSerializer, useExisting: SiteContextUrlSerializer },
    {
        provide: APP_INITIALIZER,
        useFactory: deprecatedInitSiteContextRoutesHandler,
        deps: [SiteContextRoutesHandler],
        multi: true,
    },
];
