// @spartacus/core/esm2015/src/store-finder/store/selectors/view-all-stores.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/index';
import { getStoreFinderState } from './feature.selector';
const ɵ0 = /**
 * @param {?} storesState
 * @return {?}
 */
(storesState) => storesState.viewAllStores;
/** @type {?} */
export const getViewAllStoresState = createSelector(getStoreFinderState, (ɵ0));
const ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
state => StateLoaderSelectors.loaderValueSelector(state);
/** @type {?} */
export const getViewAllStoresEntities = createSelector(getViewAllStoresState, (ɵ1));
const ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
state => StateLoaderSelectors.loaderLoadingSelector(state);
/** @type {?} */
export const getViewAllStoresLoading = createSelector(getViewAllStoresState, (ɵ2));
export { ɵ0, ɵ1, ɵ2 };
