// @spartacus/core/esm5/src/cart/connectors/voucher/cart-voucher.connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { CartVoucherAdapter } from './cart-voucher.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./cart-voucher.adapter";
var CartVoucherConnector = /** @class */ (function () {
    function CartVoucherConnector(adapter) {
        this.adapter = adapter;
    }
    /**
     * @param {?} userId
     * @param {?} cartId
     * @param {?} voucherId
     * @return {?}
     */
    CartVoucherConnector.prototype.add = /**
     * @param {?} userId
     * @param {?} cartId
     * @param {?} voucherId
     * @return {?}
     */
    function (userId, cartId, voucherId) {
        return this.adapter.add(userId, cartId, voucherId);
    };
    /**
     * @param {?} userId
     * @param {?} cartId
     * @param {?} voucherId
     * @return {?}
     */
    CartVoucherConnector.prototype.remove = /**
     * @param {?} userId
     * @param {?} cartId
     * @param {?} voucherId
     * @return {?}
     */
    function (userId, cartId, voucherId) {
        return this.adapter.remove(userId, cartId, voucherId);
    };
    CartVoucherConnector.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    CartVoucherConnector.ctorParameters = function () { return [
        { type: CartVoucherAdapter }
    ]; };
    /** @nocollapse */ CartVoucherConnector.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function CartVoucherConnector_Factory() { return new CartVoucherConnector(i0.ɵɵinject(i1.CartVoucherAdapter)); }, token: CartVoucherConnector, providedIn: "root" });
    return CartVoucherConnector;
}());
export { CartVoucherConnector };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    CartVoucherConnector.prototype.adapter;
}
