// @spartacus/core/esm2015/src/cart/store/selectors/cart.selector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createFeatureSelector, createSelector, } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/index';
import { CART_FEATURE, } from '../cart-state';
/** @type {?} */
const getCartContentSelector = (/**
 * @param {?} state
 * @return {?}
 */
(state) => state.content);
const ɵ0 = getCartContentSelector;
/** @type {?} */
const getCartRefreshSelector = (/**
 * @param {?} state
 * @return {?}
 */
(state) => state.refresh);
const ɵ1 = getCartRefreshSelector;
/** @type {?} */
const getCartEntriesSelector = (/**
 * @param {?} state
 * @return {?}
 */
(state) => state.entries);
const ɵ2 = getCartEntriesSelector;
/** @type {?} */
const getCartMergeCompleteSelector = (/**
 * @param {?} state
 * @return {?}
 */
(state) => state.cartMergeComplete);
const ɵ3 = getCartMergeCompleteSelector;
/** @type {?} */
export const getCartsState = createFeatureSelector(CART_FEATURE);
const ɵ4 = /**
 * @param {?} cartsState
 * @return {?}
 */
(cartsState) => cartsState.active;
/** @type {?} */
export const getActiveCartState = createSelector(getCartsState, (ɵ4));
const ɵ5 = /**
 * @param {?} state
 * @return {?}
 */
state => StateLoaderSelectors.loaderValueSelector(state);
/** @type {?} */
export const getCartState = createSelector(getActiveCartState, (ɵ5));
/** @type {?} */
export const getCartContent = createSelector(getCartState, getCartContentSelector);
/** @type {?} */
export const getCartRefresh = createSelector(getCartState, getCartRefreshSelector);
const ɵ6 = /**
 * @param {?} state
 * @return {?}
 */
state => (StateLoaderSelectors.loaderSuccessSelector(state) &&
    !StateLoaderSelectors.loaderLoadingSelector(state) &&
    !StateLoaderSelectors.loaderValueSelector(state).refresh) ||
    (StateLoaderSelectors.loaderErrorSelector(state) &&
        !StateLoaderSelectors.loaderLoadingSelector(state) &&
        !StateLoaderSelectors.loaderValueSelector(state).refresh);
/** @type {?} */
export const getCartLoaded = createSelector(getActiveCartState, (ɵ6));
const ɵ7 = /**
 * @param {?} state
 * @return {?}
 */
state => StateLoaderSelectors.loaderLoadingSelector(state);
/** @type {?} */
export const getCartLoading = createSelector(getActiveCartState, (ɵ7));
/** @type {?} */
export const getCartMergeComplete = createSelector(getCartState, getCartMergeCompleteSelector);
/** @type {?} */
export const getCartEntriesMap = createSelector(getCartState, getCartEntriesSelector);
/** @type {?} */
export const getCartEntrySelectorFactory = (/**
 * @param {?} productCode
 * @return {?}
 */
(productCode) => {
    return createSelector(getCartEntriesMap, (/**
     * @param {?} entries
     * @return {?}
     */
    entries => {
        if (entries) {
            return entries[productCode];
        }
    }));
});
const ɵ8 = /**
 * @param {?} entities
 * @return {?}
 */
entities => {
    return Object.keys(entities).map((/**
     * @param {?} code
     * @return {?}
     */
    code => entities[code]));
};
/** @type {?} */
export const getCartEntries = createSelector(getCartEntriesMap, (ɵ8));
const ɵ9 = /**
 * @param {?} content
 * @return {?}
 */
content => content.user;
/** @type {?} */
export const getCartUser = createSelector(getCartContent, (ɵ9));
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8, ɵ9 };
