// @spartacus/core/esm2015/src/asm/store/selectors/customer-search.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/loader/loader-group.selectors';
import { getAsmState } from './feature.selector';
const ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
(state) => state.customerSearchResult;
/** @type {?} */
export const getCustomerSearchResultsLoaderState = createSelector(getAsmState, (ɵ0));
const ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
state => StateLoaderSelectors.loaderValueSelector(state);
/** @type {?} */
export const getCustomerSearchResults = createSelector(getCustomerSearchResultsLoaderState, (ɵ1));
const ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
state => StateLoaderSelectors.loaderLoadingSelector(state);
/** @type {?} */
export const getCustomerSearchResultsLoading = createSelector(getCustomerSearchResultsLoaderState, (ɵ2));
export { ɵ0, ɵ1, ɵ2 };
