// @spartacus/core/esm5/src/kyma/store/selectors/open-id-token.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/index';
import { getKymaState } from './feature.selector';
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.openIdToken; };
/** @type {?} */
export var getOpenIdTokenState = createSelector(getKymaState, (ɵ0));
/** @type {?} */
export var getOpenIdTokenValue = createSelector(getOpenIdTokenState, StateLoaderSelectors.loaderValueSelector);
/** @type {?} */
export var getOpenIdTokenLoading = createSelector(getOpenIdTokenState, StateLoaderSelectors.loaderLoadingSelector);
/** @type {?} */
export var getOpenIdTokenSuccess = createSelector(getOpenIdTokenState, StateLoaderSelectors.loaderSuccessSelector);
/** @type {?} */
export var getOpenIdTokenError = createSelector(getOpenIdTokenState, StateLoaderSelectors.loaderErrorSelector);
export { ɵ0 };
