// @spartacus/core/esm2015/src/state/utils/processes-loader/processes-loader.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { resetMeta as loaderResetMeta, } from '../loader/loader.action';
/** @type {?} */
export const PROCESSES_INCREMENT_ACTION = '[PROCESSES LOADER] INCREMENT';
/** @type {?} */
export const PROCESSES_DECREMENT_ACTION = '[PROCESSES LOADER] DECREMENT';
/** @type {?} */
export const PROCESSES_LOADER_RESET_ACTION = '[PROCESSES LOADER] RESET';
/**
 * @record
 */
export function ProcessesLoaderMeta() { }
if (false) {
    /** @type {?} */
    ProcessesLoaderMeta.prototype.entityType;
    /** @type {?|undefined} */
    ProcessesLoaderMeta.prototype.processesCountDiff;
}
/**
 * @record
 */
export function ProcessesLoaderAction() { }
if (false) {
    /** @type {?|undefined} */
    ProcessesLoaderAction.prototype.payload;
    /** @type {?|undefined} */
    ProcessesLoaderAction.prototype.meta;
}
/**
 * @param {?} entityType
 * @return {?}
 */
export function processesIncrementMeta(entityType) {
    return {
        entityType: entityType,
        loader: undefined,
        processesCountDiff: 1,
    };
}
/**
 * @param {?} entityType
 * @return {?}
 */
export function processesDecrementMeta(entityType) {
    return {
        entityType: entityType,
        loader: undefined,
        processesCountDiff: -1,
    };
}
/**
 * @param {?} entityType
 * @return {?}
 */
export function processesLoaderResetMeta(entityType) {
    // processes reset action is a reset action for loader reducer, but not the other way around
    return Object.assign({}, loaderResetMeta(entityType), { processesCountDiff: null });
}
export class ProcessesLoaderResetAction {
    /**
     * @param {?} entityType
     */
    constructor(entityType) {
        this.type = PROCESSES_LOADER_RESET_ACTION;
        this.meta = processesLoaderResetMeta(entityType);
    }
}
if (false) {
    /** @type {?} */
    ProcessesLoaderResetAction.prototype.type;
    /** @type {?} */
    ProcessesLoaderResetAction.prototype.meta;
}
export class ProcessesIncrementAction {
    /**
     * @param {?} entityType
     */
    constructor(entityType) {
        this.type = PROCESSES_INCREMENT_ACTION;
        this.meta = processesIncrementMeta(entityType);
    }
}
if (false) {
    /** @type {?} */
    ProcessesIncrementAction.prototype.type;
    /** @type {?} */
    ProcessesIncrementAction.prototype.meta;
}
export class ProcessesDecrementAction {
    /**
     * @param {?} entityType
     */
    constructor(entityType) {
        this.type = PROCESSES_DECREMENT_ACTION;
        this.meta = processesDecrementMeta(entityType);
    }
}
if (false) {
    /** @type {?} */
    ProcessesDecrementAction.prototype.type;
    /** @type {?} */
    ProcessesDecrementAction.prototype.meta;
}
