// @spartacus/core/esm2015/src/product/store/actions/product.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PRODUCT_DETAIL_ENTITY } from '../product-state';
import { EntityScopedLoaderActions } from '../../../state/utils/scoped-loader/entity-scoped-loader.actions';
/** @type {?} */
export const LOAD_PRODUCT = '[Product] Load Product Data';
/** @type {?} */
export const LOAD_PRODUCT_FAIL = '[Product] Load Product Data Fail';
/** @type {?} */
export const LOAD_PRODUCT_SUCCESS = '[Product] Load Product Data Success';
/**
 * @record
 */
export function ProductMeta() { }
if (false) {
    /** @type {?|undefined} */
    ProductMeta.prototype.scope;
}
/**
 * @record
 */
export function EntityScopedLoaderAction() { }
if (false) {
    /** @type {?|undefined} */
    EntityScopedLoaderAction.prototype.payload;
    /** @type {?|undefined} */
    EntityScopedLoaderAction.prototype.meta;
}
export class LoadProduct extends EntityScopedLoaderActions.EntityScopedLoadAction {
    /**
     * @param {?} payload
     * @param {?=} scope
     */
    constructor(payload, scope = '') {
        super(PRODUCT_DETAIL_ENTITY, payload, scope);
        this.payload = payload;
        this.type = LOAD_PRODUCT;
    }
}
if (false) {
    /** @type {?} */
    LoadProduct.prototype.type;
    /** @type {?} */
    LoadProduct.prototype.payload;
}
export class LoadProductFail extends EntityScopedLoaderActions.EntityScopedFailAction {
    /**
     * @param {?} productCode
     * @param {?} payload
     * @param {?=} scope
     */
    constructor(productCode, payload, scope = '') {
        super(PRODUCT_DETAIL_ENTITY, productCode, scope, payload);
        this.payload = payload;
        this.type = LOAD_PRODUCT_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadProductFail.prototype.type;
    /** @type {?} */
    LoadProductFail.prototype.payload;
}
export class LoadProductSuccess extends EntityScopedLoaderActions.EntityScopedSuccessAction {
    /**
     * @param {?} payload
     * @param {?=} scope
     */
    constructor(payload, scope = '') {
        super(PRODUCT_DETAIL_ENTITY, payload.code, scope);
        this.payload = payload;
        this.type = LOAD_PRODUCT_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadProductSuccess.prototype.type;
    /** @type {?} */
    LoadProductSuccess.prototype.payload;
}
