// @spartacus/core/esm2015/src/product/store/actions/product-reviews.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const LOAD_PRODUCT_REVIEWS = '[Product] Load Product Reviews Data';
/** @type {?} */
export const LOAD_PRODUCT_REVIEWS_FAIL = '[Product] Load Product Reviews Data Fail';
/** @type {?} */
export const LOAD_PRODUCT_REVIEWS_SUCCESS = '[Product] Load Product Reviews Data Success';
/** @type {?} */
export const POST_PRODUCT_REVIEW = '[Product] Post Product Review';
/** @type {?} */
export const POST_PRODUCT_REVIEW_FAIL = '[Product] Post Product Review Fail';
/** @type {?} */
export const POST_PRODUCT_REVIEW_SUCCESS = '[Product] Post Product Review Success';
export class LoadProductReviews {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_PRODUCT_REVIEWS;
    }
}
if (false) {
    /** @type {?} */
    LoadProductReviews.prototype.type;
    /** @type {?} */
    LoadProductReviews.prototype.payload;
}
export class LoadProductReviewsFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_PRODUCT_REVIEWS_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadProductReviewsFail.prototype.type;
    /** @type {?} */
    LoadProductReviewsFail.prototype.payload;
}
export class LoadProductReviewsSuccess {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_PRODUCT_REVIEWS_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadProductReviewsSuccess.prototype.type;
    /** @type {?} */
    LoadProductReviewsSuccess.prototype.payload;
}
export class PostProductReview {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = POST_PRODUCT_REVIEW;
    }
}
if (false) {
    /** @type {?} */
    PostProductReview.prototype.type;
    /** @type {?} */
    PostProductReview.prototype.payload;
}
export class PostProductReviewFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = POST_PRODUCT_REVIEW_FAIL;
    }
}
if (false) {
    /** @type {?} */
    PostProductReviewFail.prototype.type;
    /** @type {?} */
    PostProductReviewFail.prototype.payload;
}
export class PostProductReviewSuccess {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = POST_PRODUCT_REVIEW_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    PostProductReviewSuccess.prototype.type;
    /** @type {?} */
    PostProductReviewSuccess.prototype.payload;
}
