// @spartacus/core/esm5/src/site-context/connectors/site.adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
SiteAdapter = /** @class */ (function () {
    function SiteAdapter() {
    }
    return SiteAdapter;
}());
/**
 * @abstract
 */
export { SiteAdapter };
if (false) {
    /**
     * Abstract method used to load languages.
     * @abstract
     * @return {?}
     */
    SiteAdapter.prototype.loadLanguages = function () { };
    /**
     * Abstract method used to load currencies.
     * @abstract
     * @return {?}
     */
    SiteAdapter.prototype.loadCurrencies = function () { };
    /**
     * Abstract method used to get countries with optional type.
     * @abstract
     * @param {?=} type
     * @return {?}
     */
    SiteAdapter.prototype.loadCountries = function (type) { };
    /**
     * Abstract method used to get regions for a country.
     * @abstract
     * @param {?} countryIsoCode
     * @return {?}
     */
    SiteAdapter.prototype.loadRegions = function (countryIsoCode) { };
    /**
     * Abstract method used to get base site data.
     * @abstract
     * @return {?}
     */
    SiteAdapter.prototype.loadBaseSite = function () { };
}
