// @spartacus/core/esm5/src/state/utils/loader/loader.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var LOADER_LOAD_ACTION = '[LOADER] LOAD';
/** @type {?} */
export var LOADER_FAIL_ACTION = '[LOADER] FAIL';
/** @type {?} */
export var LOADER_SUCCESS_ACTION = '[LOADER] SUCCESS';
/** @type {?} */
export var LOADER_RESET_ACTION = '[LOADER] RESET';
/**
 * @record
 */
export function LoaderMeta() { }
if (false) {
    /** @type {?} */
    LoaderMeta.prototype.entityType;
    /** @type {?} */
    LoaderMeta.prototype.loader;
}
/**
 * @record
 */
export function LoaderAction() { }
if (false) {
    /** @type {?|undefined} */
    LoaderAction.prototype.payload;
    /** @type {?|undefined} */
    LoaderAction.prototype.meta;
}
/**
 * @param {?} entityType
 * @return {?}
 */
export function loadMeta(entityType) {
    return {
        entityType: entityType,
        loader: {
            load: true,
        },
    };
}
/**
 * @param {?} entityType
 * @param {?=} error
 * @return {?}
 */
export function failMeta(entityType, error) {
    return {
        entityType: entityType,
        loader: {
            error: error ? error : true,
        },
    };
}
/**
 * @param {?} entityType
 * @return {?}
 */
export function successMeta(entityType) {
    return {
        entityType: entityType,
        loader: {
            success: true,
        },
    };
}
/**
 * @param {?} entityType
 * @return {?}
 */
export function resetMeta(entityType) {
    return {
        entityType: entityType,
        loader: {},
    };
}
var LoaderLoadAction = /** @class */ (function () {
    function LoaderLoadAction(entityType) {
        this.type = LOADER_LOAD_ACTION;
        this.meta = loadMeta(entityType);
    }
    return LoaderLoadAction;
}());
export { LoaderLoadAction };
if (false) {
    /** @type {?} */
    LoaderLoadAction.prototype.type;
    /** @type {?} */
    LoaderLoadAction.prototype.meta;
}
var LoaderFailAction = /** @class */ (function () {
    function LoaderFailAction(entityType, error) {
        this.type = LOADER_FAIL_ACTION;
        this.meta = failMeta(entityType, error);
    }
    return LoaderFailAction;
}());
export { LoaderFailAction };
if (false) {
    /** @type {?} */
    LoaderFailAction.prototype.type;
    /** @type {?} */
    LoaderFailAction.prototype.meta;
}
var LoaderSuccessAction = /** @class */ (function () {
    function LoaderSuccessAction(entityType) {
        this.type = LOADER_SUCCESS_ACTION;
        this.meta = successMeta(entityType);
    }
    return LoaderSuccessAction;
}());
export { LoaderSuccessAction };
if (false) {
    /** @type {?} */
    LoaderSuccessAction.prototype.type;
    /** @type {?} */
    LoaderSuccessAction.prototype.meta;
}
var LoaderResetAction = /** @class */ (function () {
    function LoaderResetAction(entityType) {
        this.type = LOADER_RESET_ACTION;
        this.meta = resetMeta(entityType);
    }
    return LoaderResetAction;
}());
export { LoaderResetAction };
if (false) {
    /** @type {?} */
    LoaderResetAction.prototype.type;
    /** @type {?} */
    LoaderResetAction.prototype.meta;
}
