// @spartacus/core/esm2015/src/state/utils/entity/entity.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const ENTITY_REMOVE_ACTION = '[ENTITY] REMOVE';
/** @type {?} */
export const ENTITY_REMOVE_ALL_ACTION = '[ENTITY] REMOVE ALL';
/**
 * @record
 */
export function EntityMeta() { }
if (false) {
    /** @type {?} */
    EntityMeta.prototype.entityType;
    /** @type {?} */
    EntityMeta.prototype.entityId;
    /** @type {?|undefined} */
    EntityMeta.prototype.entityRemove;
}
/**
 * @param {?} type
 * @param {?} id
 * @return {?}
 */
export function entityMeta(type, id) {
    return {
        entityType: type,
        entityId: id,
    };
}
/**
 * @param {?} type
 * @param {?} id
 * @return {?}
 */
export function entityRemoveMeta(type, id) {
    return {
        entityId: id,
        entityType: type,
        entityRemove: true,
    };
}
/**
 * @param {?} type
 * @return {?}
 */
export function entityRemoveAllMeta(type) {
    return {
        entityId: null,
        entityType: type,
        entityRemove: true,
    };
}
/**
 * @record
 */
export function EntityAction() { }
if (false) {
    /** @type {?|undefined} */
    EntityAction.prototype.payload;
    /** @type {?|undefined} */
    EntityAction.prototype.meta;
}
export class EntityRemoveAction {
    /**
     * @param {?} entityType
     * @param {?} id
     */
    constructor(entityType, id) {
        this.type = ENTITY_REMOVE_ACTION;
        this.meta = entityRemoveMeta(entityType, id);
    }
}
if (false) {
    /** @type {?} */
    EntityRemoveAction.prototype.type;
    /** @type {?} */
    EntityRemoveAction.prototype.meta;
}
export class EntityRemoveAllAction {
    /**
     * @param {?} entityType
     */
    constructor(entityType) {
        this.type = ENTITY_REMOVE_ALL_ACTION;
        this.meta = entityRemoveAllMeta(entityType);
    }
}
if (false) {
    /** @type {?} */
    EntityRemoveAllAction.prototype.type;
    /** @type {?} */
    EntityRemoveAllAction.prototype.meta;
}
