// @spartacus/core/esm5/src/user/connectors/customer-coupon/customer-coupon.connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { CustomerCouponAdapter } from './customer-coupon.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./customer-coupon.adapter";
var CustomerCouponConnector = /** @class */ (function () {
    function CustomerCouponConnector(adapter) {
        this.adapter = adapter;
    }
    /**
     * @param {?} userId
     * @param {?} pageSize
     * @param {?} currentPage
     * @param {?} sort
     * @return {?}
     */
    CustomerCouponConnector.prototype.getCustomerCoupons = /**
     * @param {?} userId
     * @param {?} pageSize
     * @param {?} currentPage
     * @param {?} sort
     * @return {?}
     */
    function (userId, pageSize, currentPage, sort) {
        return this.adapter.getCustomerCoupons(userId, pageSize, currentPage, sort);
    };
    /**
     * @param {?} userId
     * @param {?} couponCode
     * @return {?}
     */
    CustomerCouponConnector.prototype.turnOnNotification = /**
     * @param {?} userId
     * @param {?} couponCode
     * @return {?}
     */
    function (userId, couponCode) {
        return this.adapter.turnOnNotification(userId, couponCode);
    };
    /**
     * @param {?} userId
     * @param {?} couponCode
     * @return {?}
     */
    CustomerCouponConnector.prototype.turnOffNotification = /**
     * @param {?} userId
     * @param {?} couponCode
     * @return {?}
     */
    function (userId, couponCode) {
        return this.adapter.turnOffNotification(userId, couponCode);
    };
    /**
     * @param {?} userId
     * @param {?} couponCode
     * @return {?}
     */
    CustomerCouponConnector.prototype.claimCustomerCoupon = /**
     * @param {?} userId
     * @param {?} couponCode
     * @return {?}
     */
    function (userId, couponCode) {
        return this.adapter.claimCustomerCoupon(userId, couponCode);
    };
    CustomerCouponConnector.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    CustomerCouponConnector.ctorParameters = function () { return [
        { type: CustomerCouponAdapter }
    ]; };
    /** @nocollapse */ CustomerCouponConnector.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function CustomerCouponConnector_Factory() { return new CustomerCouponConnector(i0.ɵɵinject(i1.CustomerCouponAdapter)); }, token: CustomerCouponConnector, providedIn: "root" });
    return CustomerCouponConnector;
}());
export { CustomerCouponConnector };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    CustomerCouponConnector.prototype.adapter;
}
