// @spartacus/core/esm5/src/user/connectors/interests/user-interests.connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { UserInterestsAdapter } from './user-interests.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./user-interests.adapter";
var UserInterestsConnector = /** @class */ (function () {
    function UserInterestsConnector(adapter) {
        this.adapter = adapter;
    }
    /**
     * @param {?} userId
     * @param {?=} pageSize
     * @param {?=} currentPage
     * @param {?=} sort
     * @param {?=} productCode
     * @param {?=} notificationType
     * @return {?}
     */
    UserInterestsConnector.prototype.getInterests = /**
     * @param {?} userId
     * @param {?=} pageSize
     * @param {?=} currentPage
     * @param {?=} sort
     * @param {?=} productCode
     * @param {?=} notificationType
     * @return {?}
     */
    function (userId, pageSize, currentPage, sort, productCode, notificationType) {
        return this.adapter.getInterests(userId, pageSize, currentPage, sort, productCode, notificationType);
    };
    /**
     * @param {?} userId
     * @param {?} item
     * @return {?}
     */
    UserInterestsConnector.prototype.removeInterest = /**
     * @param {?} userId
     * @param {?} item
     * @return {?}
     */
    function (userId, item) {
        return this.adapter.removeInterest(userId, item);
    };
    /**
     * @param {?} userId
     * @param {?} productCode
     * @param {?} notificationType
     * @return {?}
     */
    UserInterestsConnector.prototype.addInterest = /**
     * @param {?} userId
     * @param {?} productCode
     * @param {?} notificationType
     * @return {?}
     */
    function (userId, productCode, notificationType) {
        return this.adapter.addInterest(userId, productCode, notificationType);
    };
    UserInterestsConnector.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    UserInterestsConnector.ctorParameters = function () { return [
        { type: UserInterestsAdapter }
    ]; };
    /** @nocollapse */ UserInterestsConnector.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function UserInterestsConnector_Factory() { return new UserInterestsConnector(i0.ɵɵinject(i1.UserInterestsAdapter)); }, token: UserInterestsConnector, providedIn: "root" });
    return UserInterestsConnector;
}());
export { UserInterestsConnector };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    UserInterestsConnector.prototype.adapter;
}
