// @spartacus/core/esm2015/src/cart/store/reducers/cart.reducer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as DeprecatedCartActions from '../actions/cart.action';
import { CartActions } from '../actions/index';
/** @type {?} */
export const initialState = {
    content: {},
    entries: {},
    refresh: false,
    cartMergeComplete: false,
};
/**
 * @param {?=} state
 * @param {?=} action
 * @return {?}
 */
export function reducer(state = initialState, action) {
    switch (action.type) {
        case DeprecatedCartActions.MERGE_CART: {
            return Object.assign({}, state, { cartMergeComplete: false });
        }
        case DeprecatedCartActions.MERGE_CART_SUCCESS: {
            return Object.assign({}, state, { cartMergeComplete: true, refresh: true });
        }
        case DeprecatedCartActions.LOAD_CART_SUCCESS:
        case DeprecatedCartActions.CREATE_CART_SUCCESS: {
            /** @type {?} */
            const content = Object.assign({}, action.payload);
            /** @type {?} */
            let entries = {};
            if (content.entries) {
                entries = content.entries.reduce((/**
                 * @param {?} entryMap
                 * @param {?} entry
                 * @return {?}
                 */
                (entryMap, entry) => {
                    return Object.assign({}, entryMap, { [entry.product.code]: state.entries && state.entries[entry.product.code]
                            ? Object.assign({}, state.entries[entry.product.code], entry) : entry });
                }), Object.assign({}, entries));
                delete content['entries'];
            }
            return Object.assign({}, state, { content,
                entries, refresh: false });
        }
        case CartActions.CART_ADD_VOUCHER_SUCCESS:
        case CartActions.CART_REMOVE_VOUCHER_SUCCESS:
        case CartActions.CART_REMOVE_ENTRY_SUCCESS:
        case CartActions.CART_UPDATE_ENTRY_SUCCESS:
        case CartActions.CART_ADD_ENTRY_SUCCESS:
        case DeprecatedCartActions.ADD_EMAIL_TO_CART_SUCCESS: {
            return Object.assign({}, state, { refresh: true });
        }
        case DeprecatedCartActions.RESET_CART_DETAILS: {
            return {
                content: {
                    guid: state.content.guid,
                    code: state.content.code,
                    user: state.content.user,
                },
                entries: {},
                refresh: false,
                cartMergeComplete: false,
            };
        }
        case DeprecatedCartActions.CLEAR_CART: {
            return initialState;
        }
    }
    return state;
}
