// @spartacus/core/esm2015/src/cart/store/reducers/multi-cart.reducer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CartActions } from '../actions/index';
/** @type {?} */
export const activeCartInitialState = '';
/** @type {?} */
export const wishListInitialState = '';
/**
 * @param {?=} state
 * @param {?=} action
 * @return {?}
 */
export function activeCartReducer(state = activeCartInitialState, action) {
    switch (action.type) {
        case CartActions.LOAD_MULTI_CART_SUCCESS:
        case CartActions.CREATE_MULTI_CART_SUCCESS:
        // point to `temp-${uuid}` cart when we are creating/merging cart
        case CartActions.CREATE_MULTI_CART:
            if (action.payload &&
                action.payload.extraData &&
                action.payload.extraData.active) {
                return (/** @type {?} */ (action.meta.entityId));
            }
            else {
                return state;
            }
        case CartActions.REMOVE_CART:
            if (action.payload === state) {
                return activeCartInitialState;
            }
            else {
                return state;
            }
    }
    return state;
}
/** @type {?} */
export const cartEntitiesInitialState = undefined;
/**
 * @param {?=} state
 * @param {?=} action
 * @return {?}
 */
export function cartEntitiesReducer(state = cartEntitiesInitialState, action) {
    switch (action.type) {
        case CartActions.LOAD_MULTI_CART_SUCCESS:
        case CartActions.CREATE_MULTI_CART_SUCCESS:
        case CartActions.CREATE_WISH_LIST_SUCCESS:
        case CartActions.LOAD_WISH_LIST_SUCCESS:
        case CartActions.SET_TEMP_CART:
            return action.payload.cart;
    }
    return state;
}
/**
 * @param {?=} state
 * @param {?=} action
 * @return {?}
 */
export function wishListReducer(state = wishListInitialState, action) {
    switch (action.type) {
        case CartActions.CREATE_WISH_LIST_SUCCESS:
        case CartActions.LOAD_WISH_LIST_SUCCESS:
            return (/** @type {?} */ (action.meta.entityId));
    }
    return state;
}
