// @spartacus/core/esm5/src/store-finder/store/selectors/find-stores.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/index';
import { getStoreFinderState } from './feature.selector';
var ɵ0 = /**
 * @param {?} storesState
 * @return {?}
 */
function (storesState) { return storesState.findStores; };
/** @type {?} */
export var getFindStoresState = createSelector(getStoreFinderState, (ɵ0));
var ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return StateLoaderSelectors.loaderValueSelector(state); };
/** @type {?} */
export var getFindStoresEntities = createSelector(getFindStoresState, (ɵ1));
var ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return StateLoaderSelectors.loaderLoadingSelector(state); };
/** @type {?} */
export var getStoresLoading = createSelector(getFindStoresState, (ɵ2));
export { ɵ0, ɵ1, ɵ2 };
