// @spartacus/core/esm2015/src/user/connectors/interests/user-interests.connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { UserInterestsAdapter } from './user-interests.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./user-interests.adapter";
export class UserInterestsConnector {
    /**
     * @param {?} adapter
     */
    constructor(adapter) {
        this.adapter = adapter;
    }
    /**
     * @param {?} userId
     * @param {?=} pageSize
     * @param {?=} currentPage
     * @param {?=} sort
     * @param {?=} productCode
     * @param {?=} notificationType
     * @return {?}
     */
    getInterests(userId, pageSize, currentPage, sort, productCode, notificationType) {
        return this.adapter.getInterests(userId, pageSize, currentPage, sort, productCode, notificationType);
    }
    /**
     * @param {?} userId
     * @param {?} item
     * @return {?}
     */
    removeInterest(userId, item) {
        return this.adapter.removeInterest(userId, item);
    }
    /**
     * @param {?} userId
     * @param {?} productCode
     * @param {?} notificationType
     * @return {?}
     */
    addInterest(userId, productCode, notificationType) {
        return this.adapter.addInterest(userId, productCode, notificationType);
    }
}
UserInterestsConnector.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
UserInterestsConnector.ctorParameters = () => [
    { type: UserInterestsAdapter }
];
/** @nocollapse */ UserInterestsConnector.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function UserInterestsConnector_Factory() { return new UserInterestsConnector(i0.ɵɵinject(i1.UserInterestsAdapter)); }, token: UserInterestsConnector, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    UserInterestsConnector.prototype.adapter;
}
