// @spartacus/core/esm2015/src/site-context/store/actions/base-site.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const LOAD_BASE_SITE = '[Site-context] Load BaseSite';
/** @type {?} */
export const LOAD_BASE_SITE_FAIL = '[Site-context] Load BaseSite Fail';
/** @type {?} */
export const LOAD_BASE_SITE_SUCCESS = '[Site-context] Load BaseSite Success';
/** @type {?} */
export const SET_ACTIVE_BASE_SITE = '[Site-context] Set Active BaseSite';
/** @type {?} */
export const BASE_SITE_CHANGE = '[Site-context] BaseSite Change';
export class LoadBaseSite {
    constructor() {
        this.type = LOAD_BASE_SITE;
    }
}
if (false) {
    /** @type {?} */
    LoadBaseSite.prototype.type;
}
export class LoadBaseSiteFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_BASE_SITE_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadBaseSiteFail.prototype.type;
    /** @type {?} */
    LoadBaseSiteFail.prototype.payload;
}
export class LoadBaseSiteSuccess {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_BASE_SITE_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadBaseSiteSuccess.prototype.type;
    /** @type {?} */
    LoadBaseSiteSuccess.prototype.payload;
}
export class SetActiveBaseSite {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SET_ACTIVE_BASE_SITE;
    }
}
if (false) {
    /** @type {?} */
    SetActiveBaseSite.prototype.type;
    /** @type {?} */
    SetActiveBaseSite.prototype.payload;
}
export class BaseSiteChange {
    constructor() {
        this.type = BASE_SITE_CHANGE;
    }
}
if (false) {
    /** @type {?} */
    BaseSiteChange.prototype.type;
}
