// @spartacus/core/esm2015/src/site-context/store/actions/currencies.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const LOAD_CURRENCIES = '[Site-context] Load Currencies';
/** @type {?} */
export const LOAD_CURRENCIES_FAIL = '[Site-context] Load Currencies Fail';
/** @type {?} */
export const LOAD_CURRENCIES_SUCCESS = '[Site-context] Load Currencies Success';
/** @type {?} */
export const SET_ACTIVE_CURRENCY = '[Site-context] Set Active Currency';
/** @type {?} */
export const CURRENCY_CHANGE = '[Site-context] Currency Change';
export class LoadCurrencies {
    constructor() {
        this.type = LOAD_CURRENCIES;
    }
}
if (false) {
    /** @type {?} */
    LoadCurrencies.prototype.type;
}
export class LoadCurrenciesFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_CURRENCIES_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadCurrenciesFail.prototype.type;
    /** @type {?} */
    LoadCurrenciesFail.prototype.payload;
}
export class LoadCurrenciesSuccess {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_CURRENCIES_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadCurrenciesSuccess.prototype.type;
    /** @type {?} */
    LoadCurrenciesSuccess.prototype.payload;
}
export class SetActiveCurrency {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SET_ACTIVE_CURRENCY;
    }
}
if (false) {
    /** @type {?} */
    SetActiveCurrency.prototype.type;
    /** @type {?} */
    SetActiveCurrency.prototype.payload;
}
export class CurrencyChange {
    constructor() {
        this.type = CURRENCY_CHANGE;
    }
}
if (false) {
    /** @type {?} */
    CurrencyChange.prototype.type;
}
