// @spartacus/core/esm5/src/user/services/find-product-page-meta.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { of, combineLatest } from 'rxjs';
import { filter, map, switchMap } from 'rxjs/operators';
import { PageMetaResolver, } from '../../cms/page';
import { TranslationService } from '../../i18n/translation.service';
import { PageType } from '../../model/cms.model';
import { RoutingService } from '../../routing/facade/routing.service';
import { ProductSearchService } from '../../product/facade/product-search.service';
import { AuthService } from '../../auth/facade/auth.service';
import * as i0 from "@angular/core";
import * as i1 from "../../routing/facade/routing.service";
import * as i2 from "../../product/facade/product-search.service";
import * as i3 from "../../i18n/translation.service";
import * as i4 from "../../auth/facade/auth.service";
var FindProductPageMetaResolver = /** @class */ (function (_super) {
    tslib_1.__extends(FindProductPageMetaResolver, _super);
    function FindProductPageMetaResolver(routingService, productSearchService, translation, authService) {
        var _this = _super.call(this) || this;
        _this.routingService = routingService;
        _this.productSearchService = productSearchService;
        _this.translation = translation;
        _this.authService = authService;
        _this.totalAndCode$ = combineLatest([
            _this.productSearchService.getResults().pipe(filter((/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return !!(data && data.pagination); })), map((/**
             * @param {?} results
             * @return {?}
             */
            function (results) { return results.pagination.totalResults; }))),
            _this.routingService.getRouterState().pipe(map((/**
             * @param {?} state
             * @return {?}
             */
            function (state) { return state.state.queryParams['couponcode']; })), filter(Boolean)),
        ]);
        _this.pageType = PageType.CONTENT_PAGE;
        _this.pageTemplate = 'SearchResultsListPageTemplate';
        return _this;
    }
    /**
     * @deprecated since version 1.3
     *
     * The resolve method is no longer preferred and will be removed with release 2.0.
     * The caller `PageMetaService` service is improved to expect all individual resolvers
     * instead, so that the code is easier extensible.
     */
    /**
     * @deprecated since version 1.3
     *
     * The resolve method is no longer preferred and will be removed with release 2.0.
     * The caller `PageMetaService` service is improved to expect all individual resolvers
     * instead, so that the code is easier extensible.
     * @return {?}
     */
    FindProductPageMetaResolver.prototype.resolve = /**
     * @deprecated since version 1.3
     *
     * The resolve method is no longer preferred and will be removed with release 2.0.
     * The caller `PageMetaService` service is improved to expect all individual resolvers
     * instead, so that the code is easier extensible.
     * @return {?}
     */
    function () {
        return combineLatest([this.resolveTitle(), this.resolveBreadcrumbs()]).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), title = _b[0], breadcrumbs = _b[1];
            return ({
                title: title,
                breadcrumbs: breadcrumbs,
            });
        })));
    };
    /**
     * @return {?}
     */
    FindProductPageMetaResolver.prototype.resolveBreadcrumbs = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var breadcrumbs = [{ label: 'Home', link: '/' }];
        this.authService.isUserLoggedIn().subscribe((/**
         * @param {?} login
         * @return {?}
         */
        function (login) {
            if (login)
                breadcrumbs.push({ label: 'My Coupons', link: '/my-account/coupons' });
        }));
        return of(breadcrumbs);
    };
    /**
     * @return {?}
     */
    FindProductPageMetaResolver.prototype.resolveTitle = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return this.totalAndCode$.pipe(switchMap((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), total = _b[0], code = _b[1];
            return _this.translation.translate('pageMetaResolver.search.findProductTitle', {
                count: total,
                coupon: code,
            });
        })));
    };
    /**
     * @param {?} page
     * @return {?}
     */
    FindProductPageMetaResolver.prototype.getScore = /**
     * @param {?} page
     * @return {?}
     */
    function (page) {
        /** @type {?} */
        var score = 0;
        if (this.pageType) {
            score += page.type === this.pageType ? 1 : -1;
        }
        if (this.pageTemplate) {
            score += page.template === this.pageTemplate ? 1 : -1;
        }
        this.routingService
            .getRouterState()
            .pipe(map((/**
         * @param {?} state
         * @return {?}
         */
        function (state) {
            return state.state.queryParams;
        })), filter(Boolean))
            .subscribe((/**
         * @param {?} queryParams
         * @return {?}
         */
        function (queryParams) {
            if (queryParams) {
                score += queryParams['couponcode'] ? 1 : -1;
            }
        }))
            .unsubscribe();
        return score;
    };
    FindProductPageMetaResolver.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    FindProductPageMetaResolver.ctorParameters = function () { return [
        { type: RoutingService },
        { type: ProductSearchService },
        { type: TranslationService },
        { type: AuthService }
    ]; };
    /** @nocollapse */ FindProductPageMetaResolver.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function FindProductPageMetaResolver_Factory() { return new FindProductPageMetaResolver(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i2.ProductSearchService), i0.ɵɵinject(i3.TranslationService), i0.ɵɵinject(i4.AuthService)); }, token: FindProductPageMetaResolver, providedIn: "root" });
    return FindProductPageMetaResolver;
}(PageMetaResolver));
export { FindProductPageMetaResolver };
if (false) {
    /** @type {?} */
    FindProductPageMetaResolver.prototype.totalAndCode$;
    /**
     * @type {?}
     * @protected
     */
    FindProductPageMetaResolver.prototype.routingService;
    /**
     * @type {?}
     * @protected
     */
    FindProductPageMetaResolver.prototype.productSearchService;
    /**
     * @type {?}
     * @protected
     */
    FindProductPageMetaResolver.prototype.translation;
    /**
     * @type {?}
     * @protected
     */
    FindProductPageMetaResolver.prototype.authService;
}
