// @spartacus/core/esm5/src/auth/store/actions/user-token.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var LOAD_USER_TOKEN = '[Auth] Load User Token';
/** @type {?} */
export var LOAD_USER_TOKEN_FAIL = '[Auth] Load User Token Fail';
/** @type {?} */
export var LOAD_USER_TOKEN_SUCCESS = '[Auth] Load User Token Success';
/** @type {?} */
export var REFRESH_USER_TOKEN = '[Auth] Refresh User Token';
/** @type {?} */
export var REFRESH_USER_TOKEN_FAIL = '[Auth] Refresh User Token Fail';
/** @type {?} */
export var REFRESH_USER_TOKEN_SUCCESS = '[Auth] Refresh User Token Success';
/** @type {?} */
export var REVOKE_USER_TOKEN = '[Auth] Revoke User Token';
/** @type {?} */
export var REVOKE_USER_TOKEN_FAIL = '[Auth] Revoke User Token Fail';
/** @type {?} */
export var REVOKE_USER_TOKEN_SUCCESS = '[Auth] Revoke User Token Success';
var LoadUserToken = /** @class */ (function () {
    function LoadUserToken(payload) {
        this.payload = payload;
        this.type = LOAD_USER_TOKEN;
    }
    return LoadUserToken;
}());
export { LoadUserToken };
if (false) {
    /** @type {?} */
    LoadUserToken.prototype.type;
    /** @type {?} */
    LoadUserToken.prototype.payload;
}
var LoadUserTokenFail = /** @class */ (function () {
    function LoadUserTokenFail(payload) {
        this.payload = payload;
        this.type = LOAD_USER_TOKEN_FAIL;
    }
    return LoadUserTokenFail;
}());
export { LoadUserTokenFail };
if (false) {
    /** @type {?} */
    LoadUserTokenFail.prototype.type;
    /** @type {?} */
    LoadUserTokenFail.prototype.payload;
}
var LoadUserTokenSuccess = /** @class */ (function () {
    function LoadUserTokenSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_USER_TOKEN_SUCCESS;
    }
    return LoadUserTokenSuccess;
}());
export { LoadUserTokenSuccess };
if (false) {
    /** @type {?} */
    LoadUserTokenSuccess.prototype.type;
    /** @type {?} */
    LoadUserTokenSuccess.prototype.payload;
}
var RefreshUserToken = /** @class */ (function () {
    function RefreshUserToken(payload) {
        this.payload = payload;
        this.type = REFRESH_USER_TOKEN;
    }
    return RefreshUserToken;
}());
export { RefreshUserToken };
if (false) {
    /** @type {?} */
    RefreshUserToken.prototype.type;
    /** @type {?} */
    RefreshUserToken.prototype.payload;
}
var RefreshUserTokenSuccess = /** @class */ (function () {
    function RefreshUserTokenSuccess(payload) {
        this.payload = payload;
        this.type = REFRESH_USER_TOKEN_SUCCESS;
    }
    return RefreshUserTokenSuccess;
}());
export { RefreshUserTokenSuccess };
if (false) {
    /** @type {?} */
    RefreshUserTokenSuccess.prototype.type;
    /** @type {?} */
    RefreshUserTokenSuccess.prototype.payload;
}
var RefreshUserTokenFail = /** @class */ (function () {
    function RefreshUserTokenFail(payload) {
        this.payload = payload;
        this.type = REFRESH_USER_TOKEN_FAIL;
    }
    return RefreshUserTokenFail;
}());
export { RefreshUserTokenFail };
if (false) {
    /** @type {?} */
    RefreshUserTokenFail.prototype.type;
    /** @type {?} */
    RefreshUserTokenFail.prototype.payload;
}
var RevokeUserToken = /** @class */ (function () {
    function RevokeUserToken(payload) {
        this.payload = payload;
        this.type = REVOKE_USER_TOKEN;
    }
    return RevokeUserToken;
}());
export { RevokeUserToken };
if (false) {
    /** @type {?} */
    RevokeUserToken.prototype.type;
    /** @type {?} */
    RevokeUserToken.prototype.payload;
}
var RevokeUserTokenSuccess = /** @class */ (function () {
    function RevokeUserTokenSuccess(payload) {
        this.payload = payload;
        this.type = REVOKE_USER_TOKEN_SUCCESS;
    }
    return RevokeUserTokenSuccess;
}());
export { RevokeUserTokenSuccess };
if (false) {
    /** @type {?} */
    RevokeUserTokenSuccess.prototype.type;
    /** @type {?} */
    RevokeUserTokenSuccess.prototype.payload;
}
var RevokeUserTokenFail = /** @class */ (function () {
    function RevokeUserTokenFail(payload) {
        this.payload = payload;
        this.type = REVOKE_USER_TOKEN_FAIL;
    }
    return RevokeUserTokenFail;
}());
export { RevokeUserTokenFail };
if (false) {
    /** @type {?} */
    RevokeUserTokenFail.prototype.type;
    /** @type {?} */
    RevokeUserTokenFail.prototype.payload;
}
