// @spartacus/core/esm2015/src/routing/external-routes/external-routes.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { UrlMatcherFactoryService } from '../services/url-matcher-factory.service';
import { ExternalRoutesConfig } from './external-routes-config';
import { ExternalRoutesGuard } from './external-routes.guard';
/**
 * Service that helps redirecting to different storefront systems for configured URLs
 */
export class ExternalRoutesService {
    /**
     * @param {?} config
     * @param {?} matcherFactory
     * @param {?} injector
     */
    constructor(config, matcherFactory, injector) {
        this.config = config;
        this.matcherFactory = matcherFactory;
        this.injector = injector;
    }
    /**
     * @protected
     * @return {?}
     */
    get internalUrlPatterns() {
        return ((this.config && this.config.routing && this.config.routing.internal) || []);
    }
    /**
     * Prepends routes (to the Router.config) that are responsible for redirecting to a different storefront system
     * @return {?}
     */
    addRoutes() {
        /** @type {?} */
        const router = this.injector.get(Router);
        /** @type {?} */
        const newRoutes = this.getRoutes();
        if (newRoutes.length) {
            router.resetConfig([...newRoutes, ...router.config]);
        }
    }
    /**
     * Returns routes that are responsible for redirection to different storefront systems
     * @protected
     * @return {?}
     */
    getRoutes() {
        if (!this.internalUrlPatterns.length) {
            return [];
        }
        /** @type {?} */
        const routes = [];
        routes.push({
            pathMatch: 'full',
            matcher: this.getUrlMatcher(),
            canActivate: [ExternalRoutesGuard],
            component: (/** @type {?} */ ({})),
        });
        return routes;
    }
    /**
     * Returns the URL matcher for the external route
     * @protected
     * @return {?}
     */
    getUrlMatcher() {
        /** @type {?} */
        const matcher = this.matcherFactory.getGlobUrlMatcher(this.internalUrlPatterns);
        return this.matcherFactory.getOppositeUrlMatcher(matcher); // the external route should be activated only when it's NOT an internal route
    }
}
ExternalRoutesService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ExternalRoutesService.ctorParameters = () => [
    { type: ExternalRoutesConfig },
    { type: UrlMatcherFactoryService },
    { type: Injector }
];
if (false) {
    /**
     * @type {?}
     * @protected
     */
    ExternalRoutesService.prototype.config;
    /**
     * @type {?}
     * @protected
     */
    ExternalRoutesService.prototype.matcherFactory;
    /**
     * @type {?}
     * @protected
     */
    ExternalRoutesService.prototype.injector;
}
