// @spartacus/core/esm5/src/anonymous-consents/store/selectors/anonymous-consent-templates.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/index';
import { getAnonymousConsentState } from './feature.selector';
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.templates; };
/** @type {?} */
export var getAnonymousConsentTemplatesState = createSelector(getAnonymousConsentState, (ɵ0));
/** @type {?} */
export var getAnonymousConsentTemplatesValue = createSelector(getAnonymousConsentTemplatesState, StateLoaderSelectors.loaderValueSelector);
/** @type {?} */
export var getAnonymousConsentTemplatesLoading = createSelector(getAnonymousConsentTemplatesState, StateLoaderSelectors.loaderLoadingSelector);
/** @type {?} */
export var getAnonymousConsentTemplatesSuccess = createSelector(getAnonymousConsentTemplatesState, StateLoaderSelectors.loaderSuccessSelector);
/** @type {?} */
export var getAnonymousConsentTemplatesError = createSelector(getAnonymousConsentTemplatesState, StateLoaderSelectors.loaderErrorSelector);
/** @type {?} */
export var getAnonymousConsentTemplate = (/**
 * @param {?} templateCode
 * @return {?}
 */
function (templateCode) {
    return createSelector(getAnonymousConsentTemplatesValue, (/**
     * @param {?} templates
     * @return {?}
     */
    function (templates) {
        return templates
            ? templates.find((/**
             * @param {?} template
             * @return {?}
             */
            function (template) { return template.id === templateCode; }))
            : null;
    }));
});
export { ɵ0 };
