// @spartacus/core/esm2015/src/user/store/actions/billing-countries.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const LOAD_BILLING_COUNTRIES = '[User] Load Billing Countries';
/** @type {?} */
export const LOAD_BILLING_COUNTRIES_FAIL = '[User] Load Billing Countries Fail';
/** @type {?} */
export const LOAD_BILLING_COUNTRIES_SUCCESS = '[User] Load Billing Countries Success';
export class LoadBillingCountries {
    constructor() {
        this.type = LOAD_BILLING_COUNTRIES;
    }
}
if (false) {
    /** @type {?} */
    LoadBillingCountries.prototype.type;
}
export class LoadBillingCountriesFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_BILLING_COUNTRIES_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadBillingCountriesFail.prototype.type;
    /** @type {?} */
    LoadBillingCountriesFail.prototype.payload;
}
export class LoadBillingCountriesSuccess {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_BILLING_COUNTRIES_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadBillingCountriesSuccess.prototype.type;
    /** @type {?} */
    LoadBillingCountriesSuccess.prototype.payload;
}
