// @spartacus/core/esm2015/src/user/store/actions/delivery-countries.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const LOAD_DELIVERY_COUNTRIES = '[User] Load Delivery Countries';
/** @type {?} */
export const LOAD_DELIVERY_COUNTRIES_FAIL = '[User] Load Delivery Countries Fail';
/** @type {?} */
export const LOAD_DELIVERY_COUNTRIES_SUCCESS = '[User] Load Delivery Countries Success';
export class LoadDeliveryCountries {
    constructor() {
        this.type = LOAD_DELIVERY_COUNTRIES;
    }
}
if (false) {
    /** @type {?} */
    LoadDeliveryCountries.prototype.type;
}
export class LoadDeliveryCountriesFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_DELIVERY_COUNTRIES_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadDeliveryCountriesFail.prototype.type;
    /** @type {?} */
    LoadDeliveryCountriesFail.prototype.payload;
}
export class LoadDeliveryCountriesSuccess {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_DELIVERY_COUNTRIES_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadDeliveryCountriesSuccess.prototype.type;
    /** @type {?} */
    LoadDeliveryCountriesSuccess.prototype.payload;
}
