// @spartacus/core/esm2015/src/user/store/actions/forgot-password.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const FORGOT_PASSWORD_EMAIL_REQUEST = '[User] Forgot Password Email Request';
/** @type {?} */
export const FORGOT_PASSWORD_EMAIL_REQUEST_SUCCESS = '[User] Forgot Password Email Request Success';
/** @type {?} */
export const FORGOT_PASSWORD_EMAIL_REQUEST_FAIL = '[User] Forgot Password Email Request Fail';
export class ForgotPasswordEmailRequest {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = FORGOT_PASSWORD_EMAIL_REQUEST;
    }
}
if (false) {
    /** @type {?} */
    ForgotPasswordEmailRequest.prototype.type;
    /** @type {?} */
    ForgotPasswordEmailRequest.prototype.payload;
}
export class ForgotPasswordEmailRequestFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = FORGOT_PASSWORD_EMAIL_REQUEST_FAIL;
    }
}
if (false) {
    /** @type {?} */
    ForgotPasswordEmailRequestFail.prototype.type;
    /** @type {?} */
    ForgotPasswordEmailRequestFail.prototype.payload;
}
export class ForgotPasswordEmailRequestSuccess {
    constructor() {
        this.type = FORGOT_PASSWORD_EMAIL_REQUEST_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    ForgotPasswordEmailRequestSuccess.prototype.type;
}
