// @spartacus/core/esm2015/src/user/store/actions/product-interests.actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PRODUCT_INTERESTS, REMOVE_PRODUCT_INTERESTS_PROCESS_ID, ADD_PRODUCT_INTEREST_PROCESS_ID, } from '../user-state';
import { PROCESS_FEATURE } from '../../../process/store/process-state';
import { LoaderLoadAction, LoaderFailAction, LoaderSuccessAction, LoaderResetAction, } from '../../../state/utils/loader/loader.action';
import { EntityFailAction, EntityLoadAction, EntitySuccessAction, EntityResetAction, } from '../../../state/utils/entity-loader/entity-loader.action';
/** @type {?} */
export const LOAD_PRODUCT_INTERESTS = 'Load Product Interests';
/** @type {?} */
export const LOAD_PRODUCT_INTERESTS_FAIL = 'Load Product Interests Fail';
/** @type {?} */
export const LOAD_PRODUCT_INTERESTS_SUCCESS = 'Load Product Interests Success';
/** @type {?} */
export const REMOVE_PRODUCT_INTEREST = 'Remove Product Interest';
/** @type {?} */
export const REMOVE_PRODUCT_INTEREST_SUCCESS = 'Remove Product Interest Success';
/** @type {?} */
export const REMOVE_PRODUCT_INTEREST_FAIL = 'Remove Product Interest Fail';
/** @type {?} */
export const ADD_PRODUCT_INTEREST = 'Add Product Interest';
/** @type {?} */
export const ADD_PRODUCT_INTEREST_FAIL = 'Add Product Interest Fail';
/** @type {?} */
export const ADD_PRODUCT_INTEREST_SUCCESS = 'Add Product Interest Success';
/** @type {?} */
export const ADD_PRODUCT_INTEREST_RESET = 'Add Product Interest Reset';
/** @type {?} */
export const REMOVE_PRODUCT_INTEREST_RESET = 'Remove Product Interest Reset';
/** @type {?} */
export const CLEAR_PRODUCT_INTERESTS = 'Clear Product Interests';
export class LoadProductInterests extends LoaderLoadAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(PRODUCT_INTERESTS);
        this.payload = payload;
        this.type = LOAD_PRODUCT_INTERESTS;
    }
}
if (false) {
    /** @type {?} */
    LoadProductInterests.prototype.type;
    /** @type {?} */
    LoadProductInterests.prototype.payload;
}
export class LoadProductInterestsFail extends LoaderFailAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(PRODUCT_INTERESTS, payload);
        this.payload = payload;
        this.type = LOAD_PRODUCT_INTERESTS_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadProductInterestsFail.prototype.type;
    /** @type {?} */
    LoadProductInterestsFail.prototype.payload;
}
export class LoadProductInterestsSuccess extends LoaderSuccessAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(PRODUCT_INTERESTS);
        this.payload = payload;
        this.type = LOAD_PRODUCT_INTERESTS_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadProductInterestsSuccess.prototype.type;
    /** @type {?} */
    LoadProductInterestsSuccess.prototype.payload;
}
export class RemoveProductInterest extends EntityLoadAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(PROCESS_FEATURE, REMOVE_PRODUCT_INTERESTS_PROCESS_ID);
        this.payload = payload;
        this.type = REMOVE_PRODUCT_INTEREST;
    }
}
if (false) {
    /** @type {?} */
    RemoveProductInterest.prototype.type;
    /** @type {?} */
    RemoveProductInterest.prototype.payload;
}
export class RemoveProductInterestSuccess extends EntitySuccessAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(PROCESS_FEATURE, REMOVE_PRODUCT_INTERESTS_PROCESS_ID);
        this.payload = payload;
        this.type = REMOVE_PRODUCT_INTEREST_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    RemoveProductInterestSuccess.prototype.type;
    /** @type {?} */
    RemoveProductInterestSuccess.prototype.payload;
}
export class RemoveProductInterestFail extends EntityFailAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(PROCESS_FEATURE, REMOVE_PRODUCT_INTERESTS_PROCESS_ID, payload);
        this.payload = payload;
        this.type = REMOVE_PRODUCT_INTEREST_FAIL;
    }
}
if (false) {
    /** @type {?} */
    RemoveProductInterestFail.prototype.type;
    /** @type {?} */
    RemoveProductInterestFail.prototype.payload;
}
export class AddProductInterest extends EntityLoadAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(PROCESS_FEATURE, ADD_PRODUCT_INTEREST_PROCESS_ID);
        this.payload = payload;
        this.type = ADD_PRODUCT_INTEREST;
    }
}
if (false) {
    /** @type {?} */
    AddProductInterest.prototype.type;
    /** @type {?} */
    AddProductInterest.prototype.payload;
}
export class AddProductInterestSuccess extends EntitySuccessAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(PROCESS_FEATURE, ADD_PRODUCT_INTEREST_PROCESS_ID);
        this.payload = payload;
        this.type = ADD_PRODUCT_INTEREST_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    AddProductInterestSuccess.prototype.type;
    /** @type {?} */
    AddProductInterestSuccess.prototype.payload;
}
export class AddProductInterestFail extends EntityFailAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(PROCESS_FEATURE, ADD_PRODUCT_INTEREST_PROCESS_ID, payload);
        this.payload = payload;
        this.type = ADD_PRODUCT_INTEREST_FAIL;
    }
}
if (false) {
    /** @type {?} */
    AddProductInterestFail.prototype.type;
    /** @type {?} */
    AddProductInterestFail.prototype.payload;
}
export class ResetAddInterestState extends EntityResetAction {
    constructor() {
        super(PROCESS_FEATURE, ADD_PRODUCT_INTEREST_PROCESS_ID);
        this.type = ADD_PRODUCT_INTEREST_RESET;
    }
}
if (false) {
    /** @type {?} */
    ResetAddInterestState.prototype.type;
}
export class ResetRemoveInterestState extends EntityResetAction {
    constructor() {
        super(PROCESS_FEATURE, REMOVE_PRODUCT_INTERESTS_PROCESS_ID);
        this.type = REMOVE_PRODUCT_INTEREST_RESET;
    }
}
if (false) {
    /** @type {?} */
    ResetRemoveInterestState.prototype.type;
}
export class ClearProductInterests extends LoaderResetAction {
    constructor() {
        super(PRODUCT_INTERESTS);
        this.type = CLEAR_PRODUCT_INTERESTS;
    }
}
if (false) {
    /** @type {?} */
    ClearProductInterests.prototype.type;
}
