// @spartacus/core/esm2015/src/user/store/actions/regions.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StateLoaderActions } from '../../../state/utils/index';
import { REGIONS } from '../user-state';
/** @type {?} */
export const LOAD_REGIONS = '[User] Load Regions';
/** @type {?} */
export const LOAD_REGIONS_SUCCESS = '[User] Load Regions Success';
/** @type {?} */
export const LOAD_REGIONS_FAIL = '[User] Load Regions Fail';
/** @type {?} */
export const CLEAR_REGIONS = '[User] Clear Regions';
export class LoadRegions extends StateLoaderActions.LoaderLoadAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(REGIONS);
        this.payload = payload;
        this.type = LOAD_REGIONS;
    }
}
if (false) {
    /** @type {?} */
    LoadRegions.prototype.type;
    /** @type {?} */
    LoadRegions.prototype.payload;
}
export class LoadRegionsFail extends StateLoaderActions.LoaderFailAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(REGIONS, payload);
        this.payload = payload;
        this.type = LOAD_REGIONS_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadRegionsFail.prototype.type;
    /** @type {?} */
    LoadRegionsFail.prototype.payload;
}
export class LoadRegionsSuccess extends StateLoaderActions.LoaderSuccessAction {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        super(REGIONS);
        this.payload = payload;
        this.type = LOAD_REGIONS_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadRegionsSuccess.prototype.type;
    /** @type {?} */
    LoadRegionsSuccess.prototype.payload;
}
export class ClearRegions {
    constructor() {
        this.type = CLEAR_REGIONS;
    }
}
if (false) {
    /** @type {?} */
    ClearRegions.prototype.type;
}
