// @spartacus/core/esm2015/src/user/store/actions/reset-password.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const RESET_PASSWORD = '[User] Reset Password';
/** @type {?} */
export const RESET_PASSWORD_SUCCESS = '[User] Reset Password Success';
/** @type {?} */
export const RESET_PASSWORD_FAIL = '[User] Reset Password Fail';
export class ResetPassword {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = RESET_PASSWORD;
    }
}
if (false) {
    /** @type {?} */
    ResetPassword.prototype.type;
    /** @type {?} */
    ResetPassword.prototype.payload;
}
export class ResetPasswordFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = RESET_PASSWORD_FAIL;
    }
}
if (false) {
    /** @type {?} */
    ResetPasswordFail.prototype.type;
    /** @type {?} */
    ResetPasswordFail.prototype.payload;
}
export class ResetPasswordSuccess {
    constructor() {
        this.type = RESET_PASSWORD_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    ResetPasswordSuccess.prototype.type;
}
