// @spartacus/core/esm2015/src/user/store/actions/titles.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const LOAD_TITLES = '[User] Load Tiltes';
/** @type {?} */
export const LOAD_TITLES_FAIL = '[User] Load Titles Fail';
/** @type {?} */
export const LOAD_TITLES_SUCCESS = '[User] Load Titles Success';
export class LoadTitles {
    constructor() {
        this.type = LOAD_TITLES;
    }
}
if (false) {
    /** @type {?} */
    LoadTitles.prototype.type;
}
export class LoadTitlesFail {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_TITLES_FAIL;
    }
}
if (false) {
    /** @type {?} */
    LoadTitlesFail.prototype.type;
    /** @type {?} */
    LoadTitlesFail.prototype.payload;
}
export class LoadTitlesSuccess {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = LOAD_TITLES_SUCCESS;
    }
}
if (false) {
    /** @type {?} */
    LoadTitlesSuccess.prototype.type;
    /** @type {?} */
    LoadTitlesSuccess.prototype.payload;
}
