// @spartacus/core/esm5/src/user/store/selectors/billing-countries.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { getUserState } from './feature.selector';
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.billingCountries; };
/** @type {?} */
export var getBillingCountriesState = createSelector(getUserState, (ɵ0));
var ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.entities; };
/** @type {?} */
export var getBillingCountriesEntites = createSelector(getBillingCountriesState, (ɵ1));
var ɵ2 = /**
 * @param {?} entites
 * @return {?}
 */
function (entites) { return Object.keys(entites).map((/**
 * @param {?} isocode
 * @return {?}
 */
function (isocode) { return entites[isocode]; })); };
/** @type {?} */
export var getAllBillingCountries = createSelector(getBillingCountriesEntites, (ɵ2));
export { ɵ0, ɵ1, ɵ2 };
