// @spartacus/core/esm5/src/user/store/selectors/customer-coupons.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { loaderValueSelector, loaderSuccessSelector, loaderLoadingSelector, } from '../../../state/utils/loader/loader.selectors';
import { getUserState } from './feature.selector';
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.customerCoupons; };
/** @type {?} */
export var getCustomerCouponsState = createSelector(getUserState, (ɵ0));
var ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return loaderSuccessSelector(state);
};
/** @type {?} */
export var getCustomerCouponsLoaded = createSelector(getCustomerCouponsState, (ɵ1));
var ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return loaderLoadingSelector(state);
};
/** @type {?} */
export var getCustomerCouponsLoading = createSelector(getCustomerCouponsState, (ɵ2));
var ɵ3 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return loaderValueSelector(state); };
/** @type {?} */
export var getCustomerCoupons = createSelector(getCustomerCouponsState, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
