// @spartacus/core/esm5/src/user/store/selectors/delivery-countries.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { getUserState } from './feature.selector';
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.countries; };
/** @type {?} */
export var getDeliveryCountriesState = createSelector(getUserState, (ɵ0));
var ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.entities; };
/** @type {?} */
export var getDeliveryCountriesEntites = createSelector(getDeliveryCountriesState, (ɵ1));
var ɵ2 = /**
 * @param {?} entites
 * @return {?}
 */
function (entites) { return Object.keys(entites).map((/**
 * @param {?} isocode
 * @return {?}
 */
function (isocode) { return entites[isocode]; })); };
/** @type {?} */
export var getAllDeliveryCountries = createSelector(getDeliveryCountriesEntites, (ɵ2));
/** @type {?} */
export var countrySelectorFactory = (/**
 * @param {?} isocode
 * @return {?}
 */
function (isocode) {
    return createSelector(getDeliveryCountriesEntites, (/**
     * @param {?} entities
     * @return {?}
     */
    function (entities) { return (Object.keys(entities).length !== 0 ? entities[isocode] : null); }));
});
export { ɵ0, ɵ1, ɵ2 };
