// @spartacus/core/esm5/src/user/store/selectors/notification-preference.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { getUserState } from './feature.selector';
import { loaderValueSelector, loaderLoadingSelector, } from '../../../state/utils/loader/loader.selectors';
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.notificationPreferences; };
/** @type {?} */
export var getPreferencesLoaderState = createSelector(getUserState, (ɵ0));
var ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return loaderValueSelector(state); };
/** @type {?} */
export var getPreferences = createSelector(getPreferencesLoaderState, (ɵ1));
var ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return loaderValueSelector(state).filter((/**
     * @param {?} p
     * @return {?}
     */
    function (p) { return p.enabled; }));
};
/** @type {?} */
export var getEnabledPreferences = createSelector(getPreferencesLoaderState, (ɵ2));
var ɵ3 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return loaderLoadingSelector(state); };
/** @type {?} */
export var getPreferencesLoading = createSelector(getPreferencesLoaderState, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
