// @spartacus/core/esm5/src/user/store/selectors/payment-methods.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/index';
import { getUserState } from './feature.selector';
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.payments; };
/** @type {?} */
export var getPaymentMethodsState = createSelector(getUserState, (ɵ0));
var ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return StateLoaderSelectors.loaderValueSelector(state);
};
/** @type {?} */
export var getPaymentMethods = createSelector(getPaymentMethodsState, (ɵ1));
var ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return StateLoaderSelectors.loaderLoadingSelector(state);
};
/** @type {?} */
export var getPaymentMethodsLoading = createSelector(getPaymentMethodsState, (ɵ2));
var ɵ3 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return StateLoaderSelectors.loaderSuccessSelector(state) &&
        !StateLoaderSelectors.loaderLoadingSelector(state);
};
/** @type {?} */
export var getPaymentMethodsLoadedSuccess = createSelector(getPaymentMethodsState, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
