// @spartacus/core/esm5/src/user/store/selectors/titles.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { getUserState } from './feature.selector';
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.titles; };
/** @type {?} */
export var getTitlesState = createSelector(getUserState, (ɵ0));
var ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.entities; };
/** @type {?} */
export var getTitlesEntites = createSelector(getTitlesState, (ɵ1));
var ɵ2 = /**
 * @param {?} entites
 * @return {?}
 */
function (entites) { return Object.keys(entites).map((/**
 * @param {?} code
 * @return {?}
 */
function (code) { return entites[code]; })); };
/** @type {?} */
export var getAllTitles = createSelector(getTitlesEntites, (ɵ2));
/** @type {?} */
export var titleSelectorFactory = (/**
 * @param {?} code
 * @return {?}
 */
function (code) {
    return createSelector(getTitlesEntites, (/**
     * @param {?} entities
     * @return {?}
     */
    function (entities) { return (Object.keys(entities).length !== 0 ? entities[code] : null); }));
});
export { ɵ0, ɵ1, ɵ2 };
