// @spartacus/core/esm5/src/user/store/selectors/user-addresses.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createSelector } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/index';
import { getUserState } from './feature.selector';
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.addresses; };
/** @type {?} */
export var getAddressesLoaderState = createSelector(getUserState, (ɵ0));
var ɵ1 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return StateLoaderSelectors.loaderValueSelector(state);
};
/** @type {?} */
export var getAddresses = createSelector(getAddressesLoaderState, (ɵ1));
var ɵ2 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return StateLoaderSelectors.loaderLoadingSelector(state);
};
/** @type {?} */
export var getAddressesLoading = createSelector(getAddressesLoaderState, (ɵ2));
var ɵ3 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return StateLoaderSelectors.loaderSuccessSelector(state) &&
        !StateLoaderSelectors.loaderLoadingSelector(state);
};
/** @type {?} */
export var getAddressesLoadedSuccess = createSelector(getAddressesLoaderState, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
