// @spartacus/core/esm2015/src/state/utils/scoped-loader/entity-scoped-loader.actions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ENTITY_FAIL_ACTION, ENTITY_LOAD_ACTION, ENTITY_RESET_ACTION, ENTITY_SUCCESS_ACTION, entityFailMeta, entityLoadMeta, entityResetMeta, entitySuccessMeta, } from '../entity-loader/entity-loader.action';
export var EntityScopedLoaderActions;
(function (EntityScopedLoaderActions) {
    /**
     * @record
     */
    function EntityScopedLoaderMeta() { }
    EntityScopedLoaderActions.EntityScopedLoaderMeta = EntityScopedLoaderMeta;
    if (false) {
        /** @type {?|undefined} */
        EntityScopedLoaderMeta.prototype.scope;
    }
    /**
     * @record
     */
    function EntityScopedLoaderAction() { }
    EntityScopedLoaderActions.EntityScopedLoaderAction = EntityScopedLoaderAction;
    if (false) {
        /** @type {?|undefined} */
        EntityScopedLoaderAction.prototype.payload;
        /** @type {?|undefined} */
        EntityScopedLoaderAction.prototype.meta;
    }
    /**
     * @param {?} entityType
     * @param {?} id
     * @param {?=} scope
     * @return {?}
     */
    function entityScopedLoadMeta(entityType, id, scope) {
        return Object.assign({}, entityLoadMeta(entityType, id), { scope });
    }
    EntityScopedLoaderActions.entityScopedLoadMeta = entityScopedLoadMeta;
    /**
     * @param {?} entityType
     * @param {?} id
     * @param {?=} scope
     * @param {?=} error
     * @return {?}
     */
    function entityScopedFailMeta(entityType, id, scope, error) {
        return Object.assign({}, entityFailMeta(entityType, id, error), { scope });
    }
    EntityScopedLoaderActions.entityScopedFailMeta = entityScopedFailMeta;
    /**
     * @param {?} entityType
     * @param {?} id
     * @param {?=} scope
     * @return {?}
     */
    function entityScopedSuccessMeta(entityType, id, scope) {
        return Object.assign({}, entitySuccessMeta(entityType, id), { scope });
    }
    EntityScopedLoaderActions.entityScopedSuccessMeta = entityScopedSuccessMeta;
    /**
     * @param {?} entityType
     * @param {?} id
     * @param {?=} scope
     * @return {?}
     */
    function entityScopedResetMeta(entityType, id, scope) {
        return Object.assign({}, entityResetMeta(entityType, id), { scope });
    }
    EntityScopedLoaderActions.entityScopedResetMeta = entityScopedResetMeta;
    class EntityScopedLoadAction {
        /**
         * @param {?} entityType
         * @param {?} id
         * @param {?=} scope
         */
        constructor(entityType, id, scope) {
            this.type = ENTITY_LOAD_ACTION;
            this.meta = entityScopedLoadMeta(entityType, id, scope);
        }
    }
    EntityScopedLoaderActions.EntityScopedLoadAction = EntityScopedLoadAction;
    if (false) {
        /** @type {?} */
        EntityScopedLoadAction.prototype.type;
        /** @type {?} */
        EntityScopedLoadAction.prototype.meta;
    }
    class EntityScopedFailAction {
        /**
         * @param {?} entityType
         * @param {?} id
         * @param {?=} scope
         * @param {?=} error
         */
        constructor(entityType, id, scope, error) {
            this.type = ENTITY_FAIL_ACTION;
            this.meta = entityScopedFailMeta(entityType, id, scope, error);
        }
    }
    EntityScopedLoaderActions.EntityScopedFailAction = EntityScopedFailAction;
    if (false) {
        /** @type {?} */
        EntityScopedFailAction.prototype.type;
        /** @type {?} */
        EntityScopedFailAction.prototype.meta;
    }
    class EntityScopedSuccessAction {
        /**
         * @param {?} entityType
         * @param {?} id
         * @param {?=} scope
         * @param {?=} payload
         */
        constructor(entityType, id, scope, payload) {
            this.payload = payload;
            this.type = ENTITY_SUCCESS_ACTION;
            this.meta = entityScopedSuccessMeta(entityType, id, scope);
        }
    }
    EntityScopedLoaderActions.EntityScopedSuccessAction = EntityScopedSuccessAction;
    if (false) {
        /** @type {?} */
        EntityScopedSuccessAction.prototype.type;
        /** @type {?} */
        EntityScopedSuccessAction.prototype.meta;
        /** @type {?} */
        EntityScopedSuccessAction.prototype.payload;
    }
    class EntityScopedResetAction {
        /**
         * @param {?} entityType
         * @param {?} id
         * @param {?=} scope
         */
        constructor(entityType, id, scope) {
            this.type = ENTITY_RESET_ACTION;
            this.meta = entityScopedResetMeta(entityType, id, scope);
        }
    }
    EntityScopedLoaderActions.EntityScopedResetAction = EntityScopedResetAction;
    if (false) {
        /** @type {?} */
        EntityScopedResetAction.prototype.type;
        /** @type {?} */
        EntityScopedResetAction.prototype.meta;
    }
})(EntityScopedLoaderActions || (EntityScopedLoaderActions = {}));
