// @spartacus/core/esm5/src/product/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export { ProductConnector, ProductAdapter, PRODUCT_NORMALIZER, PRODUCT_REFERENCES_NORMALIZER, ProductReferencesAdapter, ProductReferencesConnector, ProductReviewsConnector, ProductReviewsAdapter, PRODUCT_REVIEW_NORMALIZER, PRODUCT_REVIEW_SERIALIZER, ProductSearchConnector, ProductSearchAdapter, PRODUCT_SEARCH_PAGE_NORMALIZER, PRODUCT_SUGGESTION_NORMALIZER } from './connectors/index';
export { ProductReferenceService, ProductReviewService, ProductSearchService, ProductService, SearchboxService } from './facade/index';
export { ProductScope } from './model/index';
export { ProductModule } from './product.module';
export { CategoryPageMetaResolver, ProductPageMetaResolver, SearchPageMetaResolver, ProductLoadingService } from './services/index';
export { ProductActions } from './store/actions/index';
export { PRODUCT_FEATURE, PRODUCT_DETAIL_ENTITY } from './store/product-state';
export { ProductSelectors } from './store/selectors/index';
