// @spartacus/core/esm2015/src/occ/config-loader/occ-loaded-config-converter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BASE_SITE_CONTEXT_ID, CURRENCY_CONTEXT_ID, LANGUAGE_CONTEXT_ID, } from '../../site-context/providers/context-ids';
import { JavaRegExpConverter } from './java-reg-exp-converter';
import * as i0 from "@angular/core";
import * as i1 from "./java-reg-exp-converter";
export class OccLoadedConfigConverter {
    /**
     * @param {?} javaRegExpConverter
     */
    constructor(javaRegExpConverter) {
        this.javaRegExpConverter = javaRegExpConverter;
    }
    /**
     * @param {?} baseSites
     * @param {?} currentUrl
     * @return {?}
     */
    fromOccBaseSites(baseSites, currentUrl) {
        /** @type {?} */
        const baseSite = baseSites.find((/**
         * @param {?} site
         * @return {?}
         */
        site => this.isCurrentBaseSite(site, currentUrl)));
        if (!baseSite) {
            throw this.getError(`Current url (${currentUrl}) doesn't match with any of url patterns of any base site.`);
        }
        // Although `stores` property is an array, typically there is only one store. So we return the first store from the list.
        /** @type {?} */
        const baseStore = baseSite.stores && baseSite.stores[0];
        if (!baseStore) {
            throw this.getError(`Current base site (${baseSite.uid}) doesn't have any base store.`);
        }
        return {
            baseSite: baseSite.uid,
            languages: this.getIsoCodes(baseStore.languages, baseSite.defaultLanguage || baseStore.defaultLanguage),
            currencies: this.getIsoCodes(baseStore.currencies, baseStore.defaultCurrency),
            urlParameters: this.getUrlParams(baseSite.urlEncodingAttributes),
        };
    }
    /**
     * @param {?} __0
     * @return {?}
     */
    toSiteContextConfig({ baseSite, languages, currencies, urlParameters: urlEncodingAttributes, }) {
        /** @type {?} */
        const result = {
            context: {
                urlParameters: urlEncodingAttributes,
                [BASE_SITE_CONTEXT_ID]: [baseSite],
                [LANGUAGE_CONTEXT_ID]: languages,
                [CURRENCY_CONTEXT_ID]: currencies,
            },
        };
        return result;
    }
    /**
     * @param {?} __0
     * @return {?}
     */
    toI18nConfig({ languages }) {
        return { i18n: { fallbackLang: languages[0] } };
    }
    /**
     * @private
     * @param {?} site
     * @param {?} currentUrl
     * @return {?}
     */
    isCurrentBaseSite(site, currentUrl) {
        /** @type {?} */
        const index = (site.urlPatterns || []).findIndex((/**
         * @param {?} javaRegexp
         * @return {?}
         */
        javaRegexp => {
            /** @type {?} */
            const jsRegexp = this.javaRegExpConverter.toJsRegExp(javaRegexp);
            if (jsRegexp) {
                /** @type {?} */
                const result = jsRegexp.test(currentUrl);
                return result;
            }
        }));
        return index !== -1;
    }
    /**
     * Returns an array of url encoded site context parameters.
     *
     * It maps the string "storefront" (used in OCC) to the "baseSite" (used in Spartacus)
     * @private
     * @param {?} params
     * @return {?}
     */
    getUrlParams(params) {
        /** @type {?} */
        const STOREFRONT_PARAM = 'storefront';
        return (params || []).map((/**
         * @param {?} param
         * @return {?}
         */
        param => param === STOREFRONT_PARAM ? BASE_SITE_CONTEXT_ID : param));
    }
    /**
     * Returns iso codes in a array, where the first element is the default iso code.
     * @private
     * @param {?} elements
     * @param {?} defaultElement
     * @return {?}
     */
    getIsoCodes(elements, defaultElement) {
        /** @type {?} */
        const result = this.moveToFirst(elements, (/**
         * @param {?} el
         * @return {?}
         */
        el => el.isocode === defaultElement.isocode)).map((/**
         * @param {?} el
         * @return {?}
         */
        el => el.isocode));
        return result;
    }
    /**
     * Moves to the start of the array the first element that satisfies the given predicate.
     *
     * @private
     * @param {?} array array to modify
     * @param {?} predicate function called on elements
     * @return {?}
     */
    moveToFirst(array, predicate) {
        array = [...array];
        /** @type {?} */
        const index = array.findIndex(predicate);
        if (index !== -1) {
            const [el] = array.splice(index, 1);
            array.unshift(el);
        }
        return array;
    }
    /**
     * @private
     * @param {?} message
     * @return {?}
     */
    getError(message) {
        return new Error(`Error: Cannot get base site config! ${message}`);
    }
}
OccLoadedConfigConverter.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
OccLoadedConfigConverter.ctorParameters = () => [
    { type: JavaRegExpConverter }
];
/** @nocollapse */ OccLoadedConfigConverter.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function OccLoadedConfigConverter_Factory() { return new OccLoadedConfigConverter(i0.ɵɵinject(i1.JavaRegExpConverter)); }, token: OccLoadedConfigConverter, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    OccLoadedConfigConverter.prototype.javaRegExpConverter;
}
