// @spartacus/core/esm5/src/routing/store/actions/router.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var ROUTER_GO = '[Router] Go';
/** @type {?} */
export var ROUTER_GO_BY_URL = '[Router] Go By Url';
/** @type {?} */
export var ROUTER_BACK = '[Router] Back';
/** @type {?} */
export var ROUTER_FORWARD = '[Router] Forward';
var RouteGoAction = /** @class */ (function () {
    function RouteGoAction(payload) {
        this.payload = payload;
        this.type = ROUTER_GO;
    }
    return RouteGoAction;
}());
export { RouteGoAction };
if (false) {
    /** @type {?} */
    RouteGoAction.prototype.type;
    /** @type {?} */
    RouteGoAction.prototype.payload;
}
var RouteGoByUrlAction = /** @class */ (function () {
    function RouteGoByUrlAction(payload) {
        this.payload = payload;
        this.type = ROUTER_GO_BY_URL;
    }
    return RouteGoByUrlAction;
}());
export { RouteGoByUrlAction };
if (false) {
    /** @type {?} */
    RouteGoByUrlAction.prototype.type;
    /** @type {?} */
    RouteGoByUrlAction.prototype.payload;
}
var RouteBackAction = /** @class */ (function () {
    function RouteBackAction() {
        this.type = ROUTER_BACK;
    }
    return RouteBackAction;
}());
export { RouteBackAction };
if (false) {
    /** @type {?} */
    RouteBackAction.prototype.type;
}
var RouteForwardAction = /** @class */ (function () {
    function RouteForwardAction() {
        this.type = ROUTER_FORWARD;
    }
    return RouteForwardAction;
}());
export { RouteForwardAction };
if (false) {
    /** @type {?} */
    RouteForwardAction.prototype.type;
}
