// @spartacus/core/esm5/src/site-context/store/state.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var SITE_CONTEXT_FEATURE = 'siteContext';
/**
 * @record
 */
export function StateWithSiteContext() { }
if (false) {
    /* Skipping unnamed member:
    [SITE_CONTEXT_FEATURE]: SiteContextState;*/
}
/**
 * @record
 */
export function SiteContextState() { }
if (false) {
    /** @type {?} */
    SiteContextState.prototype.languages;
    /** @type {?} */
    SiteContextState.prototype.currencies;
    /** @type {?} */
    SiteContextState.prototype.baseSite;
}
/**
 * @record
 */
export function CurrencyEntities() { }
/**
 * @record
 */
export function CurrenciesState() { }
if (false) {
    /** @type {?} */
    CurrenciesState.prototype.entities;
    /** @type {?} */
    CurrenciesState.prototype.activeCurrency;
}
/**
 * @record
 */
export function LanguagesEntities() { }
/**
 * @record
 */
export function LanguagesState() { }
if (false) {
    /** @type {?} */
    LanguagesState.prototype.entities;
    /** @type {?} */
    LanguagesState.prototype.activeLanguage;
}
/**
 * @record
 */
export function BaseSiteState() { }
if (false) {
    /** @type {?} */
    BaseSiteState.prototype.activeSite;
    /** @type {?} */
    BaseSiteState.prototype.details;
}
