// @spartacus/core/esm5/src/routing/services/url-matcher-factory.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { GlobService } from '../../util/glob.service';
import * as i0 from "@angular/core";
import * as i1 from "../../util/glob.service";
var UrlMatcherFactoryService = /** @class */ (function () {
    function UrlMatcherFactoryService(globService) {
        this.globService = globService;
    }
    /**
     * Returns a matcher that is always fails
     */
    /**
     * Returns a matcher that is always fails
     * @return {?}
     */
    UrlMatcherFactoryService.prototype.getFalsyUrlMatcher = /**
     * Returns a matcher that is always fails
     * @return {?}
     */
    function () {
        return (/**
         * @return {?}
         */
        function falsyUrlMatcher() {
            return null;
        });
    };
    /**
     * Returns a matcher for given list of paths
     */
    /**
     * Returns a matcher for given list of paths
     * @param {?} paths
     * @return {?}
     */
    UrlMatcherFactoryService.prototype.getMultiplePathsUrlMatcher = /**
     * Returns a matcher for given list of paths
     * @param {?} paths
     * @return {?}
     */
    function (paths) {
        /** @type {?} */
        var self = this;
        /** @type {?} */
        var matcher = (/**
         * @param {?} segments
         * @param {?} segmentGroup
         * @param {?} route
         * @return {?}
         */
        function multiplePathsUrlMatcher(segments, segmentGroup, route) {
            for (var i = 0; i < paths.length; i++) {
                /** @type {?} */
                var result = self.getPathUrlMatcher(paths[i])(segments, segmentGroup, route);
                if (result) {
                    return result;
                }
            }
            return null;
        });
        matcher.paths = paths; // property added for easier debugging of routes
        return matcher;
    };
    /**
     * Similar to Angular's defaultUrlMatcher. Differences:
     * - the `path` comes from function's argument, not from `route.path`
     * - the empty path `''` is handled here, but in Angular is handled one level higher in the match() function
     */
    /**
     * Similar to Angular's defaultUrlMatcher. Differences:
     * - the `path` comes from function's argument, not from `route.path`
     * - the empty path `''` is handled here, but in Angular is handled one level higher in the match() function
     * @protected
     * @param {?=} path
     * @return {?}
     */
    UrlMatcherFactoryService.prototype.getPathUrlMatcher = /**
     * Similar to Angular's defaultUrlMatcher. Differences:
     * - the `path` comes from function's argument, not from `route.path`
     * - the empty path `''` is handled here, but in Angular is handled one level higher in the match() function
     * @protected
     * @param {?=} path
     * @return {?}
     */
    function (path) {
        if (path === void 0) { path = ''; }
        return (/**
         * @param {?} segments
         * @param {?} segmentGroup
         * @param {?} route
         * @return {?}
         */
        function (segments, segmentGroup, route) {
            /**
             * @license
             * The MIT License
             * Copyright (c) 2010-2019 Google LLC. http://angular.io/license
             *
             * See:
             * - https://github.com/angular/angular/blob/6f5f481fdae03f1d8db36284b64c7b82d9519d85/packages/router/src/shared.ts#L121
             */
            // use function's argument, not the `route.path`
            if (path === '') {
                if (route.pathMatch === 'full' &&
                    (segmentGroup.hasChildren() || segments.length > 0)) {
                    return null;
                }
                return { consumed: [], posParams: {} };
            }
            /** @type {?} */
            var parts = path.split('/');
            if (parts.length > segments.length) {
                // The actual URL is shorter than the config, no match
                return null;
            }
            if (route.pathMatch === 'full' &&
                (segmentGroup.hasChildren() || parts.length < segments.length)) {
                // The config is longer than the actual URL but we are looking for a full match, return null
                return null;
            }
            /** @type {?} */
            var posParams = {};
            // Check each config part against the actual URL
            for (var index = 0; index < parts.length; index++) {
                /** @type {?} */
                var part = parts[index];
                /** @type {?} */
                var segment = segments[index];
                /** @type {?} */
                var isParameter = part.startsWith(':');
                if (isParameter) {
                    posParams[part.substring(1)] = segment;
                }
                else if (part !== segment.path) {
                    // The actual URL part does not match the config, no match
                    return null;
                }
            }
            return { consumed: segments.slice(0, parts.length), posParams: posParams };
        });
    };
    /**
     * Returns URL matcher that accepts almost everything (like `**` route), but not paths accepted by the given matcher
     */
    /**
     * Returns URL matcher that accepts almost everything (like `**` route), but not paths accepted by the given matcher
     * @param {?} originalMatcher
     * @return {?}
     */
    UrlMatcherFactoryService.prototype.getOppositeUrlMatcher = /**
     * Returns URL matcher that accepts almost everything (like `**` route), but not paths accepted by the given matcher
     * @param {?} originalMatcher
     * @return {?}
     */
    function (originalMatcher) {
        /** @type {?} */
        var matcher = (/**
         * @param {?} segments
         * @param {?} group
         * @param {?} route
         * @return {?}
         */
        function oppositeUrlMatcher(segments, group, route) {
            return originalMatcher(segments, group, route)
                ? null
                : { consumed: segments, posParams: {} };
        });
        matcher.originalMatcher = originalMatcher; // property added for easier debugging of routes
        return matcher;
    };
    /**
     * Returns URL matcher for the given list of glob-like patterns. Each pattern must start with `/` or `!/`.
     */
    /**
     * Returns URL matcher for the given list of glob-like patterns. Each pattern must start with `/` or `!/`.
     * @param {?} globPatterns
     * @return {?}
     */
    UrlMatcherFactoryService.prototype.getGlobUrlMatcher = /**
     * Returns URL matcher for the given list of glob-like patterns. Each pattern must start with `/` or `!/`.
     * @param {?} globPatterns
     * @return {?}
     */
    function (globPatterns) {
        /** @type {?} */
        var globValidator = this.globService.getValidator(globPatterns);
        /** @type {?} */
        var matcher = (/**
         * @param {?} segments
         * @return {?}
         */
        function globUrlMatcher(segments) {
            /** @type {?} */
            var fullPath = "/" + segments.map((/**
             * @param {?} s
             * @return {?}
             */
            function (s) { return s.path; })).join('/');
            return globValidator(fullPath)
                ? { consumed: segments, posParams: {} }
                : null;
        });
        matcher.globPatterns = globPatterns; // property added for easier debugging of routes
        return matcher;
    };
    UrlMatcherFactoryService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    UrlMatcherFactoryService.ctorParameters = function () { return [
        { type: GlobService }
    ]; };
    /** @nocollapse */ UrlMatcherFactoryService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function UrlMatcherFactoryService_Factory() { return new UrlMatcherFactoryService(i0.ɵɵinject(i1.GlobService)); }, token: UrlMatcherFactoryService, providedIn: "root" });
    return UrlMatcherFactoryService;
}());
export { UrlMatcherFactoryService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    UrlMatcherFactoryService.prototype.globService;
}
