// @spartacus/core/esm5/src/user/connectors/consent/user-consent.connector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { UserConsentAdapter } from './user-consent.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./user-consent.adapter";
var UserConsentConnector = /** @class */ (function () {
    function UserConsentConnector(adapter) {
        this.adapter = adapter;
    }
    /**
     * @param {?} userId
     * @return {?}
     */
    UserConsentConnector.prototype.loadConsents = /**
     * @param {?} userId
     * @return {?}
     */
    function (userId) {
        return this.adapter.loadConsents(userId);
    };
    /**
     * @param {?} userId
     * @param {?} consentTemplateId
     * @param {?} consentTemplateVersion
     * @return {?}
     */
    UserConsentConnector.prototype.giveConsent = /**
     * @param {?} userId
     * @param {?} consentTemplateId
     * @param {?} consentTemplateVersion
     * @return {?}
     */
    function (userId, consentTemplateId, consentTemplateVersion) {
        return this.adapter.giveConsent(userId, consentTemplateId, consentTemplateVersion);
    };
    /**
     * @param {?} userId
     * @param {?} consentCode
     * @return {?}
     */
    UserConsentConnector.prototype.withdrawConsent = /**
     * @param {?} userId
     * @param {?} consentCode
     * @return {?}
     */
    function (userId, consentCode) {
        return this.adapter.withdrawConsent(userId, consentCode);
    };
    UserConsentConnector.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    UserConsentConnector.ctorParameters = function () { return [
        { type: UserConsentAdapter }
    ]; };
    /** @nocollapse */ UserConsentConnector.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function UserConsentConnector_Factory() { return new UserConsentConnector(i0.ɵɵinject(i1.UserConsentAdapter)); }, token: UserConsentConnector, providedIn: "root" });
    return UserConsentConnector;
}());
export { UserConsentConnector };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    UserConsentConnector.prototype.adapter;
}
