// @spartacus/core/esm5/src/user/store/effects/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BillingCountriesEffect } from './billing-countries.effect';
import { ClearMiscsDataEffect } from './clear-miscs-data.effect';
import { ConsignmentTrackingEffects } from './consignment-tracking.effect';
import { DeliveryCountriesEffects } from './delivery-countries.effect';
import { ForgotPasswordEffects } from './forgot-password.effect';
import { NotificationPreferenceEffects } from './notification-preference.effect';
import { OrderDetailsEffect } from './order-details.effect';
import { OrderReturnRequestEffect } from './order-return-request.effect';
import { UserPaymentMethodsEffects } from './payment-methods.effect';
import { RegionsEffects } from './regions.effect';
import { ResetPasswordEffects } from './reset-password.effect';
import { TitlesEffects } from './titles.effect';
import { UpdateEmailEffects } from './update-email.effect';
import { UpdatePasswordEffects } from './update-password.effect';
import { UserAddressesEffects } from './user-addresses.effect';
import { UserConsentsEffect } from './user-consents.effect';
import { UserDetailsEffects } from './user-details.effect';
import { UserOrdersEffect } from './user-orders.effect';
import { UserRegisterEffects } from './user-register.effect';
import { CustomerCouponEffects } from './customer-coupon.effect';
import { ProductInterestsEffect } from './product-interests.effect';
/** @type {?} */
export var effects = [
    ClearMiscsDataEffect,
    DeliveryCountriesEffects,
    RegionsEffects,
    TitlesEffects,
    UserDetailsEffects,
    UserAddressesEffects,
    UserPaymentMethodsEffects,
    UserRegisterEffects,
    UserOrdersEffect,
    OrderDetailsEffect,
    BillingCountriesEffect,
    ResetPasswordEffects,
    ForgotPasswordEffects,
    UpdateEmailEffects,
    UpdatePasswordEffects,
    UserConsentsEffect,
    ConsignmentTrackingEffects,
    CustomerCouponEffects,
    NotificationPreferenceEffects,
    ProductInterestsEffect,
    OrderReturnRequestEffect,
];
export { BillingCountriesEffect } from './billing-countries.effect';
export { ClearMiscsDataEffect } from './clear-miscs-data.effect';
export { ConsignmentTrackingEffects } from './consignment-tracking.effect';
export { DeliveryCountriesEffects } from './delivery-countries.effect';
export { NotificationPreferenceEffects } from './notification-preference.effect';
export { OrderDetailsEffect } from './order-details.effect';
export { OrderReturnRequestEffect } from './order-return-request.effect';
export { UserPaymentMethodsEffects } from './payment-methods.effect';
export { RegionsEffects } from './regions.effect';
export { ResetPasswordEffects } from './reset-password.effect';
export { TitlesEffects } from './titles.effect';
export { UserAddressesEffects } from './user-addresses.effect';
export { UserConsentsEffect } from './user-consents.effect';
export { UserDetailsEffects } from './user-details.effect';
export { UserOrdersEffect } from './user-orders.effect';
export { UserRegisterEffects } from './user-register.effect';
export { CustomerCouponEffects } from './customer-coupon.effect';
export { ProductInterestsEffect } from './product-interests.effect';
