// @spartacus/schematics/src/shared/utils/package-utils_spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const testing_1 = require("@angular-devkit/schematics/testing");
const path = __importStar(require("path"));
const package_utils_1 = require("./package-utils");
const collectionPath = path.join(__dirname, '../../collection.json');
const schematicRunner = new testing_1.SchematicTestRunner('schematics', collectionPath);
describe('Package utils', () => {
    let appTree;
    const workspaceOptions = {
        name: 'workspace',
        version: '0.5.0',
    };
    const appOptions = {
        name: 'schematics-test',
        inlineStyle: false,
        inlineTemplate: false,
        routing: false,
        style: 'scss',
        skipTests: false,
        projectRoot: '',
    };
    const defaultOptions = {
        project: 'schematics-test',
    };
    beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
        appTree = yield schematicRunner
            .runExternalSchematicAsync('@schematics/angular', 'workspace', workspaceOptions)
            .toPromise();
        appTree = yield schematicRunner
            .runExternalSchematicAsync('@schematics/angular', 'application', appOptions, appTree)
            .toPromise();
        appTree = yield schematicRunner
            .runSchematicAsync('add-spartacus', defaultOptions, appTree)
            .toPromise();
    }));
    describe('getAngularVersion', () => {
        it('should return angular version', () => __awaiter(void 0, void 0, void 0, function* () {
            const testVersion = '5.5.5';
            const buffer = appTree.read('package.json');
            if (buffer) {
                const packageJsonObject = JSON.parse(buffer.toString('utf-8'));
                packageJsonObject.dependencies['@angular/core'] = testVersion;
                appTree.overwrite('package.json', JSON.stringify(packageJsonObject));
                const version = package_utils_1.getAngularVersion(appTree);
                expect(version).toEqual(testVersion);
            }
        }));
    });
    describe('getSpartacusSchematicsVersion', () => {
        it('should return spartacus version', () => __awaiter(void 0, void 0, void 0, function* () {
            const version = package_utils_1.getSpartacusSchematicsVersion();
            expect(version).toBeTruthy();
            expect(version.length).toBeGreaterThanOrEqual(3);
        }));
    });
    describe('getSpartacusCurrentFeatureLevel', () => {
        it('should return feature level based on spartacus current version', () => __awaiter(void 0, void 0, void 0, function* () {
            const version = package_utils_1.getSpartacusSchematicsVersion();
            const featureLevel = package_utils_1.getSpartacusCurrentFeatureLevel();
            expect(featureLevel).toBeTruthy();
            expect(featureLevel.length).toEqual(3);
            expect(featureLevel).toEqual(version.substring(0, 3));
        }));
    });
});
