// @spartacus/core/esm5/src/asm/facade/asm.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { AsmActions } from '../store/actions/index';
import { AsmSelectors } from '../store/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
var AsmService = /** @class */ (function () {
    function AsmService(store) {
        this.store = store;
    }
    /**
     * Search for customers
     * @param options
     */
    /**
     * Search for customers
     * @param {?} options
     * @return {?}
     */
    AsmService.prototype.customerSearch = /**
     * Search for customers
     * @param {?} options
     * @return {?}
     */
    function (options) {
        this.store.dispatch(new AsmActions.CustomerSearch(options));
    };
    /**
     * Reset the customer search result data to the initial state.
     */
    /**
     * Reset the customer search result data to the initial state.
     * @return {?}
     */
    AsmService.prototype.customerSearchReset = /**
     * Reset the customer search result data to the initial state.
     * @return {?}
     */
    function () {
        this.store.dispatch(new AsmActions.CustomerSearchReset());
    };
    /**
     * Returns the customer search result data.
     */
    /**
     * Returns the customer search result data.
     * @return {?}
     */
    AsmService.prototype.getCustomerSearchResults = /**
     * Returns the customer search result data.
     * @return {?}
     */
    function () {
        return this.store.pipe(select(AsmSelectors.getCustomerSearchResults));
    };
    /**
     * Returns the customer search result loading status.
     */
    /**
     * Returns the customer search result loading status.
     * @return {?}
     */
    AsmService.prototype.getCustomerSearchResultsLoading = /**
     * Returns the customer search result loading status.
     * @return {?}
     */
    function () {
        return this.store.pipe(select(AsmSelectors.getCustomerSearchResultsLoading));
    };
    /**
     * Updates the state of the ASM UI
     */
    /**
     * Updates the state of the ASM UI
     * @param {?} asmUi
     * @return {?}
     */
    AsmService.prototype.updateAsmUiState = /**
     * Updates the state of the ASM UI
     * @param {?} asmUi
     * @return {?}
     */
    function (asmUi) {
        this.store.dispatch(new AsmActions.AsmUiUpdate(asmUi));
    };
    /**
     * Get the state of the ASM UI
     */
    /**
     * Get the state of the ASM UI
     * @return {?}
     */
    AsmService.prototype.getAsmUiState = /**
     * Get the state of the ASM UI
     * @return {?}
     */
    function () {
        return this.store.pipe(select(AsmSelectors.getAsmUi));
    };
    AsmService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    AsmService.ctorParameters = function () { return [
        { type: Store }
    ]; };
    /** @nocollapse */ AsmService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AsmService_Factory() { return new AsmService(i0.ɵɵinject(i1.Store)); }, token: AsmService, providedIn: "root" });
    return AsmService;
}());
export { AsmService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    AsmService.prototype.store;
}
