// @spartacus/core/esm5/src/checkout/store/selectors/checkout.selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createFeatureSelector, createSelector, } from '@ngrx/store';
import { StateLoaderSelectors } from '../../../state/utils/index';
import { CHECKOUT_FEATURE, } from '../checkout-state';
/** @type {?} */
var getDeliveryAddressSelector = (/**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.address; });
var ɵ0 = getDeliveryAddressSelector;
/** @type {?} */
var getDeliveryModeSelector = (/**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return state.deliveryMode;
});
var ɵ1 = getDeliveryModeSelector;
/** @type {?} */
var getPaymentDetailsSelector = (/**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return state.paymentDetails;
});
var ɵ2 = getPaymentDetailsSelector;
/** @type {?} */
var getOrderDetailsSelector = (/**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return state.orderDetails;
});
var ɵ3 = getOrderDetailsSelector;
/** @type {?} */
export var getCheckoutState = createFeatureSelector(CHECKOUT_FEATURE);
var ɵ4 = /**
 * @param {?} checkoutState
 * @return {?}
 */
function (checkoutState) { return checkoutState.steps; };
/** @type {?} */
export var getCheckoutStepsState = createSelector(getCheckoutState, (ɵ4));
var ɵ5 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return StateLoaderSelectors.loaderValueSelector(state); };
/** @type {?} */
export var getCheckoutSteps = createSelector(getCheckoutStepsState, (ɵ5));
/** @type {?} */
export var getDeliveryAddress = createSelector(getCheckoutSteps, getDeliveryAddressSelector);
/** @type {?} */
export var getDeliveryMode = createSelector(getCheckoutSteps, getDeliveryModeSelector);
var ɵ6 = /**
 * @param {?} deliveryMode
 * @return {?}
 */
function (deliveryMode) {
    return (deliveryMode &&
        Object.keys(deliveryMode.supported).map((/**
         * @param {?} code
         * @return {?}
         */
        function (code) { return deliveryMode.supported[code]; })));
};
/** @type {?} */
export var getSupportedDeliveryModes = createSelector(getDeliveryMode, (ɵ6));
var ɵ7 = /**
 * @param {?} deliveryMode
 * @return {?}
 */
function (deliveryMode) {
    return deliveryMode && deliveryMode.selected;
};
/** @type {?} */
export var getSelectedDeliveryModeCode = createSelector(getDeliveryMode, (ɵ7));
var ɵ8 = /**
 * @param {?} deliveryMode
 * @return {?}
 */
function (deliveryMode) {
    if (deliveryMode.selected !== '') {
        if (Object.keys(deliveryMode.supported).length === 0) {
            return null;
        }
        return deliveryMode.supported[deliveryMode.selected];
    }
};
/** @type {?} */
export var getSelectedDeliveryMode = createSelector(getDeliveryMode, (ɵ8));
/** @type {?} */
export var getPaymentDetails = createSelector(getCheckoutSteps, getPaymentDetailsSelector);
/** @type {?} */
export var getCheckoutOrderDetails = createSelector(getCheckoutSteps, getOrderDetailsSelector);
var ɵ9 = /**
 * @param {?} state
 * @return {?}
 */
function (state) {
    return StateLoaderSelectors.loaderSuccessSelector(state) &&
        !StateLoaderSelectors.loaderLoadingSelector(state);
};
/** @type {?} */
export var getCheckoutDetailsLoaded = createSelector(getCheckoutStepsState, (ɵ9));
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8, ɵ9 };
