// @spartacus/core/esm5/src/user/store/actions/billing-countries.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var LOAD_BILLING_COUNTRIES = '[User] Load Billing Countries';
/** @type {?} */
export var LOAD_BILLING_COUNTRIES_FAIL = '[User] Load Billing Countries Fail';
/** @type {?} */
export var LOAD_BILLING_COUNTRIES_SUCCESS = '[User] Load Billing Countries Success';
var LoadBillingCountries = /** @class */ (function () {
    function LoadBillingCountries() {
        this.type = LOAD_BILLING_COUNTRIES;
    }
    return LoadBillingCountries;
}());
export { LoadBillingCountries };
if (false) {
    /** @type {?} */
    LoadBillingCountries.prototype.type;
}
var LoadBillingCountriesFail = /** @class */ (function () {
    function LoadBillingCountriesFail(payload) {
        this.payload = payload;
        this.type = LOAD_BILLING_COUNTRIES_FAIL;
    }
    return LoadBillingCountriesFail;
}());
export { LoadBillingCountriesFail };
if (false) {
    /** @type {?} */
    LoadBillingCountriesFail.prototype.type;
    /** @type {?} */
    LoadBillingCountriesFail.prototype.payload;
}
var LoadBillingCountriesSuccess = /** @class */ (function () {
    function LoadBillingCountriesSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_BILLING_COUNTRIES_SUCCESS;
    }
    return LoadBillingCountriesSuccess;
}());
export { LoadBillingCountriesSuccess };
if (false) {
    /** @type {?} */
    LoadBillingCountriesSuccess.prototype.type;
    /** @type {?} */
    LoadBillingCountriesSuccess.prototype.payload;
}
