// @spartacus/core/esm5/src/user/store/actions/customer-coupon.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CUSTOMER_COUPONS, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, CLAIM_CUSTOMER_COUPON_PROCESS_ID, } from '../user-state';
import { LoaderLoadAction, LoaderFailAction, LoaderSuccessAction, LoaderResetAction, } from '../../../state/utils/loader/loader.action';
import { EntityFailAction, EntityLoadAction, EntityResetAction, EntitySuccessAction, } from '../../../state/utils/entity-loader/entity-loader.action';
import { PROCESS_FEATURE } from '../../../process/store';
/** @type {?} */
export var LOAD_CUSTOMER_COUPONS = '[User] Load Customer Coupons';
/** @type {?} */
export var LOAD_CUSTOMER_COUPONS_FAIL = '[User] Load Customer Coupons Fail';
/** @type {?} */
export var LOAD_CUSTOMER_COUPONS_SUCCESS = '[User] Load Customer Coupons Success';
/** @type {?} */
export var RESET_LOAD_CUSTOMER_COUPONS = '[User] Reset Load Customer Coupons';
/** @type {?} */
export var SUBSCRIBE_CUSTOMER_COUPON = '[User] Subscribe Customer Notification Coupon';
/** @type {?} */
export var SUBSCRIBE_CUSTOMER_COUPON_FAIL = '[User] Subscribe Customer Coupon Notification Fail';
/** @type {?} */
export var SUBSCRIBE_CUSTOMER_COUPON_SUCCESS = '[User] Subscribe Customer Coupon Notification Success';
/** @type {?} */
export var RESET_SUBSCRIBE_CUSTOMER_COUPON_PROCESS = '[User] Reset Subscribe Customer Coupon Process';
/** @type {?} */
export var UNSUBSCRIBE_CUSTOMER_COUPON = '[User] Unsubscribe Customer Notification Coupon';
/** @type {?} */
export var UNSUBSCRIBE_CUSTOMER_COUPON_FAIL = '[User] Unsubscribe Customer Coupon Notification Fail';
/** @type {?} */
export var UNSUBSCRIBE_CUSTOMER_COUPON_SUCCESS = '[User] Unsubscribe Customer Coupon Notification Success';
/** @type {?} */
export var RESET_UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS = '[User] Reset Unsubscribe Customer Coupon Process';
/** @type {?} */
export var CLAIM_CUSTOMER_COUPON = '[User] Claim Customer';
/** @type {?} */
export var CLAIM_CUSTOMER_COUPON_FAIL = '[User] Claim Customer Fail';
/** @type {?} */
export var CLAIM_CUSTOMER_COUPON_SUCCESS = '[User] Claim Customer Success';
var LoadCustomerCoupons = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCustomerCoupons, _super);
    function LoadCustomerCoupons(payload) {
        var _this = _super.call(this, CUSTOMER_COUPONS) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_COUPONS;
        return _this;
    }
    return LoadCustomerCoupons;
}(LoaderLoadAction));
export { LoadCustomerCoupons };
if (false) {
    /** @type {?} */
    LoadCustomerCoupons.prototype.type;
    /** @type {?} */
    LoadCustomerCoupons.prototype.payload;
}
var LoadCustomerCouponsFail = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCustomerCouponsFail, _super);
    function LoadCustomerCouponsFail(payload) {
        var _this = _super.call(this, CUSTOMER_COUPONS, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_COUPONS_FAIL;
        return _this;
    }
    return LoadCustomerCouponsFail;
}(LoaderFailAction));
export { LoadCustomerCouponsFail };
if (false) {
    /** @type {?} */
    LoadCustomerCouponsFail.prototype.type;
    /** @type {?} */
    LoadCustomerCouponsFail.prototype.payload;
}
var LoadCustomerCouponsSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(LoadCustomerCouponsSuccess, _super);
    function LoadCustomerCouponsSuccess(payload) {
        var _this = _super.call(this, CUSTOMER_COUPONS) || this;
        _this.payload = payload;
        _this.type = LOAD_CUSTOMER_COUPONS_SUCCESS;
        return _this;
    }
    return LoadCustomerCouponsSuccess;
}(LoaderSuccessAction));
export { LoadCustomerCouponsSuccess };
if (false) {
    /** @type {?} */
    LoadCustomerCouponsSuccess.prototype.type;
    /** @type {?} */
    LoadCustomerCouponsSuccess.prototype.payload;
}
var ResetLoadCustomerCoupons = /** @class */ (function (_super) {
    tslib_1.__extends(ResetLoadCustomerCoupons, _super);
    function ResetLoadCustomerCoupons() {
        var _this = _super.call(this, CUSTOMER_COUPONS) || this;
        _this.type = RESET_LOAD_CUSTOMER_COUPONS;
        return _this;
    }
    return ResetLoadCustomerCoupons;
}(LoaderResetAction));
export { ResetLoadCustomerCoupons };
if (false) {
    /** @type {?} */
    ResetLoadCustomerCoupons.prototype.type;
}
// Subscribe coupon notification actions
var 
// Subscribe coupon notification actions
SubscribeCustomerCoupon = /** @class */ (function (_super) {
    tslib_1.__extends(SubscribeCustomerCoupon, _super);
    function SubscribeCustomerCoupon(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = SUBSCRIBE_CUSTOMER_COUPON;
        return _this;
    }
    return SubscribeCustomerCoupon;
}(EntityLoadAction));
// Subscribe coupon notification actions
export { SubscribeCustomerCoupon };
if (false) {
    /** @type {?} */
    SubscribeCustomerCoupon.prototype.type;
    /** @type {?} */
    SubscribeCustomerCoupon.prototype.payload;
}
var SubscribeCustomerCouponFail = /** @class */ (function (_super) {
    tslib_1.__extends(SubscribeCustomerCouponFail, _super);
    function SubscribeCustomerCouponFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = SUBSCRIBE_CUSTOMER_COUPON_FAIL;
        return _this;
    }
    return SubscribeCustomerCouponFail;
}(EntityFailAction));
export { SubscribeCustomerCouponFail };
if (false) {
    /** @type {?} */
    SubscribeCustomerCouponFail.prototype.type;
    /** @type {?} */
    SubscribeCustomerCouponFail.prototype.payload;
}
var SubscribeCustomerCouponSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(SubscribeCustomerCouponSuccess, _super);
    function SubscribeCustomerCouponSuccess(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = SUBSCRIBE_CUSTOMER_COUPON_SUCCESS;
        return _this;
    }
    return SubscribeCustomerCouponSuccess;
}(EntitySuccessAction));
export { SubscribeCustomerCouponSuccess };
if (false) {
    /** @type {?} */
    SubscribeCustomerCouponSuccess.prototype.type;
    /** @type {?} */
    SubscribeCustomerCouponSuccess.prototype.payload;
}
var ResetSubscribeCustomerCouponProcess = /** @class */ (function (_super) {
    tslib_1.__extends(ResetSubscribeCustomerCouponProcess, _super);
    function ResetSubscribeCustomerCouponProcess() {
        var _this = _super.call(this, PROCESS_FEATURE, SUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID) || this;
        _this.type = RESET_SUBSCRIBE_CUSTOMER_COUPON_PROCESS;
        return _this;
    }
    return ResetSubscribeCustomerCouponProcess;
}(EntityResetAction));
export { ResetSubscribeCustomerCouponProcess };
if (false) {
    /** @type {?} */
    ResetSubscribeCustomerCouponProcess.prototype.type;
}
var UnsubscribeCustomerCoupon = /** @class */ (function (_super) {
    tslib_1.__extends(UnsubscribeCustomerCoupon, _super);
    function UnsubscribeCustomerCoupon(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = UNSUBSCRIBE_CUSTOMER_COUPON;
        return _this;
    }
    return UnsubscribeCustomerCoupon;
}(EntityLoadAction));
export { UnsubscribeCustomerCoupon };
if (false) {
    /** @type {?} */
    UnsubscribeCustomerCoupon.prototype.type;
    /** @type {?} */
    UnsubscribeCustomerCoupon.prototype.payload;
}
var UnsubscribeCustomerCouponFail = /** @class */ (function (_super) {
    tslib_1.__extends(UnsubscribeCustomerCouponFail, _super);
    function UnsubscribeCustomerCouponFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = UNSUBSCRIBE_CUSTOMER_COUPON_FAIL;
        return _this;
    }
    return UnsubscribeCustomerCouponFail;
}(EntityFailAction));
export { UnsubscribeCustomerCouponFail };
if (false) {
    /** @type {?} */
    UnsubscribeCustomerCouponFail.prototype.type;
    /** @type {?} */
    UnsubscribeCustomerCouponFail.prototype.payload;
}
var UnsubscribeCustomerCouponSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(UnsubscribeCustomerCouponSuccess, _super);
    function UnsubscribeCustomerCouponSuccess(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = UNSUBSCRIBE_CUSTOMER_COUPON_SUCCESS;
        return _this;
    }
    return UnsubscribeCustomerCouponSuccess;
}(EntitySuccessAction));
export { UnsubscribeCustomerCouponSuccess };
if (false) {
    /** @type {?} */
    UnsubscribeCustomerCouponSuccess.prototype.type;
    /** @type {?} */
    UnsubscribeCustomerCouponSuccess.prototype.payload;
}
var ResetUnsubscribeCustomerCouponProcess = /** @class */ (function (_super) {
    tslib_1.__extends(ResetUnsubscribeCustomerCouponProcess, _super);
    function ResetUnsubscribeCustomerCouponProcess() {
        var _this = _super.call(this, PROCESS_FEATURE, UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS_ID) || this;
        _this.type = RESET_UNSUBSCRIBE_CUSTOMER_COUPON_PROCESS;
        return _this;
    }
    return ResetUnsubscribeCustomerCouponProcess;
}(EntityResetAction));
export { ResetUnsubscribeCustomerCouponProcess };
if (false) {
    /** @type {?} */
    ResetUnsubscribeCustomerCouponProcess.prototype.type;
}
var ClaimCustomerCoupon = /** @class */ (function (_super) {
    tslib_1.__extends(ClaimCustomerCoupon, _super);
    function ClaimCustomerCoupon(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, CLAIM_CUSTOMER_COUPON_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = CLAIM_CUSTOMER_COUPON;
        return _this;
    }
    return ClaimCustomerCoupon;
}(EntityLoadAction));
export { ClaimCustomerCoupon };
if (false) {
    /** @type {?} */
    ClaimCustomerCoupon.prototype.type;
    /** @type {?} */
    ClaimCustomerCoupon.prototype.payload;
}
var ClaimCustomerCouponFail = /** @class */ (function (_super) {
    tslib_1.__extends(ClaimCustomerCouponFail, _super);
    function ClaimCustomerCouponFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, CLAIM_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = CLAIM_CUSTOMER_COUPON_FAIL;
        return _this;
    }
    return ClaimCustomerCouponFail;
}(EntityFailAction));
export { ClaimCustomerCouponFail };
if (false) {
    /** @type {?} */
    ClaimCustomerCouponFail.prototype.type;
    /** @type {?} */
    ClaimCustomerCouponFail.prototype.payload;
}
var ClaimCustomerCouponSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(ClaimCustomerCouponSuccess, _super);
    function ClaimCustomerCouponSuccess(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, CLAIM_CUSTOMER_COUPON_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = CLAIM_CUSTOMER_COUPON_SUCCESS;
        return _this;
    }
    return ClaimCustomerCouponSuccess;
}(EntitySuccessAction));
export { ClaimCustomerCouponSuccess };
if (false) {
    /** @type {?} */
    ClaimCustomerCouponSuccess.prototype.type;
    /** @type {?} */
    ClaimCustomerCouponSuccess.prototype.payload;
}
