// @spartacus/core/esm5/src/user/store/actions/delivery-countries.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var LOAD_DELIVERY_COUNTRIES = '[User] Load Delivery Countries';
/** @type {?} */
export var LOAD_DELIVERY_COUNTRIES_FAIL = '[User] Load Delivery Countries Fail';
/** @type {?} */
export var LOAD_DELIVERY_COUNTRIES_SUCCESS = '[User] Load Delivery Countries Success';
var LoadDeliveryCountries = /** @class */ (function () {
    function LoadDeliveryCountries() {
        this.type = LOAD_DELIVERY_COUNTRIES;
    }
    return LoadDeliveryCountries;
}());
export { LoadDeliveryCountries };
if (false) {
    /** @type {?} */
    LoadDeliveryCountries.prototype.type;
}
var LoadDeliveryCountriesFail = /** @class */ (function () {
    function LoadDeliveryCountriesFail(payload) {
        this.payload = payload;
        this.type = LOAD_DELIVERY_COUNTRIES_FAIL;
    }
    return LoadDeliveryCountriesFail;
}());
export { LoadDeliveryCountriesFail };
if (false) {
    /** @type {?} */
    LoadDeliveryCountriesFail.prototype.type;
    /** @type {?} */
    LoadDeliveryCountriesFail.prototype.payload;
}
var LoadDeliveryCountriesSuccess = /** @class */ (function () {
    function LoadDeliveryCountriesSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_DELIVERY_COUNTRIES_SUCCESS;
    }
    return LoadDeliveryCountriesSuccess;
}());
export { LoadDeliveryCountriesSuccess };
if (false) {
    /** @type {?} */
    LoadDeliveryCountriesSuccess.prototype.type;
    /** @type {?} */
    LoadDeliveryCountriesSuccess.prototype.payload;
}
