// @spartacus/core/esm5/src/user/store/actions/forgot-password.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var FORGOT_PASSWORD_EMAIL_REQUEST = '[User] Forgot Password Email Request';
/** @type {?} */
export var FORGOT_PASSWORD_EMAIL_REQUEST_SUCCESS = '[User] Forgot Password Email Request Success';
/** @type {?} */
export var FORGOT_PASSWORD_EMAIL_REQUEST_FAIL = '[User] Forgot Password Email Request Fail';
var ForgotPasswordEmailRequest = /** @class */ (function () {
    function ForgotPasswordEmailRequest(payload) {
        this.payload = payload;
        this.type = FORGOT_PASSWORD_EMAIL_REQUEST;
    }
    return ForgotPasswordEmailRequest;
}());
export { ForgotPasswordEmailRequest };
if (false) {
    /** @type {?} */
    ForgotPasswordEmailRequest.prototype.type;
    /** @type {?} */
    ForgotPasswordEmailRequest.prototype.payload;
}
var ForgotPasswordEmailRequestFail = /** @class */ (function () {
    function ForgotPasswordEmailRequestFail(payload) {
        this.payload = payload;
        this.type = FORGOT_PASSWORD_EMAIL_REQUEST_FAIL;
    }
    return ForgotPasswordEmailRequestFail;
}());
export { ForgotPasswordEmailRequestFail };
if (false) {
    /** @type {?} */
    ForgotPasswordEmailRequestFail.prototype.type;
    /** @type {?} */
    ForgotPasswordEmailRequestFail.prototype.payload;
}
var ForgotPasswordEmailRequestSuccess = /** @class */ (function () {
    function ForgotPasswordEmailRequestSuccess() {
        this.type = FORGOT_PASSWORD_EMAIL_REQUEST_SUCCESS;
    }
    return ForgotPasswordEmailRequestSuccess;
}());
export { ForgotPasswordEmailRequestSuccess };
if (false) {
    /** @type {?} */
    ForgotPasswordEmailRequestSuccess.prototype.type;
}
