// @spartacus/core/esm5/src/user/store/actions/notification-preference.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { StateLoaderActions } from '../../../state/utils/index';
import { PROCESS_FEATURE } from '../../../process/store/process-state';
import { EntityFailAction, EntityLoadAction, EntityResetAction, EntitySuccessAction, } from '../../../state/utils/entity-loader/entity-loader.action';
import { UPDATE_NOTIFICATION_PREFERENCES_PROCESS_ID, NOTIFICATION_PREFERENCES, } from '../user-state';
/** @type {?} */
export var LOAD_NOTIFICATION_PREFERENCES = '[User] Load Notification Preferences';
/** @type {?} */
export var LOAD_NOTIFICATION_PREFERENCES_FAIL = '[User] Load Notification Preferences Fail';
/** @type {?} */
export var LOAD_NOTIFICATION_PREFERENCES_SUCCESS = '[User] Load Notification Preferences Success';
/** @type {?} */
export var UPDATE_NOTIFICATION_PREFERENCES = '[User] Update Notification Preferences';
/** @type {?} */
export var UPDATE_NOTIFICATION_PREFERENCES_FAIL = '[User] Update Notification Preferences Fail';
/** @type {?} */
export var UPDATE_NOTIFICATION_PREFERENCES_SUCCESS = '[User] Update Notification Preferences Success';
/** @type {?} */
export var RESET_NOTIFICATION_PREFERENCES = '[User] Reset Notification Preferences';
/** @type {?} */
export var CLEAR_NOTIFICATION_PREFERENCES = '[User] Clear Notification Preferences';
var LoadNotificationPreferences = /** @class */ (function (_super) {
    tslib_1.__extends(LoadNotificationPreferences, _super);
    function LoadNotificationPreferences(payload) {
        var _this = _super.call(this, NOTIFICATION_PREFERENCES) || this;
        _this.payload = payload;
        _this.type = LOAD_NOTIFICATION_PREFERENCES;
        return _this;
    }
    return LoadNotificationPreferences;
}(StateLoaderActions.LoaderLoadAction));
export { LoadNotificationPreferences };
if (false) {
    /** @type {?} */
    LoadNotificationPreferences.prototype.type;
    /** @type {?} */
    LoadNotificationPreferences.prototype.payload;
}
var LoadNotificationPreferencesFail = /** @class */ (function (_super) {
    tslib_1.__extends(LoadNotificationPreferencesFail, _super);
    function LoadNotificationPreferencesFail(payload) {
        var _this = _super.call(this, NOTIFICATION_PREFERENCES, payload) || this;
        _this.payload = payload;
        _this.type = LOAD_NOTIFICATION_PREFERENCES_FAIL;
        return _this;
    }
    return LoadNotificationPreferencesFail;
}(StateLoaderActions.LoaderFailAction));
export { LoadNotificationPreferencesFail };
if (false) {
    /** @type {?} */
    LoadNotificationPreferencesFail.prototype.type;
    /** @type {?} */
    LoadNotificationPreferencesFail.prototype.payload;
}
var LoadNotificationPreferencesSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(LoadNotificationPreferencesSuccess, _super);
    function LoadNotificationPreferencesSuccess(payload) {
        var _this = _super.call(this, NOTIFICATION_PREFERENCES) || this;
        _this.payload = payload;
        _this.type = LOAD_NOTIFICATION_PREFERENCES_SUCCESS;
        return _this;
    }
    return LoadNotificationPreferencesSuccess;
}(StateLoaderActions.LoaderSuccessAction));
export { LoadNotificationPreferencesSuccess };
if (false) {
    /** @type {?} */
    LoadNotificationPreferencesSuccess.prototype.type;
    /** @type {?} */
    LoadNotificationPreferencesSuccess.prototype.payload;
}
var UpdateNotificationPreferences = /** @class */ (function (_super) {
    tslib_1.__extends(UpdateNotificationPreferences, _super);
    function UpdateNotificationPreferences(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_NOTIFICATION_PREFERENCES_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = UPDATE_NOTIFICATION_PREFERENCES;
        return _this;
    }
    return UpdateNotificationPreferences;
}(EntityLoadAction));
export { UpdateNotificationPreferences };
if (false) {
    /** @type {?} */
    UpdateNotificationPreferences.prototype.type;
    /** @type {?} */
    UpdateNotificationPreferences.prototype.payload;
}
var UpdateNotificationPreferencesFail = /** @class */ (function (_super) {
    tslib_1.__extends(UpdateNotificationPreferencesFail, _super);
    function UpdateNotificationPreferencesFail(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_NOTIFICATION_PREFERENCES_PROCESS_ID, payload) || this;
        _this.payload = payload;
        _this.type = UPDATE_NOTIFICATION_PREFERENCES_FAIL;
        return _this;
    }
    return UpdateNotificationPreferencesFail;
}(EntityFailAction));
export { UpdateNotificationPreferencesFail };
if (false) {
    /** @type {?} */
    UpdateNotificationPreferencesFail.prototype.type;
    /** @type {?} */
    UpdateNotificationPreferencesFail.prototype.payload;
}
var UpdateNotificationPreferencesSuccess = /** @class */ (function (_super) {
    tslib_1.__extends(UpdateNotificationPreferencesSuccess, _super);
    function UpdateNotificationPreferencesSuccess(payload) {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_NOTIFICATION_PREFERENCES_PROCESS_ID) || this;
        _this.payload = payload;
        _this.type = UPDATE_NOTIFICATION_PREFERENCES_SUCCESS;
        return _this;
    }
    return UpdateNotificationPreferencesSuccess;
}(EntitySuccessAction));
export { UpdateNotificationPreferencesSuccess };
if (false) {
    /** @type {?} */
    UpdateNotificationPreferencesSuccess.prototype.type;
    /** @type {?} */
    UpdateNotificationPreferencesSuccess.prototype.payload;
}
var ResetNotificationPreferences = /** @class */ (function (_super) {
    tslib_1.__extends(ResetNotificationPreferences, _super);
    function ResetNotificationPreferences() {
        var _this = _super.call(this, PROCESS_FEATURE, UPDATE_NOTIFICATION_PREFERENCES_PROCESS_ID) || this;
        _this.type = RESET_NOTIFICATION_PREFERENCES;
        return _this;
    }
    return ResetNotificationPreferences;
}(EntityResetAction));
export { ResetNotificationPreferences };
if (false) {
    /** @type {?} */
    ResetNotificationPreferences.prototype.type;
}
var ClearNotificationPreferences = /** @class */ (function (_super) {
    tslib_1.__extends(ClearNotificationPreferences, _super);
    function ClearNotificationPreferences() {
        var _this = _super.call(this, NOTIFICATION_PREFERENCES) || this;
        _this.type = CLEAR_NOTIFICATION_PREFERENCES;
        return _this;
    }
    return ClearNotificationPreferences;
}(StateLoaderActions.LoaderResetAction));
export { ClearNotificationPreferences };
if (false) {
    /** @type {?} */
    ClearNotificationPreferences.prototype.type;
}
