// @spartacus/core/esm5/src/user/store/actions/reset-password.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var RESET_PASSWORD = '[User] Reset Password';
/** @type {?} */
export var RESET_PASSWORD_SUCCESS = '[User] Reset Password Success';
/** @type {?} */
export var RESET_PASSWORD_FAIL = '[User] Reset Password Fail';
var ResetPassword = /** @class */ (function () {
    function ResetPassword(payload) {
        this.payload = payload;
        this.type = RESET_PASSWORD;
    }
    return ResetPassword;
}());
export { ResetPassword };
if (false) {
    /** @type {?} */
    ResetPassword.prototype.type;
    /** @type {?} */
    ResetPassword.prototype.payload;
}
var ResetPasswordFail = /** @class */ (function () {
    function ResetPasswordFail(payload) {
        this.payload = payload;
        this.type = RESET_PASSWORD_FAIL;
    }
    return ResetPasswordFail;
}());
export { ResetPasswordFail };
if (false) {
    /** @type {?} */
    ResetPasswordFail.prototype.type;
    /** @type {?} */
    ResetPasswordFail.prototype.payload;
}
var ResetPasswordSuccess = /** @class */ (function () {
    function ResetPasswordSuccess() {
        this.type = RESET_PASSWORD_SUCCESS;
    }
    return ResetPasswordSuccess;
}());
export { ResetPasswordSuccess };
if (false) {
    /** @type {?} */
    ResetPasswordSuccess.prototype.type;
}
