// @spartacus/core/esm5/src/user/store/actions/titles.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var LOAD_TITLES = '[User] Load Tiltes';
/** @type {?} */
export var LOAD_TITLES_FAIL = '[User] Load Titles Fail';
/** @type {?} */
export var LOAD_TITLES_SUCCESS = '[User] Load Titles Success';
var LoadTitles = /** @class */ (function () {
    function LoadTitles() {
        this.type = LOAD_TITLES;
    }
    return LoadTitles;
}());
export { LoadTitles };
if (false) {
    /** @type {?} */
    LoadTitles.prototype.type;
}
var LoadTitlesFail = /** @class */ (function () {
    function LoadTitlesFail(payload) {
        this.payload = payload;
        this.type = LOAD_TITLES_FAIL;
    }
    return LoadTitlesFail;
}());
export { LoadTitlesFail };
if (false) {
    /** @type {?} */
    LoadTitlesFail.prototype.type;
    /** @type {?} */
    LoadTitlesFail.prototype.payload;
}
var LoadTitlesSuccess = /** @class */ (function () {
    function LoadTitlesSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_TITLES_SUCCESS;
    }
    return LoadTitlesSuccess;
}());
export { LoadTitlesSuccess };
if (false) {
    /** @type {?} */
    LoadTitlesSuccess.prototype.type;
    /** @type {?} */
    LoadTitlesSuccess.prototype.payload;
}
