// @spartacus/core/esm5/src/routing/store/selectors/routing.selector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createFeatureSelector, createSelector, } from '@ngrx/store';
import { ROUTING_FEATURE } from '../routing-state';
/** @type {?} */
export var getRouterFeatureState = createFeatureSelector(ROUTING_FEATURE);
var ɵ0 = /**
 * @param {?} state
 * @return {?}
 */
function (state) { return state.router; };
/** @type {?} */
export var getRouterState = createSelector(getRouterFeatureState, (ɵ0));
var ɵ1 = /**
 * @param {?} routingState
 * @return {?}
 */
function (routingState) {
    return (routingState.state && routingState.state.context) || { id: '' };
};
/** @type {?} */
export var getPageContext = createSelector(getRouterState, (ɵ1));
var ɵ2 = /**
 * @param {?} routingState
 * @return {?}
 */
function (routingState) {
    return routingState.nextState && routingState.nextState.context;
};
/** @type {?} */
export var getNextPageContext = createSelector(getRouterState, (ɵ2));
var ɵ3 = /**
 * @param {?} context
 * @return {?}
 */
function (context) { return !!context; };
/** @type {?} */
export var isNavigating = createSelector(getNextPageContext, (ɵ3));
export { ɵ0, ɵ1, ɵ2, ɵ3 };
