// @spartacus/core/esm5/src/product/facade/searchbox.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { select } from '@ngrx/store';
import { ProductActions } from '../store/actions/index';
import { ProductSelectors } from '../store/selectors/index';
import { ProductSearchService } from './product-search.service';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
var SearchboxService = /** @class */ (function (_super) {
    tslib_1.__extends(SearchboxService, _super);
    function SearchboxService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * dispatch the search for the search box
     */
    /**
     * dispatch the search for the search box
     * @param {?} query
     * @param {?=} searchConfig
     * @return {?}
     */
    SearchboxService.prototype.search = /**
     * dispatch the search for the search box
     * @param {?} query
     * @param {?=} searchConfig
     * @return {?}
     */
    function (query, searchConfig) {
        this.store.dispatch(new ProductActions.SearchProducts({
            queryText: query,
            searchConfig: searchConfig,
        }, true));
    };
    /**
     * @return {?}
     */
    SearchboxService.prototype.getResults = /**
     * @return {?}
     */
    function () {
        return this.store.pipe(select(ProductSelectors.getAuxSearchResults));
    };
    /**
     * clears the products and suggestions
     */
    /**
     * clears the products and suggestions
     * @return {?}
     */
    SearchboxService.prototype.clearResults = /**
     * clears the products and suggestions
     * @return {?}
     */
    function () {
        this.store.dispatch(new ProductActions.ClearProductSearchResult({
            clearSearchboxResults: true,
        }));
    };
    /**
     * @return {?}
     */
    SearchboxService.prototype.getSuggestionResults = /**
     * @return {?}
     */
    function () {
        return this.store.pipe(select(ProductSelectors.getProductSuggestions));
    };
    /**
     * @param {?} query
     * @param {?=} searchConfig
     * @return {?}
     */
    SearchboxService.prototype.searchSuggestions = /**
     * @param {?} query
     * @param {?=} searchConfig
     * @return {?}
     */
    function (query, searchConfig) {
        this.store.dispatch(new ProductActions.GetProductSuggestions({
            term: query,
            searchConfig: searchConfig,
        }));
    };
    SearchboxService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */ SearchboxService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function SearchboxService_Factory() { return new SearchboxService(i0.ɵɵinject(i1.Store)); }, token: SearchboxService, providedIn: "root" });
    return SearchboxService;
}(ProductSearchService));
export { SearchboxService };
