// @spartacus/core/esm2015/src/product/store/product-state.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const PRODUCT_FEATURE = 'product';
/** @type {?} */
export const PRODUCT_DETAIL_ENTITY = '[Product] Detail Entity';
/**
 * @record
 */
export function StateWithProduct() { }
if (false) {
    /* Skipping unnamed member:
    [PRODUCT_FEATURE]: ProductsState;*/
}
/**
 * @record
 */
export function ProductsState() { }
if (false) {
    /** @type {?} */
    ProductsState.prototype.details;
    /** @type {?} */
    ProductsState.prototype.search;
    /** @type {?} */
    ProductsState.prototype.reviews;
    /** @type {?} */
    ProductsState.prototype.references;
}
/**
 * @record
 */
export function ProductsSearchState() { }
if (false) {
    /** @type {?} */
    ProductsSearchState.prototype.results;
    /** @type {?} */
    ProductsSearchState.prototype.suggestions;
    /** @type {?} */
    ProductsSearchState.prototype.auxResults;
}
/**
 * @record
 */
export function ProductReviewsState() { }
if (false) {
    /** @type {?} */
    ProductReviewsState.prototype.productCode;
    /** @type {?} */
    ProductReviewsState.prototype.list;
}
/**
 * @record
 */
export function ProductReferencesState() { }
if (false) {
    /** @type {?} */
    ProductReferencesState.prototype.productCode;
    /** @type {?} */
    ProductReferencesState.prototype.list;
}
