// @spartacus/core/esm5/src/site-context/store/actions/base-site.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var LOAD_BASE_SITE = '[Site-context] Load BaseSite';
/** @type {?} */
export var LOAD_BASE_SITE_FAIL = '[Site-context] Load BaseSite Fail';
/** @type {?} */
export var LOAD_BASE_SITE_SUCCESS = '[Site-context] Load BaseSite Success';
/** @type {?} */
export var SET_ACTIVE_BASE_SITE = '[Site-context] Set Active BaseSite';
/** @type {?} */
export var BASE_SITE_CHANGE = '[Site-context] BaseSite Change';
var LoadBaseSite = /** @class */ (function () {
    function LoadBaseSite() {
        this.type = LOAD_BASE_SITE;
    }
    return LoadBaseSite;
}());
export { LoadBaseSite };
if (false) {
    /** @type {?} */
    LoadBaseSite.prototype.type;
}
var LoadBaseSiteFail = /** @class */ (function () {
    function LoadBaseSiteFail(payload) {
        this.payload = payload;
        this.type = LOAD_BASE_SITE_FAIL;
    }
    return LoadBaseSiteFail;
}());
export { LoadBaseSiteFail };
if (false) {
    /** @type {?} */
    LoadBaseSiteFail.prototype.type;
    /** @type {?} */
    LoadBaseSiteFail.prototype.payload;
}
var LoadBaseSiteSuccess = /** @class */ (function () {
    function LoadBaseSiteSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_BASE_SITE_SUCCESS;
    }
    return LoadBaseSiteSuccess;
}());
export { LoadBaseSiteSuccess };
if (false) {
    /** @type {?} */
    LoadBaseSiteSuccess.prototype.type;
    /** @type {?} */
    LoadBaseSiteSuccess.prototype.payload;
}
var SetActiveBaseSite = /** @class */ (function () {
    function SetActiveBaseSite(payload) {
        this.payload = payload;
        this.type = SET_ACTIVE_BASE_SITE;
    }
    return SetActiveBaseSite;
}());
export { SetActiveBaseSite };
if (false) {
    /** @type {?} */
    SetActiveBaseSite.prototype.type;
    /** @type {?} */
    SetActiveBaseSite.prototype.payload;
}
var BaseSiteChange = /** @class */ (function () {
    function BaseSiteChange() {
        this.type = BASE_SITE_CHANGE;
    }
    return BaseSiteChange;
}());
export { BaseSiteChange };
if (false) {
    /** @type {?} */
    BaseSiteChange.prototype.type;
}
