// @spartacus/core/esm5/src/site-context/store/actions/languages.action.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var LOAD_LANGUAGES = '[Site-context] Load Languages';
/** @type {?} */
export var LOAD_LANGUAGES_FAIL = '[Site-context] Load Languages Fail';
/** @type {?} */
export var LOAD_LANGUAGES_SUCCESS = '[Site-context] Load Languages Success';
/** @type {?} */
export var SET_ACTIVE_LANGUAGE = '[Site-context] Set Active Language';
/** @type {?} */
export var LANGUAGE_CHANGE = '[Site-context] Language Change';
var LoadLanguages = /** @class */ (function () {
    function LoadLanguages() {
        this.type = LOAD_LANGUAGES;
    }
    return LoadLanguages;
}());
export { LoadLanguages };
if (false) {
    /** @type {?} */
    LoadLanguages.prototype.type;
}
var LoadLanguagesFail = /** @class */ (function () {
    function LoadLanguagesFail(payload) {
        this.payload = payload;
        this.type = LOAD_LANGUAGES_FAIL;
    }
    return LoadLanguagesFail;
}());
export { LoadLanguagesFail };
if (false) {
    /** @type {?} */
    LoadLanguagesFail.prototype.type;
    /** @type {?} */
    LoadLanguagesFail.prototype.payload;
}
var LoadLanguagesSuccess = /** @class */ (function () {
    function LoadLanguagesSuccess(payload) {
        this.payload = payload;
        this.type = LOAD_LANGUAGES_SUCCESS;
    }
    return LoadLanguagesSuccess;
}());
export { LoadLanguagesSuccess };
if (false) {
    /** @type {?} */
    LoadLanguagesSuccess.prototype.type;
    /** @type {?} */
    LoadLanguagesSuccess.prototype.payload;
}
var SetActiveLanguage = /** @class */ (function () {
    function SetActiveLanguage(payload) {
        this.payload = payload;
        this.type = SET_ACTIVE_LANGUAGE;
    }
    return SetActiveLanguage;
}());
export { SetActiveLanguage };
if (false) {
    /** @type {?} */
    SetActiveLanguage.prototype.type;
    /** @type {?} */
    SetActiveLanguage.prototype.payload;
}
var LanguageChange = /** @class */ (function () {
    function LanguageChange() {
        this.type = LANGUAGE_CHANGE;
    }
    return LanguageChange;
}());
export { LanguageChange };
if (false) {
    /** @type {?} */
    LanguageChange.prototype.type;
}
