// @spartacus/core/esm5/src/occ/services/loading-scopes.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { OccConfig } from '../config/occ-config';
import * as i0 from "@angular/core";
import * as i1 from "../config/occ-config";
var LoadingScopesService = /** @class */ (function () {
    function LoadingScopesService(config) {
        this.config = config;
    }
    /**
     * Aims to expand scopes based on loading scopes config.
     *
     * I.e. if 'details' scope includes 'list' scope by configuration, it'll return ['details', 'list']
     *
     * If scope data overlaps with each other, the data should be merged in the order of scopes provided,
     * i.e. the last scope is merged last, overwriting parts of the data already provided by preceding scope.
     * It should apply also to implicit scopes (that are included by configuration).
     *
     * @param model
     * @param scopes
     */
    /**
     * Aims to expand scopes based on loading scopes config.
     *
     * I.e. if 'details' scope includes 'list' scope by configuration, it'll return ['details', 'list']
     *
     * If scope data overlaps with each other, the data should be merged in the order of scopes provided,
     * i.e. the last scope is merged last, overwriting parts of the data already provided by preceding scope.
     * It should apply also to implicit scopes (that are included by configuration).
     *
     * @param {?} model
     * @param {?} scopes
     * @return {?}
     */
    LoadingScopesService.prototype.expand = /**
     * Aims to expand scopes based on loading scopes config.
     *
     * I.e. if 'details' scope includes 'list' scope by configuration, it'll return ['details', 'list']
     *
     * If scope data overlaps with each other, the data should be merged in the order of scopes provided,
     * i.e. the last scope is merged last, overwriting parts of the data already provided by preceding scope.
     * It should apply also to implicit scopes (that are included by configuration).
     *
     * @param {?} model
     * @param {?} scopes
     * @return {?}
     */
    function (model, scopes) {
        var e_1, _a;
        /** @type {?} */
        var scopesConfig = this.config &&
            this.config.backend &&
            this.config.backend.loadingScopes &&
            this.config.backend.loadingScopes[model];
        if (scopesConfig) {
            /** @type {?} */
            var expandedScopes = tslib_1.__spread(scopes);
            /** @type {?} */
            var i = expandedScopes.length;
            while (i > 0) {
                i--;
                /** @type {?} */
                var includedScopes = scopesConfig[expandedScopes[i]] &&
                    scopesConfig[expandedScopes[i]].include;
                if (includedScopes) {
                    try {
                        for (var includedScopes_1 = (e_1 = void 0, tslib_1.__values(includedScopes)), includedScopes_1_1 = includedScopes_1.next(); !includedScopes_1_1.done; includedScopes_1_1 = includedScopes_1.next()) {
                            var includedScope = includedScopes_1_1.value;
                            if (!expandedScopes.includes(includedScope)) {
                                expandedScopes.splice(i, 0, includedScope);
                                i++;
                            }
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (includedScopes_1_1 && !includedScopes_1_1.done && (_a = includedScopes_1.return)) _a.call(includedScopes_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
            }
            return expandedScopes;
        }
        return scopes;
    };
    /**
     * Return maxAge for product scope in milliseconds
     *
     * @param model
     * @param scope
     */
    /**
     * Return maxAge for product scope in milliseconds
     *
     * @param {?} model
     * @param {?} scope
     * @return {?}
     */
    LoadingScopesService.prototype.getMaxAge = /**
     * Return maxAge for product scope in milliseconds
     *
     * @param {?} model
     * @param {?} scope
     * @return {?}
     */
    function (model, scope) {
        /** @type {?} */
        var scopesConfig = this.config &&
            this.config.backend &&
            this.config.backend.loadingScopes &&
            this.config.backend.loadingScopes[model];
        return (scopesConfig[scope] && scopesConfig[scope].maxAge) * 1000 || 0;
    };
    LoadingScopesService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    LoadingScopesService.ctorParameters = function () { return [
        { type: OccConfig }
    ]; };
    /** @nocollapse */ LoadingScopesService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function LoadingScopesService_Factory() { return new LoadingScopesService(i0.ɵɵinject(i1.OccConfig)); }, token: LoadingScopesService, providedIn: "root" });
    return LoadingScopesService;
}());
export { LoadingScopesService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    LoadingScopesService.prototype.config;
}
